---
title: "Contents"
description: "Tutorials on some of the common steps during scRNAseq data analysis using three popular toolkits: [ {{< fa brands r-project >}} Seurat](https://satijalab.org/seurat/), [{{< fa brands r-project >}} Bioconductor](https://bioconductor.org/packages/release/bioc/html/scran.html) and [{{< fa brands python >}} Scanpy](https://scanpy.readthedocs.io/en/stable/)."
date: ""
author: ""
toc: false
number-sections: false
sidebar: false
code-tools: false
format: html
---

In this workshop we cover the topics below using three different toolkits. All analyses follow the same steps, however, there are some minor differences between the toolkits, as not all methods are implemented the same everywhere. To *set up a local environment* with Docker, see the [precourse instructions](home_precourse.qmd).

You can *run the analysis* using the toolkit of choice in one of the following ways:
- On Uppmax, using Singularity (recommended during the workshop)
- Locally, using Docker

The instructions for both options can be found [here](containers/docs/run-containers.qmd).


| Topic | {{< fa brands r-project >}} Seurat | {{< fa brands r-project >}} Bioconductor | {{< fa brands python >}} Scanpy |
|:--------|:---------------------------:|:---------------------------------:|:------------------------:|
| {{< fa clipboard-list >}} Quality Control | [{{< fa file-lines >}}](labs/seurat/seurat_01_qc.html) | [{{< fa file-lines >}}](labs/bioc/bioc_01_qc.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_01_qc.html) |
| {{< fa brands hubspot >}} Dimensionality reduction | [{{< fa file-lines >}}](labs/seurat/seurat_02_dimred.html) | [{{< fa file-lines >}}](labs/bioc/bioc_02_dimred.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_02_dimred.html) |
| {{< fa diagram-project >}} Data integration | [{{< fa file-lines >}}](labs/seurat/seurat_03_integration.html) | [{{< fa file-lines >}}](labs/bioc/bioc_03_integration.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_03_integration.html) |
| {{< fa circle-nodes >}} Clustering | [{{< fa file-lines >}}](labs/seurat/seurat_04_clustering.html) | [{{< fa file-lines >}}](labs/bioc/bioc_04_clustering.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_04_clustering.html) |
| {{< fa chart-simple >}} Differential expression | [{{< fa file-lines >}}](labs/seurat/seurat_05_dge.html) | [{{< fa file-lines >}}](labs/bioc/bioc_05_dge.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_05_dge.html) |
| {{< fa bacterium >}} Celltype prediction | [{{< fa file-lines >}}](labs/seurat/seurat_06_celltyping.html) | [{{< fa file-lines >}}](labs/bioc/bioc_06_celltyping.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_06_celltyping.html) |
| {{< fa route >}} Trajectory inference | [{{< fa file-lines >}}](labs/seurat/seurat_08_trajectory.html) | | [{{< fa file-lines >}}](labs/scanpy/scanpy_08_trajectory.html) |
| {{< fa microscope >}} Spatial transcriptomics | [{{< fa file-lines >}}](labs/seurat/seurat_07_spatial.html) | [{{< fa file-lines >}}](labs/bioc/bioc_07_spatial.html) | [{{< fa file-lines >}}](labs/scanpy/scanpy_07_spatial.html) |

Many additional information and explanations can be found in the [Single Cell Glossary](https://nbisweden.github.io/single-cell-pbl/glossary_of_terms_single_cell.html)

## Useful resources

- Single cell RNA-seq course at from Hemberg lab [{{< fa square-arrow-up-right >}}](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
- Single cell RNA-seq course in Python [{{< fa square-arrow-up-right >}}](https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about)
- Single cell RNA-seq course at Broad [{{< fa square-arrow-up-right >}}](https://broadinstitute.github.io/2019_scWorkshop/)
- Repository listing many scRNA-seq tools [{{< fa square-arrow-up-right >}}](https://github.com/seandavi/awesome-single-cell)
- SingleCellExperiment objects for many datasets [{{< fa square-arrow-up-right >}}](https://hemberg-lab.github.io/scRNA.seq.datasets/)
- Conquer datasets - many different datasets based on a salmon pipeline [{{< fa square-arrow-up-right >}}](http://imlspenticton.uzh.ch:3838/conquer/)
- The Human Cell Atlas project [{{< fa square-arrow-up-right >}}](https://www.humancellatlas.org/)
- The github repository for this course [{{< fa brands github >}}](https://github.com/NBISweden/workshop-scRNAseq)
- Bitbucket repository for QC report scripts [{{< fa brands bitbucket >}}](https://bitbucket.org/asbj/qc-summary_scrnaseq)
- Bitbucket repository for the NBIS scRNAseq pipeline [{{< fa brands bitbucket >}}](https://bitbucket.org/scilifelab-lts/lts-workflows-sm-scrnaseq)

***
