---
title: "Contents"
description: "Tutorials on some of the common steps during scRNAseq data analysis using three popular toolkits: [ {{< fa brands r-project >}} Seurat](https://satijalab.org/seurat/), [{{< fa brands r-project >}} Bioconductor](https://bioconductor.org/packages/release/bioc/html/scran.html) and [{{< fa brands python >}} Scanpy](https://scanpy.readthedocs.io/en/stable/)."
date: ""
author: ""
toc: false
number-sections: false
sidebar: false
code-tools: false
format: html
---

You can run the labs either using SciLifeLab Serve applications (recommended during course) or using Docker locally on your system. Instructions on running the labs are provided [[here]{.badge .badge-primary}](/other/containers.qmd)

A short description of the data used in the tutorials is provided [here](/other/data.qmd).

::: {.callout-tip}
We perform the same steps with all three toolkits, but there are some small differences as all methods are not implemented everywhere. It's up to you which toolkit you want to work with. To download file, *Right click > Save Link As...*.
:::

| Topic | {{< fa brands r-project >}} Seurat | {{< fa brands r-project >}} Bioconductor | {{< fa brands python >}} Scanpy |
|:--------|:---------------------------:|:---------------------------------:|:------------------------:|
| 1 {{< fa clipboard-list >}} Quality Control | [{{< fa file-lines >}}](labs/seurat/seurat_01_qc.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_01_qc.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_01_qc.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_01_qc.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_01_qc.html)  [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_01_qc.ipynb) |
| 2 {{< fa brands hubspot >}} Dimensionality reduction | [{{< fa file-lines >}}](labs/seurat/seurat_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_02_dimred.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_02_dimred.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_02_dimred.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_02_dimred.ipynb) |
| 3 {{< fa diagram-project >}} Data integration | [{{< fa file-lines >}}](labs/seurat/seurat_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_03_integration.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_03_integration.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_03_integration.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_03_integration.ipynb) |
| 4 {{< fa circle-nodes >}} Clustering | [{{< fa file-lines >}}](labs/seurat/seurat_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_04_clustering.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_04_clustering.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_04_clustering.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_04_clustering.ipynb) |
| 5 {{< fa chart-simple >}} Differential expression | [{{< fa file-lines >}}](labs/seurat/seurat_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_05_dge.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_05_dge.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_05_dge.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_05_dge.ipynb) |
| 6 {{< fa bacterium >}} Celltype prediction | [{{< fa file-lines >}}](labs/seurat/seurat_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_06_celltyping.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_06_celltyping.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_06_celltyping.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_06_celltyping.ipynb) |
| 7 {{< fa route >}} Trajectory inference | [{{< fa file-lines >}}](labs/seurat/seurat_07_trajectory.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_07_trajectory.qmd) | | [{{< fa file-lines >}}](labs/scanpy/scanpy_07_trajectory.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_07_trajectory.ipynb) |

:::{.callout-caution}
The _Spatial Transcriptomics_ labs are not covered in the workshop and are therefore optional. See [instructions](/other/containers-spatial.qmd) for running these labs.
:::

| Optional Topic | {{< fa brands r-project >}} Seurat | {{< fa brands r-project >}} Bioconductor | {{< fa brands python >}} Scanpy |
|:--------|:---------------------------:|:---------------------------------:|:------------------------:|
| 8 {{< fa microscope >}} Spatial transcriptomics | [{{< fa file-lines >}}](labs/seurat/seurat_08_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_08_spatial.qmd) | [{{< fa file-lines >}}](labs/bioc/bioc_08_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_08_spatial.qmd) | [{{< fa file-lines >}}](labs/scanpy/scanpy_08_spatial.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_08_spatial.ipynb) |

## Lecture recordings

The lectures from the 2025 course can be found on the [NBIS YouTube channel](https://www.youtube.com/playlist?list=PLBsJUKzoJTHTqY4FOO8R_8qcZb1hgB2QX]).

## Toolkit comparisons

Below are a few reports where the results from the 3 toolkits and multiple methods in the toolkits are analysed. *OBS!* Several of these scripts requires additional packages that are not available in the Docker containers used for the exercises.

* General comparison of the main steps of the pipelines: [{{< fa file-lines >}}](labs/comparison/comparison_pipelines.html) [{{< fa download >}}](https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_pipelines.qmd)
* Comparison of Variable gene selection methods: [{{< fa file-lines >}}](labs/comparison/comparison_hvg.html) [{{< fa download >}}](https://raw.githubusercont\
ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_hvg.qmd) 
* Differential gene expression detection: [{{< fa file-lines >}}](labs/comparison/comparison_dge.html) [{{< fa download >}}](https://raw.githubusercont\
ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_dge.qmd)
* Normalization methods: [{{< fa file-lines >}}](labs/comparison/comparison_normalization.html) [{{< fa download >}}](https://raw.githubusercont\
ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_normalization.qmd)

## Previous course iterations

* scRNAseq 2024 [{{< fa square-arrow-up-right >}}](https://nbisweden.github.io/workshop-scrnaseq-2024)
* scRNAseq 2023 [{{< fa square-arrow-up-right >}}](https://nbisweden.github.io/workshop-scrnaseq-2023)
* Recorded videos of lectures (from 2022) are available on Youtube [{{< fa brands youtube >}}](https://youtube.com/playlist?list=PLBsJUKzoJTHQA4Qg1yc1RRY2Km4t4vEeN&si=p37W8NwQREqf617q)

*NB: Seurat exercises in previous courses used Seurat v4*

## Useful resources

- The github repository for this course [{{< fa brands github >}}](https://github.com/NBISweden/workshop-scRNAseq)
- Single Cell Glossary [{{< fa square-arrow-up-right >}}](https://nbisweden.github.io/single-cell-pbl/glossary_of_terms_single_cell.html)
- Single cell RNA-seq course at from Hemberg lab [{{< fa square-arrow-up-right >}}](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
- Single cell RNA-seq course in Python [{{< fa square-arrow-up-right >}}](https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about)
- Single cell RNA-seq course at Broad [{{< fa square-arrow-up-right >}}](https://broadinstitute.github.io/2019_scWorkshop/)
- Repository listing many scRNA-seq tools [{{< fa square-arrow-up-right >}}](https://github.com/seandavi/awesome-single-cell)
- SingleCellExperiment objects for many datasets [{{< fa square-arrow-up-right >}}](https://hemberg-lab.github.io/scRNA.seq.datasets/)
- Conquer datasets - many different datasets based on a salmon pipeline [{{< fa square-arrow-up-right >}}](http://imlspenticton.uzh.ch:3838/conquer/)
- The Human Cell Atlas project [{{< fa square-arrow-up-right >}}](https://www.humancellatlas.org/)

***
