<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Combining single-cell gene expression data with spatial information to reveal the spatial distribution of gene activity within tissues.">

<title> Spatial Transcriptomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-microscope" aria-label="microscope"></i> Spatial Transcriptomics</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit</p>
                  <div>
        <div class="description">
          Combining single-cell gene expression data with spatial information to reveal the spatial distribution of gene activity within tissues.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">03-Jan-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-st_prep" id="toc-meta-st_prep" class="nav-link active" data-scroll-target="#meta-st_prep"><span class="header-section-number">1</span> Preparation</a></li>
  <li><a href="#meta-st_qc" id="toc-meta-st_qc" class="nav-link" data-scroll-target="#meta-st_qc"><span class="header-section-number">2</span> Quality control</a>
  <ul>
  <li><a href="#meta-st_qc_filter" id="toc-meta-st_qc_filter" class="nav-link" data-scroll-target="#meta-st_qc_filter"><span class="header-section-number">2.1</span> Filter spots</a></li>
  <li><a href="#meta-st_qc_top" id="toc-meta-st_qc_top" class="nav-link" data-scroll-target="#meta-st_qc_top"><span class="header-section-number">2.2</span> Top expressed genes</a></li>
  <li><a href="#meta-st_qc_filterg" id="toc-meta-st_qc_filterg" class="nav-link" data-scroll-target="#meta-st_qc_filterg"><span class="header-section-number">2.3</span> Filter genes</a></li>
  </ul></li>
  <li><a href="#meta-st_analysis" id="toc-meta-st_analysis" class="nav-link" data-scroll-target="#meta-st_analysis"><span class="header-section-number">3</span> Analysis</a>
  <ul>
  <li><a href="#meta-st_analysis_dimred" id="toc-meta-st_analysis_dimred" class="nav-link" data-scroll-target="#meta-st_analysis_dimred"><span class="header-section-number">3.1</span> Dimensionality reduction and clustering</a></li>
  <li><a href="#meta-st_analysis_int" id="toc-meta-st_analysis_int" class="nav-link" data-scroll-target="#meta-st_analysis_int"><span class="header-section-number">3.2</span> Integration</a></li>
  <li><a href="#meta-st_analysis_svg" id="toc-meta-st_analysis_svg" class="nav-link" data-scroll-target="#meta-st_analysis_svg"><span class="header-section-number">3.3</span> Spatially Variable Features</a></li>
  </ul></li>
  <li><a href="#meta-st_ss" id="toc-meta-st_ss" class="nav-link" data-scroll-target="#meta-st_ss"><span class="header-section-number">4</span> Single cell data</a></li>
  <li><a href="#meta-st_sub" id="toc-meta-st_sub" class="nav-link" data-scroll-target="#meta-st_sub"><span class="header-section-number">5</span> Subset ST for cortex</a>
  <ul>
  <li><a href="#integrate-with-scrnaseq" id="toc-integrate-with-scrnaseq" class="nav-link" data-scroll-target="#integrate-with-scrnaseq"><span class="header-section-number">5.1</span> Integrate with scRNAseq</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">6</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<p>Spatial transcriptomic data with the Visium platform is in many ways similar to scRNAseq data. It contains UMI counts for 5-20 cells instead of single cells, but is still quite sparse in the same way as scRNAseq data is, but with the additional information about spatial location in the tissue.<br>
Here we will first run quality control in a similar manner to scRNAseq data, then QC filtering, dimensionality reduction, integration and clustering. Then we will use scRNAseq data from mouse cortex to run LabelTransfer to predict celltypes in the Visium spots.<br>
We will use two <strong>Visium</strong> spatial transcriptomics dataset of the mouse brain (Sagittal), which are publicly available from the <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/">10x genomics website</a>. Note, that these dataset have already been filtered for spots that does not overlap with the tissue.</p>
<section id="meta-st_prep" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-st_prep"><span class="header-section-number">1</span> Preparation</h2>
<p>Load packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install('DropletUtils',update = F)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("Spaniel",update = F)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("RachelQueen1/Spaniel", ref = "Development" ,upgrade = F,dependencies = F)</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("renozao/xbioc")</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("meichendong/SCDC")</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Spaniel)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(Matrix)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(dplyr)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(scran)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(SingleR)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(scater)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(ggplot2)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(patchwork)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">require</span>(cowplot)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load ST data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>webpath <span class="ot">&lt;-</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/visium/Posterior"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path)) <span class="fu">dir.create</span>(path, <span class="at">recursive =</span> T)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.tar.gz"</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> file_list) {</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, i))) {</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Downloading "</span>, <span class="fu">paste0</span>(webpath, i), <span class="st">" to "</span>, <span class="fu">file.path</span>(path, i), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">paste0</span>(webpath, i), <span class="at">destfile =</span> <span class="fu">file.path</span>(path, i))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Uncompressing "</span>, <span class="fu">file.path</span>(path, i), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system</span>(<span class="fu">paste0</span>(<span class="st">"tar -xvzf "</span>, <span class="fu">file.path</span>(path, i), <span class="st">" -C "</span>,path))</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Uncompressing data/visium/Posterior/V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.tar.gz
Uncompressing data/visium/Posterior/V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>webpath <span class="ot">&lt;-</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/visium/Anterior"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path)) <span class="fu">dir.create</span>(path, <span class="at">recursive =</span> T)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.tar.gz"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz"</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> file_list) {</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, i))) {</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Downloading "</span>, <span class="fu">paste0</span>(webpath, i), <span class="st">" to "</span>, <span class="fu">file.path</span>(path, i), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">paste0</span>(webpath, i), <span class="at">destfile =</span> <span class="fu">file.path</span>(path, i))</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Uncompressing "</span>, <span class="fu">file.path</span>(path, i), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system</span>(<span class="fu">paste0</span>(<span class="st">"tar -xvzf "</span>, <span class="fu">file.path</span>(path, i), <span class="st">" -C "</span>, path))</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Uncompressing data/visium/Anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.tar.gz
Uncompressing data/visium/Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</code></pre>
</div>
</div>
<p>Merge the objects into one SCE object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sce.a <span class="ot">&lt;-</span> Spaniel<span class="sc">::</span><span class="fu">createVisiumSCE</span>(<span class="at">tenXDir=</span><span class="st">"data/visium/Anterior"</span>, <span class="at">resolution=</span><span class="st">"Low"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sce.p <span class="ot">&lt;-</span> Spaniel<span class="sc">::</span><span class="fu">createVisiumSCE</span>(<span class="at">tenXDir=</span><span class="st">"data/visium/Posterior"</span>, <span class="at">resolution=</span><span class="st">"Low"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sce.a, sce.p)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>Sample <span class="ot">&lt;-</span> <span class="fu">sub</span>( <span class="st">".*[/]"</span> , <span class="st">""</span> ,  <span class="fu">sub</span>(<span class="st">"/filtered_feature_bc_matrix"</span> , <span class="st">""</span>, sce<span class="sc">$</span>Sample ))</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>lll <span class="ot">&lt;-</span> <span class="fu">list</span>(sce.a , sce.p)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>lll <span class="ot">&lt;-</span> <span class="fu">lapply</span>( lll, <span class="cf">function</span>(x) x<span class="sc">@</span>metadata )</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(lll) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>,<span class="st">"Posterior"</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>metadata <span class="ot">&lt;-</span> lll</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can further convert the gene ensembl IDs to gene names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useMart</span>(<span class="at">biomart=</span><span class="st">"ENSEMBL_MART_ENSEMBL"</span>, <span class="at">dataset=</span><span class="st">"mmusculus_gene_ensembl"</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>annot <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>,<span class="st">"gene_biotype"</span>), <span class="at">mart =</span> mart, <span class="at">useCache =</span> F)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(annot, <span class="st">"data/spatial/annot.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gene_names <span class="ot">&lt;-</span> <span class="fu">as.character</span> ( annot[<span class="fu">match</span>(<span class="fu">rownames</span>(sce),annot[,<span class="st">"ensembl_gene_id"</span>]),<span class="st">"external_gene_name"</span>] )</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>gene_names[<span class="fu">is.na</span>(gene_names) ] <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[gene_names <span class="sc">!=</span> <span class="st">""</span>, ]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sce) <span class="ot">&lt;-</span> gene_names[gene_names <span class="sc">!=</span> <span class="st">""</span>]</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32053  6050</code></pre>
</div>
</div>
</section>
<section id="meta-st_qc" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-st_qc"><span class="header-section-number">2</span> Quality control</h2>
<p>Similar to scRNA-seq we use statistics on number of counts, number of features and percent mitochondria for quality control.</p>
<p>Now the counts and feature counts are calculated on the Spatial assay, so they are named <strong>nCount_Spatial</strong> and <strong>nFeature_Spatial</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mitochondrial genes</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^mt-"</span>,<span class="fu">rownames</span>(sce))]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Ribosomal genes</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>ribo_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^Rp[sl]"</span>,<span class="fu">rownames</span>(sce))]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Hemoglobin genes - includes all genes starting with HB except HBP.</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>hb_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^Hb[^(p)]"</span>,<span class="fu">rownames</span>(sce))]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(sce, <span class="at">flatten =</span> T, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mt=</span>mito_genes, <span class="at">hb=</span>hb_genes, <span class="at">ribo=</span>ribo_genes))</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(sce))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 24 columns
       Sample            Barcode   Section    Spot_Y    Spot_X   Image_Y
  &lt;character&gt;        &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
1    Anterior AAACAAGTATCTCCCA-1         1        50       102      7474
2    Anterior AAACACCAATAACTGC-1         1        59        19      8552
3    Anterior AAACAGAGCGACTCCT-1         1        14        94      3163
4    Anterior AAACAGCTTTCAGAAG-1         1        43         9      6636
5    Anterior AAACAGGGTCTATATT-1         1        47        13      7115
6    Anterior AAACATGGTGAGAGGA-1         1        62         0      8912
    Image_X   pixel_x   pixel_y       sum  detected     total       sum
  &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
1      8500   438.898   214.079     13991      4462     13991     13960
2      2788   143.959   158.417     39797      8126     39797     39742
3      7950   410.499   436.678     29951      6526     29951     29905
4      2100   108.434   257.349     42333      8190     42333     42262
5      2375   122.633   232.616     35700      8090     35700     35660
6      1480    76.420   139.828     22148      6518     22148     22096
   detected subsets_mt_sum subsets_mt_detected subsets_mt_percent
  &lt;integer&gt;      &lt;numeric&gt;           &lt;integer&gt;          &lt;numeric&gt;
1      4458           1521                  12           10.89542
2      8116           3977                  12           10.00705
3      6520           4265                  12           14.26183
4      8181           2870                  12            6.79097
5      8083           1831                  13            5.13460
6      6509           2390                  12           10.81644
  subsets_hb_sum subsets_hb_detected subsets_hb_percent subsets_ribo_sum
       &lt;numeric&gt;           &lt;integer&gt;          &lt;numeric&gt;        &lt;numeric&gt;
1             60                   4           0.429799              826
2            831                   6           2.090987             2199
3            111                   5           0.371175             1663
4            117                   5           0.276844             3129
5             73                   5           0.204711             2653
6            134                   5           0.606445             1478
  subsets_ribo_detected subsets_ribo_percent     total
              &lt;integer&gt;            &lt;numeric&gt; &lt;numeric&gt;
1                    85              5.91691     13960
2                    89              5.53319     39742
3                    88              5.56094     29905
4                    88              7.40381     42262
5                    90              7.43971     35660
6                    84              6.68899     22096</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="fu">plotColData</span>(sce,<span class="at">y =</span> <span class="st">"detected"</span>,<span class="at">x =</span> <span class="st">"Sample"</span>,<span class="at">colour_by =</span> <span class="st">"Sample"</span>),</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>          <span class="fu">plotColData</span>(sce,<span class="at">y =</span> <span class="st">"total"</span>,<span class="at">x =</span> <span class="st">"Sample"</span>,<span class="at">colour_by =</span> <span class="st">"Sample"</span>),</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>          <span class="fu">plotColData</span>(sce,<span class="at">y =</span> <span class="st">"subsets_mt_percent"</span>,<span class="at">x =</span> <span class="st">"Sample"</span>,<span class="at">colour_by =</span> <span class="st">"Sample"</span>),</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>          <span class="fu">plotColData</span>(sce,<span class="at">y =</span> <span class="st">"subsets_ribo_percent"</span>,<span class="at">x =</span> <span class="st">"Sample"</span>,<span class="at">colour_by =</span> <span class="st">"Sample"</span>),</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>          <span class="fu">plotColData</span>(sce,<span class="at">y =</span> <span class="st">"subsets_hb_percent"</span>,<span class="at">x =</span> <span class="st">"Sample"</span>,<span class="at">colour_by =</span> <span class="st">"Sample"</span>),<span class="at">ncol =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-7-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="1056"></a></p>
</figure>
</div>
</div>
</div>
<p>We can also plot the same data onto the tissue section.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"detected"</span>,<span class="st">"total"</span>,<span class="st">"subsets_mt_percent"</span>,<span class="st">"subsets_ribo_percent"</span>,<span class="st">"subsets_hb_percent"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">clusterRes=</span>j,<span class="at">customTitle =</span> j,</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-8-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the spots with low number of counts/features and high mitochondrial content are mainly towards the edges of the tissue. It is quite likely that these regions are damaged tissue. You may also see regions within a tissue with low quality if you have tears or folds in your section.<br>
But remember, for some tissue types, the amount of genes expressed and proportion mitochondria may also be a biological features, so bear in mind what tissue you are working on and what these features mean.</p>
<section id="meta-st_qc_filter" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="meta-st_qc_filter"><span class="header-section-number">2.1</span> Filter spots</h3>
<p>Select all spots with less than <strong>25%</strong> mitocondrial reads, less than <strong>20%</strong> hb-reads and <strong>500</strong> detected genes. You must judge for yourself based on your knowledge of the tissue what are appropriate filtering criteria for your dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[ , sce<span class="sc">$</span>detected <span class="sc">&gt;</span> <span class="dv">500</span> <span class="sc">&amp;</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>              sce<span class="sc">$</span>subsets_mt_percent <span class="sc">&lt;</span> <span class="dv">25</span> <span class="sc">&amp;</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>              sce<span class="sc">$</span>subsets_hb_percent <span class="sc">&lt;</span> <span class="dv">20</span>]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32053  5804</code></pre>
</div>
</div>
<p>And replot onto tissue section:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"detected"</span>,<span class="st">"total"</span>,<span class="st">"subsets_mt_percent"</span>,<span class="st">"subsets_mt_percent"</span>,<span class="st">"subsets_hb_percent"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">clusterRes=</span>j,<span class="at">customTitle =</span> j,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_qc_top" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="meta-st_qc_top"><span class="header-section-number">2.2</span> Top expressed genes</h3>
<p>As for scRNA-seq data, we will look at what the top expressed genes are.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>C <span class="ot">=</span> <span class="fu">counts</span>(sce)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>C<span class="sc">@</span>x <span class="ot">=</span> C<span class="sc">@</span>x <span class="sc">/</span> <span class="fu">rep.int</span>(<span class="fu">colSums</span>(C), <span class="fu">diff</span>(C<span class="sc">@</span>p))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>most_expressed <span class="ot">&lt;-</span> <span class="fu">order</span>(Matrix<span class="sc">::</span><span class="fu">rowSums</span>( C ),<span class="at">decreasing =</span> T)[<span class="dv">20</span><span class="sc">:</span><span class="dv">1</span>]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>( <span class="fu">as.matrix</span>(<span class="fu">t</span>(C[most_expressed,])),<span class="at">cex=</span>.<span class="dv">1</span>, <span class="at">las=</span><span class="dv">1</span>, <span class="at">xlab=</span><span class="st">"% total count per cell"</span>,<span class="at">col=</span>scales<span class="sc">::</span><span class="fu">hue_pal</span>()(<span class="dv">20</span>)[<span class="dv">20</span><span class="sc">:</span><span class="dv">1</span>],<span class="at">horizontal=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-11-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(C)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you can see, the mitochondrial genes are among the top expressed genes. Also the lncRNA gene Bc1 (brain cytoplasmic RNA 1). Also one hemoglobin gene.</p>
</section>
<section id="meta-st_qc_filterg" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="meta-st_qc_filterg"><span class="header-section-number">2.3</span> Filter genes</h3>
<p>We will remove the <em>Bc1</em> gene, hemoglobin genes (blood contamination) and the mitochondrial genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32053  5804</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Bl1</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"Bc1"</span>, <span class="fu">rownames</span>(sce)), ]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Mitocondrial</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^mt-"</span>, <span class="fu">rownames</span>(sce)), ]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Hemoglobin gene (optional if that is a problem on your data)</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^Hb.*-"</span>, <span class="fu">rownames</span>(sce)), ]</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32031  5804</code></pre>
</div>
</div>
</section>
</section>
<section id="meta-st_analysis" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="meta-st_analysis"><span class="header-section-number">3</span> Analysis</h2>
<p>We will proceed with the data in a very similar manner to scRNA-seq data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(sce, <span class="at">sizes=</span><span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">40</span>, <span class="dv">60</span>, <span class="dv">80</span>))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can plot gene expression of individual genes, the gene <em>Hpca</em> is a strong hippocampal marker and <em>Ttr</em> is a marker of the choroid plexus.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Hpca"</span>, <span class="st">"Ttr"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Gene"</span>,</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">gene=</span>j,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> j,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-14-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="624"></a></p>
</figure>
</div>
</div>
</div>
<section id="meta-st_analysis_dimred" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="meta-st_analysis_dimred"><span class="header-section-number">3.1</span> Dimensionality reduction and clustering</h3>
<p>We can then now run dimensionality reduction and clustering using the same workflow as we use for scRNA-seq analysis.</p>
<p>But make sure you run it on the <code>SCT</code> assay.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">=</span> <span class="fu">getTopHVGs</span>(var.out, <span class="at">n=</span><span class="dv">2000</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">exprs_values =</span> <span class="st">"logcounts"</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">subset_row=</span>hvgs,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">ncomponents =</span> <span class="dv">50</span>,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">ntop =</span> <span class="dv">100</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">scale =</span> T)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(sce,<span class="at">k=</span><span class="dv">5</span>,<span class="at">use.dimred=</span><span class="st">"PCA"</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>louvain_SNNk5 <span class="ot">&lt;-</span> <span class="fu">factor</span>( igraph<span class="sc">::</span><span class="fu">cluster_louvain</span>(g)<span class="sc">$</span>membership )</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(sce,<span class="at">dimred =</span> <span class="st">"PCA"</span>, <span class="at">n_dimred =</span> <span class="dv">50</span>,   <span class="at">ncomponents =</span> <span class="dv">2</span>,<span class="at">min_dist=</span><span class="fl">0.1</span>,<span class="at">spread=</span>.<span class="dv">3</span>,</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">metric=</span><span class="st">"correlation"</span>,<span class="at">name =</span> <span class="st">"UMAP_on_PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then plot clusters onto umap or onto the tissue section.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"louvain_SNNk5"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,<span class="at">clusterRes =</span> j,</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> j,</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(sce, <span class="at">dimred =</span> <span class="st">"UMAP_on_PCA"</span>, <span class="at">colour_by =</span> <span class="st">"louvain_SNNk5"</span>)</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(sce, <span class="at">dimred =</span> <span class="st">"UMAP_on_PCA"</span>, <span class="at">colour_by =</span> <span class="st">"Sample"</span>)</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="864"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-st_analysis_int" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="meta-st_analysis_int"><span class="header-section-number">3.2</span> Integration</h3>
<p>Quite often there are strong batch effects between different ST sections, so it may be a good idea to integrate the data across sections.</p>
<p>We will do a similar integration as in the Data Integration lab.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>mnn_out <span class="ot">&lt;-</span> batchelor<span class="sc">::</span><span class="fu">fastMNN</span>(sce, <span class="at">subset.row =</span> hvgs, <span class="at">batch =</span> <span class="fu">factor</span>(sce<span class="sc">$</span>Sample), <span class="at">k =</span> <span class="dv">20</span>, <span class="at">d =</span> <span class="dv">50</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(sce, <span class="st">"MNN"</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(mnn_out,<span class="st">"corrected"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mnn_out)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger   (Mb)  max used   (Mb)
Ncells  11244846  600.6   19614032 1047.6  19614032 1047.6
Vcells 193810550 1478.7  342214384 2610.9 342214384 2610.9</code></pre>
</div>
</div>
<p>Then we run dimensionality reduction and clustering as before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(sce,<span class="at">k=</span><span class="dv">5</span>,<span class="at">use.dimred=</span><span class="st">"MNN"</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>louvain_SNNk5 <span class="ot">&lt;-</span> <span class="fu">factor</span>( igraph<span class="sc">::</span><span class="fu">cluster_louvain</span>(g)<span class="sc">$</span>membership )</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(sce,<span class="at">dimred =</span> <span class="st">"MNN"</span>, <span class="at">n_dimred =</span> <span class="dv">50</span>,   <span class="at">ncomponents =</span> <span class="dv">2</span>,<span class="at">min_dist=</span><span class="fl">0.1</span>,<span class="at">spread=</span>.<span class="dv">3</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">metric=</span><span class="st">"correlation"</span>,<span class="at">name =</span> <span class="st">"UMAP_on_MNN"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"louvain_SNNk5"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,<span class="at">clusterRes =</span> j,</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> j,</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(sce, <span class="at">dimred =</span> <span class="st">"UMAP_on_MNN"</span>, <span class="at">colour_by =</span> <span class="st">"louvain_SNNk5"</span>)</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> <span class="fu">plotReducedDim</span>(sce, <span class="at">dimred =</span> <span class="st">"UMAP_on_MNN"</span>, <span class="at">colour_by =</span> <span class="st">"Sample"</span>)</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-19-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="864"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you see any differences between the integrated and non-integrated clustering? Judge for yourself, which of the clusterings do you think looks best? As a reference, you can compare to brain regions in the <a href="https://mouse.brain-map.org/experiment/thumbnails/100042147?image_type=atlas">Allen brain atlas</a>.</p>
</div>
</div>
</section>
<section id="meta-st_analysis_svg" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="meta-st_analysis_svg"><span class="header-section-number">3.3</span> Spatially Variable Features</h3>
<p>There are two main workflows to identify molecular features that correlate with spatial location within a tissue. The first is to perform differential expression based on spatially distinct clusters, the other is to find features that have spatial patterning without taking clusters or spatial annotation into account. First, we will do differential expression between clusters just as we did for the scRNAseq data before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># differential expression between cluster 4 and cluster 6</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>cell_selection <span class="ot">&lt;-</span> sce[ , sce<span class="sc">$</span>louvain_SNNk5 <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">6</span>) ]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>cell_selection<span class="sc">$</span>louvain_SNNk5 <span class="ot">&lt;-</span> <span class="fu">factor</span>(cell_selection<span class="sc">$</span>louvain_SNNk5)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>markers_genes <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( <span class="at">x =</span> cell_selection,</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">groups =</span> cell_selection<span class="sc">$</span>louvain_SNNk5,</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">lfc=</span>.<span class="dv">25</span>,</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">pval.type =</span> <span class="st">"all"</span>,</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co">#List of dataFrames with the results for each cluster</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>top5_cell_selection <span class="ot">&lt;-</span> <span class="fu">lapply</span>( <span class="fu">names</span>(markers_genes), <span class="cf">function</span>(x) { temp <span class="ot">&lt;-</span> markers_genes[[x]][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] ; temp<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(markers_genes[[x]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] ; temp<span class="sc">$</span>cluster <span class="ot">&lt;-</span> x ; <span class="fu">return</span>(temp) } )</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>top5_cell_selection <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">do.call</span>(rbind, top5_cell_selection))</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>top5_cell_selection</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["p.value"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["gene"],"name":[3],"type":["chr"],"align":["left"]},{"label":["cluster"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"3.702790e-144","2":"1.186041e-139","3":"Gng4","4":"4"},{"1":"1.290997e-119","2":"2.067596e-115","3":"Gpsm1","4":"4"},{"1":"1.008635e-108","2":"1.072991e-104","3":"Pcbp3","4":"4"},{"1":"1.339940e-108","2":"1.072991e-104","3":"Synpr","4":"4"},{"1":"1.240753e-99","2":"7.948509e-96","3":"Meis2","4":"4"},{"1":"1.463072e-164","2":"4.686366e-160","3":"Ptgds","4":"6"},{"1":"2.786430e-115","2":"4.462607e-111","3":"Atp1a2","4":"6"},{"1":"1.081121e-88","2":"1.154313e-84","3":"Id3","4":"6"},{"1":"7.657854e-85","2":"6.132218e-81","3":"Apoe","4":"6"},{"1":"4.193939e-83","2":"2.686721e-79","3":"Mgp","4":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot top markers</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Anterior"</span>, <span class="st">"Posterior"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>to_plot <span class="ot">&lt;-</span> top5_cell_selection<span class="sc">$</span>gene[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> to_plot){</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> samples){</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> i]</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">@</span>metadata <span class="ot">&lt;-</span> temp<span class="sc">@</span>metadata[[i]]</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>    plist[[n]]<span class="ot">&lt;-</span><span class="fu">spanielPlot</span>(<span class="at">object =</span> temp,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Gene"</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">gene=</span>j,</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> j,</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">&lt;-</span>n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-20-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="meta-st_ss" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-st_ss"><span class="header-section-number">4</span> Single cell data</h2>
<p>We can use a scRNA-seq dataset as a reference to predict the proportion of different celltypes in the Visium spots. Keep in mind that it is important to have a reference that contains all the celltypes you expect to find in your spots. Ideally it should be a scRNA-seq reference from the exact same tissue. We will use a reference scRNA-seq dataset of ~14,000 adult mouse cortical cell taxonomy from the Allen Institute, generated with the SMART-Seq2 protocol.</p>
<p>First dowload the seurat data from: https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1 to folder <code>data/spatial/</code> with command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>webpath <span class="ot">&lt;-</span> <span class="st">"https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1"</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="st">"data/spatial/allen_cortex.rds"</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(path)){  </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="st">"data/spatial/"</span>,<span class="at">recursive =</span> T)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">options</span>(<span class="at">timeout=</span><span class="dv">10000</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>( <span class="at">url =</span> webpath , <span class="at">destfile =</span> path)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">options</span>(<span class="at">timeout=</span><span class="dv">60</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For speed, and for a more fair comparison of the celltypes, we will subsample all celltypes to a maximum of 200 cells per class (<code>subclass</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>allen_reference <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/spatial/allen_cortex.rds"</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>allen_reference_sce <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.SingleCellExperiment</span>(allen_reference)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check number of cells per subclass</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>allen_reference_sce<span class="sc">$</span>subclass <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"/"</span>,<span class="st">"_"</span>,<span class="fu">sub</span>(<span class="st">" "</span>,<span class="st">"_"</span>,allen_reference_sce<span class="sc">$</span>subclass))</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(allen_reference_sce<span class="sc">$</span>subclass)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Astro         CR       Endo    L2_3_IT         L4      L5_IT      L5_PT 
       368          7         94        982       1401        880        544 
     L6_CT      L6_IT        L6b      Lamp5 Macrophage      Meis2         NP 
       960       1872        358       1122         51         45        362 
     Oligo       Peri      Pvalb   Serpinf1        SMC       Sncg        Sst 
        91         32       1337         27         55        125       1741 
       Vip       VLMC 
      1728         67 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select 20 cells per subclass, fist set subclass ass active.ident</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>subset_cells <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(allen_reference_sce<span class="sc">$</span>subclass),<span class="cf">function</span>(x){</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>( <span class="fu">sum</span>(allen_reference_sce<span class="sc">$</span>subclass <span class="sc">==</span> x) <span class="sc">&gt;</span> <span class="dv">20</span> ){</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">colnames</span>(allen_reference_sce)[allen_reference_sce<span class="sc">$</span>subclass <span class="sc">==</span> x],<span class="at">size =</span> <span class="dv">20</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> <span class="fu">colnames</span>(allen_reference_sce)[allen_reference_sce<span class="sc">$</span>subclass <span class="sc">==</span> x]</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  } })</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>allen_reference_sce <span class="ot">&lt;-</span> allen_reference_sce[,<span class="fu">unlist</span>(subset_cells)]</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="co"># check again number of cells per subclass</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(allen_reference_sce<span class="sc">$</span>subclass)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Astro         CR       Endo    L2_3_IT         L4      L5_IT      L5_PT 
        20          7         20         20         20         20         20 
     L6_CT      L6_IT        L6b      Lamp5 Macrophage      Meis2         NP 
        20         20         20         20         20         20         20 
     Oligo       Peri      Pvalb   Serpinf1        SMC       Sncg        Sst 
        20         20         20         20         20         20         20 
       Vip       VLMC 
        20         20 </code></pre>
</div>
</div>
<p>Then run normalization and dimensionality reduction.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>allen_reference_sce <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(allen_reference_sce, <span class="at">sizes=</span><span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">40</span>, <span class="dv">60</span>, <span class="dv">80</span>))</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>allen_reference_sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(allen_reference_sce)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>allen.var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(allen_reference_sce, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>allen.hvgs <span class="ot">=</span> <span class="fu">getTopHVGs</span>(allen.var.out, <span class="at">n=</span><span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-st_sub" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-st_sub"><span class="header-section-number">5</span> Subset ST for cortex</h2>
<p>Since the scRNAseq dataset was generated from the mouse cortex, we will subset the visium dataset in order to select mainly the spots part of the cortex. Note that the integration can also be performed on the whole brain slice, but it would give rise to false positive cell type assignments and therefore it should be interpreted with more care.</p>
<section id="integrate-with-scrnaseq" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="integrate-with-scrnaseq"><span class="header-section-number">5.1</span> Integrate with scRNAseq</h3>
<p>Here, will use SingleR for prediciting which cell types are present in the dataset. We can first select the anterior part as an example (to speed up predictions).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>sce.anterior <span class="ot">&lt;-</span> sce[,sce<span class="sc">$</span>Sample <span class="sc">==</span> <span class="st">"Anterior"</span>]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>sce.anterior<span class="sc">@</span>metadata <span class="ot">&lt;-</span> sce.anterior<span class="sc">@</span>metadata[[<span class="st">"Anterior"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we select the highly variable genes that are present in both datasets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Find common highly variable genes</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>common_hvgs <span class="ot">&lt;-</span> allen.hvgs[allen.hvgs <span class="sc">%in%</span> hvgs]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Predict cell classes</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>pred.grun <span class="ot">&lt;-</span> <span class="fu">SingleR</span>(<span class="at">test=</span>sce.anterior[common_hvgs,],</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">ref=</span>allen_reference_sce[common_hvgs,],</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">labels=</span>allen_reference_sce<span class="sc">$</span>subclass)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Transfer the classes to the SCE object</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>sce.anterior<span class="sc">$</span>cell_prediction <span class="ot">&lt;-</span> pred.grun<span class="sc">$</span>labels</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>sce.anterior<span class="sc">@</span>colData <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sce.anterior<span class="sc">@</span>colData,</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">as.data.frame.matrix</span>(<span class="fu">table</span>(<span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(sce.anterior),sce.anterior<span class="sc">$</span>cell_prediction))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can plot the predicted cell populations back to tissue.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot cell predictions</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">spanielPlot</span>(<span class="at">object =</span> sce.anterior,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">clusterRes =</span> <span class="st">"cell_prediction"</span>,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> <span class="st">"cell_prediction"</span>,</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="528"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">1</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"L2_3_IT"</span>,<span class="st">"L4"</span>,<span class="st">"L5_IT"</span>,<span class="st">"L6_IT"</span>)){</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  plist[[n]] <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> sce.anterior,</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Cluster"</span>,</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">clusterRes =</span> i,</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> i,</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,<span class="at">ptSize =</span> .<span class="dv">3</span>,</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSizeMax =</span> <span class="dv">1</span>,<span class="at">ptSizeMin =</span> .<span class="dv">1</span>)</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> n<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">ncol=</span><span class="dv">2</span>, <span class="at">plotlist =</span> plist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-26-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="864"></a></p>
</figure>
</div>
</div>
</div>
<p>Keep in mind, that the scores are “just” prediction scores, and do not correspond to proportion of cells that are of a certain celltype or similar. It mainly tell you that gene expression in a certain spot is hihgly similar/dissimilar to gene expression of a celltype. If we look at the scores, we see that some spots got really clear predictions by celltype, while others did not have high scores for any of the celltypes.</p>
<p>We can also plot the gene expression and add filters together, too:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">spanielPlot</span>(<span class="at">object =</span> sce.anterior,</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">plotType =</span> <span class="st">"Gene"</span>,</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">gene=</span><span class="st">"Wfs1"</span>,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">showFilter =</span> sce.anterior<span class="sc">$</span>L4 ,</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">customTitle =</span> <span class="st">""</span>,</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">techType =</span> <span class="st">"Visium"</span>,</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">ptSize=</span><span class="dv">0</span>,<span class="at">ptSizeMin =</span> <span class="sc">-</span>.<span class="dv">3</span>,<span class="at">ptSizeMax =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_07_spatial_files/figure-html/unnamed-chunk-27-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="bioc_07_spatial_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="624"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">6</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] cowplot_1.1.2               patchwork_1.1.3            
 [3] scater_1.30.1               ggplot2_3.4.4              
 [5] SingleR_2.4.0               scran_1.30.0               
 [7] scuttle_1.12.0              dplyr_1.1.4                
 [9] Matrix_1.6-4                SingleCellExperiment_1.24.0
[11] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[13] GenomicRanges_1.54.1        GenomeInfoDb_1.38.2        
[15] IRanges_2.36.0              S4Vectors_0.40.2           
[17] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[19] matrixStats_1.2.0           biomaRt_2.58.0             
[21] Spaniel_1.16.0             

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.0-3     bitops_1.0-7             
  [3] httr_1.4.7                RColorBrewer_1.1-3       
  [5] tools_4.3.2               sctransform_0.4.1        
  [7] ResidualMatrix_1.12.0     utf8_1.2.4               
  [9] R6_2.5.1                  HDF5Array_1.30.0         
 [11] lazyeval_0.2.2            uwot_0.1.16              
 [13] rhdf5filters_1.14.1       withr_2.5.2              
 [15] sp_2.1-2                  prettyunits_1.2.0        
 [17] gridExtra_2.3             progressr_0.14.0         
 [19] cli_3.6.2                 spatstat.explore_3.2-5   
 [21] fastDummies_1.7.3         labeling_0.4.3           
 [23] Seurat_5.0.1              spatstat.data_3.0-3      
 [25] ggridges_0.5.5            pbapply_1.7-2            
 [27] R.utils_2.12.3            parallelly_1.36.0        
 [29] limma_3.58.1              RSQLite_2.3.4            
 [31] generics_0.1.3            ica_1.0-3                
 [33] spatstat.random_3.2-2     ggbeeswarm_0.7.2         
 [35] fansi_1.0.6               abind_1.4-5              
 [37] R.methodsS3_1.8.2         lifecycle_1.0.4          
 [39] yaml_2.3.8                edgeR_4.0.3              
 [41] rhdf5_2.46.1              SparseArray_1.2.3        
 [43] BiocFileCache_2.10.1      Rtsne_0.17               
 [45] grid_4.3.2                blob_1.2.4               
 [47] promises_1.2.1            dqrng_0.3.2              
 [49] crayon_1.5.2              miniUI_0.1.1.1           
 [51] lattice_0.21-9            beachmat_2.18.0          
 [53] KEGGREST_1.42.0           pillar_1.9.0             
 [55] knitr_1.45                metapod_1.10.1           
 [57] future.apply_1.11.1       codetools_0.2-19         
 [59] leiden_0.4.3.1            glue_1.6.2               
 [61] data.table_1.14.10        vctrs_0.6.5              
 [63] png_0.1-8                 spam_2.10-0              
 [65] gtable_0.3.4              cachem_1.0.8             
 [67] xfun_0.41                 S4Arrays_1.2.0           
 [69] mime_0.12                 DropletUtils_1.22.0      
 [71] survival_3.5-7            statmod_1.5.0            
 [73] bluster_1.12.0            ellipsis_0.3.2           
 [75] fitdistrplus_1.1-11       ROCR_1.0-11              
 [77] nlme_3.1-163              bit64_4.0.5              
 [79] progress_1.2.3            filelock_1.0.3           
 [81] RcppAnnoy_0.0.21          irlba_2.3.5.1            
 [83] vipor_0.4.7               KernSmooth_2.23-22       
 [85] colorspace_2.1-0          DBI_1.1.3                
 [87] tidyselect_1.2.0          bit_4.0.5                
 [89] compiler_4.3.2            curl_5.2.0               
 [91] BiocNeighbors_1.20.1      xml2_1.3.6               
 [93] DelayedArray_0.28.0       plotly_4.10.3            
 [95] scales_1.3.0              lmtest_0.9-40            
 [97] rappdirs_0.3.3            stringr_1.5.1            
 [99] digest_0.6.33             goftest_1.2-3            
[101] spatstat.utils_3.0-4      rmarkdown_2.25           
[103] XVector_0.42.0            htmltools_0.5.7          
[105] pkgconfig_2.0.3           sparseMatrixStats_1.14.0 
[107] dbplyr_2.4.0              fastmap_1.1.1            
[109] rlang_1.1.2               htmlwidgets_1.6.4        
[111] shiny_1.8.0               DelayedMatrixStats_1.24.0
[113] farver_2.1.1              zoo_1.8-12               
[115] jsonlite_1.8.8            BiocParallel_1.36.0      
[117] R.oo_1.25.0               BiocSingular_1.18.0      
[119] RCurl_1.98-1.13           magrittr_2.0.3           
[121] GenomeInfoDbData_1.2.11   dotCall64_1.1-1          
[123] Rhdf5lib_1.24.1           munsell_0.5.0            
[125] Rcpp_1.0.11               viridis_0.6.4            
[127] reticulate_1.34.0         stringi_1.8.3            
[129] zlibbioc_1.48.0           MASS_7.3-60              
[131] plyr_1.8.9                parallel_4.3.2           
[133] listenv_0.9.0             ggrepel_0.9.4            
[135] deldir_2.0-2              Biostrings_2.70.1        
[137] splines_4.3.2             tensor_1.5               
[139] hms_1.1.3                 locfit_1.5-9.8           
[141] igraph_1.6.0              spatstat.geom_3.2-7      
[143] RcppHNSW_0.5.0            reshape2_1.4.4           
[145] ScaledMatrix_1.10.0       XML_3.99-0.16            
[147] evaluate_0.23             SeuratObject_5.0.1       
[149] httpuv_1.6.13             batchelor_1.18.0         
[151] RANN_2.6.1                tidyr_1.3.0              
[153] purrr_1.0.2               polyclip_1.10-6          
[155] future_1.33.1             scattermore_1.2          
[157] rsvd_1.0.5                xtable_1.8-4             
[159] RSpectra_0.16-1           later_1.3.2              
[161] viridisLite_0.4.2         tibble_3.2.1             
[163] memoise_2.0.1             beeswarm_0.4.0           
[165] AnnotationDbi_1.64.1      cluster_2.1.4            
[167] globals_0.16.2           </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"descPosition":"bottom","closeEffect":"zoom","loop":true,"selector":".lightbox","openEffect":"zoom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>