<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Identify genes that are significantly over or under-expressed between conditions in specific cell populations.">

<title> Differential gene expression</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-chart-simple" aria-label="chart-simple"></i> Differential gene expression</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit</p>
                  <div>
        <div class="description">
          Identify genes that are significantly over or under-expressed between conditions in specific cell populations.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">10-Jan-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-dge_numbers" id="toc-meta-dge_numbers" class="nav-link active" data-scroll-target="#meta-dge_numbers"><span class="header-section-number">1</span> <strong>?meta:dge_numbers</strong></a></li>
  <li><a href="#meta-dge_cmg" id="toc-meta-dge_cmg" class="nav-link" data-scroll-target="#meta-dge_cmg"><span class="header-section-number">2</span> Cell marker genes</a></li>
  <li><a href="#meta-dge_cond" id="toc-meta-dge_cond" class="nav-link" data-scroll-target="#meta-dge_cond"><span class="header-section-number">3</span> DGE across conditions</a></li>
  <li><a href="#meta-dge_gsa" id="toc-meta-dge_gsa" class="nav-link" data-scroll-target="#meta-dge_gsa"><span class="header-section-number">4</span> Gene Set Analysis (GSA)</a>
  <ul>
  <li><a href="#meta-dge_gsa_hyper" id="toc-meta-dge_gsa_hyper" class="nav-link" data-scroll-target="#meta-dge_gsa_hyper"><span class="header-section-number">4.1</span> Hypergeometric enrichment test</a></li>
  </ul></li>
  <li><a href="#meta-dge_gsea" id="toc-meta-dge_gsea" class="nav-link" data-scroll-target="#meta-dge_gsea"><span class="header-section-number">5</span> Gene Set Enrichment Analysis (GSEA)</a></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">6</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<p>In this tutorial we will cover differential gene expression, which comprises an extensive range of topics and methods. In single cell, differential expresison can have multiple functionalities such as identifying marker genes for cell populations, as well as identifying differentially regulated genes across conditions (healthy vs control). We will also cover controlling batch effect in your test.</p>
<p>We can first load the data from the clustering session. Moreover, we can already decide which clustering resolution to use. First let’s define using the <code>louvain</code> clustering to identifying differentially expressed genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scater)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># library(venn)</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pheatmap)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(igraph)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download pre-computed data if missing or long compute</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>fetch_data <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># url for source and intermediate data</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>path_data <span class="ot">&lt;-</span> <span class="st">"https://nextcloud.dc.scilifelab.se/public.php/webdav"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>curl_upass <span class="ot">&lt;-</span> <span class="st">"-u zbC5fr2LbEZ9rSE:scRNAseq2025"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>path_file <span class="ot">&lt;-</span> <span class="st">"data/covid/results/bioc_covid_qc_dr_int_cl.rds"</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">dirname</span>(path_file))) <span class="fu">dir.create</span>(<span class="fu">dirname</span>(path_file), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (fetch_data <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">file.exists</span>(path_file)) <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">file.path</span>(path_data, <span class="st">"covid/results_bioc/bioc_covid_qc_dr_int_cl.rds"</span>), <span class="at">destfile =</span> path_file, <span class="at">method =</span> <span class="st">"curl"</span>, <span class="at">extra =</span> curl_upass)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(path_file)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">reducedDims</span>(sce))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 17
names(17): PCA UMAP tSNE_on_PCA UMAP_on_PCA ... UMAP_on_Scanorama SNN harmony2</code></pre>
</div>
</div>
<section id="meta-dge_numbers" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-dge_numbers"><span class="header-section-number">1</span> <strong>?meta:dge_numbers</strong></h2>
<p><strong>?meta:dge_numbers_1</strong></p>
<p>However, <code>findMarker</code> in Scran is implemented so that the tests are run in a pariwise manner, e.g.&nbsp;each cluster is tested agains all the others individually. Then a combined p-value is calculated across all the tests using <code>combineMarkers</code>. So for this method, one large cluster will not influence the results in the same way as <code>FindMarkers</code> in Seurat or <code>rank_genes_groups</code> in Scanpy.</p>
</section>
<section id="meta-dge_cmg" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-dge_cmg"><span class="header-section-number">2</span> Cell marker genes</h2>
<p>Let us first compute a ranking for the highly differential genes in each cluster. There are many different tests and parameters to be chosen that can be used to refine your results. When looking for marker genes, we want genes that are positively expressed in a cell type and possibly not expressed in others.</p>
<p>In the scran function <code>findMarkers</code> t-test, wilcoxon test and binomial test implemented.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differentiall expression</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>markers_genes <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> sce,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">groups =</span> <span class="fu">as.character</span>(sce<span class="sc">$</span>leiden_k20),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"wilcox"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">lfc =</span> .<span class="dv">5</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pval.type =</span> <span class="st">"all"</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">direction =</span> <span class="st">"up"</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co"># List of dataFrames with the results for each cluster</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>markers_genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 15
names(15): 1 10 11 12 13 14 15 2 3 4 5 6 7 8 9</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing the expression of one</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(markers_genes[[<span class="st">"1"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 17 columns
         p.value       FDR summary.AUC    AUC.10    AUC.11    AUC.12    AUC.13
       &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
IFITM3         1         1    0.299559  0.699555  0.622212  0.761834 0.0542719
IL1B           1         1    0.292927  0.427379  0.406347  0.424464 0.3889630
ISG15          1         1    0.282864  0.423660  0.467255  0.438522 0.2551512
IFI6           1         1    0.279036  0.511673  0.514570  0.467627 0.3105327
G0S2           1         1    0.279071  0.407772  0.379221  0.407413 0.3268747
SGK1           1         1    0.272416  0.511194  0.476678  0.464195 0.4109583
          AUC.14    AUC.15     AUC.2     AUC.3     AUC.4     AUC.5     AUC.6
       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
IFITM3  0.215658  0.525895  0.780363  0.299559  0.110508  0.786835  0.800302
IL1B    0.393321  0.415498  0.416701  0.292927  0.201328  0.427371  0.425130
ISG15   0.354853  0.413541  0.416615  0.282864  0.178454  0.487900  0.450810
IFI6    0.425910  0.525959  0.526322  0.279036  0.144124  0.574253  0.561719
G0S2    0.347510  0.404422  0.397823  0.279071  0.223041  0.409048  0.401366
SGK1    0.462013  0.470234  0.479035  0.272416  0.174514  0.495060  0.476501
           AUC.7     AUC.8     AUC.9
       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
IFITM3  0.784298  0.689206  0.505839
IL1B    0.427014  0.426359  0.391358
ISG15   0.438798  0.376307  0.523902
IFI6    0.520653  0.436423  0.560452
G0S2    0.406407  0.403665  0.378647
SGK1    0.510243  0.508500  0.463580</code></pre>
</div>
</div>
<p>We can now select the top 25 overexpressed genes for plotting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Colect the top 25 genes for each cluster and put the into a single table</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>top25 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(markers_genes), <span class="cf">function</span>(x) {</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> markers_genes[[x]][<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(markers_genes[[x]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">$</span>cluster <span class="ot">&lt;-</span> x</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(temp)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>top25 <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">do.call</span>(rbind, top25))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>top25<span class="sc">$</span>p.value[top25<span class="sc">$</span>p.value <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="fl">1e-300</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>top25</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["p.value"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["gene"],"name":[3],"type":["chr"],"align":["left"]},{"label":["cluster"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"1.000000e+00","2":"1.000000e+00","3":"IFITM3","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IL1B","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ISG15","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IFI6","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"G0S2","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SGK1","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MT2A","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BCL2A1","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RETN","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CLU","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"S100A8","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PLAC8","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IER3","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SAMSN1","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MAFB","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RFLNB","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RPS4Y1","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"S100A12","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DUSP6","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MX1","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"APOBEC3A","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ALOX5AP","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZBTB43","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HMGB2","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HLA-DRB5","4":"1"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SPON2","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CLIC3","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KLRF1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TXK","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TRDC","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KLRB1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IL2RB","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"GNPTAB","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IGFBP7","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"S1PR5","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CEP78","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SORL1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TTC38","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD7","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MYOM2","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ADGRG1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HSH2D","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"METRNL","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SYNGR1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"OGA","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CDC42SE1","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"UTRN","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"GK5","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"XIST","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PRSS23","4":"10"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TUBA1B","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HSP90B1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"H2AFZ","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SUB1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HMGB1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PPIB","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IGKC","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TXN","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HMGN2","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"LY6E","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MYDGF","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HIST1H4C","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RAN","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SEC61B","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"UQCRH","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IFITM3","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ANP32B","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PDIA6","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SET","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HSP90AA1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SSR3","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"JPT1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"COX6A1","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SLC25A5","4":"11"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ATP5MC3","4":"11"},{"1":"8.021102e-17","2":"1.512299e-12","3":"IL7R","4":"12"},{"1":"3.529417e-16","2":"3.327182e-12","3":"CD3D","4":"12"},{"1":"1.210895e-13","2":"7.610072e-10","3":"TRAC","4":"12"},{"1":"8.039365e-12","2":"3.789355e-08","3":"PIK3IP1","4":"12"},{"1":"5.111201e-10","2":"1.927332e-06","3":"CD3G","4":"12"},{"1":"5.298512e-06","2":"1.664969e-02","3":"C12orf57","4":"12"},{"1":"1.470916e-05","2":"3.961807e-02","3":"TCF7","4":"12"},{"1":"2.935836e-05","2":"6.919032e-02","3":"TRBC2","4":"12"},{"1":"5.399074e-05","2":"1.131046e-01","3":"LTB","4":"12"},{"1":"5.788220e-02","2":"1.000000e+00","3":"RCAN3","4":"12"},{"1":"1.102573e-01","2":"1.000000e+00","3":"SPOCK2","4":"12"},{"1":"1.357187e-01","2":"1.000000e+00","3":"MAL","4":"12"},{"1":"3.922826e-01","2":"1.000000e+00","3":"LEPROTL1","4":"12"},{"1":"4.204512e-01","2":"1.000000e+00","3":"ARHGAP15","4":"12"},{"1":"4.702124e-01","2":"1.000000e+00","3":"GIMAP7","4":"12"},{"1":"5.163191e-01","2":"1.000000e+00","3":"LDHB","4":"12"},{"1":"7.310596e-01","2":"1.000000e+00","3":"ACAP1","4":"12"},{"1":"7.365297e-01","2":"1.000000e+00","3":"GCC2","4":"12"},{"1":"7.458233e-01","2":"1.000000e+00","3":"PRDX2","4":"12"},{"1":"7.727854e-01","2":"1.000000e+00","3":"MGAT4A","4":"12"},{"1":"7.733861e-01","2":"1.000000e+00","3":"MZT2A","4":"12"},{"1":"7.742281e-01","2":"1.000000e+00","3":"CD3E","4":"12"},{"1":"8.320509e-01","2":"1.000000e+00","3":"CLEC2D","4":"12"},{"1":"8.606724e-01","2":"1.000000e+00","3":"EVL","4":"12"},{"1":"8.706461e-01","2":"1.000000e+00","3":"NOSIP","4":"12"},{"1":"8.157540e-08","2":"1.538023e-03","3":"FCGR3A","4":"13"},{"1":"6.699128e-06","2":"6.315268e-02","3":"CDKN1C","4":"13"},{"1":"1.099195e-03","2":"1.000000e+00","3":"RHOC","4":"13"},{"1":"1.228896e-03","2":"1.000000e+00","3":"HES4","4":"13"},{"1":"1.307328e-01","2":"1.000000e+00","3":"MTSS1","4":"13"},{"1":"2.416265e-01","2":"1.000000e+00","3":"TCF7L2","4":"13"},{"1":"2.676374e-01","2":"1.000000e+00","3":"RRAS","4":"13"},{"1":"3.959442e-01","2":"1.000000e+00","3":"SIGLEC10","4":"13"},{"1":"4.480870e-01","2":"1.000000e+00","3":"ZNF703","4":"13"},{"1":"4.927634e-01","2":"1.000000e+00","3":"ST3GAL5","4":"13"},{"1":"7.188462e-01","2":"1.000000e+00","3":"RHOB","4":"13"},{"1":"8.198827e-01","2":"1.000000e+00","3":"AC004687.1","4":"13"},{"1":"8.233909e-01","2":"1.000000e+00","3":"SPN","4":"13"},{"1":"8.274438e-01","2":"1.000000e+00","3":"CSF1R","4":"13"},{"1":"8.495061e-01","2":"1.000000e+00","3":"ABI3","4":"13"},{"1":"8.709117e-01","2":"1.000000e+00","3":"BATF3","4":"13"},{"1":"9.022146e-01","2":"1.000000e+00","3":"UNC119","4":"13"},{"1":"9.033666e-01","2":"1.000000e+00","3":"CXCL16","4":"13"},{"1":"9.118425e-01","2":"1.000000e+00","3":"PAG1","4":"13"},{"1":"9.119901e-01","2":"1.000000e+00","3":"AL139246.5","4":"13"},{"1":"9.334029e-01","2":"1.000000e+00","3":"IL3RA","4":"13"},{"1":"9.389182e-01","2":"1.000000e+00","3":"CTSL","4":"13"},{"1":"9.455676e-01","2":"1.000000e+00","3":"SYTL1","4":"13"},{"1":"9.643092e-01","2":"1.000000e+00","3":"FAM89B","4":"13"},{"1":"9.650408e-01","2":"1.000000e+00","3":"MRPS35","4":"13"},{"1":"1.000000e+00","2":"1.000000e+00","3":"LFNG","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KRAS","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"FAM110A","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ADGRE1","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MLX","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CALHM6","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"FAM49A","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CCM2","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD86","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZDHHC18","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BFAR","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CORO1B","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SNN","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MMP24OS","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SCIMP","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PTPRJ","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"INSIG1","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DPEP2","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MFSD12","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TESC","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MGST3","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CDH23","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HES4","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TADA3","4":"14"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DOCK5","4":"14"},{"1":"1.142040e-02","2":"1.000000e+00","3":"FCER1A","4":"15"},{"1":"9.854367e-02","2":"1.000000e+00","3":"ENHO","4":"15"},{"1":"1.726599e-01","2":"1.000000e+00","3":"CLEC10A","4":"15"},{"1":"2.064708e-01","2":"1.000000e+00","3":"HLA-DQA1","4":"15"},{"1":"4.162365e-01","2":"1.000000e+00","3":"CD1C","4":"15"},{"1":"9.236658e-01","2":"1.000000e+00","3":"KCNK6","4":"15"},{"1":"9.474608e-01","2":"1.000000e+00","3":"FABP5","4":"15"},{"1":"9.670637e-01","2":"1.000000e+00","3":"NDRG2","4":"15"},{"1":"9.901580e-01","2":"1.000000e+00","3":"PLD4","4":"15"},{"1":"9.930307e-01","2":"1.000000e+00","3":"IGFBP7","4":"15"},{"1":"9.957735e-01","2":"1.000000e+00","3":"IL18","4":"15"},{"1":"9.959208e-01","2":"1.000000e+00","3":"RUFY3","4":"15"},{"1":"9.959551e-01","2":"1.000000e+00","3":"TMEM109","4":"15"},{"1":"9.968679e-01","2":"1.000000e+00","3":"SPINT2","4":"15"},{"1":"9.970842e-01","2":"1.000000e+00","3":"BUD23","4":"15"},{"1":"9.981824e-01","2":"1.000000e+00","3":"DDOST","4":"15"},{"1":"9.986664e-01","2":"1.000000e+00","3":"TTC3","4":"15"},{"1":"9.989323e-01","2":"1.000000e+00","3":"C1QBP","4":"15"},{"1":"9.989389e-01","2":"1.000000e+00","3":"DDX27","4":"15"},{"1":"9.991129e-01","2":"1.000000e+00","3":"FAM162A","4":"15"},{"1":"9.994895e-01","2":"1.000000e+00","3":"EIF2AK4","4":"15"},{"1":"9.996141e-01","2":"1.000000e+00","3":"MAT2A","4":"15"},{"1":"9.996496e-01","2":"1.000000e+00","3":"LSM5","4":"15"},{"1":"9.997821e-01","2":"1.000000e+00","3":"OSTC","4":"15"},{"1":"9.997839e-01","2":"1.000000e+00","3":"PHB2","4":"15"},{"1":"1.089415e-19","2":"2.053984e-15","3":"CD79A","4":"2"},{"1":"3.013909e-15","2":"2.841212e-11","3":"MS4A1","4":"2"},{"1":"4.468069e-14","2":"2.808032e-10","3":"TNFRSF13C","4":"2"},{"1":"4.148575e-13","2":"1.955431e-09","3":"BANK1","4":"2"},{"1":"6.809328e-12","2":"2.567661e-08","3":"IGHM","4":"2"},{"1":"5.185271e-11","2":"1.629385e-07","3":"LINC00926","4":"2"},{"1":"1.560949e-09","2":"4.204304e-06","3":"CD79B","4":"2"},{"1":"6.537976e-08","2":"1.540837e-04","3":"IGHD","4":"2"},{"1":"2.653064e-05","2":"5.557875e-02","3":"IGKC","4":"2"},{"1":"8.300776e-05","2":"1.565028e-01","3":"RALGPS2","4":"2"},{"1":"3.531253e-03","2":"1.000000e+00","3":"CXCR4","4":"2"},{"1":"1.029802e-02","2":"1.000000e+00","3":"P2RX5","4":"2"},{"1":"1.044297e-01","2":"1.000000e+00","3":"ISG20","4":"2"},{"1":"1.343794e-01","2":"1.000000e+00","3":"CD22","4":"2"},{"1":"1.446398e-01","2":"1.000000e+00","3":"AFF3","4":"2"},{"1":"2.403883e-01","2":"1.000000e+00","3":"JUN","4":"2"},{"1":"3.068676e-01","2":"1.000000e+00","3":"TCF4","4":"2"},{"1":"4.417767e-01","2":"1.000000e+00","3":"GNG7","4":"2"},{"1":"4.639890e-01","2":"1.000000e+00","3":"SWAP70","4":"2"},{"1":"5.646430e-01","2":"1.000000e+00","3":"CD40","4":"2"},{"1":"8.374622e-01","2":"1.000000e+00","3":"TMEM243","4":"2"},{"1":"8.792524e-01","2":"1.000000e+00","3":"IFT57","4":"2"},{"1":"8.815627e-01","2":"1.000000e+00","3":"CD69","4":"2"},{"1":"9.346551e-01","2":"1.000000e+00","3":"NFKBID","4":"2"},{"1":"9.506568e-01","2":"1.000000e+00","3":"BLK","4":"2"},{"1":"9.999962e-01","2":"1.000000e+00","3":"RPN1","4":"3"},{"1":"9.999992e-01","2":"1.000000e+00","3":"RETN","4":"3"},{"1":"9.999997e-01","2":"1.000000e+00","3":"MXD1","4":"3"},{"1":"9.999999e-01","2":"1.000000e+00","3":"AHR","4":"3"},{"1":"9.999999e-01","2":"1.000000e+00","3":"CORO1C","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ARF3","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CCR1","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ENTPD1","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SEC31A","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SLA","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CYP1B1","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CPEB4","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ETS2","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TMEM170B","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ATP11A","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RNASE6","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NORAD","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MRPL33","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BCL6","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CLEC4E","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ALKBH7","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"FBXO9","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RUNX1","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"AKR1A1","4":"3"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ISG15","4":"3"},{"1":"1.097525e-14","2":"4.900089e-12","3":"APLP2","4":"4"},{"1":"1.132269e-14","2":"4.900089e-12","3":"FGR","4":"4"},{"1":"1.145026e-14","2":"4.900089e-12","3":"AP1S2","4":"4"},{"1":"1.162606e-14","2":"4.900089e-12","3":"GRN","4":"4"},{"1":"1.168037e-14","2":"4.900089e-12","3":"COX8A","4":"4"},{"1":"1.168037e-14","2":"4.900089e-12","3":"GMFG","4":"4"},{"1":"1.168037e-14","2":"4.900089e-12","3":"TYMP","4":"4"},{"1":"1.171672e-14","2":"4.900089e-12","3":"ATP6V0B","4":"4"},{"1":"1.171672e-14","2":"4.900089e-12","3":"SAT1","4":"4"},{"1":"1.171672e-14","2":"4.900089e-12","3":"SPI1","4":"4"},{"1":"1.171672e-14","2":"4.900089e-12","3":"EIF3K","4":"4"},{"1":"1.171672e-14","2":"4.900089e-12","3":"ITGB2","4":"4"},{"1":"1.173857e-14","2":"4.900089e-12","3":"HLA-E","4":"4"},{"1":"1.173857e-14","2":"4.900089e-12","3":"UQCRB","4":"4"},{"1":"1.173857e-14","2":"4.900089e-12","3":"LCP1","4":"4"},{"1":"1.174951e-14","2":"4.900089e-12","3":"GNAI2","4":"4"},{"1":"1.174951e-14","2":"4.900089e-12","3":"NEAT1","4":"4"},{"1":"1.174951e-14","2":"4.900089e-12","3":"UBC","4":"4"},{"1":"1.174951e-14","2":"4.900089e-12","3":"LGALS1","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"S100A10","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"TMSB10","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"RACK1","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"ACTB","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"RPL7","4":"4"},{"1":"1.175316e-14","2":"4.900089e-12","3":"RPL8","4":"4"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NAP1L4","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SATB1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BCL11B","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"EPB41","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CCR7","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RCAN3","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"XIST","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TTC3","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DNAJB1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZNF101","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ARHGAP15","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZNF292","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NDFIP1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TPR","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"AC016831.5","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZC3HAV1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ITK","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"FAM107B","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD6","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"HSPB1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CAMK4","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"MPHOSPH8","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PHF3","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KMT2A","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SORL1","4":"5"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD69","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IGLC2","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"FAM129C","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"SMARCB1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"IGKC","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PHACTR1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ORAI2","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NT5C","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TCOF1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PMAIP1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ZCCHC7","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NFKBID","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"EPB41L4A-AS1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NFKB2","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CREBRF","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BIRC3","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"NCOA3","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TTC3","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"YBX3","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TCEA1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CCDC50","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"VPREB3","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"BCL11A","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"PAX5","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CHPT1","4":"6"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD8A","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD8B","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"LYAR","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TRGC2","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DUSP2","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TNFAIP3","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"TUBA4A","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KLRG1","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"XIST","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD6","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"EML4","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CD2","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ATG2A","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"DNAJB1","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"AC058791.1","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"ITGA4","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CNOT6L","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"GZMK","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RNF125","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KIAA1551","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"GYPC","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RORA","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"RASGRP1","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"CYTOR","4":"7"},{"1":"1.000000e+00","2":"1.000000e+00","3":"KMT2A","4":"7"},{"1":"2.534613e-21","2":"4.778760e-17","3":"CST7","4":"8"},{"1":"4.528149e-20","2":"4.268686e-16","3":"NKG7","4":"8"},{"1":"1.127202e-18","2":"7.084087e-15","3":"CCL5","4":"8"},{"1":"1.550105e-16","2":"7.306419e-13","3":"GZMA","4":"8"},{"1":"3.397259e-16","2":"1.281038e-12","3":"CTSW","4":"8"},{"1":"9.071240e-16","2":"2.850486e-12","3":"GZMM","4":"8"},{"1":"7.487552e-13","2":"2.016719e-09","3":"GZMH","4":"8"},{"1":"8.395778e-10","2":"1.978675e-06","3":"KLRD1","4":"8"},{"1":"5.480946e-08","2":"1.148197e-04","3":"GZMB","4":"8"},{"1":"9.915894e-08","2":"1.869543e-04","3":"GNLY","4":"8"},{"1":"1.197720e-07","2":"2.052892e-04","3":"HOPX","4":"8"},{"1":"1.050871e-06","2":"1.651094e-03","3":"C12orf75","4":"8"},{"1":"2.286169e-06","2":"3.315649e-03","3":"FGFBP2","4":"8"},{"1":"1.518347e-03","2":"1.000000e+00","3":"CD247","4":"8"},{"1":"1.748512e-03","2":"1.000000e+00","3":"PRF1","4":"8"},{"1":"4.925277e-03","2":"1.000000e+00","3":"CCL4","4":"8"},{"1":"2.069177e-02","2":"1.000000e+00","3":"SYNE1","4":"8"},{"1":"2.366028e-01","2":"1.000000e+00","3":"CMC1","4":"8"},{"1":"2.912062e-01","2":"1.000000e+00","3":"ARL4C","4":"8"},{"1":"2.942688e-01","2":"1.000000e+00","3":"SYTL3","4":"8"},{"1":"3.014639e-01","2":"1.000000e+00","3":"RUNX3","4":"8"},{"1":"4.549195e-01","2":"1.000000e+00","3":"APMAP","4":"8"},{"1":"6.096589e-01","2":"1.000000e+00","3":"MATK","4":"8"},{"1":"7.564139e-01","2":"1.000000e+00","3":"ABHD17A","4":"8"},{"1":"9.559005e-01","2":"1.000000e+00","3":"TRDC","4":"8"},{"1":"4.245295e-06","2":"8.004080e-02","3":"PF4","4":"9"},{"1":"1.528578e-05","2":"1.440991e-01","3":"PPBP","4":"9"},{"1":"3.878298e-03","2":"1.000000e+00","3":"GNG11","4":"9"},{"1":"5.633827e-03","2":"1.000000e+00","3":"GP9","4":"9"},{"1":"3.328087e-02","2":"1.000000e+00","3":"CMTM5","4":"9"},{"1":"8.668065e-02","2":"1.000000e+00","3":"TREML1","4":"9"},{"1":"1.683333e-01","2":"1.000000e+00","3":"NRGN","4":"9"},{"1":"2.729799e-01","2":"1.000000e+00","3":"ITGA2B","4":"9"},{"1":"2.928266e-01","2":"1.000000e+00","3":"MYL9","4":"9"},{"1":"2.984135e-01","2":"1.000000e+00","3":"HIST1H2AC","4":"9"},{"1":"3.215610e-01","2":"1.000000e+00","3":"TUBB1","4":"9"},{"1":"3.390438e-01","2":"1.000000e+00","3":"CLEC1B","4":"9"},{"1":"4.106979e-01","2":"1.000000e+00","3":"CAVIN2","4":"9"},{"1":"5.615710e-01","2":"1.000000e+00","3":"GRAP2","4":"9"},{"1":"5.688858e-01","2":"1.000000e+00","3":"MTURN","4":"9"},{"1":"6.724945e-01","2":"1.000000e+00","3":"TMEM40","4":"9"},{"1":"7.286170e-01","2":"1.000000e+00","3":"ACRBP","4":"9"},{"1":"7.729976e-01","2":"1.000000e+00","3":"PDLIM1","4":"9"},{"1":"9.527698e-01","2":"1.000000e+00","3":"RGS18","4":"9"},{"1":"9.607897e-01","2":"1.000000e+00","3":"SPARC","4":"9"},{"1":"9.702716e-01","2":"1.000000e+00","3":"SNCA","4":"9"},{"1":"9.712176e-01","2":"1.000000e+00","3":"PTGS1","4":"9"},{"1":"9.771752e-01","2":"1.000000e+00","3":"PRKAR2B","4":"9"},{"1":"9.930153e-01","2":"1.000000e+00","3":"BEX3","4":"9"},{"1":"9.960745e-01","2":"1.000000e+00","3":"CLU","4":"9"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can plot them as barplots per cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>), <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">3</span>, <span class="dv">1</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">unique</span>(top25<span class="sc">$</span>cluster)) {</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">barplot</span>(<span class="fu">sort</span>(<span class="fu">setNames</span>(<span class="sc">-</span><span class="fu">log10</span>(top25<span class="sc">$</span>p.value), top25<span class="sc">$</span>gene)[top25<span class="sc">$</span>cluster <span class="sc">==</span> i], F),</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">horiz =</span> T, <span class="at">las =</span> <span class="dv">1</span>, <span class="at">main =</span> <span class="fu">paste0</span>(i, <span class="st">" vs. rest"</span>), <span class="at">border =</span> <span class="st">"white"</span>, <span class="at">yaxs =</span> <span class="st">"i"</span>, <span class="at">xlab =</span> <span class="st">"-log10pval"</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>)), <span class="at">lty =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-5-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-5-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-5-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-5-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-5-3.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>We can visualize them as a heatmap. Here we are selecting the top 5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>top25 <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_min</span>(p.value, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="ot">-&gt;</span> top5</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotHeatmap</span>(sce[, <span class="fu">order</span>(sce<span class="sc">$</span>leiden_k20)],</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">unique</span>(top5<span class="sc">$</span>gene),</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">center =</span> T, <span class="at">zlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_columns_by =</span> <span class="st">"leiden_k20"</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_colnames =</span> F, <span class="at">cluster_cols =</span> F,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">fontsize_row =</span> <span class="dv">6</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"purple"</span>, <span class="st">"black"</span>, <span class="st">"yellow"</span>))(<span class="dv">90</span>)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-6-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
<p>We can also plot a violin plot for each gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotExpression</span>(sce, <span class="at">features =</span> <span class="fu">unique</span>(top5<span class="sc">$</span>gene), <span class="at">x =</span> <span class="st">"leiden_k20"</span>, <span class="at">ncol =</span> <span class="dv">5</span>, <span class="at">colour_by =</span> <span class="st">"leiden_k20"</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-7-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="1248"></a></p>
</figure>
</div>
</div>
</div>
<p>Another way is by representing the overall group expression and detection rates in a dot-plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(sce, <span class="at">features =</span> <span class="fu">unique</span>(top5<span class="sc">$</span>gene), <span class="at">group=</span><span class="st">"leiden_k20"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-8-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-dge_cond" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="meta-dge_cond"><span class="header-section-number">3</span> DGE across conditions</h2>
<p>The second way of computing differential expression is to answer which genes are differentially expressed within a cluster. For example, in our case we have libraries comming from patients and controls and we would like to know which genes are influenced the most in a particular cell type. For this end, we will first subset our data for the desired cell cluster, then change the cell identities to the variable of comparison (which now in our case is the <strong>type</strong>, e.g.&nbsp;Covid/Ctrl).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter cells from that cluster</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>cell_selection <span class="ot">&lt;-</span> sce[, sce<span class="sc">$</span>leiden_k20 <span class="sc">==</span> <span class="dv">2</span>]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differentiall expression</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>DGE_cell_selection <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> cell_selection,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">groups =</span> cell_selection<span class="sc">@</span>colData<span class="sc">$</span>type,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">lfc =</span> .<span class="dv">25</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">pval.type =</span> <span class="st">"all"</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">direction =</span> <span class="st">"any"</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>top5_cell_selection <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(DGE_cell_selection), <span class="cf">function</span>(x) {</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    temp <span class="ot">&lt;-</span> DGE_cell_selection[[x]][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(DGE_cell_selection[[x]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    temp<span class="sc">$</span>cluster <span class="ot">&lt;-</span> x</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(temp)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>top5_cell_selection <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">do.call</span>(rbind, top5_cell_selection))</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>top5_cell_selection</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["p.value"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["gene"],"name":[3],"type":["chr"],"align":["left"]},{"label":["cluster"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"1.631149e-23","2":"3.075368e-19","3":"RPS4Y1","4":"Control"},{"1":"1.715742e-11","2":"1.617430e-07","3":"XIST","4":"Control"},{"1":"4.260841e-10","2":"2.642720e-06","3":"ISG15","4":"Control"},{"1":"5.606704e-10","2":"2.642720e-06","3":"CD69","4":"Control"},{"1":"1.125015e-08","2":"4.242205e-05","3":"HLA-DRB5","4":"Control"},{"1":"1.631149e-23","2":"3.075368e-19","3":"RPS4Y1","4":"Covid"},{"1":"1.715742e-11","2":"1.617430e-07","3":"XIST","4":"Covid"},{"1":"4.260841e-10","2":"2.642720e-06","3":"ISG15","4":"Covid"},{"1":"5.606704e-10","2":"2.642720e-06","3":"CD69","4":"Covid"},{"1":"1.125015e-08","2":"4.242205e-05","3":"HLA-DRB5","4":"Covid"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can now plot the expression across the <strong>type</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>scater<span class="sc">::</span><span class="fu">plotExpression</span>(cell_selection, <span class="at">features =</span> <span class="fu">unique</span>(top5_cell_selection<span class="sc">$</span>gene), <span class="at">x =</span> <span class="st">"type"</span>, <span class="at">ncol =</span> <span class="dv">5</span>, <span class="at">colour_by =</span> <span class="st">"type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Or we can plot them as dotplots to see the expression in each individual sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDots</span>(cell_selection, <span class="fu">unique</span>(top5_cell_selection<span class="sc">$</span>gene), <span class="at">group=</span><span class="st">"sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-11-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="1248"></a></p>
</figure>
</div>
</div>
</div>
<p>Clearly many of the top Covid genes are only high in the covid_17 sample, and not a general feature of covid patients. In this case using a method that can control for pseudreplication or pseudobulk methods would be appropriate. We have a more thorough discussion on this issue with sample batch effects in differential expression across conditions in the <a href="https://nbisweden.github.io/workshop-scRNAseq/labs/seurat/seurat_05_dge.html#meta-dge_cond">Seurat tutorial</a>.</p>
</section>
<section id="meta-dge_gsa" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-dge_gsa"><span class="header-section-number">4</span> Gene Set Analysis (GSA)</h2>
<section id="meta-dge_gsa_hyper" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="meta-dge_gsa_hyper"><span class="header-section-number">4.1</span> Hypergeometric enrichment test</h3>
<p>Having a defined list of differentially expressed genes, you can now look for their combined function using hypergeometric test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load additional packages</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichR)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check available databases to perform enrichment (then choose one)</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>enrichR<span class="sc">::</span><span class="fu">listEnrichrDbs</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["geneCoverage"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["genesPerTerm"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["libraryName"],"name":[3],"type":["chr"],"align":["left"]},{"label":["link"],"name":[4],"type":["chr"],"align":["left"]},{"label":["numTerms"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["appyter"],"name":[6],"type":["chr"],"align":["left"]},{"label":["categoryId"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"13362","2":"275","3":"Genome_Browser_PWMs","4":"http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","5":"615","6":"ea115789fcbf12797fd692cec6df0ab4dbc79c6a","7":"1"},{"1":"27884","2":"1284","3":"TRANSFAC_and_JASPAR_PWMs","4":"http://jaspar.genereg.net/html/DOWNLOAD/","5":"326","6":"7d42eb43a64a4e3b20d721fc7148f685b53b6b30","7":"1"},{"1":"6002","2":"77","3":"Transcription_Factor_PPIs","4":"","5":"290","6":"849f222220618e2599d925b6b51868cf1dab3763","7":"1"},{"1":"47172","2":"1370","3":"ChEA_2013","4":"http://amp.pharm.mssm.edu/lib/cheadownload.jsp","5":"353","6":"7ebe772afb55b63b41b79dd8d06ea0fdd9fa2630","7":"7"},{"1":"47107","2":"509","3":"Drug_Perturbations_from_GEO_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"701","6":"ad270a6876534b7cb063e004289dcd4d3164f342","7":"7"},{"1":"21493","2":"3713","3":"ENCODE_TF_ChIP-seq_2014","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"498","6":"497787ebc418d308045efb63b8586f10c526af51","7":"7"},{"1":"1295","2":"18","3":"BioCarta_2013","4":"https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"249","6":"4a293326037a5229aedb1ad7b2867283573d8bcd","7":"7"},{"1":"3185","2":"73","3":"Reactome_2013","4":"http://www.reactome.org/download/index.html","5":"78","6":"b343994a1b68483b0122b08650201c9b313d5c66","7":"7"},{"1":"2854","2":"34","3":"WikiPathways_2013","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"199","6":"5c307674c8b97e098f8399c92f451c0ff21cbf68","7":"7"},{"1":"15057","2":"300","3":"Disease_Signatures_from_GEO_up_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"142","6":"248c4ed8ea28352795190214713c86a39fd7afab","7":"7"},{"1":"4128","2":"48","3":"KEGG_2013","4":"http://www.kegg.jp/kegg/download/","5":"200","6":"eb26f55d3904cb0ea471998b6a932a9bf65d8e50","7":"7"},{"1":"34061","2":"641","3":"TF-LOF_Expression_from_GEO","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"269","6":"","7":"1"},{"1":"7504","2":"155","3":"TargetScan_microRNA","4":"http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","5":"222","6":"f4029bf6a62c91ab29401348e51df23b8c44c90f","7":"7"},{"1":"16399","2":"247","3":"PPI_Hub_Proteins","4":"http://amp.pharm.mssm.edu/X2K","5":"385","6":"69c0cfe07d86f230a7ef01b365abcc7f6e52f138","7":"2"},{"1":"12753","2":"57","3":"GO_Molecular_Function_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"1136","6":"f531ac2b6acdf7587a54b79b465a5f4aab8f00f9","7":"7"},{"1":"23726","2":"127","3":"GeneSigDB","4":"https://pubmed.ncbi.nlm.nih.gov/22110038/","5":"2139","6":"6d655e0aa3408a7accb3311fbda9b108681a8486","7":"4"},{"1":"32740","2":"85","3":"Chromosome_Location","4":"http://software.broadinstitute.org/gsea/msigdb/index.jsp","5":"386","6":"8dab0f96078977223646ff63eb6187e0813f1433","7":"7"},{"1":"13373","2":"258","3":"Human_Gene_Atlas","4":"http://biogps.org/downloads/","5":"84","6":"0741451470203d7c40a06274442f25f74b345c9c","7":"5"},{"1":"19270","2":"388","3":"Mouse_Gene_Atlas","4":"http://biogps.org/downloads/","5":"96","6":"31191bfadded5f96983f93b2a113cf2110ff5ddb","7":"5"},{"1":"13236","2":"82","3":"GO_Cellular_Component_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"641","6":"e1d004d5797cbd2363ef54b1c3b361adb68795c6","7":"7"},{"1":"14264","2":"58","3":"GO_Biological_Process_2015","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"5192","6":"bf120b6e11242b1a64c80910d8e89f87e618e235","7":"7"},{"1":"3096","2":"31","3":"Human_Phenotype_Ontology","4":"http://www.human-phenotype-ontology.org/","5":"1779","6":"17a138b0b70aa0e143fe63c14f82afb70bc3ed0a","7":"3"},{"1":"22288","2":"4368","3":"Epigenomics_Roadmap_HM_ChIP-seq","4":"http://www.roadmapepigenomics.org/","5":"383","6":"e1bc8a398e9b21f9675fb11bef18087eda21b1bf","7":"1"},{"1":"4533","2":"37","3":"KEA_2013","4":"http://amp.pharm.mssm.edu/lib/keacommandline.jsp","5":"474","6":"462045609440fa1e628a75716b81a1baa5bd9145","7":"7"},{"1":"10231","2":"158","3":"NURSA_Human_Endogenous_Complexome","4":"https://www.nursa.org/nursa/index.jsf","5":"1796","6":"7d3566b12ebc23dd23d9ca9bb97650f826377b16","7":"2"},{"1":"2741","2":"5","3":"CORUM","4":"http://mips.helmholtz-muenchen.de/genre/proj/corum/","5":"1658","6":"d047f6ead7831b00566d5da7a3b027ed9196e104","7":"2"},{"1":"5655","2":"342","3":"SILAC_Phosphoproteomics","4":"http://amp.pharm.mssm.edu/lib/keacommandline.jsp","5":"84","6":"54dcd9438b33301deb219866e162b0f9da7e63a0","7":"2"},{"1":"10406","2":"715","3":"MGI_Mammalian_Phenotype_Level_3","4":"http://www.informatics.jax.org/","5":"71","6":"c3bfc90796cfca8f60cba830642a728e23a53565","7":"7"},{"1":"10493","2":"200","3":"MGI_Mammalian_Phenotype_Level_4","4":"http://www.informatics.jax.org/","5":"476","6":"0b09a9a1aa0af4fc7ea22d34a9ae644d45864bd6","7":"7"},{"1":"11251","2":"100","3":"Old_CMAP_up","4":"http://www.broadinstitute.org/cmap/","5":"6100","6":"9041f90cccbc18479138330228b336265e09021c","7":"7"},{"1":"8695","2":"100","3":"Old_CMAP_down","4":"http://www.broadinstitute.org/cmap/","5":"6100","6":"ebc0d905b3b3142f936d400c5f2a4ff926c81c37","7":"7"},{"1":"1759","2":"25","3":"OMIM_Disease","4":"http://www.omim.org/downloads","5":"90","6":"cb2b92578a91e023d0498a334923ee84add34eca","7":"4"},{"1":"2178","2":"89","3":"OMIM_Expanded","4":"http://www.omim.org/downloads","5":"187","6":"27eca242904d8e12a38cf8881395bc50d57a03e1","7":"4"},{"1":"851","2":"15","3":"VirusMINT","4":"http://mint.bio.uniroma2.it/download.html","5":"85","6":"5abad1fc36216222b0420cadcd9be805a0dda63e","7":"4"},{"1":"10061","2":"106","3":"MSigDB_Computational","4":"http://www.broadinstitute.org/gsea/msigdb/collections.jsp","5":"858","6":"e4cdcc7e259788fdf9b25586cce3403255637064","7":"4"},{"1":"11250","2":"166","3":"MSigDB_Oncogenic_Signatures","4":"http://www.broadinstitute.org/gsea/msigdb/collections.jsp","5":"189","6":"c76f5319c33c4833c71db86a30d7e33cd63ff8cf","7":"4"},{"1":"15406","2":"300","3":"Disease_Signatures_from_GEO_down_2014","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"142","6":"aabdf7017ae55ae75a004270924bcd336653b986","7":"7"},{"1":"17711","2":"300","3":"Virus_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"323","6":"45268b7fc680d05dd9a29743c2f2b2840a7620bf","7":"4"},{"1":"17576","2":"300","3":"Virus_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"323","6":"5f531580ccd168ee4acc18b02c6bdf8200e19d08","7":"4"},{"1":"15797","2":"176","3":"Cancer_Cell_Line_Encyclopedia","4":"https://portals.broadinstitute.org/ccle/home\\n","5":"967","6":"eb38dbc3fb20adafa9d6f9f0b0e36f378e75284f","7":"5"},{"1":"12232","2":"343","3":"NCI-60_Cancer_Cell_Lines","4":"http://biogps.org/downloads/","5":"93","6":"75c81676d8d6d99d262c9660edc024b78cfb07c9","7":"5"},{"1":"13572","2":"301","3":"Tissue_Protein_Expression_from_ProteomicsDB","4":"https://www.proteomicsdb.org/","5":"207","6":"","7":"7"},{"1":"6454","2":"301","3":"Tissue_Protein_Expression_from_Human_Proteome_Map","4":"http://www.humanproteomemap.org/index.php","5":"30","6":"49351dc989f9e6ca97c55f8aca7778aa3bfb84b9","7":"5"},{"1":"3723","2":"47","3":"HMDB_Metabolites","4":"http://www.hmdb.ca/downloads","5":"3906","6":"1905132115d22e4119bce543bdacaab074edb363","7":"6"},{"1":"7588","2":"35","3":"Pfam_InterPro_Domains","4":"ftp://ftp.ebi.ac.uk/pub/databases/interpro/","5":"311","6":"e2b4912cfb799b70d87977808c54501544e4cdc9","7":"6"},{"1":"7682","2":"78","3":"GO_Biological_Process_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"941","6":"5216d1ade194ffa5a6c00f105e2b1899f64f45fe","7":"7"},{"1":"7324","2":"172","3":"GO_Cellular_Component_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"205","6":"fd1332a42395e0bc1dba82868b39be7983a48cc5","7":"7"},{"1":"8469","2":"122","3":"GO_Molecular_Function_2013","4":"http://www.geneontology.org/GO.downloads.annotations.shtml","5":"402","6":"7e3e99e5aae02437f80b0697b197113ce3209ab0","7":"7"},{"1":"13121","2":"305","3":"Allen_Brain_Atlas_up","4":"http://www.brain-map.org/","5":"2192","6":"3804715a63a308570e47aa1a7877f01147ca6202","7":"5"},{"1":"26382","2":"1811","3":"ENCODE_TF_ChIP-seq_2015","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"816","6":"56b6adb4dc8a2f540357ef992d6cd93dfa2907e5","7":"1"},{"1":"29065","2":"2123","3":"ENCODE_Histone_Modifications_2015","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"412","6":"55b56cd8cf2ff04b26a09b9f92904008b82f3a6f","7":"1"},{"1":"280","2":"9","3":"Phosphatase_Substrates_from_DEPOD","4":"http://www.koehn.embl.de/depod/","5":"59","6":"d40701e21092b999f4720d1d2b644dd0257b6259","7":"2"},{"1":"13877","2":"304","3":"Allen_Brain_Atlas_down","4":"http://www.brain-map.org/","5":"2192","6":"ea67371adec290599ddf484ced2658cfae259304","7":"5"},{"1":"15852","2":"912","3":"ENCODE_Histone_Modifications_2013","4":"http://genome.ucsc.edu/ENCODE/downloads.html","5":"109","6":"c209ae527bc8e98e4ccd27a668d36cd2c80b35b4","7":"7"},{"1":"4320","2":"129","3":"Achilles_fitness_increase","4":"http://www.broadinstitute.org/achilles","5":"216","6":"98366496a75f163164106e72439fb2bf2f77de4e","7":"4"},{"1":"4271","2":"128","3":"Achilles_fitness_decrease","4":"http://www.broadinstitute.org/achilles","5":"216","6":"83a710c1ff67fd6b8af0d80fa6148c40dbd9bc64","7":"4"},{"1":"10496","2":"201","3":"MGI_Mammalian_Phenotype_2013","4":"http://www.informatics.jax.org/","5":"476","6":"a4c6e217a81a4a58ff5a1c9fc102b70beab298e9","7":"7"},{"1":"1678","2":"21","3":"BioCarta_2015","4":"https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"239","6":"70e4eb538daa7688691acfe5d9c3c19022be832b","7":"7"},{"1":"756","2":"12","3":"HumanCyc_2015","4":"http://humancyc.org/","5":"125","6":"711f0c02b23f5e02a01207174943cfeee9d3ea9c","7":"7"},{"1":"3800","2":"48","3":"KEGG_2015","4":"http://www.kegg.jp/kegg/download/","5":"179","6":"e80d25c56de53c704791ddfdc6ab5eec28ae7243","7":"7"},{"1":"1918","2":"39","3":"Panther_2015","4":"http://www.pantherdb.org/","5":"104","6":"ab824aeeff0712bab61f372e43aebb870d1677a9","7":"7"},{"1":"5863","2":"51","3":"WikiPathways_2015","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"404","6":"1f7eea2f339f37856522c1f1c70ec74c7b25325f","7":"7"},{"1":"6768","2":"47","3":"Reactome_2015","4":"http://www.reactome.org/download/index.html","5":"1389","6":"36e541bee015eddb8d53827579549e30fe7a3286","7":"7"},{"1":"25651","2":"807","3":"ESCAPE","4":"http://www.maayanlab.net/ESCAPE/","5":"315","6":"a7acc741440264717ff77751a7e5fed723307835","7":"5"},{"1":"19129","2":"1594","3":"HomoloGene","4":"http://www.ncbi.nlm.nih.gov/homologene","5":"12","6":"663b665b75a804ef98add689f838b68e612f0d2a","7":"6"},{"1":"23939","2":"293","3":"Disease_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"839","6":"0f412e0802d76efa0374504c2c9f5e0624ff7f09","7":"8"},{"1":"23561","2":"307","3":"Disease_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"839","6":"9ddc3902fb01fb9eaf1a2a7c2ff3acacbb48d37e","7":"8"},{"1":"23877","2":"302","3":"Drug_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"906","6":"068623a05ecef3e4a5e0b4f8db64bb8faa3c897f","7":"8"},{"1":"15886","2":"9","3":"Genes_Associated_with_NIH_Grants","4":"https://grants.nih.gov/grants/oer.htm\\n","5":"32876","6":"76fc5ec6735130e287e62bae6770a3c5ee068645","7":"6"},{"1":"24350","2":"299","3":"Drug_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"906","6":"c9c2155b5ac81ac496854fa61ba566dcae06cc80","7":"8"},{"1":"3102","2":"25","3":"KEA_2015","4":"http://amp.pharm.mssm.edu/Enrichr","5":"428","6":"18a081774e6e0aaf60b1a4be7fd20afcf9e08399","7":"2"},{"1":"31132","2":"298","3":"Gene_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"2460","6":"53dedc29ce3100930d68e506f941ef59de05dc6b","7":"8"},{"1":"30832","2":"302","3":"Gene_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"2460","6":"499882af09c62dd6da545c15cb51c1dc5e234f78","7":"8"},{"1":"48230","2":"1429","3":"ChEA_2015","4":"http://amp.pharm.mssm.edu/Enrichr","5":"395","6":"712eb7b6edab04658df153605ec6079fa89fb5c7","7":"7"},{"1":"5613","2":"36","3":"dbGaP","4":"http://www.ncbi.nlm.nih.gov/gap","5":"345","6":"010f1267055b1a1cb036e560ea525911c007a666","7":"4"},{"1":"9559","2":"73","3":"LINCS_L1000_Chem_Pert_up","4":"https://clue.io/","5":"33132","6":"5e678b3debe8d8ea95187d0cd35c914017af5eb3","7":"7"},{"1":"9448","2":"63","3":"LINCS_L1000_Chem_Pert_down","4":"https://clue.io/","5":"33132","6":"fedbf5e221f45ee60ebd944f92569b5eda7f2330","7":"7"},{"1":"16725","2":"1443","3":"GTEx_Tissue_Expression_Down","4":"http://www.gtexportal.org/","5":"2918","6":"74b818bd299a9c42c1750ffe43616aa9f7929f02","7":"5"},{"1":"19249","2":"1443","3":"GTEx_Tissue_Expression_Up","4":"http://www.gtexportal.org/","5":"2918","6":"103738763d89cae894bec9f145ac28167a90e611","7":"5"},{"1":"15090","2":"282","3":"Ligand_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"261","6":"1eb3c0426140340527155fd0ef67029db2a72191","7":"8"},{"1":"16129","2":"292","3":"Aging_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"286","6":"cd95fe1b505ba6f28cd722cfba50fdea979d3b4c","7":"8"},{"1":"15309","2":"308","3":"Aging_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"286","6":"74c4f0a0447777005b2a5c00c9882a56dfc62d7c","7":"8"},{"1":"15103","2":"318","3":"Ligand_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"261","6":"31baa39da2931ddd5f7aedf2d0bbba77d2ba7b46","7":"8"},{"1":"15022","2":"290","3":"MCF7_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"401","6":"555f68aef0a29a67b614a0d7e20b6303df9069c6","7":"8"},{"1":"15676","2":"310","3":"MCF7_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"401","6":"1bc2ba607f1ff0dda44e2a15f32a2c04767da18c","7":"8"},{"1":"15854","2":"279","3":"Microbe_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"312","6":"9e613dba78ef7e60676b13493a9dc49ccd3c8b3f","7":"8"},{"1":"15015","2":"321","3":"Microbe_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"312","6":"d0c3e2a68e8c611c669098df2c87b530cec3e132","7":"8"},{"1":"3788","2":"159","3":"LINCS_L1000_Ligand_Perturbations_down","4":"https://clue.io/","5":"96","6":"957846cb05ef31fc8514120516b73cc65af7980e","7":"7"},{"1":"3357","2":"153","3":"LINCS_L1000_Ligand_Perturbations_up","4":"https://clue.io/","5":"96","6":"3bd494146c98d8189898a947f5ef5710f1b7c4b2","7":"7"},{"1":"12668","2":"300","3":"L1000_Kinase_and_GPCR_Perturbations_down","4":"https://clue.io/","5":"3644","6":"1ccc5bce553e0c2279f8e3f4ddcfbabcf566623b","7":"7"},{"1":"12638","2":"300","3":"L1000_Kinase_and_GPCR_Perturbations_up","4":"https://clue.io/","5":"3644","6":"b54a0d4ba525eac4055c7314ca9d9312adcb220c","7":"7"},{"1":"8973","2":"64","3":"Reactome_2016","4":"http://www.reactome.org/download/index.html","5":"1530","6":"1f54638e8f45075fb79489f0e0ef906594cb0678","7":"7"},{"1":"7010","2":"87","3":"KEGG_2016","4":"http://www.kegg.jp/kegg/download/","5":"293","6":"43f56da7540195ba3c94eb6e34c522a699b36da9","7":"7"},{"1":"5966","2":"51","3":"WikiPathways_2016","4":"http://www.wikipathways.org/index.php/Download_Pathways","5":"437","6":"340be98b444cad50bb974df69018fd598e23e5e1","7":"7"},{"1":"15562","2":"887","3":"ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","4":"","5":"104","6":"5426f7747965c23ef32cff46fabf906e2cd76bfa","7":"1"},{"1":"17850","2":"300","3":"Kinase_Perturbations_from_GEO_down","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"285","6":"bb9682d78b8fc43be842455e076166fcd02cefc3","7":"2"},{"1":"17660","2":"300","3":"Kinase_Perturbations_from_GEO_up","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"285","6":"78618915009cac3a0663d6f99d359e39a31b6660","7":"2"},{"1":"1348","2":"19","3":"BioCarta_2016","4":"http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","5":"237","6":"13d9ab18921d5314a5b2b366f6142b78ab0ff6aa","7":"2"},{"1":"934","2":"13","3":"HumanCyc_2016","4":"http://humancyc.org/","5":"152","6":"d6a502ef9b4c789ed5e73ca5a8de372796e5c72a","7":"2"},{"1":"2541","2":"39","3":"NCI-Nature_2016","4":"https://www.ndexbio.org/index.html#/networkset/7bc65b82-2a2f-11ed-ac45-0ac135e8bacf","5":"209","6":"3c1e1f7d1a651d9aaa198e73704030716fc09431","7":"2"},{"1":"2041","2":"42","3":"Panther_2016","4":"http://www.pantherdb.org/pathway/","5":"112","6":"ca5f6abf7f75d9baae03396e84d07300bf1fd051","7":"2"},{"1":"5209","2":"300","3":"DrugMatrix","4":"https://ntp.niehs.nih.gov/drugmatrix/","5":"7876","6":"255c3db820d612f34310f22a6985dad50e9fe1fe","7":"4"},{"1":"49238","2":"1550","3":"ChEA_2016","4":"http://amp.pharm.mssm.edu/Enrichr","5":"645","6":"af271913344aa08e6a755af1d433ef15768d749a","7":"7"},{"1":"2243","2":"19","3":"huMAP","4":"http://proteincomplexes.org/","5":"995","6":"249247d2f686d3eb4b9e4eb976c51159fac80a89","7":"2"},{"1":"19586","2":"545","3":"Jensen_TISSUES","4":"http://tissues.jensenlab.org/","5":"1842","6":"e8879ab9534794721614d78fe2883e9e564d7759","7":"3"},{"1":"22440","2":"505","3":"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"1302","6":"f0752e4d7f5198f86446678966b260c530d19d78","7":"8"},{"1":"8184","2":"24","3":"MGI_Mammalian_Phenotype_2017","4":"http://www.informatics.jax.org/","5":"5231","6":"0705e59bff98deda6e9cbe00cfcdd871c85e7d04","7":"7"},{"1":"18329","2":"161","3":"Jensen_COMPARTMENTS","4":"http://compartments.jensenlab.org/","5":"2283","6":"56ec68c32d4e83edc2ee83bea0e9f6a3829b2279","7":"3"},{"1":"15755","2":"28","3":"Jensen_DISEASES","4":"http://diseases.jensenlab.org/","5":"1811","6":"3045dff8181367c1421627bb8e4c5a32c6d67f98","7":"3"},{"1":"10271","2":"22","3":"BioPlex_2017","4":"http://bioplex.hms.harvard.edu/","5":"3915","6":"b8620b1a9d0d271d1a2747d8cfc63589dba39991","7":"2"},{"1":"10427","2":"38","3":"GO_Cellular_Component_2017","4":"http://www.geneontology.org/","5":"636","6":"8fed21d22dfcc3015c05b31d942fdfc851cc8e04","7":"7"},{"1":"10601","2":"25","3":"GO_Molecular_Function_2017","4":"http://www.geneontology.org/","5":"972","6":"b4018906e0a8b4e81a1b1afc51e0a2e7655403eb","7":"7"},{"1":"13822","2":"21","3":"GO_Biological_Process_2017","4":"http://www.geneontology.org/","5":"3166","6":"d9da4dba4a3eb84d4a28a3835c06dfbbe5811f92","7":"7"},{"1":"8002","2":"143","3":"GO_Cellular_Component_2017b","4":"http://www.geneontology.org/","5":"816","6":"ecf39c41fa5bc7deb625a2b5761a708676e9db7c","7":"7"},{"1":"10089","2":"45","3":"GO_Molecular_Function_2017b","4":"http://www.geneontology.org/","5":"3271","6":"8d8340361dd36a458f1f0a401f1a3141de1f3200","7":"7"},{"1":"13247","2":"49","3":"GO_Biological_Process_2017b","4":"http://www.geneontology.org/","5":"10125","6":"6404c38bffc2b3732de4e3fbe417b5043009fe34","7":"7"},{"1":"21809","2":"2316","3":"ARCHS4_Tissues","4":"http://amp.pharm.mssm.edu/archs4","5":"108","6":"4126374338235650ab158ba2c61cd2e2383b70df","7":"5"},{"1":"23601","2":"2395","3":"ARCHS4_Cell-lines","4":"http://amp.pharm.mssm.edu/archs4","5":"125","6":"5496ef9c9ae9429184d0b9485c23ba468ee522a8","7":"5"},{"1":"20883","2":"299","3":"ARCHS4_IDG_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"352","6":"ce60be284fdd5a9fc6240a355421a9e12b1ee84a","7":"4"},{"1":"19612","2":"299","3":"ARCHS4_Kinases_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"498","6":"6721c5ed97b7772e4a19fdc3f797110df0164b75","7":"2"},{"1":"25983","2":"299","3":"ARCHS4_TFs_Coexp","4":"http://amp.pharm.mssm.edu/archs4","5":"1724","6":"8a468c3ae29fa68724f744cbef018f4f3b61c5ab","7":"1"},{"1":"19500","2":"137","3":"SysMyo_Muscle_Gene_Sets","4":"http://sys-myo.rhcloud.com/","5":"1135","6":"","7":"8"},{"1":"14893","2":"128","3":"miRTarBase_2017","4":"http://mirtarbase.mbc.nctu.edu.tw/","5":"3240","6":"6b7c7fe2a97b19aecbfba12d8644af6875ad99c4","7":"1"},{"1":"17598","2":"1208","3":"TargetScan_microRNA_2017","4":"http://www.targetscan.org/","5":"683","6":"79d13fb03d2fa6403f9be45c90eeda0f6822e269","7":"1"},{"1":"5902","2":"109","3":"Enrichr_Libraries_Most_Popular_Genes","4":"http://amp.pharm.mssm.edu/Enrichr","5":"121","6":"e9b7d8ee237d0a690bd79d970a23a9fa849901ed","7":"6"},{"1":"12486","2":"299","3":"Enrichr_Submissions_TF-Gene_Coocurrence","4":"http://amp.pharm.mssm.edu/Enrichr","5":"1722","6":"be2ca8ef5a8c8e17d7e7bd290e7cbfe0951396c0","7":"1"},{"1":"1073","2":"100","3":"Data_Acquisition_Method_Most_Popular_Genes","4":"http://amp.pharm.mssm.edu/Enrichr","5":"12","6":"17ce5192b9eba7d109b6d228772ea8ab222e01ef","7":"6"},{"1":"19513","2":"117","3":"DSigDB","4":"http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","5":"4026","6":"287476538ab98337dbe727b3985a436feb6d192a","7":"4"},{"1":"14433","2":"36","3":"GO_Biological_Process_2018","4":"http://www.geneontology.org/","5":"5103","6":"b5b77681c46ac58cd050e60bcd4ad5041a9ab0a9","7":"7"},{"1":"8655","2":"61","3":"GO_Cellular_Component_2018","4":"http://www.geneontology.org/","5":"446","6":"e9ebe46188efacbe1056d82987ff1c70218fa7ae","7":"7"},{"1":"11459","2":"39","3":"GO_Molecular_Function_2018","4":"http://www.geneontology.org/","5":"1151","6":"79ff80ae9a69dd00796e52569e41422466fa0bee","7":"7"},{"1":"19741","2":"270","3":"TF_Perturbations_Followed_by_Expression","4":"http://www.ncbi.nlm.nih.gov/geo/","5":"1958","6":"34d08a4878c19584aaf180377f2ea96faa6a6eb1","7":"1"},{"1":"27360","2":"802","3":"Chromosome_Location_hg19","4":"http://hgdownload.cse.ucsc.edu/downloads.html","5":"36","6":"fdab39c467ba6b0fb0288df1176d7dfddd7196d5","7":"6"},{"1":"13072","2":"26","3":"NIH_Funded_PIs_2017_Human_GeneRIF","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"5687","6":"859b100fac3ca774ad84450b1fbb65a78fcc6b12","7":"6"},{"1":"13464","2":"45","3":"NIH_Funded_PIs_2017_Human_AutoRIF","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"12558","6":"fc5bf033b932cf173633e783fc8c6228114211f8","7":"6"},{"1":"13787","2":"200","3":"Rare_Diseases_AutoRIF_ARCHS4_Predictions","4":"https://amp.pharm.mssm.edu/geneshot/","5":"3725","6":"375ff8cdd64275a916fa24707a67968a910329bb","7":"4"},{"1":"13929","2":"200","3":"Rare_Diseases_GeneRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/gene/about-generif","5":"2244","6":"0f7fb7f347534779ecc6c87498e96b5460a8d652","7":"4"},{"1":"16964","2":"200","3":"NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"12558","6":"f77de51aaf0979dd6f56381cf67ba399b4640d28","7":"6"},{"1":"17258","2":"200","3":"NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions","4":"https://www.ncbi.nlm.nih.gov/pubmed/","5":"5684","6":"25fa899b715cd6a9137f6656499f89cd25144029","7":"6"},{"1":"10352","2":"58","3":"Rare_Diseases_GeneRIF_Gene_Lists","4":"https://www.ncbi.nlm.nih.gov/gene/about-generif","5":"2244","6":"0fb9ac92dbe52024661c088f71a1134f00567a8b","7":"4"},{"1":"10471","2":"76","3":"Rare_Diseases_AutoRIF_Gene_Lists","4":"https://amp.pharm.mssm.edu/geneshot/","5":"3725","6":"ee3adbac2da389959410260b280e7df1fd3730df","7":"4"},{"1":"12419","2":"491","3":"SubCell_BarCode","4":"http://www.subcellbarcode.org/","5":"104","6":"b50bb9480d8a77103fb75b331fd9dd927246939a","7":"2"},{"1":"19378","2":"37","3":"GWAS_Catalog_2019","4":"https://www.ebi.ac.uk/gwas","5":"1737","6":"fef3864bcb5dd9e60cee27357eff30226116c49b","7":"7"},{"1":"6201","2":"45","3":"WikiPathways_2019_Human","4":"https://www.wikipathways.org/","5":"472","6":"b0c9e9ebb9014f14561e896008087725a2db24b7","7":"7"},{"1":"4558","2":"54","3":"WikiPathways_2019_Mouse","4":"https://www.wikipathways.org/","5":"176","6":"e7750958da20f585c8b6d5bc4451a5a4305514ba","7":"7"},{"1":"3264","2":"22","3":"TRRUST_Transcription_Factors_2019","4":"https://www.grnpedia.org/trrust/","5":"571","6":"5f8cf93e193d2bcefa5a37ccdf0eefac576861b0","7":"1"},{"1":"7802","2":"92","3":"KEGG_2019_Human","4":"https://www.kegg.jp/","5":"308","6":"3477bc578c4ea5d851dcb934fe2a41e9fd789bb4","7":"7"},{"1":"8551","2":"98","3":"KEGG_2019_Mouse","4":"https://www.kegg.jp/","5":"303","6":"187eb44b2d6fa154ebf628eba1f18537f64e797c","7":"7"},{"1":"12444","2":"23","3":"InterPro_Domains_2019","4":"https://www.ebi.ac.uk/interpro/","5":"1071","6":"18dd5ec520fdf589a93d6a7911289c205e1ddf22","7":"6"},{"1":"9000","2":"20","3":"Pfam_Domains_2019","4":"https://pfam.xfam.org/","5":"608","6":"a6325ed264f9ac9e6518796076c46a1d885cca7a","7":"6"},{"1":"7744","2":"363","3":"DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","4":"https://depmap.org/","5":"558","6":"0b08b32b20854ac8a738458728a9ea50c2e04800","7":"7"},{"1":"6204","2":"387","3":"DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","4":"https://depmap.org/","5":"325","6":"b7c4ead26d0eb64f1697c030d31682b581c8bb56","7":"4"},{"1":"13420","2":"32","3":"MGI_Mammalian_Phenotype_Level_4_2019","4":"http://www.informatics.jax.org/","5":"5261","6":"f1bed632e89ebc054da44236c4815cdce03ef5ee","7":"7"},{"1":"14148","2":"122","3":"UK_Biobank_GWAS_v1","4":"https://www.ukbiobank.ac.uk/tag/gwas/","5":"857","6":"958fb52e6215626673a5acf6e9289a1b84d11b4a","7":"4"},{"1":"9813","2":"49","3":"BioPlanet_2019","4":"https://tripod.nih.gov/bioplanet/","5":"1510","6":"e110851dfc763d30946f2abedcc2cd571ac357a0","7":"2"},{"1":"1397","2":"13","3":"ClinVar_2019","4":"https://www.ncbi.nlm.nih.gov/clinvar/","5":"182","6":"0a95303f8059bec08836ecfe02ce3da951150547","7":"4"},{"1":"9116","2":"22","3":"PheWeb_2019","4":"http://pheweb.sph.umich.edu/","5":"1161","6":"6a7c7321b6b72c5285b722f7902d26a2611117cb","7":"4"},{"1":"17464","2":"63","3":"DisGeNET","4":"https://www.disgenet.org","5":"9828","6":"3c261626478ce9e6bf2c7f0a8014c5e901d43dc0","7":"4"},{"1":"394","2":"73","3":"HMS_LINCS_KinomeScan","4":"http://lincs.hms.harvard.edu/kinomescan/","5":"148","6":"47ba06cdc92469ac79400fc57acd84ba343ba616","7":"2"},{"1":"11851","2":"586","3":"CCLE_Proteomics_2020","4":"https://portals.broadinstitute.org/ccle","5":"378","6":"7094b097ae2301a1d6a5bd856a193b084cca993d","7":"5"},{"1":"8189","2":"421","3":"ProteomicsDB_2020","4":"https://www.proteomicsdb.org/","5":"913","6":"8c87c8346167bac2ba68195a32458aba9b1acfd1","7":"5"},{"1":"18704","2":"100","3":"lncHUB_lncRNA_Co-Expression","4":"https://amp.pharm.mssm.edu/lnchub/","5":"3729","6":"45b597d7efa5693b7e4172b09c0ed2dda3305582","7":"1"},{"1":"5605","2":"39","3":"Virus-Host_PPI_P-HIPSTer_2020","4":"http://phipster.org/","5":"6715","6":"a592eed13e8e9496aedbab63003b965574e46a65","7":"2"},{"1":"5718","2":"31","3":"Elsevier_Pathway_Collection","4":"http://www.transgene.ru/disease-pathways/","5":"1721","6":"9196c760e3bcae9c9de1e3f87ad81f96bde24325","7":"2"},{"1":"14156","2":"40","3":"Table_Mining_of_CRISPR_Studies","4":"","5":"802","6":"ad580f3864fa8ff69eaca11f6d2e7f9b86378d08","7":"6"},{"1":"16979","2":"295","3":"COVID-19_Related_Gene_Sets","4":"https://amp.pharm.mssm.edu/covid19","5":"205","6":"72b0346849570f66a77a6856722601e711596cb4","7":"7"},{"1":"4383","2":"146","3":"MSigDB_Hallmark_2020","4":"https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp","5":"50","6":"6952efda94663d4bd8db09bf6eeb4e67d21ef58c","7":"2"},{"1":"54974","2":"483","3":"Enrichr_Users_Contributed_Lists_2020","4":"https://maayanlab.cloud/Enrichr","5":"1482","6":"8dc362703b38b30ac3b68b6401a9b20a58e7d3ef","7":"6"},{"1":"12118","2":"448","3":"TG_GATES_2020","4":"https://toxico.nibiohn.go.jp/english/","5":"1190","6":"9e32560437b11b4628b00ccf3d584360f7f7daee","7":"4"},{"1":"12361","2":"124","3":"Allen_Brain_Atlas_10x_scRNA_2021","4":"https://portal.brain-map.org/","5":"766","6":"46f8235cb585829331799a71aec3f7c082170219","7":"5"},{"1":"9763","2":"139","3":"Descartes_Cell_Types_and_Tissue_2021","4":"https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/","5":"172","6":"","7":"5"},{"1":"8078","2":"102","3":"KEGG_2021_Human","4":"https://www.kegg.jp/","5":"320","6":"","7":"2"},{"1":"7173","2":"43","3":"WikiPathway_2021_Human","4":"https://www.wikipathways.org/","5":"622","6":"","7":"7"},{"1":"5833","2":"100","3":"HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","4":"https://hubmapconsortium.github.io/ccf-asct-reporter/","5":"344","6":"","7":"5"},{"1":"14937","2":"33","3":"GO_Biological_Process_2021","4":"http://www.geneontology.org/","5":"6036","6":"","7":"7"},{"1":"11497","2":"80","3":"GO_Cellular_Component_2021","4":"http://www.geneontology.org/","5":"511","6":"","7":"7"},{"1":"11936","2":"34","3":"GO_Molecular_Function_2021","4":"http://www.geneontology.org/","5":"1274","6":"","7":"7"},{"1":"9767","2":"33","3":"MGI_Mammalian_Phenotype_Level_4_2021","4":"http://www.informatics.jax.org/","5":"4601","6":"","7":"7"},{"1":"14167","2":"80","3":"CellMarker_Augmented_2021","4":"http://biocc.hrbmu.edu.cn/CellMarker/","5":"1097","6":"","7":"5"},{"1":"17851","2":"102","3":"Orphanet_Augmented_2021","4":"http://www.orphadata.org/","5":"3774","6":"","7":"4"},{"1":"16853","2":"360","3":"COVID-19_Related_Gene_Sets_2021","4":"https://maayanlab.cloud/covid19/","5":"478","6":"","7":"4"},{"1":"6654","2":"136","3":"PanglaoDB_Augmented_2021","4":"https://panglaodb.se/","5":"178","6":"","7":"5"},{"1":"1683","2":"10","3":"Azimuth_Cell_Types_2021","4":"https://azimuth.hubmapconsortium.org/","5":"341","6":"","7":"5"},{"1":"20414","2":"112","3":"PhenGenI_Association_2021","4":"https://www.ncbi.nlm.nih.gov/gap/phegeni","5":"950","6":"","7":"4"},{"1":"26076","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Human_Down","4":"https://maayanlab.cloud/archs4/","5":"4269","6":"","7":"8"},{"1":"26338","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Human_Up","4":"https://maayanlab.cloud/archs4/","5":"4269","6":"","7":"8"},{"1":"25381","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Mouse_Down","4":"https://maayanlab.cloud/archs4/","5":"4216","6":"","7":"8"},{"1":"25409","2":"250","3":"RNAseq_Automatic_GEO_Signatures_Mouse_Up","4":"https://maayanlab.cloud/archs4/","5":"4216","6":"","7":"8"},{"1":"11980","2":"250","3":"GTEx_Aging_Signatures_2021","4":"https://gtexportal.org/","5":"270","6":"","7":"4"},{"1":"31158","2":"805","3":"HDSigDB_Human_2021","4":"https://www.hdinhd.org/","5":"2564","6":"","7":"4"},{"1":"30006","2":"815","3":"HDSigDB_Mouse_2021","4":"https://www.hdinhd.org/","5":"2579","6":"","7":"4"},{"1":"13366","2":"103","3":"HuBMAP_ASCTplusB_augmented_2022","4":"https://hubmapconsortium.github.io/ccf-asct-reporter/","5":"777","6":"","7":"5"},{"1":"13697","2":"343","3":"FANTOM6_lncRNA_KD_DEGs","4":"https://fantom.gsc.riken.jp/6/","5":"350","6":"","7":"1"},{"1":"2183","2":"18","3":"MAGMA_Drugs_and_Diseases","4":"https://github.com/nybell/drugsets/tree/main/DATA/GENESETS","5":"1395","6":"","7":"4"},{"1":"12765","2":"13","3":"PFOCR_Pathways","4":"https://pfocr.wikipathways.org/","5":"17326","6":"","7":"7"},{"1":"1509","2":"100","3":"Tabula_Sapiens","4":"https://tabula-sapiens-portal.ds.czbiohub.org/","5":"469","6":"","7":"5"},{"1":"18365","2":"1214","3":"ChEA_2022","4":"https://maayanlab.cloud/chea3/","5":"757","6":"","7":"1"},{"1":"13525","2":"175","3":"Diabetes_Perturbations_GEO_2022","4":"https://appyters.maayanlab.cloud/#/Gene_Expression_T2D_Signatures","5":"601","6":"","7":"4"},{"1":"9525","2":"245","3":"LINCS_L1000_Chem_Pert_Consensus_Sigs","4":"https://maayanlab.cloud/sigcom-lincs/#/Download","5":"10850","6":"","7":"4"},{"1":"9440","2":"245","3":"LINCS_L1000_CRISPR_KO_Consensus_Sigs","4":"https://maayanlab.cloud/sigcom-lincs/#/Download","5":"10424","6":"","7":"4"},{"1":"3857","2":"80","3":"Tabula_Muris","4":"https://tabula-muris.ds.czbiohub.org/","5":"106","6":"","7":"5"},{"1":"10489","2":"61","3":"Reactome_2022","4":"https://reactome.org/download-data","5":"1818","6":"","7":"7"},{"1":"1198","2":"23","3":"SynGO_2022","4":"https://www.syngoportal.org/","5":"118","6":"","7":"3"},{"1":"1882","2":"47","3":"GlyGen_Glycosylated_Proteins_2022","4":"https://www.glygen.org/","5":"338","6":"","7":"2"},{"1":"1552","2":"16","3":"IDG_Drug_Targets_2022","4":"https://drugcentral.org/","5":"888","6":"","7":"4"},{"1":"6713","2":"68","3":"KOMP2_Mouse_Phenotypes_2022","4":"https://www.mousephenotype.org/","5":"529","6":"","7":"3"},{"1":"936","2":"15","3":"Metabolomics_Workbench_Metabolites_2022","4":"https://www.metabolomicsworkbench.org/","5":"233","6":"","7":"2"},{"1":"8220","2":"146","3":"Proteomics_Drug_Atlas_2023","4":"https://www.nature.com/articles/s41587-022-01539-0","5":"1748","6":"","7":"4"},{"1":"9021","2":"793","3":"The_Kinase_Library_2023","4":"https://kinase-library.phosphosite.org/site","5":"303","6":"","7":"7"},{"1":"8076","2":"96","3":"GTEx_Tissues_V8_2023","4":"https://gtexportal.org/home/","5":"511","6":"","7":"5"},{"1":"14698","2":"33","3":"GO_Biological_Process_2023","4":"http://www.geneontology.org/","5":"5407","6":"","7":"3"},{"1":"10972","2":"85","3":"GO_Cellular_Component_2023","4":"http://www.geneontology.org/","5":"474","6":"","7":"3"},{"1":"12126","2":"38","3":"GO_Molecular_Function_2023","4":"http://www.geneontology.org/","5":"1147","6":"","7":"3"},{"1":"13662","2":"12","3":"PFOCR_Pathways_2023","4":"https://pfocr.wikipathways.org/","5":"21845","6":"","7":"2"},{"1":"18290","2":"34","3":"GWAS_Catalog_2023","4":"https://www.ebi.ac.uk/gwas","5":"5271","6":"","7":"4"},{"1":"12081","2":"50","3":"GeDiPNet_2023","4":"http://gedipnet.bicnirrh.res.in/","5":"2388","6":"","7":"4"},{"1":"12853","2":"485","3":"MAGNET_2023","4":"https://magnet-winterlab.herokuapp.com/","5":"72","6":"","7":"5"},{"1":"3712","2":"9","3":"Azimuth_2023","4":"https://azimuth.hubmapconsortium.org/","5":"1425","6":"","7":"5"},{"1":"19178","2":"218","3":"Rummagene_kinases","4":"https://rummagene.com/","5":"3810","6":"","7":"2"},{"1":"19434","2":"369","3":"Rummagene_signatures","4":"https://rummagene.com/","5":"11130","6":"","7":"8"},{"1":"19379","2":"250","3":"Rummagene_transcription_factors","4":"https://rummagene.com/","5":"7244","6":"","7":"1"},{"1":"10428","2":"115","3":"MoTrPAC_2023","4":"https://motrpac-data.org/","5":"225","6":"","7":"4"},{"1":"8044","2":"42","3":"WikiPathway_2023_Human","4":"https://www.wikipathways.org/","5":"801","6":"","7":"7"},{"1":"9711","2":"414","3":"DepMap_CRISPR_GeneDependency_CellLines_2023","4":"https://depmap.org/","5":"1093","6":"","7":"4"},{"1":"1555","2":"27","3":"SynGO_2024","4":"https://www.syngoportal.org/","5":"134","6":"","7":"3"},{"1":"12642","2":"30","3":"CellMarker_2024","4":"http://bio-bigdata.hrbmu.edu.cn/CellMarker/","5":"1692","6":"","7":"5"},{"1":"2513","2":"20","3":"DGIdb_Drug_Targets_2024","4":"https://dgidb.org/","5":"659","6":"","7":"4"},{"1":"11625","2":"36","3":"MGI_Mammalian_Phenotype_Level_4_2024","4":"http://www.informatics.jax.org/","5":"5003","6":"","7":"3"},{"1":"4494","2":"51","3":"WikiPathways_2024_Mouse","4":"https://www.wikipathways.org/","5":"188","6":"","7":"2"},{"1":"8281","2":"43","3":"WikiPathways_2024_Human","4":"https://www.wikipathways.org/","5":"829","6":"","7":"2"},{"1":"9122","2":"627","3":"The_Kinase_Library_2024","4":"https://kinase-library.phosphosite.org/site","5":"392","6":"","7":"2"},{"1":"16971","2":"76","3":"PerturbAtlas","4":"https://perturbatlas.kratoss.site/#/","5":"12841","6":"","7":"4"},{"1":"11671","2":"62","3":"Reactome_Pathways_2024","4":"https://reactome.org/download-data","5":"2105","6":"","7":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform enrichment</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>top_DGE <span class="ot">&lt;-</span> DGE_cell_selection<span class="sc">$</span>Covid[(DGE_cell_selection<span class="sc">$</span>Covid<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">&amp;</span> (<span class="fu">abs</span>(DGE_cell_selection<span class="sc">$</span>Covid[, <span class="fu">grep</span>(<span class="st">"logFC.C"</span>, <span class="fu">colnames</span>(DGE_cell_selection<span class="sc">$</span>Covid))]) <span class="sc">&gt;</span> <span class="fl">0.25</span>), ]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>enrich_results <span class="ot">&lt;-</span> <span class="fu">enrichr</span>(</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes     =</span> <span class="fu">rownames</span>(top_DGE),</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">databases =</span> <span class="st">"GO_Biological_Process_2017b"</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>)[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Uploading data to Enrichr... Done.
  Querying GO_Biological_Process_2017b... Done.
Parsing results... Done.</code></pre>
</div>
</div>
<p>Some databases of interest:<br>
<code>GO_Biological_Process_2017b</code><code>KEGG_2019_Human</code><code>KEGG_2019_Mouse</code><code>WikiPathways_2019_Human</code><code>WikiPathways_2019_Mouse</code><br>
You visualize your results using a simple barplot, for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">25</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">barplot</span>(</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">height =</span> <span class="sc">-</span><span class="fu">log10</span>(enrich_results<span class="sc">$</span>P.value)[<span class="dv">10</span><span class="sc">:</span><span class="dv">1</span>],</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> enrich_results<span class="sc">$</span>Term[<span class="dv">10</span><span class="sc">:</span><span class="dv">1</span>],</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">horiz =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">border =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex.names =</span> .<span class="dv">6</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>)), <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">0</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-13-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="meta-dge_gsea" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-dge_gsea"><span class="header-section-number">5</span> Gene Set Enrichment Analysis (GSEA)</h2>
<p>Besides the enrichment using hypergeometric test, we can also perform gene set enrichment analysis (GSEA), which scores ranked genes list (usually based on fold changes) and computes permutation test to check if a particular gene set is more present in the Up-regulated genes, among the DOWN_regulated genes or not differentially regulated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a gene rank based on the gene expression fold change</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(DGE_cell_selection<span class="sc">$</span>Covid[, <span class="fu">grep</span>(<span class="st">"logFC.C"</span>, <span class="fu">colnames</span>(DGE_cell_selection<span class="sc">$</span>Covid))], <span class="fu">casefold</span>(<span class="fu">rownames</span>(DGE_cell_selection<span class="sc">$</span>Covid), <span class="at">upper =</span> T))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once our list of genes are sorted, we can proceed with the enrichment itself. We can use the package to get gene set from the Molecular Signature Database (MSigDB) and select KEGG pathways as an example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msigdbr)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Download gene sets</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>msigdbgmt <span class="ot">&lt;-</span> msigdbr<span class="sc">::</span><span class="fu">msigdbr</span>(<span class="st">"Homo sapiens"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>msigdbgmt <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(msigdbgmt)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># List available gene sets</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(msigdbgmt<span class="sc">$</span>gs_subcat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "MIR:MIR_Legacy"  "TFT:TFT_Legacy"  "CGP"             "TFT:GTRD"       
 [5] ""                "VAX"             "CP:BIOCARTA"     "CGN"            
 [9] "GO:BP"           "GO:CC"           "IMMUNESIGDB"     "GO:MF"          
[13] "HPO"             "CP:KEGG"         "MIR:MIRDB"       "CM"             
[17] "CP"              "CP:PID"          "CP:REACTOME"     "CP:WIKIPATHWAYS"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset which gene set you want to use.</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>msigdbgmt_subset <span class="ot">&lt;-</span> msigdbgmt[msigdbgmt<span class="sc">$</span>gs_subcat <span class="sc">==</span> <span class="st">"CP:WIKIPATHWAYS"</span>, ]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>gmt <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(msigdbgmt_subset<span class="sc">$</span>gs_name), <span class="cf">function</span>(x) {</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    msigdbgmt_subset[msigdbgmt_subset<span class="sc">$</span>gs_name <span class="sc">==</span> x, <span class="st">"gene_symbol"</span>]</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gmt) <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">paste0</span>(msigdbgmt_subset<span class="sc">$</span>gs_name, <span class="st">"_"</span>, msigdbgmt_subset<span class="sc">$</span>gs_exact_source))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we will run GSEA. This will result in a table containing information for several pathways. We can then sort and filter those pathways to visualize only the top ones. You can select/filter them by either <code>p-value</code> or normalized enrichment score (<code>NES</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fgsea)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform enrichemnt analysis</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(<span class="at">pathways =</span> gmt, <span class="at">stats =</span> gene_rank, <span class="at">minSize =</span> <span class="dv">15</span>, <span class="at">maxSize =</span> <span class="dv">500</span>, <span class="at">nperm =</span> <span class="dv">10000</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>fgseaRes <span class="ot">&lt;-</span> fgseaRes[<span class="fu">order</span>(fgseaRes<span class="sc">$</span>NES, <span class="at">decreasing =</span> T), ]</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the results table to show only the top 10 UP or DOWN regulated processes (optional)</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>top10_UP <span class="ot">&lt;-</span> fgseaRes<span class="sc">$</span>pathway[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Nice summary table (shown as a plot)</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGseaTable</span>(gmt[top10_UP], gene_rank, fgseaRes, <span class="at">gseaParam =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_05_dge_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="bioc_05_dge_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which KEGG pathways are upregulated in this cluster? Which KEGG pathways are dowregulated in this cluster? Change the pathway source to another gene set (e.g.&nbsp;CP:WIKIPATHWAYS or CP:REACTOME or CP:BIOCARTA or GO:BP) and check the if you get similar results?</p>
</div>
</div>
<p>Finally, let’s save the integrated data for further analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sce, <span class="st">"data/covid/results/bioc_covid_qc_dr_int_cl_dge.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-session" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">6</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/asabjor/miniconda3/envs/seurat5_u/lib/libopenblasp-r0.3.28.dylib;  LAPACK version 3.12.0

locale:
[1] sv_SE.UTF-8/sv_SE.UTF-8/sv_SE.UTF-8/C/sv_SE.UTF-8/sv_SE.UTF-8

time zone: Europe/Stockholm
tzcode source: system (macOS)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] fgsea_1.28.0                msigdbr_7.5.1              
 [3] enrichR_3.2                 dplyr_1.1.4                
 [5] igraph_2.0.3                pheatmap_1.0.12            
 [7] patchwork_1.3.0             scran_1.30.0               
 [9] scater_1.30.1               ggplot2_3.5.1              
[11] scuttle_1.12.0              SingleCellExperiment_1.24.0
[13] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[15] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
[17] IRanges_2.36.0              S4Vectors_0.40.2           
[19] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[21] matrixStats_1.5.0          

loaded via a namespace (and not attached):
 [1] bitops_1.0-9              gridExtra_2.3            
 [3] rlang_1.1.4               magrittr_2.0.3           
 [5] compiler_4.3.3            DelayedMatrixStats_1.24.0
 [7] vctrs_0.6.5               pkgconfig_2.0.3          
 [9] crayon_1.5.3              fastmap_1.2.0            
[11] XVector_0.42.0            labeling_0.4.3           
[13] rmarkdown_2.29            ggbeeswarm_0.7.2         
[15] xfun_0.50                 bluster_1.12.0           
[17] WriteXLS_6.7.0            zlibbioc_1.48.0          
[19] beachmat_2.18.0           jsonlite_1.8.9           
[21] DelayedArray_0.28.0       BiocParallel_1.36.0      
[23] irlba_2.3.5.1             parallel_4.3.3           
[25] cluster_2.1.8             R6_2.5.1                 
[27] RColorBrewer_1.1-3        limma_3.58.1             
[29] Rcpp_1.0.13-1             knitr_1.49               
[31] Matrix_1.6-5              tidyselect_1.2.1         
[33] abind_1.4-5               yaml_2.3.10              
[35] viridis_0.6.5             codetools_0.2-20         
[37] curl_6.0.1                lattice_0.22-6           
[39] tibble_3.2.1              withr_3.0.2              
[41] evaluate_1.0.1            pillar_1.10.1            
[43] generics_0.1.3            RCurl_1.98-1.16          
[45] sparseMatrixStats_1.14.0  munsell_0.5.1            
[47] scales_1.3.0              glue_1.8.0               
[49] metapod_1.10.0            tools_4.3.3              
[51] BiocNeighbors_1.20.0      data.table_1.15.4        
[53] ScaledMatrix_1.10.0       locfit_1.5-9.10          
[55] babelgene_22.9            fastmatch_1.1-6          
[57] cowplot_1.1.3             grid_4.3.3               
[59] edgeR_4.0.16              colorspace_2.1-1         
[61] GenomeInfoDbData_1.2.11   beeswarm_0.4.0           
[63] BiocSingular_1.18.0       vipor_0.4.7              
[65] cli_3.6.3                 rsvd_1.0.5               
[67] S4Arrays_1.2.0            viridisLite_0.4.2        
[69] gtable_0.3.6              digest_0.6.37            
[71] SparseArray_1.2.2         ggrepel_0.9.6            
[73] dqrng_0.3.2               rjson_0.2.23             
[75] htmlwidgets_1.6.4         farver_2.1.2             
[77] htmltools_0.5.8.1         lifecycle_1.0.4          
[79] httr_1.4.7                statmod_1.5.0            </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"openEffect":"zoom","selector":".lightbox","closeEffect":"zoom","loop":true,"descPosition":"bottom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>