<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Reconstructing developmental or differentiation pathways from individual cell gene expression profiles to understand cellular transitions and relationships.">

<title> Trajectory inference using PAGA</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-route" aria-label="route"></i> Trajectory inference using PAGA</h1>
            <p class="subtitle lead"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Reconstructing developmental or differentiation pathways from individual cell gene expression profiles to understand cellular transitions and relationships.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">07-Feb-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#loading-libraries" id="toc-loading-libraries" class="nav-link active" data-scroll-target="#loading-libraries"><span class="header-section-number">1</span> Loading libraries</a></li>
  <li><a href="#preparing-data" id="toc-preparing-data" class="nav-link" data-scroll-target="#preparing-data"><span class="header-section-number">2</span> Preparing data</a></li>
  <li><a href="#reading-data" id="toc-reading-data" class="nav-link" data-scroll-target="#reading-data"><span class="header-section-number">3</span> Reading data</a></li>
  <li><a href="#explore-the-data" id="toc-explore-the-data" class="nav-link" data-scroll-target="#explore-the-data"><span class="header-section-number">4</span> Explore the data</a></li>
  <li><a href="#rerun-analysis-in-scanpy" id="toc-rerun-analysis-in-scanpy" class="nav-link" data-scroll-target="#rerun-analysis-in-scanpy"><span class="header-section-number">5</span> Rerun analysis in Scanpy</a></li>
  <li><a href="#run-paga" id="toc-run-paga" class="nav-link" data-scroll-target="#run-paga"><span class="header-section-number">6</span> Run PAGA</a></li>
  <li><a href="#filtering-graph-edges" id="toc-filtering-graph-edges" class="nav-link" data-scroll-target="#filtering-graph-edges"><span class="header-section-number">7</span> Filtering graph edges</a>
  <ul>
  <li><a href="#rerun-paga-again-on-the-data" id="toc-rerun-paga-again-on-the-data" class="nav-link" data-scroll-target="#rerun-paga-again-on-the-data"><span class="header-section-number">7.1</span> Rerun PAGA again on the data</a></li>
  </ul></li>
  <li><a href="#embedding-using-paga-initialization" id="toc-embedding-using-paga-initialization" class="nav-link" data-scroll-target="#embedding-using-paga-initialization"><span class="header-section-number">8</span> Embedding using PAGA-initialization</a></li>
  <li><a href="#gene-changes" id="toc-gene-changes" class="nav-link" data-scroll-target="#gene-changes"><span class="header-section-number">9</span> Gene changes</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">10</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run Python commands unless it starts with <code>%%bash</code>, in which case, those chunks run shell commands.</p>
</div>
</div>
<p>Partly following this PAGA <a href="https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html">tutorial</a> with some modifications.</p>
<section id="loading-libraries" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="loading-libraries"><span class="header-section-number">1</span> Loading libraries</h2>
<div id="libraries" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> pl</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> rcParams</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">Warning</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># verbosity: errors (0), warnings (1), info (2), hints (3)</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">100</span>, frameon<span class="op">=</span><span class="va">False</span>, figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">5</span>), facecolor<span class="op">=</span><span class="st">'white'</span>, color_map <span class="op">=</span> <span class="st">'viridis_r'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="preparing-data" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="preparing-data"><span class="header-section-number">2</span> Preparing data</h2>
<p>In order to speed up the computations during the exercises, we will be using a subset of a bone marrow dataset (originally containing about 100K cells). The bone marrow is the source of adult immune cells, and contains virtually all differentiation stages of cell from the immune system which later circulate in the blood to all other organs.</p>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="../figs/hematopoiesis.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="../figs/hematopoiesis.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<p>If you have been using the <strong>Seurat</strong>, <strong>Bioconductor</strong> or <strong>Scanpy</strong> toolkits with your own data, you need to reach to the point where can find get:</p>
<ul>
<li>A dimensionality reduction where to perform the trajectory (for example: PCA, ICA, MNN, harmony, Diffusion Maps, UMAP)</li>
<li>The cell clustering information (for example: from Louvain, k-means)</li>
<li>A KNN/SNN graph (this is useful to inspect and sanity-check your trajectories)</li>
</ul>
<p>In this case, all the data has been preprocessed with Seurat with standard pipelines. In addition there was some manual filtering done to remove clusters that are disconnected and cells that are hard to cluster, which can be seen in this <a href="https://github.com/NBISweden/workshop-scRNAseq/blob/master/scripts/data_processing/slingshot_preprocessing.Rmd">script</a></p>
<p>The file trajectory_scanpy_filtered.h5ad was converted from the Seurat object using the SeuratDisk package. For more information on how it was done, have a look at the script: <a href="https://github.com/NBISweden/workshop-scRNAseq/blob/master/scripts/data_processing/convert_to_h5ad.R">convert_to_h5ad.R</a> in the github repo.</p>
<p>You can download the data with the commands:</p>
<div id="fetch-data" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># download pre-computed data if missing or long compute</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>fetch_data <span class="op">=</span> <span class="va">True</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># url for source and intermediate data</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>path_data <span class="op">=</span> <span class="st">"https://nextcloud.dc.scilifelab.se/public.php/webdav"</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>curl_upass <span class="op">=</span> <span class="st">"zbC5fr2LbEZ9rSE:scRNAseq2025"</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>path_results <span class="op">=</span> <span class="st">"data/trajectory"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(path_results):</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    os.makedirs(path_results, exist_ok<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>path_file <span class="op">=</span> <span class="st">"data/trajectory/trajectory_seurat_filtered.h5ad"</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(path_file):</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    file_url <span class="op">=</span> os.path.join(path_data, <span class="st">"trajectory/trajectory_seurat_filtered.h5ad"</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    subprocess.call([<span class="st">"curl"</span>, <span class="st">"-u"</span>, curl_upass, <span class="st">"-o"</span>, path_file, file_url ])    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="reading-data" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="reading-data"><span class="header-section-number">3</span> Reading data</h2>
<p>We already have pre-computed and subsetted the dataset (with 6688 cells and 3585 genes) following the analysis steps in this course. We then saved the objects, so you can use common tools to open and start to work with them (either in R or Python).</p>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"data/trajectory/trajectory_seurat_filtered.h5ad"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>adata.var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="read-data" class="cell-output cell-output-display" data-execution_count="3">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">features</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0610040J01Rik</td>
<td>0610040J01Rik</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1190007I07Rik</td>
<td>1190007I07Rik</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1500009L16Rik</td>
<td>1500009L16Rik</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1700012B09Rik</td>
<td>1700012B09Rik</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1700020L24Rik</td>
<td>1700020L24Rik</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Sqor</td>
<td>Sqor</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Sting1</td>
<td>Sting1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Tent5a</td>
<td>Tent5a</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Tlcd4</td>
<td>Tlcd4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Znrd2</td>
<td>Znrd2</td>
</tr>
</tbody>
</table>

<p>3585 rows × 1 columns</p>
</div>
</div>
</div>
<div id="check-data" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check what you have in the X matrix, should be lognormalized counts.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.X[:<span class="dv">10</span>,:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;Compressed Sparse Row sparse matrix of dtype 'float64'
    with 25 stored elements and shape (10, 10)&gt;
  Coords    Values
  (0, 4)    0.11622072805743532
  (0, 8)    0.4800893970571722
  (1, 8)    0.2478910541698065
  (1, 9)    0.17188973970230348
  (2, 1)    0.09413397843954842
  (2, 7)    0.18016412971724202
  (3, 1)    0.08438841021254412
  (3, 4)    0.08438841021254412
  (3, 7)    0.08438841021254412
  (3, 8)    0.3648216463668793
  (4, 1)    0.14198147850903975
  (4, 8)    0.14198147850903975
  (5, 1)    0.17953169693896723
  (5, 8)    0.17953169693896723
  (5, 9)    0.17953169693896723
  (6, 4)    0.2319546390006887
  (6, 8)    0.42010741700351195
  (7, 1)    0.1775659421407816
  (7, 8)    0.39593115482156394
  (7, 9)    0.09271901219711086
  (8, 1)    0.12089079757716388
  (8, 8)    0.22873058755480363
  (9, 1)    0.08915380247493314
  (9, 4)    0.08915380247493314
  (9, 8)    0.38270398718590104</code></pre>
</div>
</div>
</section>
<section id="explore-the-data" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="explore-the-data"><span class="header-section-number">4</span> Explore the data</h2>
<p>There is a umap and clusters provided with the object, first plot some information from the previous analysis onto the umap.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> [<span class="st">'clusters'</span>,<span class="st">'dataset'</span>,<span class="st">'batches'</span>,<span class="st">'Phase'</span>],legend_loc <span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>, ncols <span class="op">=</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-umap">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-umap-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="scanpy_07_trajectory_files/figure-html/plot-umap-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<p>It is crucial that you performing analysis of a dataset understands what is going on, what are the clusters you see in your data and most importantly How are the clusters related to each other?. Well, let’s explore the data a bit. With the help of this table, write down which cluster numbers in your dataset express these key markers.</p>
<table class="table">
<thead>
<tr class="header">
<th>Marker</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Cd34</td>
<td>HSC progenitor</td>
</tr>
<tr class="even">
<td>Ms4a1</td>
<td>B cell lineage</td>
</tr>
<tr class="odd">
<td>Cd3e</td>
<td>T cell lineage</td>
</tr>
<tr class="even">
<td>Ltf</td>
<td>Granulocyte lineage</td>
</tr>
<tr class="odd">
<td>Cst3</td>
<td>Monocyte lineage</td>
</tr>
<tr class="even">
<td>Mcpt8</td>
<td>Mast Cell lineage</td>
</tr>
<tr class="odd">
<td>Alas2</td>
<td>RBC lineage</td>
</tr>
<tr class="even">
<td>Siglech</td>
<td>Dendritic cell lineage</td>
</tr>
<tr class="odd">
<td>C1qc</td>
<td>Macrophage cell lineage</td>
</tr>
<tr class="even">
<td>Pf4</td>
<td>Megakaryocyte cell lineage</td>
</tr>
</tbody>
</table>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> [<span class="st">"Cd34"</span>,<span class="st">"Alas2"</span>,<span class="st">"Pf4"</span>,<span class="st">"Mcpt8"</span>,<span class="st">"Ltf"</span>,<span class="st">"Cst3"</span>, <span class="st">"Siglech"</span>, <span class="st">"C1qc"</span>, <span class="st">"Ms4a1"</span>, <span class="st">"Cd3e"</span>, ]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> markers, use_raw <span class="op">=</span> <span class="va">False</span>, ncols <span class="op">=</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-markers">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-markers-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="scanpy_07_trajectory_files/figure-html/plot-markers-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
</section>
<section id="rerun-analysis-in-scanpy" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="rerun-analysis-in-scanpy"><span class="header-section-number">5</span> Rerun analysis in Scanpy</h2>
<p>Redo clustering and umap using the basic Scanpy pipeline. Use the provided “X_harmony_Phase” dimensionality reduction as the staring point.</p>
<div id="process" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first, store the old umap with a new name so it is not overwritten</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'X_umap_old'</span>] <span class="op">=</span> adata.obsm[<span class="st">'X_umap'</span>]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_pcs <span class="op">=</span> <span class="dv">30</span>, n_neighbors <span class="op">=</span> <span class="dv">20</span>, use_rep<span class="op">=</span><span class="st">"X_harmony_Phase"</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata, min_dist<span class="op">=</span><span class="fl">0.4</span>, spread<span class="op">=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:10)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm)
    'umap', UMAP parameters (adata.uns) (0:00:10)</code></pre>
</div>
</div>
<div id="cluster" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> [<span class="st">'clusters'</span>],legend_loc <span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>, edges <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> markers, use_raw <span class="op">=</span> <span class="va">False</span>, ncols <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Redo clustering as well</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added <span class="op">=</span> <span class="st">"leiden_1.0"</span>, resolution <span class="op">=</span> <span class="fl">1.0</span>) <span class="co"># default resolution in 1.0</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added <span class="op">=</span> <span class="st">"leiden_1.2"</span>, resolution <span class="op">=</span> <span class="fl">1.2</span>) <span class="co"># default resolution in 1.0</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added <span class="op">=</span> <span class="st">"leiden_1.4"</span>, resolution <span class="op">=</span> <span class="fl">1.4</span>) <span class="co"># default resolution in 1.0</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.tl.louvain(adata, key_added = "leiden_1.0") # default resolution in 1.0</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> [<span class="st">'leiden_1.0'</span>, <span class="st">'leiden_1.2'</span>, <span class="st">'leiden_1.4'</span>,<span class="st">'clusters'</span>],legend_loc <span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>, ncols <span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="cluster-1">
<p><a href="scanpy_07_trajectory_files/figure-html/cluster-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="scanpy_07_trajectory_files/figure-html/cluster-output-1.png" class="img-fluid"></a></p>
</div>
</div>
<div class="cell-output cell-output-display">
<div id="cluster-2">
<p><a href="scanpy_07_trajectory_files/figure-html/cluster-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="scanpy_07_trajectory_files/figure-html/cluster-output-2.png" class="img-fluid"></a></p>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>running Leiden clustering
    finished: found 16 clusters and added
    'leiden_1.0', the cluster labels (adata.obs, categorical) (0:00:00)
running Leiden clustering
    finished: found 17 clusters and added
    'leiden_1.2', the cluster labels (adata.obs, categorical) (0:00:00)
running Leiden clustering
    finished: found 19 clusters and added
    'leiden_1.4', the cluster labels (adata.obs, categorical) (0:00:00)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="cluster-3">
<p><a href="scanpy_07_trajectory_files/figure-html/cluster-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="scanpy_07_trajectory_files/figure-html/cluster-output-4.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<div id="annotate" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Rename clusters with really clear markers, the rest are left unlabelled.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>annot <span class="op">=</span> pd.DataFrame(adata.obs[<span class="st">'leiden_1.4'</span>].astype(<span class="st">'string'</span>))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'10'</span>] <span class="op">=</span> <span class="st">'10_megakaryo'</span> <span class="co">#Pf4</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'17'</span>] <span class="op">=</span> <span class="st">'17_macro'</span>  <span class="co">#C1qc</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'11'</span>] <span class="op">=</span> <span class="st">'11_eryth'</span> <span class="co">#Alas2</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'18'</span>] <span class="op">=</span> <span class="st">'18_dend'</span> <span class="co">#Siglech</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'13'</span>] <span class="op">=</span> <span class="st">'13_mast'</span> <span class="co">#Mcpt8</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'0'</span>] <span class="op">=</span> <span class="st">'0_mono'</span> <span class="co">#Cts3</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'1'</span>] <span class="op">=</span> <span class="st">'1_gran'</span> <span class="co">#Ltf</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'9'</span>] <span class="op">=</span> <span class="st">'9_gran'</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'14'</span>] <span class="op">=</span> <span class="st">'14_TC'</span> <span class="co">#Cd3e</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'16'</span>] <span class="op">=</span> <span class="st">'16_BC'</span> <span class="co">#Ms4a1</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'8'</span>] <span class="op">=</span> <span class="st">'8_progen'</span>  <span class="co"># Cd34</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'4'</span>] <span class="op">=</span> <span class="st">'4_progen'</span> </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>annot[annot[<span class="st">'leiden_1.4'</span>] <span class="op">==</span> <span class="st">'5'</span>] <span class="op">=</span> <span class="st">'5_progen'</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'annot'</span>]<span class="op">=</span>annot[<span class="st">'leiden_1.4'</span>].astype(<span class="st">'category'</span>)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color <span class="op">=</span> <span class="st">'annot'</span>,legend_loc <span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>, ncols <span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>annot.value_counts()</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="co">#type(annot)</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="co"># astype('category')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="annotate-1">
<p><a href="scanpy_07_trajectory_files/figure-html/annotate-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="scanpy_07_trajectory_files/figure-html/annotate-output-1.png" class="img-fluid"></a></p>
</div>
</div>
<div id="annotate-2" class="cell-output cell-output-display" data-execution_count="9">
<pre><code>leiden_1.4  
0_mono          509
1_gran          487
2               479
3               463
4_progen        387
5_progen        384
7               368
6               368
8_progen        367
9_gran          366
10_megakaryo    301
11_eryth        294
12              276
13_mast         159
14_TC           151
15              128
16_BC           124
17_macro        116
18_dend         101
dtype: int64</code></pre>
</div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot onto the Seurat embedding:</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata, basis<span class="op">=</span><span class="st">'X_umap_old'</span>, color <span class="op">=</span> <span class="st">'annot'</span>,legend_loc <span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>, ncols <span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-annot">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-annot-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="scanpy_07_trajectory_files/figure-html/plot-annot-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
</section>
<section id="run-paga" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="run-paga"><span class="header-section-number">6</span> Run PAGA</h2>
<p>Use the clusters from leiden clustering with leiden_1.4 and run PAGA. First we create the graph and initialize the positions using the umap.</p>
<div id="paga" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the umap to initialize the graph layout.</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sc.tl.draw_graph(adata, init_pos<span class="op">=</span><span class="st">'X_umap'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span><span class="st">'annot'</span>, legend_loc<span class="op">=</span><span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>sc.tl.paga(adata, groups<span class="op">=</span><span class="st">'annot'</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>sc.pl.paga(adata, color<span class="op">=</span><span class="st">'annot'</span>, edge_width_scale <span class="op">=</span> <span class="fl">0.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>drawing single-cell graph using layout 'fa'
WARNING: Package 'fa2-modified' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2-modified' (`pip install fa2-modified`).
    finished: added
    'X_draw_graph_fr', graph_drawing coordinates (adata.obsm) (0:00:03)
running PAGA
    finished: added
    'paga/connectivities', connectivities adjacency (adata.uns)
    'paga/connectivities_tree', connectivities subtree (adata.uns) (0:00:00)
--&gt; added 'pos', the PAGA positions (adata.uns['paga'])</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="paga-1">
<p><a href="scanpy_07_trajectory_files/figure-html/paga-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="scanpy_07_trajectory_files/figure-html/paga-output-2.png" class="img-fluid"></a></p>
</div>
</div>
<div class="cell-output cell-output-display">
<div id="paga-2">
<p><a href="scanpy_07_trajectory_files/figure-html/paga-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="scanpy_07_trajectory_files/figure-html/paga-output-3.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<p>As you can see, we have edges between many clusters that we know are are unrelated, so we may need to clean up the data a bit more.</p>
</section>
<section id="filtering-graph-edges" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="filtering-graph-edges"><span class="header-section-number">7</span> Filtering graph edges</h2>
<p>First, lets explore the graph a bit. So we plot the umap with the graph connections on top.</p>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, edges<span class="op">=</span><span class="va">True</span>, color <span class="op">=</span> <span class="st">'annot'</span>, legend_loc<span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize<span class="op">=</span> <span class="st">'xx-small'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-graph">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-graph-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="scanpy_07_trajectory_files/figure-html/plot-graph-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<p>We have many edges in the graph between unrelated clusters, so lets try with fewer neighbors.</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">5</span>,  use_rep <span class="op">=</span> <span class="st">'X_harmony_Phase'</span>, n_pcs <span class="op">=</span> <span class="dv">30</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, edges<span class="op">=</span><span class="va">True</span>, color <span class="op">=</span> <span class="st">'annot'</span>, legend_loc<span class="op">=</span> <span class="st">'on data'</span>, legend_fontsize<span class="op">=</span> <span class="st">'xx-small'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:00)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="redo-graph">
<p><a href="scanpy_07_trajectory_files/figure-html/redo-graph-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="scanpy_07_trajectory_files/figure-html/redo-graph-output-2.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<section id="rerun-paga-again-on-the-data" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="rerun-paga-again-on-the-data"><span class="header-section-number">7.1</span> Rerun PAGA again on the data</h3>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sc.tl.draw_graph(adata, init_pos<span class="op">=</span><span class="st">'X_umap'</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span><span class="st">'annot'</span>, legend_loc<span class="op">=</span><span class="st">'on data'</span>, legend_fontsize <span class="op">=</span> <span class="st">'xx-small'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>drawing single-cell graph using layout 'fa'
WARNING: Package 'fa2-modified' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2-modified' (`pip install fa2-modified`).
    finished: added
    'X_draw_graph_fr', graph_drawing coordinates (adata.obsm) (0:00:02)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="draw-graph">
<p><a href="scanpy_07_trajectory_files/figure-html/draw-graph-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="scanpy_07_trajectory_files/figure-html/draw-graph-output-2.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc.tl.paga(adata, groups<span class="op">=</span><span class="st">'annot'</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sc.pl.paga(adata, color<span class="op">=</span><span class="st">'annot'</span>, edge_width_scale <span class="op">=</span> <span class="fl">0.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>running PAGA
    finished: added
    'paga/connectivities', connectivities adjacency (adata.uns)
    'paga/connectivities_tree', connectivities subtree (adata.uns) (0:00:00)
--&gt; added 'pos', the PAGA positions (adata.uns['paga'])</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="paga2">
<p><a href="scanpy_07_trajectory_files/figure-html/paga2-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="scanpy_07_trajectory_files/figure-html/paga2-output-2.png" class="img-fluid"></a></p>
</div>
</div>
</div>
</section>
</section>
<section id="embedding-using-paga-initialization" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="embedding-using-paga-initialization"><span class="header-section-number">8</span> Embedding using PAGA-initialization</h2>
<p>We can now redraw the graph using another starting position from the paga layout. The following is just as well possible for a UMAP.</p>
<div id="draw-graph-paga" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sc.tl.draw_graph(adata, init_pos<span class="op">=</span><span class="st">'paga'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>drawing single-cell graph using layout 'fa'
WARNING: Package 'fa2-modified' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2-modified' (`pip install fa2-modified`).
    finished: added
    'X_draw_graph_fr', graph_drawing coordinates (adata.obsm) (0:00:14)</code></pre>
</div>
</div>
<p>Now we can see all marker genes also at single-cell resolution in a meaningful layout.</p>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span>[<span class="st">'annot'</span>], legend_loc<span class="op">=</span><span class="st">'on data'</span>, legend_fontsize<span class="op">=</span>  <span class="st">'xx-small'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-graph-paga">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-graph-paga-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="scanpy_07_trajectory_files/figure-html/plot-graph-paga-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<p>Compare the 2 graphs</p>
<div id="paga-compare" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sc.pl.paga_compare(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    adata, threshold<span class="op">=</span><span class="fl">0.03</span>, title<span class="op">=</span><span class="st">''</span>, right_margin<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">10</span>, edge_width_scale<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    legend_fontsize<span class="op">=</span><span class="dv">12</span>, fontsize<span class="op">=</span><span class="dv">12</span>, frameon<span class="op">=</span><span class="va">False</span>, edges<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--&gt; added 'pos', the PAGA positions (adata.uns['paga'])</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="paga-compare-1">
<p><a href="scanpy_07_trajectory_files/figure-html/paga-compare-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="scanpy_07_trajectory_files/figure-html/paga-compare-output-2.png" class="img-fluid"></a></p>
</div>
</div>
<div id="paga-compare-2" class="cell-output cell-output-display" data-execution_count="18">
<pre><code>[&lt;Axes: xlabel='UMAP1', ylabel='UMAP2'&gt;, &lt;Axes: &gt;]</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Does this graph fit the biological expectations given what you know of hematopoesis. Please have a look at the figure in Section 2 and compare to the paths you now have.</p>
</div>
</div>
</section>
<section id="gene-changes" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="gene-changes"><span class="header-section-number">9</span> Gene changes</h2>
<p>We can reconstruct gene changes along PAGA paths for a given set of genes</p>
<p>Choose a root cell for diffusion pseudotime. We have 3 progenitor clusters, but cluster 5 seems the most clear.</p>
<div id="pseudotime" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>adata.uns[<span class="st">'iroot'</span>] <span class="op">=</span> np.flatnonzero(adata.obs[<span class="st">'annot'</span>]  <span class="op">==</span> <span class="st">'5_progen'</span>)[<span class="dv">0</span>]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>sc.tl.dpt(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.
computing Diffusion Maps using n_comps=15(=n_dcs)
computing transitions
    finished (0:00:00)
    eigenvalues of transition matrix
    [1.         0.9989591  0.997628   0.9970365  0.9956704  0.99334306
     0.9918951  0.9915921  0.99013233 0.98801893 0.9870309  0.9861044
     0.9851118  0.9845008  0.9839531 ]
    finished: added
    'X_diffmap', diffmap coordinates (adata.obsm)
    'diffmap_evals', eigenvalues of transition matrix (adata.uns) (0:00:00)
computing Diffusion Pseudotime using n_dcs=10
    finished: added
    'dpt_pseudotime', the pseudotime (adata.obs) (0:00:00)</code></pre>
</div>
</div>
<p>Use the full raw data for visualization.</p>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span>[<span class="st">'annot'</span>, <span class="st">'dpt_pseudotime'</span>], legend_loc<span class="op">=</span><span class="st">'on data'</span>, legend_fontsize<span class="op">=</span> <span class="st">'x-small'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-pt">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-pt-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="scanpy_07_trajectory_files/figure-html/plot-pt-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>The pseudotime represents the distance of every cell to the starting cluster. Have a look at the pseudotime plot, how well do you think it represents actual developmental time? What does it represent?</p>
</div>
</div>
<p>By looking at the different know lineages and the layout of the graph we define manually some paths to the graph that corresponds to specific lineages.</p>
<div id="paths" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define paths</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>paths <span class="op">=</span> [(<span class="st">'erythrocytes'</span>, [<span class="st">'5_progen'</span>, <span class="st">'8_progen'</span>, <span class="st">'6'</span>, <span class="st">'3'</span>, <span class="st">'7'</span>, <span class="st">'11_eryth'</span>]),</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>         (<span class="st">'lympoid'</span>, [<span class="st">'5_progen'</span>, <span class="st">'12'</span>, <span class="st">'16_BC'</span>, <span class="st">'14_TC'</span>]),</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>         (<span class="st">'granulo'</span>, [<span class="st">'5_progen'</span>, <span class="st">'4_progen'</span>, <span class="st">'2'</span>, <span class="st">'9_gran'</span>, <span class="st">'1_gran'</span>]),</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>         (<span class="st">'mono'</span>, [<span class="st">'5_progen'</span>, <span class="st">'4_progen'</span>, <span class="st">'0_mono'</span>, <span class="st">'18_dend'</span>, <span class="st">'17_macro'</span>])</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>         ]</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'distance'</span>] <span class="op">=</span> adata.obs[<span class="st">'dpt_pseudotime'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we select some genes that can vary in the lineages and plot onto the paths.</p>
<div id="select-genes" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>gene_names <span class="op">=</span> [<span class="st">'Gata2'</span>, <span class="st">'Gata1'</span>, <span class="st">'Klf1'</span>, <span class="st">'Epor'</span>, <span class="st">'Hba-a2'</span>,  <span class="co"># erythroid</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>              <span class="st">'Elane'</span>, <span class="st">'Cebpe'</span>, <span class="st">'Gfi1'</span>,                    <span class="co"># neutrophil</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>              <span class="st">'Irf8'</span>, <span class="st">'Csf1r'</span>, <span class="st">'Ctsg'</span>,                     <span class="co"># monocyte</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">'Itga2b'</span>,<span class="st">'Prss34'</span>,<span class="st">'Cma1'</span>,<span class="st">'Procr'</span>,            <span class="co"># Megakaryo,Basophil,Mast,HPC</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>              <span class="st">'C1qc'</span>,<span class="st">'Siglech'</span>,<span class="st">'Ms4a1'</span>,<span class="st">'Cd3e'</span>,<span class="st">'Cd34'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>_, axs <span class="op">=</span> pl.subplots(ncols<span class="op">=</span><span class="dv">4</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>), gridspec_kw<span class="op">=</span>{</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">'wspace'</span>: <span class="fl">0.05</span>, <span class="st">'left'</span>: <span class="fl">0.12</span>})</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>pl.subplots_adjust(left<span class="op">=</span><span class="fl">0.05</span>, right<span class="op">=</span><span class="fl">0.98</span>, top<span class="op">=</span><span class="fl">0.82</span>, bottom<span class="op">=</span><span class="fl">0.2</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ipath, (descr, path) <span class="kw">in</span> <span class="bu">enumerate</span>(paths):</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    sc.pl.paga_path(</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>        adata<span class="op">=</span>adata, </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>        nodes<span class="op">=</span>path, </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>        keys<span class="op">=</span>gene_names,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>        show_node_names<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>        ax<span class="op">=</span>axs[ipath],</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>        ytick_fontsize<span class="op">=</span><span class="dv">12</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>        left_margin<span class="op">=</span><span class="fl">0.15</span>,</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>        n_avg<span class="op">=</span><span class="dv">50</span>,</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>        annotations<span class="op">=</span>[<span class="st">'distance'</span>],</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>        show_yticks<span class="op">=</span><span class="va">True</span> <span class="cf">if</span> ipath <span class="op">==</span> <span class="dv">0</span> <span class="cf">else</span> <span class="va">False</span>,</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>        show_colorbar<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>        color_map<span class="op">=</span><span class="st">'Greys'</span>,</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>        groups_key<span class="op">=</span><span class="st">'annot'</span>,</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>        color_maps_annotations<span class="op">=</span>{<span class="st">'distance'</span>: <span class="st">'viridis'</span>},</span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>        title<span class="op">=</span><span class="st">'</span><span class="sc">{}</span><span class="st"> path'</span>.<span class="bu">format</span>(descr),</span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>        return_data<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>        use_raw<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>        show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>pl.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="plot-genes">
<p><a href="scanpy_07_trajectory_files/figure-html/plot-genes-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="scanpy_07_trajectory_files/figure-html/plot-genes-output-1.png" class="img-fluid"></a></p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>As you can see, we can manipulate the trajectory quite a bit by selecting different number of neighbors, components etc. to fit with our assumptions on the development of these celltypes.</p>
<p>Please explore further how you can tweak the trajectory. For instance, can you create a PAGA trajectory using the orignial umap from Seurat instead? Hint, you first need to compute the neighbors on the umap.</p>
</div>
</div>
</section>
<section id="session-info" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="session-info"><span class="header-section-number">10</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div id="session" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sc.logging.print_versions()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-----
anndata     0.10.8
scanpy      1.10.3
-----
PIL                 11.1.0
asttokens           NA
cffi                1.17.1
colorama            0.4.6
comm                0.2.2
cycler              0.12.1
cython_runtime      NA
dateutil            2.9.0.post0
debugpy             1.8.12
decorator           5.1.1
exceptiongroup      1.2.2
executing           2.1.0
h5py                3.12.1
igraph              0.11.6
ipykernel           6.29.5
jedi                0.19.2
joblib              1.4.2
kiwisolver          1.4.7
legacy_api_wrap     NA
leidenalg           0.10.2
llvmlite            0.43.0
matplotlib          3.9.2
matplotlib_inline   0.1.7
mpl_toolkits        NA
natsort             8.4.0
networkx            3.4
numba               0.60.0
numpy               1.26.4
packaging           24.2
pandas              1.5.3
parso               0.8.4
pickleshare         0.7.5
platformdirs        4.3.6
prompt_toolkit      3.0.50
psutil              6.1.1
pure_eval           0.2.3
pycparser           2.22
pydev_ipython       NA
pydevconsole        NA
pydevd              3.2.3
pydevd_file_utils   NA
pydevd_plugins      NA
pydevd_tracing      NA
pygments            2.19.1
pynndescent         0.5.13
pyparsing           3.2.1
pytz                2024.2
scipy               1.14.1
session_info        1.0.0
six                 1.17.0
sklearn             1.6.1
sparse              0.15.5
stack_data          0.6.3
texttable           1.7.0
threadpoolctl       3.5.0
torch               2.5.1.post207
torchgen            NA
tornado             6.4.2
tqdm                4.67.1
traitlets           5.14.3
typing_extensions   NA
umap                0.5.7
wcwidth             0.2.13
yaml                6.0.2
zmq                 26.2.0
zoneinfo            NA
-----
IPython             8.31.0
jupyter_client      8.6.3
jupyter_core        5.7.2
-----
Python 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
Linux-6.10.14-linuxkit-x86_64-with-glibc2.35
-----
Session information updated at 2025-02-27 15:26</code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"selector":".lightbox","closeEffect":"zoom","descPosition":"bottom","openEffect":"zoom","loop":true});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>