<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Overview of all three pipeline results.">

<title>Comparison of HVGs</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Comparison of HVGs</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat Toolkit <i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit <i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Overview of all three pipeline results.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">05-Feb-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-data" id="toc-load-data" class="nav-link active" data-scroll-target="#load-data"><span class="header-section-number">1</span> Load data</a></li>
  <li><a href="#umaps" id="toc-umaps" class="nav-link" data-scroll-target="#umaps"><span class="header-section-number">2</span> Umaps</a></li>
  <li><a href="#variable-features" id="toc-variable-features" class="nav-link" data-scroll-target="#variable-features"><span class="header-section-number">3</span> Variable features</a>
  <ul>
  <li><a href="#compare-dispersions" id="toc-compare-dispersions" class="nav-link" data-scroll-target="#compare-dispersions"><span class="header-section-number">3.1</span> Compare dispersions</a></li>
  </ul></li>
  <li><a href="#for-one-sample" id="toc-for-one-sample" class="nav-link" data-scroll-target="#for-one-sample"><span class="header-section-number">4</span> For one sample</a>
  <ul>
  <li><a href="#seurat" id="toc-seurat" class="nav-link" data-scroll-target="#seurat"><span class="header-section-number">4.1</span> Seurat</a></li>
  <li><a href="#bioc" id="toc-bioc" class="nav-link" data-scroll-target="#bioc"><span class="header-section-number">4.2</span> Bioc</a></li>
  <li><a href="#scanpy" id="toc-scanpy" class="nav-link" data-scroll-target="#scanpy"><span class="header-section-number">4.3</span> Scanpy</a></li>
  <li><a href="#all-together" id="toc-all-together" class="nav-link" data-scroll-target="#all-together"><span class="header-section-number">4.4</span> All together</a></li>
  <li><a href="#top-genes" id="toc-top-genes" class="nav-link" data-scroll-target="#top-genes"><span class="header-section-number">4.5</span> Top genes</a></li>
  <li><a href="#expression-levels" id="toc-expression-levels" class="nav-link" data-scroll-target="#expression-levels"><span class="header-section-number">4.6</span> Expression levels</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion"><span class="header-section-number">5</span> Discussion</a>
  <ul>
  <li><a href="#scry" id="toc-scry" class="nav-link" data-scroll-target="#scry"><span class="header-section-number">5.1</span> scry</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">6</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>With data in place, now we can start loading libraries we will use in this tutorial.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(zellkonverter)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Matrix)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(basilisk)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">source_url</span>(<span class="st">"https://raw.githubusercontent.com/asabjorklund/single_cell_R_scripts/main/overlap_phyper_v2.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-data" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="load-data"><span class="header-section-number">1</span> Load data</h2>
<p>OBS! Zellkonverter installs conda env with basilisk! Takes a while to run first time!!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>path_results <span class="ot">&lt;-</span> <span class="st">"data/covid/results"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_results)) <span class="fu">dir.create</span>(path_results, <span class="at">recursive =</span> T)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>path_seurat <span class="ot">=</span> <span class="st">"../seurat/data/covid/results/"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>path_bioc <span class="ot">=</span> <span class="st">"../bioc/data/covid/results/"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>path_scanpy <span class="ot">=</span> <span class="st">"../scanpy/data/covid/results/"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>path_results <span class="ot">&lt;-</span> <span class="st">"data/covid/results"</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_results)) <span class="fu">dir.create</span>(path_results, <span class="at">recursive =</span> T)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>path_seurat <span class="ot">=</span> <span class="st">"../seurat/data/covid/results/"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>path_bioc <span class="ot">=</span> <span class="st">"../bioc/data/covid/results/"</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>path_scanpy <span class="ot">=</span> <span class="st">"../scanpy/data/covid/results/"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># fetch the files with qc and dimred for each </span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_seurat,<span class="st">"seurat_covid_qc_dr_int_cl.rds"</span>))</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># bioc</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_bioc,<span class="st">"bioc_covid_qc_dr_int_cl.rds"</span>))</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>bioc <span class="ot">=</span> <span class="fu">as.Seurat</span>(sce)</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="co"># scanpy</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>scanpy.sce <span class="ot">=</span> <span class="fu">readH5AD</span>(<span class="fu">file.path</span>(path_scanpy, <span class="st">"scanpy_covid_qc_dr_scanorama_cl.h5ad"</span>))</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>scanpy <span class="ot">=</span> <span class="fu">as.Seurat</span>(scanpy.sce, <span class="at">counts =</span> <span class="cn">NULL</span>, <span class="at">data =</span> <span class="st">"X"</span>) <span class="co"># only have the var.genes data that is scaled.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="umaps" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="umaps"><span class="header-section-number">2</span> Umaps</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat"</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(bioc, <span class="at">group.by =</span> <span class="st">"sample"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc"</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(scanpy, <span class="at">group.by =</span> <span class="st">"sample"</span>, <span class="at">reduction =</span> <span class="st">"X_umap_uncorr"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy"</span>),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/umaps-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="comparison_hvg_files/figure-html/umaps-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>Create one dataset with the cells that are present in all samples. Also add in umap from all 3 pipelines.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>meta.seurat <span class="ot">=</span> sobj<span class="sc">@</span>meta.data</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>meta.scanpy <span class="ot">=</span> scanpy<span class="sc">@</span>meta.data</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>meta.bioc <span class="ot">=</span> bioc<span class="sc">@</span>meta.data</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>meta.bioc<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">rownames</span>(meta.bioc)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>meta.scanpy<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">rownames</span>(meta.scanpy), <span class="cf">function</span>(x) <span class="fu">substr</span>(x,<span class="dv">1</span>,<span class="fu">nchar</span>(x)<span class="sc">-</span><span class="dv">2</span>))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>meta.seurat<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">rownames</span>(meta.seurat),<span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">3</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>in.all <span class="ot">=</span> <span class="fu">intersect</span>(<span class="fu">intersect</span>(meta.scanpy<span class="sc">$</span>cell, meta.seurat<span class="sc">$</span>cell), meta.bioc<span class="sc">$</span>cell)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>tmp1 <span class="ot">=</span> meta.bioc[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp1) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp1),<span class="st">"_bioc"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">=</span> meta.scanpy[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp2) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp2),<span class="st">"_scpy"</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> sobj[,<span class="fu">match</span>(in.all, meta.seurat<span class="sc">$</span>cell)]</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>meta.all <span class="ot">=</span> <span class="fu">cbind</span>(all<span class="sc">@</span>meta.data, tmp1,tmp2)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>all<span class="sc">@</span>meta.data <span class="ot">=</span> meta.all</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "pca"               "umap"              "tsne"             
 [4] "UMAP10_on_PCA"     "UMAP_on_ScaleData" "integrated_cca"   
 [7] "umap_cca"          "tsne_cca"          "harmony"          
[10] "umap_harmony"      "scanorama"         "scanoramaC"       
[13] "umap_scanorama"    "umap_scanoramaC"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>reductions<span class="sc">$</span>UMAP_on_PCA<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_bioc"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapbioc_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>reductions<span class="sc">$</span>X_umap_uncorr<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_scpy"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapscpy_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "pca"               "umap"              "tsne"             
 [4] "UMAP10_on_PCA"     "UMAP_on_ScaleData" "integrated_cca"   
 [7] "umap_cca"          "tsne_cca"          "harmony"          
[10] "umap_harmony"      "scanorama"         "scanoramaC"       
[13] "umap_scanorama"    "umap_scanoramaC"   "umap_bioc"        
[16] "umap_scpy"        </code></pre>
</div>
</div>
</section>
<section id="variable-features" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="variable-features"><span class="header-section-number">3</span> Variable features</h2>
<p>In Seurat, <code>FindVariableFeatures</code> is not batch aware unless the data is split into layers by samples, here we have the variable genes created with layers. In Bioc <code>modelGeneVar</code> we used sample as a blocking parameter, e.g calculates the variable genes per sample and combines the variances. Similar in scanpy we used the samples as <code>batch_key</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>seurat <span class="ot">=</span> <span class="fu">VariableFeatures</span>(sobj)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>bioc <span class="ot">=</span> sce<span class="sc">@</span>metadata<span class="sc">$</span>hvgs</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># scanpy has no strict number on selection, instead uses dispersion cutoff. So select instead top 2000 dispersion genes.</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>scpy.hvg <span class="ot">=</span> <span class="fu">rowData</span>(scanpy.sce)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>hvgs_scanpy <span class="ot">=</span> <span class="fu">rownames</span>(scpy.hvg)[scpy.hvg<span class="sc">$</span>highly_variable]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>scanpy <span class="ot">=</span> <span class="fu">rownames</span>(scpy.hvg)[<span class="fu">order</span>(scpy.hvg<span class="sc">$</span>dispersions_norm, <span class="at">decreasing =</span> T)][<span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>]</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>cmat <span class="ot">=</span> <span class="fu">make_comb_mat</span>(hvgs)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A combination matrix with 3 sets and 7 combinations.
  ranges of combination set size: c(245, 869).
  mode for the combination size: distinct.
  sets are on rows.

Combination sets are:
  seurat bioc scanpy code size
       x    x      x  111  607
       x    x         110  279
       x           x  101  245
            x      x  011  344
       x              100  869
            x         010  770
                   x  001  804

Sets are:
     set size
  seurat 2000
    bioc 2000
  scanpy 2000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UpSet</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/hvg-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="comparison_hvg_files/figure-html/hvg-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Surprisingly low overlap between the methods and many genes that are unique to one pipeline. With discrepancies in the doublet filtering the cells used differ to some extent, but otherwise the variation should be similar. Even if it is estimated in different ways.</p>
<p>Is the differences more due to the combination of ranks/dispersions or also found within a single dataset?</p>
<p>Only Seurat have the dispersions for each individual dataset stored in the object.</p>
<section id="compare-dispersions" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="compare-dispersions"><span class="header-section-number">3.1</span> Compare dispersions</h3>
<p>Recalculate for bioc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce, <span class="at">block =</span> sce<span class="sc">$</span>sample)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>hvgs_bioc <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(var.out, <span class="at">n =</span> <span class="dv">2000</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>cutoff <span class="ot">&lt;-</span> <span class="fu">rownames</span>(var.out) <span class="sc">%in%</span> hvgs_bioc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge all hvg info</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>all.genes <span class="ot">=</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(var.out), <span class="fu">rownames</span>(scpy.hvg))</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>scpy.hvg <span class="ot">=</span> <span class="fu">rowData</span>(scanpy.sce)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(scpy.hvg) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(scpy.hvg), <span class="st">"_scpy"</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(var.out) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(var.out), <span class="st">"_bioc"</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>seu.hvg <span class="ot">=</span> <span class="fu">HVFInfo</span>(all) </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>all.hvg <span class="ot">=</span> <span class="fu">cbind</span>(seu.hvg[all.genes,], scpy.hvg[all.genes,], var.out[all.genes,])</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(all.hvg<span class="sc">$</span>means_scpy, all.hvg<span class="sc">$</span>dispersions_norm_scpy, <span class="at">main =</span> <span class="st">"Scanpy"</span>,<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(all.hvg<span class="sc">$</span>means_scpy[all.hvg<span class="sc">$</span>highly_variable_scpy], all.hvg<span class="sc">$</span>dispersions_norm_scpy[all.hvg<span class="sc">$</span>highly_variable_scpy], <span class="at">col=</span><span class="st">"red"</span>,<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(all.hvg<span class="sc">$</span>mean_bioc, all.hvg<span class="sc">$</span>bio_bioc, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.4</span>, <span class="at">main =</span> <span class="st">"Bioc"</span>)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(all.hvg<span class="sc">$</span>mean_bioc[<span class="fu">rownames</span>(all.hvg) <span class="sc">%in%</span> hvgs<span class="sc">$</span>bioc], all.hvg<span class="sc">$</span>bio_bioc[<span class="fu">rownames</span>(all.hvg) <span class="sc">%in%</span> hvgs<span class="sc">$</span>bioc], <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> .<span class="dv">6</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">log1p</span>(all.hvg<span class="sc">$</span>mean), all.hvg<span class="sc">$</span>variance.standardized, <span class="at">main =</span> <span class="st">"Seurat"</span>,<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)    </span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="fu">log1p</span>(all.hvg<span class="sc">$</span>mean)[<span class="fu">match</span>(hvgs<span class="sc">$</span>seurat, <span class="fu">rownames</span>(all.hvg))], all.hvg<span class="sc">$</span>variance.standardized[<span class="fu">match</span>(hvgs<span class="sc">$</span>seurat, <span class="fu">rownames</span>(all.hvg))],<span class="at">col=</span><span class="st">"red"</span>,<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.4</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/dispersions-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="comparison_hvg_files/figure-html/dispersions-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Scanpy uses min_disp=0.5, min_mean=0.0125, max_mean=3, so the highly expressed ones are not included.</p>
<p>Seurat uses <code>variance.standardized</code> to rank the genes. Bioc uses the <code>bio</code> slot with estimated biological variation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>all.hvg<span class="sc">$</span>mean_log <span class="ot">=</span> <span class="fu">log1p</span>(all.hvg<span class="sc">$</span>mean)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sel.means <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"mean"</span>,<span class="st">"mean_log"</span>,<span class="st">"means_scpy"</span>, <span class="st">"mean_bioc"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(all.hvg[,sel.means])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/pairs-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="comparison_hvg_files/figure-html/pairs-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sel.disp <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"variance.standardized"</span>,<span class="st">"dispersions_norm_scpy"</span>, <span class="st">"bio_bioc"</span>,<span class="st">"total_bioc"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(all.hvg[,sel.disp])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/pairs-plot-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="comparison_hvg_files/figure-html/pairs-plot-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Difference in what cells were used, and also in how the dispersions across the samples is combined into one value. Do for just one sample instead.</p>
</section>
</section>
<section id="for-one-sample" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="for-one-sample"><span class="header-section-number">4</span> For one sample</h2>
<p>Run hvg selection for ctrl.13 sample, so that the handling of batches is not influencing results. Use default settings in all the different methods.</p>
<section id="seurat" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="seurat"><span class="header-section-number">4.1</span> Seurat</h3>
<p>Try the different methods implemented in Seurat. From the help section:</p>
<pre><code>* “vst”: First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).
* “mean.var.plot” (mvp): First, uses a function to calculate average expression (mean.function) and dispersion (dispersion.function) for each feature. Next, divides features into num.bin (deafult 20) bins based on their average expression, and calculates z-scores for dispersion within each bin. The purpose of this is to identify variable features while controlling for the strong relationship between variability and average expression
* “dispersion” (disp): selects the genes with the highest dispersion values</code></pre>
<p>Feature selection for individual datasets</p>
<p>In each dataset, we next aimed to identify a subset of features (e.g., genes) exhibiting high variability across cells, and therefore represent heterogeneous features to prioritize for downstream analysis. Choosing genes solely based on their log-normalized single-cell variance fails to account for the mean-variance relationship that is inherent to single-cell RNA-seq. Therefore, we first applied a variance-stabilizing transformation to correct for this [Mayer et al., 2018, Hafemeister and Satija, 2019]. To learn the mean-variance relationship from the data, we computed the mean and variance of each gene using the unnormalized data (i.e., UMI or counts matrix), and applied -transformation to both. We then fit a curve to predict the variance of each gene as a function of its mean, by calculating a local fitting of polynomials of degree 2 (R function loess, span = 0.3). This global fit provided us with a regularized estimator of variance given the mean of a feature. As such, we could use it to standardize feature counts without removing higher-than-expected variation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">=</span> all[,all<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"ctrl_13"</span>]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data[,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data)] <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(ctrl)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>hvg_seu <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>hvg_seu<span class="sc">$</span>vst <span class="ot">=</span> <span class="fu">VariableFeatures</span>(ctrl)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#top20 &lt;- head(hvg_seu$vst, 20)</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">#LabelPoints(plot = VariableFeaturePlot(ctrl), points = top20, repel = TRUE)</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(ctrl, <span class="at">selection.method =</span> <span class="st">"mean.var.plot"</span>)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>hvg_seu<span class="sc">$</span>mvp <span class="ot">=</span> <span class="fu">VariableFeatures</span>(ctrl)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co">#top20 &lt;- head(hvg_seu$mvp, 20)</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="co">#LabelPoints(plot = VariableFeaturePlot(ctrl), points = top20, repel = TRUE)</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(ctrl, <span class="at">selection.method =</span> <span class="st">"dispersion"</span>)</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>hvg_seu<span class="sc">$</span>disp <span class="ot">=</span> <span class="fu">VariableFeatures</span>(ctrl)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="co">#top20 &lt;- head(hvg_seu$disp, 20)</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="co">#LabelPoints(plot = VariableFeaturePlot(ctrl), points = top20, repel = TRUE)</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>cmat <span class="ot">=</span> <span class="fu">make_comb_mat</span>(hvg_seu)</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>cmat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A combination matrix with 3 sets and 7 combinations.
  ranges of combination set size: c(31, 1295).
  mode for the combination size: distinct.
  sets are on rows.

Combination sets are:
  vst mvp disp code size
    x   x    x  111  401
    x   x       110   31
    x        x  101  273
        x    x  011  549
    x           100 1295
        x       010  149
             x  001  777

Sets are:
   set size
   vst 2000
   mvp 1130
  disp 2000</code></pre>
</div>
</div>
<p>Very little overlap between the methods in Seurat. Mvp and disp are calculated on <code>data</code>, while vst is done on <code>counts</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>vinfo.seu <span class="ot">=</span> ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(vinfo.seu) <span class="ot">=</span> <span class="fu">rownames</span>(ctrl)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>vg_vst <span class="ot">=</span> <span class="fu">rownames</span>(vinfo.seu) <span class="sc">%in%</span> hvg_seu<span class="sc">$</span>vst</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>vg_disp <span class="ot">=</span> <span class="fu">rownames</span>(vinfo.seu) <span class="sc">%in%</span> hvg_seu<span class="sc">$</span>disp</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>vg_mvp <span class="ot">=</span> <span class="fu">rownames</span>(vinfo.seu) <span class="sc">%in%</span> hvg_seu<span class="sc">$</span>mvp</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>vg_vst_disp <span class="ot">=</span> vinfo.seu<span class="sc">$</span>vg_vst <span class="sc">+</span> vinfo.seu<span class="sc">$</span>vg_disp <span class="sc">==</span> <span class="dv">2</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>hvinfo <span class="ot">=</span> <span class="fu">ifelse</span>(vinfo.seu<span class="sc">$</span>vg_vst, <span class="st">"VST"</span>,<span class="cn">NA</span>)</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>hvinfo[vinfo.seu<span class="sc">$</span>vg_mvp] <span class="ot">=</span> <span class="st">"MVP"</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>hvinfo[vinfo.seu<span class="sc">$</span>vg_disp] <span class="ot">=</span> <span class="st">"DISP"</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>vinfo.seu<span class="sc">$</span>hvinfo[vinfo.seu<span class="sc">$</span>vg_vst_disp] <span class="ot">=</span> <span class="st">"VST/DISP"</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>means <span class="ot">=</span> <span class="fu">colnames</span>(vinfo.seu)[<span class="fu">grepl</span>(<span class="st">"mean"</span>, <span class="fu">colnames</span>(vinfo.seu))]</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>disp <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"vf_vst_counts.ctrl_13_variance.standardized"</span>,<span class="st">"vf_disp_data.ctrl_13_mvp.dispersion.scaled"</span>,<span class="st">"vf_mvp_data.ctrl_13_mvp.dispersion.scaled"</span>)</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(vinfo.seu[,means])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/plot-seurat-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="comparison_hvg_files/figure-html/plot-seurat-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>plot_hvg <span class="ot">=</span> <span class="cf">function</span>(df, m,d,vg, <span class="at">log=</span><span class="cn">FALSE</span>){</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  top20 <span class="ot">=</span> <span class="fu">head</span>(vg,<span class="dv">20</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">=</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>.data[[m]], <span class="at">y=</span>.data[[d]], <span class="at">color=</span>hvinfo)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (log) { p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">scale_x_log10</span>()}</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LabelPoints</span>(<span class="at">plot =</span> p, <span class="at">points =</span> top20, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>p1S <span class="ot">=</span>   <span class="fu">plot_hvg</span>(vinfo.seu, <span class="st">"vf_vst_counts.ctrl_13_mean"</span>, <span class="st">"vf_vst_counts.ctrl_13_variance.standardized"</span>,hvg_seu<span class="sc">$</span>vst, <span class="at">log=</span><span class="cn">TRUE</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>p2S <span class="ot">=</span>   <span class="fu">plot_hvg</span>(vinfo.seu, <span class="st">"vf_disp_data.ctrl_13_mvp.mean"</span>, <span class="st">"vf_disp_data.ctrl_13_mvp.dispersion.scaled"</span>,hvg_seu<span class="sc">$</span>disp)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>p3S <span class="ot">=</span>   <span class="fu">plot_hvg</span>(vinfo.seu, <span class="st">"vf_mvp_data.ctrl_13_mvp.mean"</span>, <span class="st">"vf_mvp_data.ctrl_13_mvp.dispersion.scaled"</span>,hvg_seu<span class="sc">$</span>mvp)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(p1S,p2S,p3S,   <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/plot-seurat-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="comparison_hvg_files/figure-html/plot-seurat-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="bioc" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="bioc"><span class="header-section-number">4.2</span> Bioc</h3>
<p>Run the bioconductor detection method with <code>modelGeneVar</code> and <code>getTopHVGs</code>. Plot both total variance and “bio” variance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ctrl.sce <span class="ot">=</span> <span class="fu">as.SingleCellExperiment</span>(ctrl)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>var.out <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(ctrl.sce)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>hvgs_bioc <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(var.out, <span class="at">n =</span> <span class="dv">2000</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>var.out.df <span class="ot">=</span> <span class="fu">data.frame</span>(var.out)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>var.out.df<span class="sc">$</span>hvg <span class="ot">=</span> <span class="fu">rownames</span>(var.out) <span class="sc">%in%</span> hvgs_bioc</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ggplot</span>(var.out.df, <span class="fu">aes</span>(<span class="at">x=</span>mean, <span class="at">y=</span>total, <span class="at">colour =</span> hvg)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>()</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>top20 <span class="ot">&lt;-</span> <span class="fu">head</span>(hvgs_bioc, <span class="dv">20</span>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>pB <span class="ot">=</span> <span class="fu">LabelPoints</span>(<span class="at">plot =</span> p, <span class="at">points =</span> top20, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>pB</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/bioc2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="comparison_hvg_files/figure-html/bioc2-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">ggplot</span>(var.out.df, <span class="fu">aes</span>(<span class="at">x=</span>mean, <span class="at">y=</span>bio, <span class="at">colour =</span> hvg)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>()</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>top20 <span class="ot">&lt;-</span> <span class="fu">head</span>(hvgs_bioc, <span class="dv">20</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>pB2 <span class="ot">=</span> <span class="fu">LabelPoints</span>(<span class="at">plot =</span> p2, <span class="at">points =</span> top20, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>pB2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/bioc2-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="comparison_hvg_files/figure-html/bioc2-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="scanpy" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="scanpy"><span class="header-section-number">4.3</span> Scanpy</h3>
<p>Run with both <code>seurat</code> (on lognorm data) and <code>seurat_v3</code> (on counts, is same as vst)</p>
<p>For the dispersion-based methods (flavor=‘seurat’ Satija et al.&nbsp;[2015] and flavor=‘cell_ranger’ Zheng et al.&nbsp;[2017]), the normalized dispersion is obtained by scaling with the mean and standard deviation of the dispersions for genes falling into a given bin for mean expression of genes. This means that for each bin of mean expression, highly variable genes are selected.</p>
<p>For flavor=‘seurat_v3’/‘seurat_v3_paper’ [Stuart et al., 2019], a normalized variance for each gene is computed. First, the data are standardized (i.e., z-score normalization per feature) with a regularized standard deviation. Next, the normalized variance is computed as the variance of each gene after the transformation. Genes are ranked by the normalized variance. Only if batch_key is not None, the two flavors differ: For flavor=‘seurat_v3’, genes are first sorted by the median (across batches) rank, with ties broken by the number of batches a gene is a HVG. For flavor=‘seurat_v3_paper’, genes are first sorted by the number of batches a gene is a HVG, with ties broken by the median (across batches) rank.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>penv <span class="ot">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/scanpy_2024_nopip"</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>hvg.scanpy <span class="ot">=</span> <span class="fu">basiliskRun</span>(<span class="at">env=</span>penv, <span class="at">fun=</span><span class="cf">function</span>(counts) {</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    scanpy <span class="ot">&lt;-</span> reticulate<span class="sc">::</span><span class="fu">import</span>(<span class="st">"scanpy"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    ad <span class="ot">=</span> reticulate<span class="sc">::</span><span class="fu">import</span>(<span class="st">"anndata"</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    adata <span class="ot">=</span> ad<span class="sc">$</span><span class="fu">AnnData</span>(counts)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    var1 <span class="ot">=</span> scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">highly_variable_genes</span>(adata, <span class="at">flavor =</span> <span class="st">"seurat_v3"</span>, <span class="at">inplace=</span><span class="cn">FALSE</span>)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">normalize_per_cell</span>(adata, <span class="at">counts_per_cell_after=</span><span class="fl">1e4</span>)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">log1p</span>(adata)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">highly_variable_genes</span>(adata)</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">disp=</span>adata<span class="sc">$</span>var, <span class="at">vst=</span>var1))</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>}, <span class="at">counts =</span> <span class="fu">t</span>(ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts.ctrl_13),  <span class="at">testload=</span><span class="st">"scanpy"</span>)</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co">#flavor</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="co">#Literal['seurat', 'cell_ranger', 'seurat_v3', 'seurat_v3_paper'] (default: 'seurat')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>disp) <span class="ot">=</span> <span class="fu">rownames</span>(ctrl)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>vst) <span class="ot">=</span> <span class="fu">rownames</span>(ctrl)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>top20 <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>disp)[<span class="fu">order</span>(hvg.scanpy<span class="sc">$</span>disp<span class="sc">$</span>dispersions_norm, <span class="at">decreasing =</span> T)], <span class="dv">20</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ggplot</span>(hvg.scanpy<span class="sc">$</span>disp, <span class="fu">aes</span>(<span class="at">x=</span>means, <span class="at">y=</span>dispersions_norm, <span class="at">colour =</span> highly_variable)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy disp"</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>pS <span class="ot">=</span> <span class="fu">LabelPoints</span>(<span class="at">plot =</span> p, <span class="at">points =</span> top20, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>pS </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/scanpy2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="comparison_hvg_files/figure-html/scanpy2-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>top20 <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>vst)[<span class="fu">order</span>(hvg.scanpy<span class="sc">$</span>vst<span class="sc">$</span>variances_norm, <span class="at">decreasing =</span> T)], <span class="dv">20</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">ggplot</span>(hvg.scanpy<span class="sc">$</span>vst, <span class="fu">aes</span>(<span class="at">x=</span>means, <span class="at">y=</span>variances_norm, <span class="at">colour =</span> highly_variable)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy vst"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>()</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>pS2 <span class="ot">=</span> <span class="fu">LabelPoints</span>(<span class="at">plot =</span> p2, <span class="at">points =</span> top20, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>pS2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/scanpy2-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="comparison_hvg_files/figure-html/scanpy2-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="all-together" class="level3" data-number="4.4">
<h3 data-number="4.4" class="anchored" data-anchor-id="all-together"><span class="header-section-number">4.4</span> All together</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(hvg.scanpy<span class="sc">$</span>disp) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(hvg.scanpy<span class="sc">$</span>disp), <span class="st">"_scpyD"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(hvg.scanpy<span class="sc">$</span>vst) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(hvg.scanpy<span class="sc">$</span>vst), <span class="st">"_scpyV"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(var.out) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(var.out), <span class="st">"_bioc"</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>ctrl.hvg <span class="ot">=</span> <span class="fu">cbind</span>(vinfo.seu, var.out, hvg.scanpy<span class="sc">$</span>disp, hvg.scanpy<span class="sc">$</span>vst)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(p1S <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat vst"</span>),p2S <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat disp"</span>),p3S <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat mvp"</span>),pB <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc total"</span>),pS,pS2, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/plot-all-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="comparison_hvg_files/figure-html/plot-all-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"vf_vst_counts.ctrl_13_variance.standardized"</span>,<span class="st">"vf_disp_data.ctrl_13_mvp.dispersion.scaled"</span>,<span class="st">"bio_bioc"</span>, <span class="st">"dispersions_norm_scpyD"</span>, <span class="st">"variances_norm_scpyV"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(ctrl.hvg[,sel])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/disps-all-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="comparison_hvg_files/figure-html/disps-all-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"vf_vst_counts.ctrl_13_mean_log"</span>,<span class="st">"vf_disp_data.ctrl_13_mvp.mean"</span>, <span class="st">"mean_bioc"</span>, <span class="st">"means_scpyD"</span>, <span class="st">"means_scpyV_log"</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>vf_vst_counts.ctrl_13_mean_log <span class="ot">=</span> <span class="fu">log1p</span>(ctrl.hvg<span class="sc">$</span>vf_vst_counts.ctrl_13_mean)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>means_scpyV_log <span class="ot">=</span> <span class="fu">log1p</span>(ctrl.hvg<span class="sc">$</span>means_scpyV)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(ctrl.hvg[,sel])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/means-all-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="comparison_hvg_files/figure-html/means-all-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">=</span> hvg_seu</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>scpyD <span class="ot">=</span> <span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>disp)[hvg.scanpy<span class="sc">$</span>disp<span class="sc">$</span>highly_variable_scpyD]</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>scpyV <span class="ot">=</span> <span class="fu">rownames</span>(hvg.scanpy<span class="sc">$</span>disp)[hvg.scanpy<span class="sc">$</span>vst<span class="sc">$</span>highly_variable_scpyV]</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>bioc <span class="ot">=</span> hvgs_bioc</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(hvgs,hvgs, <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/overlap-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="comparison_hvg_files/figure-html/overlap-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Mean Bioc stands out the most.</p>
<p>Vst in seurat and scanpy is identical and very different from the other methods. But it is also based on counts instead of lognorm data.</p>
<p>The variance estimate for the same sample is quite different. Even the top variable genes are not the same and are very different gene groups.</p>
</section>
<section id="top-genes" class="level3" data-number="4.5">
<h3 data-number="4.5" class="anchored" data-anchor-id="top-genes"><span class="header-section-number">4.5</span> Top genes</h3>
<p>Explore top genes with the different methods.</p>
<p>disp and mvp are identical, remove mvp.</p>
<p>ScanpyV and vst are identical, remove scpyV.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(hvgs, head, <span class="dv">10</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># same for mvp and disp, remove one of them.</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>topG<span class="sc">$</span>mvp <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>topG<span class="sc">$</span>scpyV <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the genes for scanpy.</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>topG<span class="sc">$</span>scpyD <span class="ot">=</span> <span class="fu">rownames</span>(ctrl.hvg)[<span class="fu">order</span>(ctrl.hvg<span class="sc">$</span>dispersions_norm_scpyD, <span class="at">decreasing =</span> T)][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co"># rank the genes</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>allG <span class="ot">=</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(topG))</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>ranks <span class="ot">=</span> <span class="fu">Reduce</span>(rbind, <span class="fu">lapply</span>(topG, <span class="cf">function</span>(x) {</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>  r <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(r) <span class="ot">=</span> x</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>  r2 <span class="ot">=</span> r[allG]</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(r2)</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ranks) <span class="ot">=</span> allG</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ranks) <span class="ot">=</span> <span class="fu">names</span>(topG)</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(ranks, <span class="at">display_numbers =</span> ranks, <span class="at">main=</span><span class="st">"Rank of top 10 gens"</span>, <span class="at">cluster_rows =</span> F, <span class="at">cluster_cols =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/top10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="comparison_hvg_files/figure-html/top10-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Plot onto umap for the unique ones.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>topG<span class="sc">$</span>disp <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>t <span class="ot">=</span> <span class="fu">table</span>(<span class="fu">unlist</span>(topG))</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>selG <span class="ot">=</span> <span class="fu">names</span>(t)[t<span class="sc">==</span><span class="dv">1</span>]</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>selG <span class="ot">=</span> <span class="fu">sapply</span>(selG, <span class="cf">function</span>(y) <span class="fu">names</span>(<span class="fu">which</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(topG, <span class="cf">function</span>(x) <span class="fu">any</span>(x<span class="sc">==</span>y))))))</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>selG <span class="ot">=</span> <span class="fu">sort</span>(selG)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="co">#small.leg &lt;- theme(legend.text = element_text(size=3), legend.key.size = unit(0.5,"point"))</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> <span class="fu">names</span>(selG)){</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>  plots[[g]] <span class="ot">=</span> <span class="fu">FeaturePlot</span>(ctrl, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">features =</span> g, <span class="at">order =</span> T)  <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="fu">paste</span>(g,selG[g], <span class="at">collapse =</span> <span class="st">" - "</span>)) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="co">#+ small.leg</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/top10-umap-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="comparison_hvg_files/figure-html/top10-umap-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>More clear celltype genes in the Bioc selection, but also higher expressed genes.. More B-cell genes for vst.</p>
</section>
<section id="expression-levels" class="level3" data-number="4.6">
<h3 data-number="4.6" class="anchored" data-anchor-id="expression-levels"><span class="header-section-number">4.6</span> Expression levels</h3>
<p>Expression levels of the variable genes. As total counts or mean expression across all cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>nC <span class="ot">=</span> <span class="fu">rowSums</span>(ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts.ctrl_13<span class="sc">&gt;</span><span class="dv">0</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>meanE <span class="ot">=</span> <span class="fu">rowMeans</span>(ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data.ctrl_13)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>vg_bioc <span class="ot">=</span> <span class="fu">rownames</span>(ctrl.hvg) <span class="sc">%in%</span> hvgs_bioc</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">fill=</span>vg_vst)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"VST"</span>),</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">fill=</span>vg_disp)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Disp"</span>), </span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">fill=</span>highly_variable_scpyD)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy Disp"</span>),</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">fill=</span>highly_variable_scpyV)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy VST"</span>),</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">fill=</span>vg_bioc)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc"</span>),</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">fill=</span>vg_vst)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"VST"</span>), </span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">fill=</span>vg_disp)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Disp"</span>), </span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">fill=</span>highly_variable_scpyD)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy Disp"</span>), </span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">fill=</span>highly_variable_scpyV)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy VST"</span>), </span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">fill=</span>vg_bioc)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>()<span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc"</span>), </span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a><span class="at">ncol =</span><span class="dv">5</span></span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/exprs-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="comparison_hvg_files/figure-html/exprs-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>BioC distibution shifted to more highly expressed genes.</p>
<p>Dispersion gives the very highly expressed genes, but also has a spread among all levels of expression.</p>
</section>
</section>
<section id="discussion" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="discussion"><span class="header-section-number">5</span> Discussion</h2>
<ul>
<li><p>Seurat v3 paper suggests vst on counts, but then uses the variable genes in lognorm space, how is this the best option_</p></li>
<li><p>From SC best practices, suggests scry for HVG selection. https://www.sc-best-practices.org/preprocessing_visualization/feature_selection.html</p></li>
</ul>
<section id="scry" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="scry"><span class="header-section-number">5.1</span> scry</h3>
<p>Run scry deviance function and select top 2k genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>ctrl.sce<span class="ot">&lt;-</span>scry<span class="sc">::</span><span class="fu">devianceFeatureSelection</span>(ctrl.sce, <span class="at">assay=</span><span class="st">"counts"</span>, <span class="at">sorted=</span><span class="cn">TRUE</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">rowData</span>(ctrl.sce)<span class="sc">$</span>binomial_deviance, <span class="at">type=</span><span class="st">"l"</span>, <span class="at">xlab=</span><span class="st">"ranked genes"</span>,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"binomial deviance"</span>, <span class="at">main=</span><span class="st">"Feature Selection with Deviance"</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">2000</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">col=</span><span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/scry-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="comparison_hvg_files/figure-html/scry-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(ctrl.sce),<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 10 rows and 1 column
        binomial_deviance
                &lt;numeric&gt;
S100A9            74343.0
FTL               53573.3
GNLY              52928.0
LYZ               48572.1
S100A8            42210.5
NKG7              36843.2
FTH1              35595.1
B2M               23773.7
CCL5              22340.5
HLA-DRA           21938.7</code></pre>
</div>
</div>
<p>Top genes are similar to the other methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>deviance <span class="ot">=</span> <span class="fu">rowData</span>(ctrl.sce)<span class="sc">$</span>binomial_deviance[<span class="fu">match</span>(<span class="fu">rownames</span>(ctrl.hvg), <span class="fu">rownames</span>(<span class="fu">rowData</span>(ctrl.sce)))]</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>ctrl.hvg<span class="sc">$</span>devG <span class="ot">=</span> <span class="fu">rownames</span>(ctrl.hvg) <span class="sc">%in%</span> <span class="fu">head</span>(<span class="fu">rownames</span>(<span class="fu">rowData</span>(ctrl.sce)),<span class="dv">2000</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">y=</span>deviance, <span class="at">color=</span>devG)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"nCell vs deviance"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>(), </span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>meanE, <span class="at">y=</span>deviance, <span class="at">color=</span>devG)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"mean expression vs deviance"</span>)<span class="sc">+</span> <span class="fu">NoLegend</span>(), </span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>dispersions_norm_scpyD, <span class="at">y=</span>deviance, <span class="at">color=</span>devG)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Dispersion vs deviance"</span>)<span class="sc">+</span> <span class="fu">NoLegend</span>(), </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>variances_norm_scpyV, <span class="at">y=</span>deviance, <span class="at">color=</span>devG)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"VST vs deviance"</span>)<span class="sc">+</span> <span class="fu">NoLegend</span>(), </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl.hvg, <span class="fu">aes</span>(<span class="at">x=</span>bio_bioc, <span class="at">y=</span>deviance, <span class="at">color=</span>devG)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"bioc bio vs deviance"</span>)<span class="sc">+</span> <span class="fu">NoLegend</span>(), </span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a><span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/scry-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="comparison_hvg_files/figure-html/scry-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Selects all genes with high expression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>deviance <span class="ot">=</span> <span class="fu">head</span>(<span class="fu">rownames</span>(<span class="fu">rowData</span>(ctrl.sce)),<span class="dv">2000</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(hvgs,hvgs, <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_hvg_files/figure-html/scry-overlap-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="comparison_hvg_files/figure-html/scry-overlap-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">6</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/asabjor/miniconda3/envs/seurat5_u/lib/libopenblasp-r0.3.28.dylib;  LAPACK version 3.12.0

locale:
[1] sv_SE.UTF-8/sv_SE.UTF-8/sv_SE.UTF-8/C/sv_SE.UTF-8/sv_SE.UTF-8

time zone: Europe/Stockholm
tzcode source: system (macOS)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] pheatmap_1.0.12             basilisk_1.14.1            
 [3] ComplexHeatmap_2.18.0       scran_1.30.0               
 [5] scuttle_1.12.0              SingleCellExperiment_1.24.0
 [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             
 [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
[11] IRanges_2.36.0              S4Vectors_0.40.2           
[13] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[15] matrixStats_1.5.0           patchwork_1.3.0            
[17] ggplot2_3.5.1               Matrix_1.6-5               
[19] zellkonverter_1.12.1        Seurat_5.1.0               
[21] SeuratObject_5.0.2          sp_2.1-4                   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.3.3            
  [3] later_1.4.1               bitops_1.0-9             
  [5] filelock_1.0.3            tibble_3.2.1             
  [7] polyclip_1.10-7           basilisk.utils_1.14.1    
  [9] fastDummies_1.7.4         lifecycle_1.0.4          
 [11] doParallel_1.0.17         edgeR_4.0.16             
 [13] globals_0.16.3            lattice_0.22-6           
 [15] MASS_7.3-60.0.1           magrittr_2.0.3           
 [17] limma_3.58.1              plotly_4.10.4            
 [19] rmarkdown_2.29            remotes_2.5.0            
 [21] yaml_2.3.10               metapod_1.10.0           
 [23] httpuv_1.6.15             sctransform_0.4.1        
 [25] spam_2.11-0               sessioninfo_1.2.2        
 [27] pkgbuild_1.4.5            spatstat.sparse_3.1-0    
 [29] reticulate_1.40.0         cowplot_1.1.3            
 [31] pbapply_1.7-2             RColorBrewer_1.1-3       
 [33] pkgload_1.4.0             abind_1.4-5              
 [35] zlibbioc_1.48.0           Rtsne_0.17               
 [37] purrr_1.0.2               RCurl_1.98-1.16          
 [39] circlize_0.4.16           GenomeInfoDbData_1.2.11  
 [41] ggrepel_0.9.6             scry_1.14.0              
 [43] irlba_2.3.5.1             listenv_0.9.1            
 [45] spatstat.utils_3.1-2      goftest_1.2-3            
 [47] RSpectra_0.16-2           spatstat.random_3.3-2    
 [49] dqrng_0.3.2               fitdistrplus_1.2-2       
 [51] parallelly_1.41.0         DelayedMatrixStats_1.24.0
 [53] leiden_0.4.3.1            codetools_0.2-20         
 [55] DelayedArray_0.28.0       shape_1.4.6.1            
 [57] tidyselect_1.2.1          farver_2.1.2             
 [59] ScaledMatrix_1.10.0       spatstat.explore_3.3-4   
 [61] jsonlite_1.8.9            GetoptLong_1.0.5         
 [63] BiocNeighbors_1.20.0      ellipsis_0.3.2           
 [65] progressr_0.15.1          iterators_1.0.14         
 [67] ggridges_0.5.6            survival_3.8-3           
 [69] foreach_1.5.2             tools_4.3.3              
 [71] ica_1.0-3                 Rcpp_1.0.13-1            
 [73] glue_1.8.0                gridExtra_2.3            
 [75] SparseArray_1.2.2         xfun_0.50                
 [77] usethis_3.1.0             dplyr_1.1.4              
 [79] withr_3.0.2               fastmap_1.2.0            
 [81] bluster_1.12.0            digest_0.6.37            
 [83] rsvd_1.0.5                R6_2.5.1                 
 [85] mime_0.12                 colorspace_2.1-1         
 [87] Cairo_1.6-2               scattermore_1.2          
 [89] tensor_1.5                spatstat.data_3.1-4      
 [91] tidyr_1.3.1               generics_0.1.3           
 [93] data.table_1.15.4         httr_1.4.7               
 [95] htmlwidgets_1.6.4         S4Arrays_1.2.0           
 [97] uwot_0.1.16               pkgconfig_2.0.3          
 [99] gtable_0.3.6              lmtest_0.9-40            
[101] XVector_0.42.0            htmltools_0.5.8.1        
[103] profvis_0.4.0             dotCall64_1.2            
[105] clue_0.3-66               scales_1.3.0             
[107] png_0.1-8                 spatstat.univar_3.1-1    
[109] knitr_1.49                rjson_0.2.23             
[111] reshape2_1.4.4            curl_6.0.1               
[113] nlme_3.1-165              cachem_1.1.0             
[115] GlobalOptions_0.1.2       zoo_1.8-12               
[117] stringr_1.5.1             KernSmooth_2.23-26       
[119] parallel_4.3.3            miniUI_0.1.1.1           
[121] pillar_1.10.1             vctrs_0.6.5              
[123] RANN_2.6.2                urlchecker_1.0.1         
[125] promises_1.3.2            BiocSingular_1.18.0      
[127] beachmat_2.18.0           xtable_1.8-4             
[129] cluster_2.1.8             evaluate_1.0.1           
[131] cli_3.6.3                 locfit_1.5-9.10          
[133] compiler_4.3.3            rlang_1.1.4              
[135] crayon_1.5.3              future.apply_1.11.2      
[137] labeling_0.4.3            fs_1.6.5                 
[139] plyr_1.8.9                stringi_1.8.4            
[141] viridisLite_0.4.2         deldir_2.0-4             
[143] BiocParallel_1.36.0       munsell_0.5.1            
[145] lazyeval_0.2.2            devtools_2.4.5           
[147] spatstat.geom_3.3-4       dir.expiry_1.10.0        
[149] RcppHNSW_0.6.0            sparseMatrixStats_1.14.0 
[151] future_1.34.0             statmod_1.5.0            
[153] shiny_1.10.0              ROCR_1.0-11              
[155] memoise_2.0.1             igraph_2.0.3             </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"selector":".lightbox","loop":true,"openEffect":"zoom","descPosition":"bottom","closeEffect":"zoom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>