<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Overview of different normalization methods">

<title>Comparison of normalization</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Comparison of normalization</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat Toolkit <i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit <i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Overview of different normalization methods
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">05-Feb-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-data" id="toc-load-data" class="nav-link active" data-scroll-target="#load-data"><span class="header-section-number">1</span> Load data</a></li>
  <li><a href="#normalize" id="toc-normalize" class="nav-link" data-scroll-target="#normalize"><span class="header-section-number">2</span> Normalize</a>
  <ul>
  <li><a href="#sct-per-layer" id="toc-sct-per-layer" class="nav-link" data-scroll-target="#sct-per-layer"><span class="header-section-number">2.1</span> SCT per layer</a></li>
  <li><a href="#annotation" id="toc-annotation" class="nav-link" data-scroll-target="#annotation"><span class="header-section-number">2.2</span> Annotation</a></li>
  <li><a href="#sct-merged" id="toc-sct-merged" class="nav-link" data-scroll-target="#sct-merged"><span class="header-section-number">2.3</span> SCT merged</a></li>
  <li><a href="#subsample" id="toc-subsample" class="nav-link" data-scroll-target="#subsample"><span class="header-section-number">2.4</span> Subsample</a></li>
  <li><a href="#scalefactors-lognorm" id="toc-scalefactors-lognorm" class="nav-link" data-scroll-target="#scalefactors-lognorm"><span class="header-section-number">2.5</span> Scalefactors lognorm</a>
  <ul class="collapse">
  <li><a href="#sf-10k" id="toc-sf-10k" class="nav-link" data-scroll-target="#sf-10k"><span class="header-section-number">2.5.1</span> SF 10K</a></li>
  <li><a href="#sf-1k" id="toc-sf-1k" class="nav-link" data-scroll-target="#sf-1k"><span class="header-section-number">2.5.2</span> SF 1K</a></li>
  <li><a href="#sf-100k" id="toc-sf-100k" class="nav-link" data-scroll-target="#sf-100k"><span class="header-section-number">2.5.3</span> SF 100K</a></li>
  <li><a href="#dispersions" id="toc-dispersions" class="nav-link" data-scroll-target="#dispersions"><span class="header-section-number">2.5.4</span> Dispersions</a></li>
  </ul></li>
  <li><a href="#logcounts" id="toc-logcounts" class="nav-link" data-scroll-target="#logcounts"><span class="header-section-number">2.6</span> Logcounts</a></li>
  <li><a href="#tmm" id="toc-tmm" class="nav-link" data-scroll-target="#tmm"><span class="header-section-number">2.7</span> TMM</a></li>
  <li><a href="#quantile" id="toc-quantile" class="nav-link" data-scroll-target="#quantile"><span class="header-section-number">2.8</span> Quantile</a></li>
  <li><a href="#deconvolution" id="toc-deconvolution" class="nav-link" data-scroll-target="#deconvolution"><span class="header-section-number">2.9</span> Deconvolution</a></li>
  <li><a href="#pearson-residuals" id="toc-pearson-residuals" class="nav-link" data-scroll-target="#pearson-residuals"><span class="header-section-number">2.10</span> Pearson residuals</a></li>
  <li><a href="#cd3-violins" id="toc-cd3-violins" class="nav-link" data-scroll-target="#cd3-violins"><span class="header-section-number">2.11</span> CD3 violins</a></li>
  </ul></li>
  <li><a href="#distributions" id="toc-distributions" class="nav-link" data-scroll-target="#distributions"><span class="header-section-number">3</span> Distributions</a></li>
  <li><a href="#cell-correlations" id="toc-cell-correlations" class="nav-link" data-scroll-target="#cell-correlations"><span class="header-section-number">4</span> Cell correlations</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span class="header-section-number">5</span> Clustering</a>
  <ul>
  <li><a href="#pcs-stats" id="toc-pcs-stats" class="nav-link" data-scroll-target="#pcs-stats"><span class="header-section-number">5.0.1</span> PCs stats</a>
  <ul class="collapse">
  <li><a href="#contribution-to-variance" id="toc-contribution-to-variance" class="nav-link" data-scroll-target="#contribution-to-variance"><span class="header-section-number">5.0.1.1</span> Contribution to variance</a></li>
  <li><a href="#pc-correlation-to-different-stats" id="toc-pc-correlation-to-different-stats" class="nav-link" data-scroll-target="#pc-correlation-to-different-stats"><span class="header-section-number">5.0.1.2</span> PC correlation to different stats</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">6</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>Run different methods of normalization and look at how the distributions look after norm.</p>
<ul>
<li>Lognorm with different scale factors.</li>
<li>SCT with all cells or per sample. How are residuals effected?</li>
<li>TMM, Quantile.</li>
<li>Deconvolution</li>
<li>Pearson residuals in scanpy</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Matrix)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(basilisk)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">source_url</span>(<span class="st">"https://raw.githubusercontent.com/asabjorklund/single_cell_R_scripts/main/overlap_phyper_v2.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-data" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Load data</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>path_seurat <span class="ot">=</span> <span class="st">"../seurat/data/covid/results/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_seurat,<span class="st">"seurat_covid_qc_dr_int_cl.rds"</span>))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>sobj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
18854 features across 7134 samples within 1 assay 
Active assay: RNA (18854 features, 2000 variable features)
 17 layers present: counts.covid_1, counts.covid_15, counts.covid_16, counts.covid_17, counts.ctrl_5, counts.ctrl_13, counts.ctrl_14, counts.ctrl_19, scale.data, data.covid_1, data.covid_15, data.covid_16, data.covid_17, data.ctrl_5, data.ctrl_13, data.ctrl_14, data.ctrl_19
 14 dimensional reductions calculated: pca, umap, tsne, UMAP10_on_PCA, UMAP_on_ScaleData, integrated_cca, umap_cca, tsne_cca, harmony, umap_harmony, scanorama, scanoramaC, umap_scanorama, umap_scanoramaC</code></pre>
</div>
</div>
<p>The object has the data split into layers, both counts and data. One single scale.data layer.</p>
<p>Plot integrated data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"RNA_snn_res.0.5"</span>, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FeaturePlot</span>(sobj,  <span class="st">"nFeature_RNA"</span>, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(), </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FeaturePlot</span>(sobj,  <span class="st">"percent_mito"</span>, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(), </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/umaps-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="comparison_normalization_files/figure-html/umaps-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SetIdent</span>(sobj, <span class="at">value =</span> <span class="st">"RNA_snn_res.0.5"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>sobj<span class="sc">@</span>active.assay <span class="ot">=</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Select a set of variable genes to analyse across all normalizations, 5K genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(sobj, <span class="at">nfeatures =</span> <span class="dv">5000</span>, <span class="at">verbose =</span> F)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>hvg <span class="ot">=</span> <span class="fu">VariableFeatures</span>(sobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="normalize" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Normalize</h1>
<section id="sct-per-layer" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sct-per-layer"><span class="header-section-number">2.1</span> SCT per layer</h2>
<p>First, run SCT on each layer, e.g.&nbsp;each sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SCTransform</span>(sobj, <span class="at">new.assay.name =</span> <span class="st">"SCTL"</span>, <span class="at">verbose =</span> F)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15161  7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15161  7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>scale.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6545 7134</code></pre>
</div>
</div>
<p>First runs SCT for each layer separately, then creates one merged counts, data, scale.data.</p>
<p>Fewer genes in the SCT counts/data matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>vg <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>SCTModel.list)){ </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  vg[[n]] <span class="ot">=</span> <span class="fu">rownames</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>SCTModel.list[[n]]<span class="sc">@</span>feature.attributes)[sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>SCTModel.list[[n]]<span class="sc">@</span>feature.attributes<span class="sc">$</span>genes_log_gmean_step1]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(vg,vg, <span class="at">bg =</span> <span class="fu">nrow</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sctl-var-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="comparison_normalization_files/figure-html/sctl-var-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>all.vg <span class="ot">=</span> <span class="fu">unique</span>(<span class="fu">unlist</span>(vg))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(all.vg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8820</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>scale.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6545 7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">rownames</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>scale.data) <span class="sc">%in%</span> all.vg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4437</code></pre>
</div>
</div>
<p>Total variable genes from each dataset is more than the size of <code>scale.data</code>. The function first runs SCT for each dataset, then runs Seurat VariableFeatures on each object. and takes the union!</p>
<pre><code># from the seurat function:
      vf.list &lt;- lapply(X  = sct.assay.list, FUN = function(object.i) VariableFeatures(object = object.i))
      variable.features.union &lt;- Reduce(f = union, x = vf.list)</code></pre>
<p>Those genes are used to calculate the residuals in scale.data. Still, among those genes, only 4437 are among the first set of 5000 variable genes.</p>
<p>Calculate residuals for all genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">PrepSCTFindMarkers</span>(sobj, <span class="at">assay =</span> <span class="st">"SCTL"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co">#, anchor.features = intersect(rownames(sobj), hvg))$s</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="annotation" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="annotation"><span class="header-section-number">2.2</span> Annotation</h2>
<p>Merge the layers and run all normalization with all genes together. Filter genes with less than 5 cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sobj<span class="sc">@</span>active.assay <span class="ot">=</span> <span class="st">"RNA"</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(<span class="at">object =</span> sobj, <span class="at">layers =</span> <span class="fu">c</span>(<span class="st">"data"</span>,<span class="st">"counts"</span>))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># remove genes with less than 5 cells.</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>nC <span class="ot">=</span> <span class="fu">rowSums</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> sobj[nC<span class="sc">&gt;</span><span class="dv">4</span>,]</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">NormalizeData</span>(sobj, <span class="at">verbose =</span> F)</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co"># add the normalized to a list </span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>ndata <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>C <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(C) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(C) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>counts <span class="ot">=</span> C</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm10k <span class="ot">=</span> D</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co"># also add in the SCTL data</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sctl_counts <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>counts</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sctl_data <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>data</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sctl_scaledata <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>scale.data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Add celltype annotation from Azimuth. Define as broad celltypes based on the clusters for visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>celltype.file <span class="ot">=</span> <span class="st">"data/celltype_azimuth.csv"</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(celltype.file)){</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  ct <span class="ot">=</span> <span class="fu">read.csv</span>(celltype.file, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  sobj<span class="sc">$</span>celltype <span class="ot">=</span> ct<span class="sc">$</span>celltype</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  sobj<span class="sc">$</span>predicted.celltype.l1 <span class="ot">=</span> ct<span class="sc">$</span>predicted.celltype.l1</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Azimuth)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  sobj <span class="ot">&lt;-</span> <span class="fu">RunAzimuth</span>(sobj, <span class="at">reference =</span> <span class="st">"pbmcref"</span>)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  annot <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"0"</span><span class="ot">=</span><span class="st">"Mono"</span>,<span class="st">"5"</span><span class="ot">=</span><span class="st">"Mono"</span>,<span class="st">"8"</span><span class="ot">=</span><span class="st">"Other"</span>,<span class="st">"9"</span><span class="ot">=</span><span class="st">"Mix"</span>,<span class="st">"3"</span><span class="ot">=</span><span class="st">"BC"</span>,<span class="st">"6"</span><span class="ot">=</span><span class="st">"BC"</span>,<span class="st">"1"</span><span class="ot">=</span><span class="st">"NK"</span>,<span class="st">"2"</span><span class="ot">=</span><span class="st">"TC"</span>,<span class="st">"4"</span><span class="ot">=</span><span class="st">"TC"</span>, <span class="st">"7"</span><span class="ot">=</span><span class="st">"TC"</span>)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  sobj<span class="sc">$</span>celltype <span class="ot">=</span> <span class="fu">unname</span>(<span class="fu">unlist</span>(annot[<span class="fu">as.character</span>(sobj<span class="sc">@</span>active.ident)]))</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  ct <span class="ot">=</span> sobj<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">"celltype"</span>,<span class="st">"predicted.celltype.l1"</span>)]</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(ct, <span class="at">file=</span>celltype.file)</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">=</span> <span class="fu">DimPlot</span>(sobj, <span class="at">label =</span> T, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>)  <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">DimPlot</span>(sobj, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>, <span class="at">group.by =</span> <span class="st">"predicted.celltype.l1"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">=</span> <span class="fu">DimPlot</span>(sobj, <span class="at">reduction =</span> <span class="st">"umap_cca"</span>, <span class="at">group.by =</span> <span class="st">"celltype"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(p1,p2,p3, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/celltypes-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="comparison_normalization_files/figure-html/celltypes-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SetIdent</span>(sobj, <span class="at">value =</span> <span class="st">"celltype"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sct-merged" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sct-merged"><span class="header-section-number">2.3</span> SCT merged</h2>
<p>Run with all cells considered a single sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SCTransform</span>(sobj, <span class="at">new.assay.name =</span> <span class="st">"SCT"</span>, <span class="at">verbose =</span> F)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sct_data <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>data</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sct_counts <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>counts</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sobj, <span class="fu">c</span>(<span class="st">"CD3E"</span>,<span class="st">"CD14"</span>,<span class="st">"CCR7"</span>,<span class="st">"RPL10"</span>), <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="at">slot =</span> <span class="st">"data"</span>, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="comparison_normalization_files/figure-html/sct-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sobj, <span class="fu">c</span>(<span class="st">"CD3E"</span>,<span class="st">"CD14"</span>,<span class="st">"CCR7"</span>,<span class="st">"RPL10"</span>), <span class="at">assay =</span> <span class="st">"SCTL"</span>, <span class="at">slot =</span> <span class="st">"data"</span>, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="comparison_normalization_files/figure-html/sct-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Returns a Seurat object with a new assay (named SCT by default) with counts being (corrected) counts, data being log1p(counts), scale.data being pearson residuals</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17513  7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17513  7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>scale.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3000 7134</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(sobj, <span class="st">"nCount_RNA"</span>,<span class="st">"nCount_SCT"</span>, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>(),</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(sobj, <span class="st">"nFeature_RNA"</span>,<span class="st">"nFeature_SCT"</span>, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>(),</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(sobj, <span class="st">"nCount_RNA"</span>,<span class="st">"nCount_SCTL"</span>, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>(),</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(sobj, <span class="st">"nFeature_RNA"</span>,<span class="st">"nFeature_SCTL"</span>, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>)<span class="sc">+</span> <span class="fu">NoLegend</span>(),</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="comparison_normalization_files/figure-html/sct-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(sobj<span class="sc">$</span>nFeature_RNA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  201 6853</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(sobj<span class="sc">$</span>nFeature_SCT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  661 3205</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sobj, <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>,<span class="st">"nCount_SCT"</span>, <span class="st">"nCount_SCTL"</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-plot-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="comparison_normalization_files/figure-html/sct-plot-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sobj, <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>,<span class="st">"nFeature_SCT"</span>, <span class="st">"nFeature_SCTL"</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-plot-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="comparison_normalization_files/figure-html/sct-plot-3.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Much lower number of features per cell, especially in the high nF cells. But also higher lowest number of features. How can the nFeatures increase?</p>
<p>Seems like it is mainly genes that are high in most cells that get “imputed” counts in SCT, but they are still low.</p>
<p>Plot stats per gene:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>df.genes <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">nCell_RNA=</span><span class="fu">rowSums</span>(ndata<span class="sc">$</span>counts<span class="sc">&gt;</span><span class="dv">0</span>),</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">nUMI_RNA=</span><span class="fu">rowSums</span>(ndata<span class="sc">$</span>counts),</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">nCell_SCT=</span><span class="fu">rowSums</span>(ndata<span class="sc">$</span>sct_counts<span class="sc">&gt;</span><span class="dv">0</span>),</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">nUMI_SCT=</span><span class="fu">rowSums</span>(ndata<span class="sc">$</span>sct_counts))</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df.genes, <span class="fu">aes</span>(<span class="at">x=</span>nCell_RNA,<span class="at">y=</span>nCell_SCT)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>(),</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df.genes, <span class="fu">aes</span>(<span class="at">x=</span>nUMI_RNA,<span class="at">y=</span>nUMI_SCT)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>(),</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df.genes, <span class="fu">aes</span>(<span class="at">x=</span>nUMI_RNA,<span class="at">y=</span>nCell_RNA)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">scale_x_log10</span>(),</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df.genes, <span class="fu">aes</span>(<span class="at">x=</span>nUMI_SCT,<span class="at">y=</span>nCell_SCT)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">scale_x_log10</span>(),</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sct-gene-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="comparison_normalization_files/figure-html/sct-gene-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Each gene is found in a similar number of cells, but the nUMI per gene is much lower with SCT</p>
<p>Get residuals for all of the hvgs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sobj<span class="sc">@</span>active.assay <span class="ot">=</span> <span class="st">"SCT"</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">PrepSCTIntegration</span>(<span class="fu">list</span>(<span class="at">s =</span> sobj), <span class="at">anchor.features =</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(sobj), hvg))<span class="sc">$</span>s</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># some of the hvgs are not in the SCT assay!</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>sct_scaledata <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>scale.data</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>scale.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4972 7134</code></pre>
</div>
</div>
<p>Many cells have counts for genes where there was no detection. Check for instance chrY genes and the XIST gene. That we know should mainly be in males/females.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>genes_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_seurat, <span class="st">"genes_table.csv"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>genes.table <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(genes_file)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>genes.table <span class="ot">&lt;-</span> genes.table[genes.table<span class="sc">$</span>external_gene_name <span class="sc">%in%</span> <span class="fu">rownames</span>(sobj), ]</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>par1 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">10001</span>, <span class="dv">2781479</span>)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>par2 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">56887903</span>, <span class="dv">57217415</span>)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>p1.gene <span class="ot">=</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>start_position <span class="sc">&gt;</span> par1[<span class="dv">1</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>start_position <span class="sc">&lt;</span> par1[<span class="dv">2</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>p2.gene <span class="ot">=</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>start_position <span class="sc">&gt;</span> par2[<span class="dv">1</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>start_position <span class="sc">&lt;</span> par2[<span class="dv">2</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>chrY.gene <span class="ot">&lt;-</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>chrY.gene <span class="ot">=</span> <span class="fu">setdiff</span>(chrY.gene, <span class="fu">c</span>(p1.gene, p2.gene))</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sobj, <span class="at">features =</span> chrY.gene, <span class="at">col.name =</span> <span class="st">"pct_chrY_SCT"</span>, <span class="at">assay =</span> <span class="st">"SCT"</span>)</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTL has fewer genes in the count matrix, </span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sobj, <span class="at">features =</span> <span class="fu">intersect</span>(chrY.gene, <span class="fu">rownames</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>SCTL<span class="sc">@</span>counts)), <span class="at">col.name =</span> <span class="st">"pct_chrY_SCTL"</span>, <span class="at">assay =</span> <span class="st">"SCTL"</span>)</span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sobj, <span class="at">features =</span> chrY.gene, <span class="at">col.name =</span> <span class="st">"pct_chrY_RNA"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"pct_chrY_RNA"</span>, <span class="st">"pct_chrY_SCT"</span>,<span class="st">"pct_chrY_SCTL"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/chrYgenes-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="comparison_normalization_files/figure-html/chrYgenes-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>ChrY genes have low counts in the female cells with SCT. But not in SCTL which is run per sample, so SCT borrows information across the samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">=</span> <span class="fu">VlnPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">features =</span> <span class="st">"XIST"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">slot =</span> <span class="st">"data"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">VlnPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">features =</span> <span class="st">"XIST"</span>, <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="at">slot =</span> <span class="st">"data"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">=</span> <span class="fu">VlnPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">features =</span> <span class="st">"XIST"</span>, <span class="at">assay =</span> <span class="st">"SCTL"</span>, <span class="at">slot =</span> <span class="st">"data"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/sex-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="comparison_normalization_files/figure-html/sex-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Same for XIST expression.</p>
</section>
<section id="subsample" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="subsample"><span class="header-section-number">2.4</span> Subsample</h2>
<p>From now, use a smaller dataset.</p>
<p>Select 10 cells per sample and cluster to make a smaller dataset for the comparison. Use resolution 1 with 11 clusters.</p>
<p>Then filter again for genes detected in at least 5 cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17513  7134</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>sobj<span class="sc">@</span>active.assay <span class="ot">=</span> <span class="st">"RNA"</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">DietSeurat</span>(sobj, <span class="at">assays =</span> <span class="st">"RNA"</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>sobj<span class="sc">$</span>clust_sample <span class="ot">=</span> <span class="fu">paste0</span>(sobj<span class="sc">$</span>RNA_snn_res<span class="fl">.1</span>, sobj<span class="sc">$</span>orig.ident)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SetIdent</span>(sobj, <span class="at">value =</span> <span class="st">"clust_sample"</span>)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">subset</span>(sobj, <span class="at">cells =</span> <span class="fu">WhichCells</span>(sobj, <span class="at">downsample =</span> <span class="dv">10</span>))</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>nC <span class="ot">=</span> <span class="fu">rowSums</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> sobj[nC<span class="sc">&gt;</span><span class="dv">4</span>,]</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 13363   928</code></pre>
</div>
</div>
<p>Also subset all normdata for the same cells/genes, and also the subset of variable genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>hvg <span class="ot">=</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(sobj),hvg)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(ndata)){</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">intersect</span>(hvg, <span class="fu">rownames</span>(ndata[[n]]))</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  ndata[[n]] <span class="ot">=</span> ndata[[n]][tmp,<span class="fu">colnames</span>(sobj)]</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="co"># clean memory </span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           used  (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
Ncells  9499379 507.4   16614425  887.4         NA  16614425  887.4
Vcells 81517207 622.0  586667556 4476.0      65536 733229785 5594.1</code></pre>
</div>
</div>
</section>
<section id="scalefactors-lognorm" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="scalefactors-lognorm"><span class="header-section-number">2.5</span> Scalefactors lognorm</h2>
<p>First plot counts and default lognorm (sf=10K) and calculate scale.data for the different normalizations.</p>
<p>Also calculate hvgs to get dispersions at different scale factors.</p>
<section id="sf-10k" class="level3" data-number="2.5.1">
<h3 data-number="2.5.1" class="anchored" data-anchor-id="sf-10k"><span class="header-section-number">2.5.1</span> SF 10K</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">SetIdent</span>(sobj, <span class="at">value =</span>  <span class="st">"RNA_snn_res.0.5"</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">#VlnPlot(sobj, c("CD3E","CD14","CCR7","RPL10"), slot = "counts", ncol=4)</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co">#VlnPlot(sobj, c("CD3E","CD14","CCR7","RPL10"), slot = "data", ncol=4)</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co"># vagenes for 10k</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>vinfo <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(sobj, <span class="at">selection.method =</span> <span class="st">"mean.var.plot"</span>, <span class="at">verbose =</span> F, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>vinfo[[<span class="st">"10k"</span>]] <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">ScaleData</span>(sobj, <span class="at">features =</span> hvg, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>scale.data</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> hvg</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm10k_scaledata <span class="ot">=</span> D</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>cd3plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>cd3plots[[<span class="dv">1</span>]] <span class="ot">=</span> <span class="fu">VlnPlot</span>(sobj, <span class="at">features =</span> <span class="st">"CD3E"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"10K"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sf-1k" class="level3" data-number="2.5.2">
<h3 data-number="2.5.2" class="anchored" data-anchor-id="sf-1k"><span class="header-section-number">2.5.2</span> SF 1K</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">NormalizeData</span>(sobj, <span class="at">scale.factor =</span> <span class="dv">1000</span>, <span class="at">verbose =</span> F)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(tmp, <span class="at">selection.method =</span> <span class="st">"mean.var.plot"</span>, <span class="at">verbose =</span> F, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>vinfo[[<span class="st">"1k"</span>]] <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm1k <span class="ot">=</span> D[hvg,]</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a><span class="co">#VlnPlot(tmp, c("CD3E","CD14","CCR7","RPL10"), slot = "data", ncol=4)</span></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">ScaleData</span>(tmp, <span class="at">features =</span> hvg, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>scale.data</span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> hvg</span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm1k_scaledata <span class="ot">=</span> D</span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>cd3plots[[<span class="dv">2</span>]] <span class="ot">=</span> <span class="fu">VlnPlot</span>(tmp, <span class="at">features =</span> <span class="st">"CD3E"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"1K"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sf-100k" class="level3" data-number="2.5.3">
<h3 data-number="2.5.3" class="anchored" data-anchor-id="sf-100k"><span class="header-section-number">2.5.3</span> SF 100K</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">NormalizeData</span>(sobj, <span class="at">scale.factor =</span> <span class="dv">100000</span>, <span class="at">verbose =</span> F)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm100k <span class="ot">=</span> D[hvg,]</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">ScaleData</span>(tmp, <span class="at">features =</span> hvg, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>scale.data</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> hvg</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>lognorm100k_scaledata <span class="ot">=</span> D</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(tmp, <span class="at">selection.method =</span> <span class="st">"mean.var.plot"</span>, <span class="at">verbose =</span> F, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>vinfo[[<span class="st">"100k"</span>]] <span class="ot">=</span> tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.data</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a><span class="co">#VlnPlot(tmp, c("CD3E","CD14","CCR7","RPL10"), slot = "data", ncol=4)</span></span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a>cd3plots[[<span class="dv">3</span>]] <span class="ot">=</span> <span class="fu">VlnPlot</span>(tmp, <span class="at">features =</span> <span class="st">"CD3E"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"100K"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(cd3plots, <span class="at">ncol =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/lognorm100k-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="comparison_normalization_files/figure-html/lognorm100k-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Clearly gives a larger difference to zero with larger size factors. So the dispersion will be higher.</p>
<p>In principle is the same as changing the pseudocount which is now 1.</p>
</section>
<section id="dispersions" class="level3" data-number="2.5.4">
<h3 data-number="2.5.4" class="anchored" data-anchor-id="dispersions"><span class="header-section-number">2.5.4</span> Dispersions</h3>
<p>Plot the estimated dispersions with mvp vs mean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>vinfo <span class="ot">=</span> vinfo[<span class="fu">sort</span>(<span class="fu">names</span>(vinfo))]</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(vinfo)){</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  plots[[n]] <span class="ot">=</span> <span class="fu">ggplot</span>(vinfo[[n]], <span class="fu">aes</span>(<span class="at">x=</span>vf_mvp_data_mvp.mean, <span class="at">y=</span>vf_mvp_data_mvp.dispersion.scaled, <span class="at">color =</span> vf_mvp_data_variable)) <span class="sc">+</span> <span class="fu">geom_point</span>()<span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="fu">paste0</span>(n, <span class="st">" - "</span>, <span class="fu">sum</span>(vinfo[[n]]<span class="sc">$</span>vf_mvp_data_variable), <span class="st">" hvgs"</span>)) <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() </span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/dispersions-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="comparison_normalization_files/figure-html/dispersions-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Much higher dispersions and more variable genes with 100K, very low with 1k.</p>
<p>The cutoffs are dispersions over 1 and mean 0.1-8</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>disp <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(cbind,<span class="fu">lapply</span>(vinfo, <span class="cf">function</span>(x) x<span class="sc">$</span>vf_mvp_data_mvp.dispersion.scaled)))</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(disp) <span class="ot">=</span> <span class="fu">names</span>(vinfo)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(disp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/dispersions-pair-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="comparison_normalization_files/figure-html/dispersions-pair-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Still similar trend on the dispersions, the same genes have high dispersion.</p>
</section>
</section>
<section id="logcounts" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="logcounts"><span class="header-section-number">2.6</span> Logcounts</h2>
<p>Log of counts without depth normalization</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>logcounts <span class="ot">=</span> <span class="fu">log1p</span>(ndata<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tmm" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="tmm"><span class="header-section-number">2.7</span> TMM</h2>
<p>Run EdgeR TMM,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># TMM</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">=</span> edgeR<span class="sc">::</span><span class="fu">DGEList</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">normLibSizes</span>(dge, <span class="at">method =</span> <span class="st">"TMM"</span>)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>tmp  <span class="ot">=</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge) <span class="co"># is with log = FALSE</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> tmp[<span class="fu">rownames</span>(sobj) <span class="sc">%in%</span> hvg,]</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> hvg</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>tmm <span class="ot">=</span> tmp</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(dge, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> tmp[<span class="fu">rownames</span>(sobj) <span class="sc">%in%</span> hvg,]</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> hvg</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>tmm_log <span class="ot">=</span> tmp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quantile" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="quantile"><span class="header-section-number">2.8</span> Quantile</h2>
<p>limma-voom quantile</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># should be run on the lognorm values. use the full dataset to calculate.</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> limma<span class="sc">::</span><span class="fu">voom</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data,<span class="at">normalize.method =</span> <span class="st">"quantile"</span>)<span class="sc">$</span>E</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>quantile <span class="ot">=</span> tmp[hvg,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="deconvolution" class="level2" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="deconvolution"><span class="header-section-number">2.9</span> Deconvolution</h2>
<p>Run <code>quickCluster</code> and then <code>computeSumFactors</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">as.SingleCellExperiment</span>(sobj)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">=</span> <span class="fu">quickCluster</span>(sce)</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cl
  1   2   3   4   5   6 
172 157 159 186 126 128 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(sce, <span class="at">cluster=</span>cl, <span class="at">min.mean=</span><span class="fl">0.1</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">assayNames</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"    "logcounts"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">assay</span>(sce, <span class="st">"logcounts"</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> tmp[hvg,]</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>deconv <span class="ot">=</span> tmp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pearson-residuals" class="level2" data-number="2.10">
<h2 data-number="2.10" class="anchored" data-anchor-id="pearson-residuals"><span class="header-section-number">2.10</span> Pearson residuals</h2>
<p>Using the method implemented in scanpy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>penv <span class="ot">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/scanpy_2024_nopip"</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>norm.scanpy <span class="ot">=</span> <span class="fu">basiliskRun</span>(<span class="at">env=</span>penv, <span class="at">fun=</span><span class="cf">function</span>(counts) {</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>    scanpy <span class="ot">&lt;-</span> reticulate<span class="sc">::</span><span class="fu">import</span>(<span class="st">"scanpy"</span>)</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    ad <span class="ot">=</span> reticulate<span class="sc">::</span><span class="fu">import</span>(<span class="st">"anndata"</span>)</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    adata <span class="ot">=</span> ad<span class="sc">$</span><span class="fu">AnnData</span>(counts)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#sc.experimental.pp.normalize_pearson_residuals(adata)</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>experimental<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">normalize_pearson_residuals</span>(adata)</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">norm=</span>adata<span class="sc">$</span>X))</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>}, <span class="at">counts =</span> <span class="fu">t</span>(sobj<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts),  <span class="at">testload=</span><span class="st">"scanpy"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>10 x 10 sparse Matrix of class "dgCMatrix"
                         
 [1,] . . . . . . . . . .
 [2,] . . . . . . . . . .
 [3,] . . . . . . . . . .
 [4,] . . . 3 . . . 2 . .
 [5,] . . . . . . 1 . . .
 [6,] . . . . . . . . . .
 [7,] . . . . . . . 1 . .
 [8,] . . . . . . . . . .
 [9,] . . . 1 . . . . . .
[10,] . . 1 . . . . . . .</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(norm.scanpy<span class="sc">$</span>norm) <span class="ot">=</span> <span class="fu">colnames</span>(sobj)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(norm.scanpy<span class="sc">$</span>norm) <span class="ot">=</span> <span class="fu">rownames</span>(sobj)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>ndata<span class="sc">$</span>pearson <span class="ot">=</span> <span class="fu">t</span>(norm.scanpy<span class="sc">$</span>norm[,hvg])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="cd3-violins" class="level2" data-number="2.11">
<h2 data-number="2.11" class="anchored" data-anchor-id="cd3-violins"><span class="header-section-number">2.11</span> CD3 violins</h2>
<p>Plot cd3 violins for these different normalizations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>cd3 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(cbind, <span class="fu">lapply</span>(ndata, <span class="cf">function</span>(x) x[<span class="st">"CD3E"</span>,])))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cd3)<span class="ot">=</span><span class="fu">names</span>(ndata)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>cd3<span class="sc">$</span>clusters <span class="ot">=</span> sobj<span class="sc">@</span>active.ident</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(cd3, <span class="at">id =</span> <span class="st">"clusters"</span>)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>variable <span class="ot">=</span> <span class="fu">factor</span>(df<span class="sc">$</span>variable, <span class="at">levels =</span> <span class="fu">sort</span>(<span class="fu">names</span>(ndata)))</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>clusters, <span class="at">y=</span>value, <span class="at">fill=</span>clusters)) <span class="sc">+</span> <span class="fu">geom_violin</span>(<span class="at">trim =</span> T, <span class="at">scale =</span> <span class="st">"width"</span> , <span class="at">adjust =</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/cd3-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="comparison_normalization_files/figure-html/cd3-1.png" class="img-fluid figure-img" width="1344"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="distributions" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Distributions</h1>
<p>Plot the expression distribution for a few single cells using each method. Select cells with highest/medium/lowest total counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>cs <span class="ot">=</span> <span class="fu">colSums</span>(ndata<span class="sc">$</span>counts)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>r <span class="ot">=</span> <span class="fu">order</span>(cs)</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">=</span> r[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">460</span>,<span class="fu">length</span>(r))]</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>cs[sel]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  covid_1_TTGTGTTAGACATAAC-1   covid_1_CGCATGGTCGATACAC-1 
                          90                         1226 
covid_15_CACGAATAGAAGCCAC-15 
                       20443 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>subset <span class="ot">=</span> <span class="fu">lapply</span>(ndata, <span class="cf">function</span>(x) x[,sel])</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ds <span class="cf">in</span> <span class="fu">sort</span>(<span class="fu">names</span>(subset))){</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(<span class="fu">data.frame</span>(subset[[ds]]))</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  plots[[ds]] <span class="ot">=</span> <span class="fu">ggplot</span>(l, <span class="fu">aes</span>(<span class="at">x=</span>value)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">200</span>) <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable, <span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span> <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(ds) </span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], <span class="at">ncol=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/distributions-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="comparison_normalization_files/figure-html/distributions-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots[<span class="dv">7</span><span class="sc">:</span><span class="dv">12</span>], <span class="at">ncol=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/distributions-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="comparison_normalization_files/figure-html/distributions-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots[<span class="dv">13</span><span class="sc">:</span><span class="fu">length</span>(plots)], <span class="at">ncol=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/distributions-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="comparison_normalization_files/figure-html/distributions-3.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cell-correlations" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Cell correlations</h1>
<p>For all the different normalizations, calculate the correlation between cells. For a fair comparison, use the same set of genes.</p>
<p>Except for the scaledata for SCTL that has too few genes!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># corSparse function copied from Signac package utilities.R</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>corSparse <span class="ot">&lt;-</span> <span class="cf">function</span>(X, <span class="at">Y =</span> <span class="cn">NULL</span>, <span class="at">cov =</span> <span class="cn">FALSE</span>) {</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  X <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="at">object =</span> X, <span class="at">Class =</span> <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(<span class="at">x =</span> X)</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>  muX <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(<span class="at">x =</span> X)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(<span class="at">x =</span> Y)) {</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(<span class="at">x =</span> X) <span class="sc">!=</span> <span class="fu">nrow</span>(<span class="at">x =</span> Y)) {</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">stop</span>(<span class="st">"Matrices must contain the same number of rows"</span>)</span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>    Y <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="at">object =</span> Y, <span class="at">Class =</span> <span class="st">"CsparseMatrix"</span>)</span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>    muY <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(<span class="at">x =</span> Y)</span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a>    covmat <span class="ot">&lt;-</span> ( <span class="fu">as.matrix</span>(<span class="at">x =</span> <span class="fu">crossprod</span>(<span class="at">x =</span> X, <span class="at">y =</span> Y)) <span class="sc">-</span> n <span class="sc">*</span> <span class="fu">tcrossprod</span>(<span class="at">x =</span> muX, <span class="at">y =</span> muY) ) <span class="sc">/</span> (n<span class="dv">-1</span>)</span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a>    sdvecX <span class="ot">&lt;-</span> <span class="fu">sqrt</span>( (<span class="fu">colSums</span>(<span class="at">x =</span> X<span class="sc">^</span><span class="dv">2</span>) <span class="sc">-</span> n<span class="sc">*</span>muX<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> (n<span class="dv">-1</span>) )</span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a>    sdvecY <span class="ot">&lt;-</span> <span class="fu">sqrt</span>( (<span class="fu">colSums</span>(<span class="at">x =</span> Y<span class="sc">^</span><span class="dv">2</span>) <span class="sc">-</span> n<span class="sc">*</span>muY<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> (n<span class="dv">-1</span>) )</span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a>    cormat <span class="ot">&lt;-</span> covmat <span class="sc">/</span> <span class="fu">tcrossprod</span>(<span class="at">x =</span> sdvecX, <span class="at">y =</span> sdvecY)</span>
<span id="cb73-18"><a href="#cb73-18" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {      </span>
<span id="cb73-19"><a href="#cb73-19" aria-hidden="true" tabindex="-1"></a>    covmat <span class="ot">&lt;-</span> ( <span class="fu">as.matrix</span>(<span class="fu">crossprod</span>(<span class="at">x =</span> X)) <span class="sc">-</span> n <span class="sc">*</span> <span class="fu">tcrossprod</span>(<span class="at">x =</span> muX) ) <span class="sc">/</span> (n<span class="dv">-1</span>)</span>
<span id="cb73-20"><a href="#cb73-20" aria-hidden="true" tabindex="-1"></a>    sdvec <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="at">x =</span> <span class="fu">diag</span>(<span class="at">x =</span> covmat))</span>
<span id="cb73-21"><a href="#cb73-21" aria-hidden="true" tabindex="-1"></a>    cormat <span class="ot">&lt;-</span> covmat <span class="sc">/</span> <span class="fu">tcrossprod</span>(<span class="at">x =</span> sdvec)</span>
<span id="cb73-22"><a href="#cb73-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb73-23"><a href="#cb73-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (cov) {</span>
<span id="cb73-24"><a href="#cb73-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dimnames</span>(<span class="at">x =</span> covmat) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb73-25"><a href="#cb73-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(covmat)</span>
<span id="cb73-26"><a href="#cb73-26" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb73-27"><a href="#cb73-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dimnames</span>(<span class="at">x =</span> cormat) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb73-28"><a href="#cb73-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(cormat)</span>
<span id="cb73-29"><a href="#cb73-29" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb73-30"><a href="#cb73-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># have fewer genes for sctl-scaledata, still use the same set for all other samples ignoring that sctl has fewer!</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>t <span class="ot">=</span> <span class="fu">table</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(ndata, rownames)))</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>selG <span class="ot">=</span> <span class="fu">names</span>(t)[t <span class="sc">&gt;=</span> (<span class="fu">length</span>(ndata)<span class="sc">-</span><span class="dv">1</span>)] </span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>ndata <span class="ot">=</span> <span class="fu">lapply</span>(ndata, <span class="cf">function</span>(x) x[<span class="fu">intersect</span>(<span class="fu">rownames</span>(x),selG),])</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>cors <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">sort</span>(<span class="fu">names</span>(ndata))){</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(n)</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>  cors[[n]] <span class="ot">=</span> <span class="fu">corSparse</span>(ndata[[n]])</span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"
[1] "deconv"
[1] "logcounts"
[1] "lognorm100k"
[1] "lognorm100k_scaledata"
[1] "lognorm10k"
[1] "lognorm10k_scaledata"
[1] "lognorm1k"
[1] "lognorm1k_scaledata"
[1] "pearson"
[1] "quantile"
[1] "sct_counts"
[1] "sct_data"
[1] "sct_scaledata"
[1] "sctl_counts"
[1] "sctl_data"
[1] "sctl_scaledata"
[1] "tmm"
[1] "tmm_log"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">#cors = lapply(ndata, function(x) corSparse(x))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>breaks <span class="ot">=</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="dv">1</span>,<span class="fl">0.01</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>cc <span class="ot">=</span> <span class="fu">lapply</span>(cors, <span class="cf">function</span>(x) </span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hist</span>(x[<span class="fu">upper.tri</span>(x)],<span class="at">breaks=</span>breaks,<span class="at">plot=</span>F)<span class="sc">$</span>counts)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>count.hist <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(cbind,cc))</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count.hist) <span class="ot">=</span> <span class="fu">names</span>(ndata)</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>count.hist<span class="sc">$</span>breaks <span class="ot">=</span> breaks[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>c <span class="ot">=</span> <span class="fu">Reduce</span>(cbind, <span class="fu">lapply</span>(cors, <span class="cf">function</span>(x) x[<span class="fu">upper.tri</span>(x)]))</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(c) <span class="ot">=</span> <span class="fu">names</span>(cors)</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>long <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(c)</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(long, <span class="fu">aes</span>(<span class="at">x=</span>value)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">100</span>) <span class="sc">+</span> </span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Var2, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span><span class="dv">0</span>, <span class="at">color=</span><span class="st">"red"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/correl-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="comparison_normalization_files/figure-html/correl-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">=</span> <span class="fu">apply</span>(c, <span class="dv">2</span>, summary)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                               Min.     1st Qu.       Median          Mean
counts                -5.088535e-03  0.23230169  0.359057137  0.3876623832
deconv                 8.952432e-04  0.26950424  0.339649442  0.3531730611
logcounts              7.587383e-04  0.27896786  0.341350609  0.3563585870
lognorm100k           -1.893007e-04  0.21737202  0.269881348  0.2823120771
lognorm100k_scaledata -1.358677e-01 -0.04628789 -0.009354206 -0.0006189914
lognorm10k             5.124563e-04  0.25871941  0.326215623  0.3391142020
lognorm10k_scaledata  -1.526215e-01 -0.04516235 -0.010395184 -0.0008814526
lognorm1k              9.819781e-04  0.31014511  0.394088563  0.4114224291
lognorm1k_scaledata   -1.701344e-01 -0.04335472 -0.011190967 -0.0008133232
pearson               -2.726683e-01 -0.06906390 -0.003071477  0.0081909333
quantile               1.156027e-03  0.23670468  0.296213401  0.3091003740
sct_counts             5.518484e-03  0.27753797  0.393958830  0.4186517500
sct_data               1.465805e-01  0.32142427  0.368884265  0.3945946018
sct_scaledata         -2.323186e-01 -0.07915305 -0.010833788  0.0067053275
sctl_counts            4.692851e-03  0.26913139  0.385389128  0.4121448393
sctl_data              1.394500e-01  0.31327415  0.359751049  0.3863770802
sctl_scaledata        -3.169345e-01 -0.09450794 -0.015246042  0.0047925770
tmm                   -5.088535e-03  0.23230169  0.359057137  0.3876623832
tmm_log               -1.111245e-05  0.29547126  0.367618514  0.3826587382
                         3rd Qu.      Max.
counts                0.51888427 0.9979323
deconv                0.41745624 0.8869059
logcounts             0.41555708 0.8166209
lognorm100k           0.33086334 0.7858652
lognorm100k_scaledata 0.03124555 0.6449176
lognorm10k            0.40025879 0.8698278
lognorm10k_scaledata  0.02789470 0.6953211
lognorm1k             0.49042499 0.9429184
lognorm1k_scaledata   0.02594518 0.7124583
pearson               0.06466442 0.6892138
quantile              0.36465779 0.8355153
sct_counts            0.54749577 0.9975654
sct_data              0.46084588 0.7892639
sct_scaledata         0.06634267 0.7227993
sctl_counts           0.54059869 0.9971110
sctl_data             0.45209032 0.8058592
sctl_scaledata        0.07696892 0.7767212
tmm                   0.51888427 0.9979323
tmm_log               0.44942367 0.8848754</code></pre>
</div>
</div>
</section>
<section id="clustering" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Clustering</h1>
<p>Run a quick umap/clustering with the different normalizations. All with same set of genes, pca with 20 PCs,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need to run the commands to get all setting correct in the seurat object. </span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>sobj.all <span class="ot">=</span>  <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> ndata<span class="sc">$</span>counts[selG,])</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>sobj.all <span class="ot">=</span> <span class="fu">NormalizeData</span>(sobj.all, <span class="at">verbose =</span> F)</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>sobj.all <span class="ot">=</span> <span class="fu">ScaleData</span>(sobj.all, <span class="at">verbose =</span> F)</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>sobj.all<span class="sc">$</span>old_clust <span class="ot">=</span> sobj<span class="sc">$</span>RNA_snn_res.<span class="fl">0.5</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>sobj.all<span class="sc">$</span>sample <span class="ot">=</span> sobj<span class="sc">$</span>orig.ident</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a><span class="co"># then run per dataset</span></span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ds <span class="cf">in</span> <span class="fu">sort</span>(<span class="fu">names</span>(ndata))){</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ds <span class="sc">==</span> <span class="st">"sctl_scaledata"</span>) { <span class="cf">next</span> }</span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>  d <span class="ot">=</span> <span class="fu">as</span>(ndata[[ds]][selG,], <span class="st">"matrix"</span>)</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>  sobj.all<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>scale.data <span class="ot">=</span> d</span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>  sobj.all <span class="ot">=</span> sobj.all <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">npcs=</span><span class="dv">50</span>,<span class="at">verbose =</span> F, <span class="at">features =</span> selG) <span class="sc">%&gt;%</span> <span class="fu">RunUMAP</span>(<span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">FindNeighbors</span>(<span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">FindClusters</span>(<span class="at">resolution =</span> <span class="fl">0.6</span>, <span class="at">verbose =</span> F)</span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>  sobj.all[[<span class="fu">paste0</span>(<span class="st">"umap_"</span>,ds)]] <span class="ot">=</span> sobj.all[[<span class="st">"umap"</span>]]</span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>  sobj.all[[<span class="fu">paste0</span>(<span class="st">"clusters_"</span>,ds)]] <span class="ot">=</span> sobj.all[[<span class="st">"seurat_clusters"</span>]]</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>  sobj.all[[<span class="fu">paste0</span>(<span class="st">"pca_"</span>,ds)]] <span class="ot">=</span> sobj.all[[<span class="st">"pca"</span>]]</span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Gene detection in all:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ds <span class="cf">in</span> <span class="fu">sort</span>(<span class="fu">names</span>(ndata))){</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ds <span class="sc">==</span> <span class="st">"sctl_scaledata"</span>) { <span class="cf">next</span> }</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  plots[[ds]] <span class="ot">=</span> <span class="fu">FeaturePlot</span>(sobj.all, <span class="at">features =</span> <span class="st">"nFeature_RNA"</span>, <span class="at">reduction =</span> <span class="fu">paste0</span>(<span class="st">"umap_"</span>,ds), <span class="at">pt.size =</span> .<span class="dv">2</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(ds) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>))</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/features-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="comparison_normalization_files/figure-html/features-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Celltype predictions in all:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>sobj.all<span class="sc">$</span>celltype <span class="ot">=</span> sobj[,<span class="fu">colnames</span>(sobj.all)]<span class="sc">$</span>celltype</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ds <span class="cf">in</span> <span class="fu">sort</span>(<span class="fu">names</span>(ndata))){</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ds <span class="sc">==</span> <span class="st">"sctl_scaledata"</span>) { <span class="cf">next</span> }</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  plots[[ds]] <span class="ot">=</span> <span class="fu">DimPlot</span>(sobj.all, <span class="at">group.by =</span> <span class="st">"celltype"</span>, <span class="at">label =</span> T, <span class="at">label.size =</span> <span class="dv">2</span>, <span class="at">reduction =</span> <span class="fu">paste0</span>(<span class="st">"umap_"</span>,ds), <span class="at">pt.size =</span> .<span class="dv">2</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(ds) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>))</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/celltype-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="comparison_normalization_files/figure-html/celltype-plot-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<section id="pcs-stats" class="level3" data-number="5.0.1">
<h3 data-number="5.0.1" class="anchored" data-anchor-id="pcs-stats"><span class="header-section-number">5.0.1</span> PCs stats</h3>
<section id="contribution-to-variance" class="level4" data-number="5.0.1.1">
<h4 data-number="5.0.1.1" class="anchored" data-anchor-id="contribution-to-variance"><span class="header-section-number">5.0.1.1</span> Contribution to variance</h4>
<p>How much variance is in each pc, among the 50 PCs calculated. Plot for first 10 PCs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>nplot <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>pcs <span class="ot">=</span> <span class="fu">names</span>(sobj.all<span class="sc">@</span>reductions)[<span class="fu">grepl</span>(<span class="st">"pca_"</span>, <span class="fu">names</span>(sobj.all<span class="sc">@</span>reductions))]</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>std <span class="ot">=</span> <span class="fu">lapply</span>(pcs, <span class="cf">function</span>(x) <span class="fu">Stdev</span>(sobj.all, <span class="at">reduction =</span> x))</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>variance_explained <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(cbind,<span class="fu">lapply</span>(std, <span class="cf">function</span>(x) (x<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>(x<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> <span class="dv">100</span>)[<span class="dv">1</span><span class="sc">:</span>nplot])))</span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(variance_explained) <span class="ot">=</span> pcs</span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>variance_explained<span class="sc">$</span>PC <span class="ot">=</span> <span class="fu">factor</span>(<span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(variance_explained)), <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(variance_explained)))</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(variance_explained[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,])</span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>PC, <span class="at">y=</span>value)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/pcs-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="comparison_normalization_files/figure-html/pcs-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>For TMM and quantile, pretty much all variance is in PC1.</p>
<p>More flat for the scaled data and for residuals.</p>
<p>Also more flat with lower size factors.</p>
</section>
<section id="pc-correlation-to-different-stats" class="level4" data-number="5.0.1.2">
<h4 data-number="5.0.1.2" class="anchored" data-anchor-id="pc-correlation-to-different-stats"><span class="header-section-number">5.0.1.2</span> PC correlation to different stats</h4>
<p>Correlation to nFeatures - plot as R-squared.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>emb <span class="ot">=</span> <span class="fu">lapply</span>(pcs, <span class="cf">function</span>(x) <span class="fu">Embeddings</span>(sobj.all, <span class="at">reduction =</span> x))</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>plot_pc_correl <span class="ot">=</span> <span class="cf">function</span>(emb, feature){</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>  f <span class="ot">=</span> <span class="fu">as.matrix</span>(sobj[[feature]])</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>  cors <span class="ot">=</span> <span class="fu">Reduce</span>(rbind,<span class="fu">lapply</span>(emb, <span class="cf">function</span>(x) {</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">apply</span>(x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>],<span class="dv">2</span>,<span class="cf">function</span>(y) <span class="fu">summary</span>(<span class="fu">lm</span>(y <span class="sc">~</span> f))<span class="sc">$</span>r.squared)</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>    }))</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">t</span>(cors))</span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(df) <span class="ot">=</span> pcs</span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>PC <span class="ot">=</span> <span class="fu">factor</span>(<span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>), <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  df2 <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(df)</span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>  df2<span class="sc">$</span>value <span class="ot">=</span> <span class="fu">abs</span>(df2<span class="sc">$</span>value)</span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-16"><a href="#cb84-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(df2, <span class="fu">aes</span>(<span class="at">x=</span>PC, <span class="at">y=</span>value)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable)</span>
<span id="cb84-17"><a href="#cb84-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb84-18"><a href="#cb84-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-19"><a href="#cb84-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pc_correl</span>(emb, <span class="st">"nFeature_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/pcs-features-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="comparison_normalization_files/figure-html/pcs-features-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Most methods have PC1 dominated by nFeatures, but more so for logcounts, lognorm with large size factors and bulk methods.</p>
<p>Same for mito genes and total counts:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pc_correl</span>(emb, <span class="st">"percent_mito"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/pcs-qc-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="comparison_normalization_files/figure-html/pcs-qc-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pc_correl</span>(emb, <span class="st">"nCount_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/pcs-qc-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26"><img src="comparison_normalization_files/figure-html/pcs-qc-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Vs celltypes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pc_correl</span>(emb, <span class="st">"celltype"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_normalization_files/figure-html/pcs-celltype-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-27"><img src="comparison_normalization_files/figure-html/pcs-celltype-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="meta-session" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Session info</h1>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/asabjor/miniconda3/envs/seurat5_u/lib/libopenblasp-r0.3.28.dylib;  LAPACK version 3.12.0

locale:
[1] sv_SE.UTF-8/sv_SE.UTF-8/sv_SE.UTF-8/C/sv_SE.UTF-8/sv_SE.UTF-8

time zone: Europe/Stockholm
tzcode source: system (macOS)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] pheatmap_1.0.12             dplyr_1.1.4                
 [3] basilisk_1.14.1             scran_1.30.0               
 [5] scuttle_1.12.0              SingleCellExperiment_1.24.0
 [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             
 [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
[11] IRanges_2.36.0              S4Vectors_0.40.2           
[13] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[15] matrixStats_1.5.0           patchwork_1.3.0            
[17] ggplot2_3.5.1               Matrix_1.6-5               
[19] Seurat_5.1.0                SeuratObject_5.0.2         
[21] sp_2.1-4                   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.3.3            
  [3] later_1.4.1               bitops_1.0-9             
  [5] filelock_1.0.3            tibble_3.2.1             
  [7] polyclip_1.10-7           basilisk.utils_1.14.1    
  [9] fastDummies_1.7.4         lifecycle_1.0.4          
 [11] edgeR_4.0.16              globals_0.16.3           
 [13] lattice_0.22-6            MASS_7.3-60.0.1          
 [15] magrittr_2.0.3            limma_3.58.1             
 [17] plotly_4.10.4             rmarkdown_2.29           
 [19] remotes_2.5.0             yaml_2.3.10              
 [21] metapod_1.10.0            httpuv_1.6.15            
 [23] glmGamPoi_1.14.0          sctransform_0.4.1        
 [25] spam_2.11-0               sessioninfo_1.2.2        
 [27] pkgbuild_1.4.5            spatstat.sparse_3.1-0    
 [29] reticulate_1.40.0         cowplot_1.1.3            
 [31] pbapply_1.7-2             RColorBrewer_1.1-3       
 [33] pkgload_1.4.0             abind_1.4-5              
 [35] zlibbioc_1.48.0           Rtsne_0.17               
 [37] purrr_1.0.2               RCurl_1.98-1.16          
 [39] GenomeInfoDbData_1.2.11   ggrepel_0.9.6            
 [41] irlba_2.3.5.1             listenv_0.9.1            
 [43] spatstat.utils_3.1-2      goftest_1.2-3            
 [45] RSpectra_0.16-2           spatstat.random_3.3-2    
 [47] dqrng_0.3.2               fitdistrplus_1.2-2       
 [49] parallelly_1.41.0         DelayedMatrixStats_1.24.0
 [51] leiden_0.4.3.1            codetools_0.2-20         
 [53] DelayedArray_0.28.0       tidyselect_1.2.1         
 [55] farver_2.1.2              ScaledMatrix_1.10.0      
 [57] spatstat.explore_3.3-4    jsonlite_1.8.9           
 [59] BiocNeighbors_1.20.0      ellipsis_0.3.2           
 [61] progressr_0.15.1          ggridges_0.5.6           
 [63] survival_3.8-3            tools_4.3.3              
 [65] ica_1.0-3                 Rcpp_1.0.13-1            
 [67] glue_1.8.0                gridExtra_2.3            
 [69] SparseArray_1.2.2         xfun_0.50                
 [71] usethis_3.1.0             withr_3.0.2              
 [73] fastmap_1.2.0             bluster_1.12.0           
 [75] digest_0.6.37             rsvd_1.0.5               
 [77] R6_2.5.1                  mime_0.12                
 [79] colorspace_2.1-1          scattermore_1.2          
 [81] tensor_1.5                spatstat.data_3.1-4      
 [83] tidyr_1.3.1               generics_0.1.3           
 [85] data.table_1.15.4         httr_1.4.7               
 [87] htmlwidgets_1.6.4         S4Arrays_1.2.0           
 [89] uwot_0.1.16               pkgconfig_2.0.3          
 [91] gtable_0.3.6              lmtest_0.9-40            
 [93] XVector_0.42.0            htmltools_0.5.8.1        
 [95] profvis_0.4.0             dotCall64_1.2            
 [97] scales_1.3.0              png_0.1-8                
 [99] spatstat.univar_3.1-1     knitr_1.49               
[101] reshape2_1.4.4            curl_6.0.1               
[103] nlme_3.1-165              cachem_1.1.0             
[105] zoo_1.8-12                stringr_1.5.1            
[107] KernSmooth_2.23-26        vipor_0.4.7              
[109] parallel_4.3.3            miniUI_0.1.1.1           
[111] ggrastr_1.0.2             pillar_1.10.1            
[113] grid_4.3.3                vctrs_0.6.5              
[115] RANN_2.6.2                urlchecker_1.0.1         
[117] promises_1.3.2            BiocSingular_1.18.0      
[119] beachmat_2.18.0           xtable_1.8-4             
[121] cluster_2.1.8             beeswarm_0.4.0           
[123] evaluate_1.0.1            cli_3.6.3                
[125] locfit_1.5-9.10           compiler_4.3.3           
[127] rlang_1.1.4               crayon_1.5.3             
[129] future.apply_1.11.2       labeling_0.4.3           
[131] ggbeeswarm_0.7.2          fs_1.6.5                 
[133] plyr_1.8.9                stringi_1.8.4            
[135] viridisLite_0.4.2         deldir_2.0-4             
[137] BiocParallel_1.36.0       munsell_0.5.1            
[139] lazyeval_0.2.2            devtools_2.4.5           
[141] spatstat.geom_3.3-4       dir.expiry_1.10.0        
[143] RcppHNSW_0.6.0            sparseMatrixStats_1.14.0 
[145] future_1.34.0             statmod_1.5.0            
[147] shiny_1.10.0              ROCR_1.0-11              
[149] memoise_2.0.1             igraph_2.0.3             </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"loop":true,"openEffect":"zoom","selector":".lightbox","descPosition":"bottom","closeEffect":"zoom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>