<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Overview of all three pipeline results.">

<title>Comparison of DGE results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Comparison of DGE results</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat Toolkit <i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit <i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Overview of all three pipeline results.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">20-Nov-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#subsample" id="toc-subsample" class="nav-link active" data-scroll-target="#subsample"><span class="header-section-number">0.1</span> Subsample</a></li>
  <li><a href="#seurat-dge" id="toc-seurat-dge" class="nav-link" data-scroll-target="#seurat-dge"><span class="header-section-number">1</span> Seurat DGE</a></li>
  <li><a href="#bioc-findmarkers" id="toc-bioc-findmarkers" class="nav-link" data-scroll-target="#bioc-findmarkers"><span class="header-section-number">2</span> Bioc findMarkers</a></li>
  <li><a href="#bioc-scoremarkers" id="toc-bioc-scoremarkers" class="nav-link" data-scroll-target="#bioc-scoremarkers"><span class="header-section-number">3</span> Bioc ScoreMarkers</a></li>
  <li><a href="#scanpy-dge" id="toc-scanpy-dge" class="nav-link" data-scroll-target="#scanpy-dge"><span class="header-section-number">4</span> Scanpy DGE</a></li>
  <li><a href="#compare-significant-genes." id="toc-compare-significant-genes." class="nav-link" data-scroll-target="#compare-significant-genes."><span class="header-section-number">5</span> Compare significant genes.</a></li>
  <li><a href="#compare-top-genes" id="toc-compare-top-genes" class="nav-link" data-scroll-target="#compare-top-genes"><span class="header-section-number">6</span> Compare top genes</a>
  <ul>
  <li><a href="#top-50" id="toc-top-50" class="nav-link" data-scroll-target="#top-50"><span class="header-section-number">6.1</span> Top 50</a></li>
  <li><a href="#top-200" id="toc-top-200" class="nav-link" data-scroll-target="#top-200"><span class="header-section-number">6.2</span> Top 200</a></li>
  <li><a href="#expression-distribution" id="toc-expression-distribution" class="nav-link" data-scroll-target="#expression-distribution"><span class="header-section-number">6.3</span> Expression distribution</a></li>
  <li><a href="#separation-of-clusters" id="toc-separation-of-clusters" class="nav-link" data-scroll-target="#separation-of-clusters"><span class="header-section-number">6.4</span> Separation of clusters</a></li>
  </ul></li>
  <li><a href="#for-one-cluster" id="toc-for-one-cluster" class="nav-link" data-scroll-target="#for-one-cluster"><span class="header-section-number">7</span> For one cluster</a></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">8</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>force_rerun <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (force_rerun){</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  reticulate<span class="sc">::</span><span class="fu">use_condaenv</span>(<span class="st">"/Users/asabjor/miniconda3/envs/scanpy_2024_nopip/"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  reticulate<span class="sc">::</span><span class="fu">py_config</span>()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Matrix)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(basilisk)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(basilisk.utils)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(zellkonverter)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co">#source("~/projects/sc-devop/single_cell_R_scripts/overlap_phyper_v2.R")</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">source_url</span>(<span class="st">"https://raw.githubusercontent.com/asabjorklund/single_cell_R_scripts/main/overlap_phyper_v2.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>path_results <span class="ot">&lt;-</span> <span class="st">"data/covid/results"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_results,<span class="st">"merged_all.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat harmony"</span>),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap_bioc_harmony"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc harmony"</span>),</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap_scpy_harmony"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy harmony"</span>),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-4-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="comparison_dge_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>Select one clustering and use that for DGE detection in all the different methods. Use clustering with Seurat and <code>louvain_0.5</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sel.clust <span class="ot">=</span> <span class="st">"RNA_snn_res.0.5"</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> sel.clust, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat harmony"</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> sel.clust, <span class="at">reduction =</span> <span class="st">"umap_bioc_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc harmony"</span>),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> sel.clust, <span class="at">reduction =</span> <span class="st">"umap_scpy_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy harmony"</span>),</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides=</span><span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-5-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="comparison_dge_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Rename clusters by celltypes to easier follow annotations.</p>
<ul>
<li>0 - Mono/DC</li>
<li>1 - NK</li>
<li>2 - CD8T</li>
<li>3 - Bcell</li>
<li>4 - CD4T</li>
<li>5 - Mono</li>
<li>6 - Bcell</li>
<li>7 - CD4T</li>
<li>8 - unclear</li>
<li>9 - unclear</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>trans <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"M-0"</span>,<span class="st">"NK-1"</span>,<span class="st">"CD8-2"</span>,<span class="st">"B-3"</span>,<span class="st">"CD4-4"</span>,<span class="st">"M-5"</span>,<span class="st">"B-6"</span>,<span class="st">"CD4-7"</span>,<span class="st">"U-8"</span>,<span class="st">"U-9"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(trans) <span class="ot">=</span> <span class="fu">as.character</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">9</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">factor</span>(trans[<span class="fu">as.character</span>(all[[sel.clust]][,<span class="dv">1</span>])], <span class="at">levels =</span> <span class="fu">sort</span>(trans))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>all<span class="sc">$</span>cluster <span class="ot">=</span> tmp</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">SetIdent</span>(all, <span class="at">value =</span> <span class="st">"cluster"</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"cluster"</span>, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat harmony"</span>),</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"cluster"</span>, <span class="at">reduction =</span> <span class="st">"umap_bioc_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc harmony"</span>),</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"cluster"</span>, <span class="at">reduction =</span> <span class="st">"umap_scpy_harmony"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy harmony"</span>),</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides=</span><span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-6-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="comparison_dge_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Also merge the layers</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(<span class="at">object =</span> all, <span class="at">layers =</span> <span class="fu">c</span>(<span class="st">"data"</span>,<span class="st">"counts"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="subsample" class="level3" data-number="0.1">
<h3 data-number="0.1" class="anchored" data-anchor-id="subsample"><span class="header-section-number">0.1</span> Subsample</h3>
<p>Subsample to 200 cells per cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(all<span class="sc">@</span>active.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  B-3   B-6 CD4-4 CD4-7 CD8-2   M-0   M-5  NK-1   U-8   U-9 
  637   343   491   328   836  2071   338  1109   244   143 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>all.sub <span class="ot">=</span> all[, <span class="fu">WhichCells</span>(all, <span class="at">downsample =</span> <span class="dv">200</span>)]</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(all.sub<span class="sc">@</span>active.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  B-3   B-6 CD4-4 CD4-7 CD8-2   M-0   M-5  NK-1   U-8   U-9 
  200   200   200   200   200   200   200   200   200   143 </code></pre>
</div>
</div>
</section>
<section id="seurat-dge" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="seurat-dge"><span class="header-section-number">1</span> Seurat DGE</h2>
<p>Run seurat <code>FindAllMarkers</code> with Wilcoxon, both the new implementation (<code>wilcox</code>) and similar to seurat v4 (<code>wilcox_limma</code>), <code>MAST</code> and <code>T-test</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>outfile <span class="ot">=</span> <span class="fu">file.path</span>(path_results, <span class="st">"seurat_dge.rds"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(outfile) <span class="sc">&amp;</span> <span class="sc">!</span>force_rerun ){ </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  markersS <span class="ot">=</span> <span class="fu">readRDS</span>(outfile)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  markersS <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  markersS<span class="sc">$</span>wilc <span class="ot">=</span> <span class="fu">FindAllMarkers</span>(all.sub, <span class="at">test.use =</span> <span class="st">"wilcox"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  markersS<span class="sc">$</span>wilcL <span class="ot">=</span> <span class="fu">FindAllMarkers</span>(all.sub, <span class="at">test.use =</span> <span class="st">"wilcox_limma"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  markersS<span class="sc">$</span>Ttest <span class="ot">=</span> <span class="fu">FindAllMarkers</span>(all.sub, <span class="at">test.use =</span> <span class="st">"t"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  markersS<span class="sc">$</span>mast <span class="ot">=</span> <span class="fu">FindAllMarkers</span>(all.sub, <span class="at">test.use =</span> <span class="st">"MAST"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(markersS, <span class="at">file =</span> outfile)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="bioc-findmarkers" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="bioc-findmarkers"><span class="header-section-number">2</span> Bioc findMarkers</h2>
<p>For Bioc no need to do the subsampling, if not for speed. It runs all pairwise comparisons and then combines the p-values across all tests.</p>
<p>Can run DGE with “all”, “some” and “any”.</p>
<ul>
<li>“any” - require significance in one pw test.</li>
<li>“all” - dge in all pw tests. A</li>
<li>“some” - combined p-value is calculated by taking the middlemost value of the Holm-corrected p-values for each gene. (By default, this the median for odd numbers of contrasts and one-after-the-median for even numbers, but the exact proportion can be changed by setting min.prop - see ?combineParallelPValues.) Here, the null hypothesis is that the gene is not DE in at least half of the contrasts.</li>
</ul>
<p>If pval.type=“all”, the null hypothesis is that the gene is not DE in all contrasts. A combined p-value for each gene is computed using Berger’s intersection union test (IUT). Ranking based on the IUT p-value will focus on genes that are DE in that cluster compared to all other clusters. This strategy is particularly effective when dealing with distinct clusters that have a unique expression profile. In such cases, it yields a highly focused marker set that concisely captures the differences between clusters.</p>
<p>Still use the subsampled object for comparable numbers of cells.</p>
<p>Run with Wilcoxon, T-test and Binomial test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">as.SingleCellExperiment</span>(all.sub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>outfile <span class="ot">=</span> <span class="fu">file.path</span>(path_results, <span class="st">"bioc_dge.rds"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(outfile) <span class="sc">&amp;</span> <span class="sc">!</span>force_rerun ){ </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  markersB <span class="ot">=</span> <span class="fu">readRDS</span>(outfile)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  markersB <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  markersB<span class="sc">$</span>wilc<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"wilcox"</span>, <span class="at">pval.type =</span> <span class="st">"all"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  markersB<span class="sc">$</span>Ttest<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"t"</span>, <span class="at">pval.type =</span> <span class="st">"all"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>  markersB<span class="sc">$</span>binom<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"binom"</span>, <span class="at">pval.type =</span> <span class="st">"all"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(markersB, <span class="at">file =</span> outfile)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>outfile <span class="ot">=</span> <span class="fu">file.path</span>(path_results, <span class="st">"bioc_dge_some.rds"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(outfile) <span class="sc">&amp;</span> <span class="sc">!</span>force_rerun ){ </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  markersBsome <span class="ot">=</span> <span class="fu">readRDS</span>(outfile)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  markersBsome <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  markersBsome<span class="sc">$</span>wilc<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"wilcox"</span>, <span class="at">pval.type =</span> <span class="st">"some"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  markersBsome<span class="sc">$</span>Ttest<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"t"</span>, <span class="at">pval.type =</span> <span class="st">"some"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  markersBsome<span class="sc">$</span>binom<span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>( sce, <span class="at">groups =</span> sce<span class="sc">$</span>cluster,</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.type =</span> <span class="st">"binom"</span>, <span class="at">pval.type =</span> <span class="st">"some"</span>,     <span class="at">direction =</span> <span class="st">"up"</span>)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(markersBsome, <span class="at">file =</span> outfile)</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="bioc-scoremarkers" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="bioc-scoremarkers"><span class="header-section-number">3</span> Bioc ScoreMarkers</h2>
<p>From help section:</p>
<p>Compared to findMarkers, this function represents a simpler and more intuitive summary of the differences between the groups. We do this by realizing that the p-values for these types of comparisons are largely meaningless; individual cells are not meaningful units of experimental replication, while the groups themselves are defined from the data. Thus, by discarding the p-values, we can simplify our marker selection by focusing only on the effect sizes between groups.</p>
<p>Here, the strategy is to perform pairwise comparisons between each pair of groups to obtain various effect sizes. For each group X, we summarize the effect sizes across all pairwise comparisons involving that group, e.g., mean, min, max and so on. This yields a DataFrame for each group where each column contains a different summarized effect and each row corresponds to a gene in x. Reordering the rows by the summary of choice can yield a ranking of potential marker genes for downstream analyses.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>outfile <span class="ot">=</span> <span class="fu">file.path</span>(path_results, <span class="st">"bioc_dge_score.rds"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(outfile) <span class="sc">&amp;</span> <span class="sc">!</span>force_rerun ){ </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  markersBS <span class="ot">=</span> <span class="fu">readRDS</span>(outfile)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  markersBS <span class="ot">&lt;-</span> <span class="fu">scoreMarkers</span>(sce, sce<span class="sc">$</span>cluster)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(markersBS, <span class="at">file =</span> outfile)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Extract top ranked per cluster for <code>mean.AUC</code>, <code>mean.logFC.cohen</code>, <code>mean.logFC.detected</code></p>
<p>The logFC.cohen columns contain the standardized log-fold change, i.e., Cohen’s d.&nbsp;For each pairwise comparison, this is defined as the difference in the mean log-expression for each group scaled by the average standard deviation across the two groups. (Technically, we should use the pooled variance; however, this introduces some unpleasant asymmetry depending on the variance of the larger group, so we take a simple average instead.) Cohen’s d is analogous to the t-statistic in a two-sample t-test and avoids spuriously large effect sizes from comparisons between highly variable groups. We can also interpret Cohen’s d as the number of standard deviations between the two group means.</p>
<p>The AUC columns contain the area under the curve. This is the probability that a randomly chosen observation in one group is greater than a randomly chosen observation in the other group. The AUC is closely related to the U-statistic used in the Wilcoxon rank sum test. Values greater than 0.5 indicate that a gene is upregulated in the first group.</p>
<p>The key difference between the AUC and Cohen’s d is that the former is less sensitive to the variance within each group</p>
<p>Finally, the logFC.detected columns contain the log-fold change in the proportion of cells with detected (i.e., non-zero) expression between groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>Sstats <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"mean.AUC"</span>, <span class="st">"mean.logFC.cohen"</span>, <span class="st">"mean.logFC.detected"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(Sstats) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"bioc_AUC"</span>, <span class="st">"bioc_Cohen"</span>, <span class="st">"bioc_det"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">=</span> <span class="fu">levels</span>(all<span class="sc">$</span>cluster)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>rankedBS <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(Sstats)){</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  rankedBS[[n]] <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (cl <span class="cf">in</span> clusters){</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    rankedBS[[n]][[cl]] <span class="ot">=</span> <span class="fu">rownames</span>(markersBS[[cl]])[<span class="fu">order</span>(markersBS[[cl]][,Sstats[n]], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="scanpy-dge" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="scanpy-dge"><span class="header-section-number">4</span> Scanpy DGE</h2>
<p>Running tests <code>t-test</code>, <code>t-test_overestim_var</code>, <code>wilcoxon</code>, <code>logreg</code>.</p>
<p>First, save SCE as an h5ad matrix. with zellkonverter. Then run python code with basilisk.</p>
<pre><code>## Python code example
import scanpy 

fpath = "/Users/asabjor/courses/course_git/temp/workshop-scRNAseq/labs/comparison/data/covid/results/merged_all.h5ad"
adata = scanpy.read_h5ad(fpath)

print(adata.shape)
print(adata.X[1:10,1:10])

scanpy.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)
scanpy.pp.log1p(adata)
print(adata.shape)
print(adata.X[1:10,1:10])
    
scanpy.tl.rank_genes_groups(adata, 'cluster', method='t-test', key_added = "Ttest")
print(adata.uns.Ttest.names[1:10])
scanpy$tl$rank_genes_groups(adata, 'cluster', method='t-test_overestim_var', key_added = "Ttest_o")
scanpy$tl$rank_genes_groups(adata, 'cluster', method='wilcoxon', key_added = "wilc")
scanpy$pp$scale(adata) # scale before logreg.
scanpy$tl$rank_genes_groups(adata, 'cluster', method='logreg', key_added = "logreg")
</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>outfile <span class="ot">=</span> <span class="fu">file.path</span>(path_results, <span class="st">"scanpy_dge.rds"</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(outfile) <span class="sc">&amp;</span> <span class="sc">!</span>force_rerun ){ </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  dge.scanpy <span class="ot">=</span> <span class="fu">readRDS</span>(outfile)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>zellkonverter<span class="sc">::</span><span class="fu">writeH5AD</span>(sce, <span class="fu">file.path</span>(path_results,<span class="st">"merged_all.h5ad"</span>))</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">py_config</span>()</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>penv <span class="ot">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/scanpy_2024_nopip"</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>dge.scanpy <span class="ot">=</span> <span class="fu">basiliskRun</span>(<span class="at">env=</span>penv, <span class="at">fun=</span><span class="cf">function</span>(fpath) {</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    scanpy <span class="ot">&lt;-</span> reticulate<span class="sc">::</span><span class="fu">import</span>(<span class="st">"scanpy"</span>)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    adata <span class="ot">=</span> scanpy<span class="sc">$</span><span class="fu">read_h5ad</span>(fpath)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    output <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>    output<span class="sc">$</span>shape1 <span class="ot">=</span> adata<span class="sc">$</span>shape</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>    output<span class="sc">$</span>head1 <span class="ot">=</span> adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>shape)</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">normalize_per_cell</span>(adata, <span class="at">counts_per_cell_after=</span><span class="fl">1e4</span>)</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">log1p</span>(adata)</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>shape)</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>    output<span class="sc">$</span>shape2 <span class="ot">=</span> adata<span class="sc">$</span>shape</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>    output<span class="sc">$</span>head2 <span class="ot">=</span> adata<span class="sc">$</span>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>tl<span class="sc">$</span><span class="fu">rank_genes_groups</span>(adata, <span class="st">'cluster'</span>, <span class="at">method=</span><span class="st">'t-test'</span>, <span class="at">key_added =</span> <span class="st">"Ttest"</span>)</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata<span class="sc">$</span>uns[<span class="st">'Ttest'</span>][<span class="st">'names'</span>])</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>tl<span class="sc">$</span><span class="fu">rank_genes_groups</span>(adata, <span class="st">'cluster'</span>, <span class="at">method=</span><span class="st">'t-test_overestim_var'</span>, <span class="at">key_added =</span> <span class="st">"Ttest_o"</span>)</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>tl<span class="sc">$</span><span class="fu">rank_genes_groups</span>(adata, <span class="st">'cluster'</span>, <span class="at">method=</span><span class="st">'wilcoxon'</span>, <span class="at">key_added =</span> <span class="st">"wilc"</span>)</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>pp<span class="sc">$</span><span class="fu">scale</span>(adata) <span class="co"># scale before logreg.</span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>    scanpy<span class="sc">$</span>tl<span class="sc">$</span><span class="fu">rank_genes_groups</span>(adata, <span class="st">'cluster'</span>, <span class="at">method=</span><span class="st">'logreg'</span>, <span class="at">key_added =</span> <span class="st">"logreg"</span>)</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    output<span class="sc">$</span>Ttest <span class="ot">=</span> adata<span class="sc">$</span>uns[<span class="st">'Ttest'</span>]</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(adata<span class="sc">$</span>uns)</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>}, <span class="at">fpath =</span> <span class="st">"/Users/asabjor/courses/course_git/temp/workshop-scRNAseq/labs/comparison/data/covid/results/merged_all.h5ad"</span>,  <span class="at">testload=</span><span class="st">"scanpy"</span>)</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">py_config</span>()</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dge.scanpy, <span class="at">file =</span> outfile)</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Parse all scanpy results into dataframes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>scanpy.tests <span class="ot">=</span> <span class="fu">names</span>(dge.scanpy)[<span class="sc">-</span><span class="dv">1</span><span class="sc">:-</span><span class="dv">2</span>]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">unique</span>(all<span class="sc">$</span>cluster))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># convert from lists to dfs.</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>markers_scanpy <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (tname <span class="cf">in</span> scanpy.tests){</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">=</span> dge.scanpy[[tname]]</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  markers_scanpy[[tname]] <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(clusters)){ </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(cbind, <span class="fu">lapply</span>(x[<span class="sc">-</span><span class="dv">1</span><span class="sc">:-</span><span class="dv">2</span>], <span class="cf">function</span>(y) { y[,i]})))</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(df) <span class="ot">=</span> <span class="fu">names</span>(x[<span class="sc">-</span><span class="dv">1</span><span class="sc">:-</span><span class="dv">2</span>])</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(df) <span class="ot">=</span> x<span class="sc">$</span>names[,i]</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    markers_scanpy[[tname]][[i]] <span class="ot">=</span> df</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="compare-significant-genes." class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="compare-significant-genes."><span class="header-section-number">5</span> Compare significant genes.</h2>
<p>Create a list object with all the significant dge genes.</p>
<ul>
<li>FDR from BioC <code>findMarkers</code> is the BH-adjusted p-value.</li>
<li>Adjusted p-value from scanpy is BH-adjusted p-value. Scanpy logreg does not have p-values.</li>
<li>P-values in Seurat are also BH.</li>
<li><strong>OBS!</strong> BioC <code>scoreMarkers</code> does not provide p-values in the same way, cannot compare here.</li>
</ul>
<p>Not sure how they define the BH background, all genes, or genes after prefiltering.</p>
<p>Set cutoff for significant at 0.01.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pval.cut <span class="ot">=</span> <span class="fl">0.01</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>sign.scanpy <span class="ot">=</span> <span class="fu">lapply</span>(markers_scanpy, <span class="cf">function</span>(x) { </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">lapply</span>(x, <span class="cf">function</span>(y) y[y<span class="sc">$</span>pvals_adj <span class="sc">&lt;=</span> pval.cut,])</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(tmp) <span class="ot">=</span> <span class="fu">as.character</span>(clusters) </span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(tmp)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>sign.dge <span class="ot">=</span> <span class="fu">unlist</span>(sign.scanpy, <span class="at">recursive =</span> F)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sign.dge) <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"Ttest_o"</span>, <span class="st">"TtestO"</span>, <span class="fu">names</span>(sign.dge))</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sign.dge) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"scpy_"</span>, <span class="fu">names</span>(sign.dge))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sign.bioc <span class="ot">=</span> <span class="fu">lapply</span>(markersB, <span class="cf">function</span>(x) { </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(x, <span class="cf">function</span>(y) y[y<span class="sc">$</span>FDR <span class="sc">&lt;=</span> pval.cut,])</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sign.bioc) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"bioc_"</span>, <span class="fu">names</span>(sign.bioc))</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>sign.bioc <span class="ot">=</span> <span class="fu">unlist</span>(sign.bioc, <span class="at">recursive =</span> F)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>sign.biocS <span class="ot">=</span> <span class="fu">lapply</span>(markersBsome, <span class="cf">function</span>(x) { </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(x, <span class="cf">function</span>(y) y[y<span class="sc">$</span>FDR <span class="sc">&lt;=</span> pval.cut,])</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sign.biocS) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"bioc_SOME-"</span>, <span class="fu">names</span>(sign.biocS))</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>sign.biocS <span class="ot">=</span> <span class="fu">unlist</span>(sign.biocS, <span class="at">recursive =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sign.seu <span class="ot">=</span> <span class="fu">lapply</span>(markersS, <span class="cf">function</span>(x){</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">=</span> x[x<span class="sc">$</span>p_val_adj <span class="sc">&lt;=</span> pval.cut,]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">=</span> <span class="fu">split</span>(x, x<span class="sc">$</span>cluster)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">=</span> <span class="fu">lapply</span>(x, <span class="cf">function</span>(y) { </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(y) <span class="ot">=</span> y<span class="sc">$</span>gene</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(y) })    </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  x</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sign.seu) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"seu_"</span>, <span class="fu">names</span>(sign.seu))</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>sign.seu <span class="ot">=</span> <span class="fu">unlist</span>(sign.seu, <span class="at">recursive =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sign.dge <span class="ot">=</span> <span class="fu">c</span>(sign.dge, sign.seu, sign.bioc, sign.biocS)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sign.dge, <span class="at">file =</span> <span class="fu">file.path</span>(path_results, <span class="st">"all_sign_dge.Rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Number of dge genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>nG <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(sign.dge, nrow))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>pipeline <span class="ot">=</span> <span class="fu">Reduce</span>(rbind,<span class="fu">strsplit</span>(<span class="fu">names</span>(nG),<span class="st">"_"</span>))[,<span class="dv">1</span>]</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">8</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(nG, <span class="at">las =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="fu">factor</span>(pipeline), <span class="at">main =</span> <span class="st">"Number of significant genes"</span>, <span class="at">cex.names =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-21-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="comparison_dge_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Turns out that the two different wilcoxon runs in Seurat have identical results, so remove one. sign.dge</p>
<p>Overlap of DGE genes is calculated with a phyper test using all expressed genes as the background. The heatmaps display the <code>-log10(p-value)</code> from the phyper test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sign.dge <span class="ot">=</span> sign.dge[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"wilcL"</span>,<span class="fu">names</span>(sign.dge))]</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">lapply</span>(sign.dge, rownames)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(tmp, tmp, <span class="at">nsize =</span> <span class="dv">5</span>, <span class="at">bg =</span> <span class="fu">nrow</span>(all), <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-22-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="comparison_dge_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Cluster and plot with annotation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">names</span>(sign.dge)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">Reduce</span>(rbind, <span class="fu">strsplit</span>(n, <span class="st">"[</span><span class="sc">\\</span><span class="st">._]"</span>))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>annot.tests <span class="ot">=</span> <span class="fu">data.frame</span>(s)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annot.tests) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Pipe"</span>,<span class="st">"Test"</span>,<span class="st">"Cluster"</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annot.tests) <span class="ot">=</span> <span class="fu">names</span>(sign.dge)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>annot.tests<span class="sc">$</span>TestName <span class="ot">=</span> <span class="fu">paste</span>(s[,<span class="dv">1</span>],s[,<span class="dv">2</span>], <span class="at">sep=</span> <span class="st">"_"</span>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="sc">-</span><span class="fu">log10</span>(o<span class="sc">$</span>P[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]<span class="sc">+</span><span class="fl">1e-230</span>),<span class="at">annotation_col =</span> annot.tests)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-23-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="comparison_dge_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Very different number of genes, so unfair comparison.</p>
</section>
<section id="compare-top-genes" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="compare-top-genes"><span class="header-section-number">6</span> Compare top genes</h2>
<p>Create a list object with all the dge genes as ranked lists to compare top X genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ranked.scanpy <span class="ot">=</span> <span class="fu">lapply</span>(markers_scanpy, <span class="cf">function</span>(x) { </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">lapply</span>(x, rownames)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(tmp) <span class="ot">=</span> clusters</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(tmp)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ranked.scanpy) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"scpy_"</span>, <span class="fu">names</span>(ranked.scanpy))</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ranked.scanpy) <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"Ttest_o"</span>, <span class="st">"TtestO"</span>, <span class="fu">names</span>(ranked.scanpy))</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>ranked.scanpy <span class="ot">=</span> <span class="fu">unlist</span>(ranked.scanpy, <span class="at">recursive =</span> F)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>getRankSeu <span class="ot">=</span> <span class="cf">function</span>(x) {</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">=</span> <span class="fu">split</span>(x, x<span class="sc">$</span>cluster)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(x, <span class="cf">function</span>(y) { </span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    y<span class="sc">$</span>gene } )</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>rank.seu <span class="ot">=</span> <span class="fu">lapply</span>(markersS, getRankSeu)</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(rank.seu) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"seu_"</span>, <span class="fu">names</span>(rank.seu))</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>rank.seu <span class="ot">=</span> <span class="fu">unlist</span>(rank.seu, <span class="at">recursive =</span> F)</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a><span class="co"># remove wilcoxonL</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>rank.seu <span class="ot">=</span> rank.seu[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"wilcL"</span>, <span class="fu">names</span>(rank.seu))]</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>rank.bioc <span class="ot">=</span> <span class="fu">lapply</span>(markersB, <span class="cf">function</span>(x) { </span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(x, rownames)</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(rank.bioc) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"bioc_"</span>, <span class="fu">names</span>(rank.bioc))</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>rank.bioc <span class="ot">=</span> <span class="fu">unlist</span>(rank.bioc, <span class="at">recursive =</span> F)</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>rank.biocSOME <span class="ot">=</span> <span class="fu">lapply</span>(markersBsome, <span class="cf">function</span>(x) { </span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(x, rownames)</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(rank.biocSOME) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"bioc_SOME-"</span>, <span class="fu">names</span>(rank.biocSOME))</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>rank.biocSOME <span class="ot">=</span> <span class="fu">unlist</span>(rank.biocSOME, <span class="at">recursive =</span> F)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>rank.biocS <span class="ot">=</span> <span class="fu">unlist</span>(rankedBS, <span class="at">recursive =</span> F)</span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>all.rank <span class="ot">=</span> <span class="fu">c</span>(ranked.scanpy, rank.seu, rank.bioc, rank.biocSOME,rank.biocS )</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(all.rank, <span class="at">file =</span> <span class="fu">file.path</span>(path_results, <span class="st">"all_ranked_dge.Rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="top-50" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="top-50"><span class="header-section-number">6.1</span> Top 50</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">names</span>(all.rank)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">Reduce</span>(rbind, <span class="fu">strsplit</span>(n, <span class="st">"[</span><span class="sc">\\</span><span class="st">._]"</span>))</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>annot.tests <span class="ot">=</span> <span class="fu">data.frame</span>(s)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annot.tests) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Pipe"</span>,<span class="st">"Test"</span>,<span class="st">"Cluster"</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annot.tests) <span class="ot">=</span> n</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>annot.tests<span class="sc">$</span>TestName <span class="ot">=</span> <span class="fu">paste</span>(s[,<span class="dv">1</span>],s[,<span class="dv">2</span>], <span class="at">sep=</span> <span class="st">"_"</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(all.rank, <span class="cf">function</span>(x) x[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>])</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(topG,topG, <span class="at">nsize =</span> <span class="dv">5</span>, <span class="at">bg =</span> <span class="fu">nrow</span>(all), <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="comparison_dge_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>pvals <span class="ot">=</span> o<span class="sc">$</span>P[<span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(o<span class="sc">$</span>P)<span class="sc">-</span><span class="dv">1</span>), <span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(o<span class="sc">$</span>P)<span class="sc">-</span><span class="dv">1</span>)]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="sc">-</span><span class="fu">log10</span>(pvals<span class="fl">+1e-320</span>),<span class="at">annotation_col =</span> annot.tests)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-25-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="comparison_dge_files/figure-html/unnamed-chunk-25-2.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Clearly more unique genes per cluster with the bioC methods.</p>
<p>Merge all top genes from all clusters.</p>
<p>OBS! Skip using background, all overlaps are significant, so we will get a false scale of the p-values, but is more visual.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">split</span>(topG, annot.tests<span class="sc">$</span>TestName)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">lapply</span>(tmp, unlist)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">lapply</span>(tmp, unique)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(tmp,tmp, <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-26-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="comparison_dge_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with grouping.</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="sc">-</span><span class="fu">log10</span>(o<span class="sc">$</span>P <span class="sc">+</span> <span class="fl">1e-320</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp), <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp)], <span class="at">display_numbers =</span> o<span class="sc">$</span>M[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp), <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-26-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="comparison_dge_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Binomial test stands out, but some similarty to the other bioC tests. Also, all unique genes per cluster.</p>
</section>
<section id="top-200" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="top-200"><span class="header-section-number">6.2</span> Top 200</h3>
<p>Same but with top200 genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(all.rank, <span class="cf">function</span>(x) x[<span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>])</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">split</span>(topG, annot.tests<span class="sc">$</span>TestName)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">lapply</span>(tmp, unlist)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">lapply</span>(tmp, unique)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(tmp,tmp, <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-27-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="comparison_dge_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with grouping.</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="sc">-</span><span class="fu">log10</span>(o<span class="sc">$</span>P <span class="sc">+</span> <span class="fl">1e-320</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp), <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp)], <span class="at">display_numbers =</span> o<span class="sc">$</span>M[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp), <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tmp)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-27-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="comparison_dge_files/figure-html/unnamed-chunk-27-2.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>BioC stands out from the rest. Wilcoxon in Seurat and Scanpy are different.</p>
</section>
<section id="expression-distribution" class="level3" data-number="6.3">
<h3 data-number="6.3" class="anchored" data-anchor-id="expression-distribution"><span class="header-section-number">6.3</span> Expression distribution</h3>
<p>Expression distribution of the top 200 genes, both as number of cells expressed or as mean lognorm expression.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>stats.df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> <span class="fu">rownames</span>(all.sub))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>stats.df<span class="sc">$</span>nC <span class="ot">=</span> <span class="fu">rowSums</span>(all.sub<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts<span class="sc">&gt;</span><span class="dv">0</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>stats.df<span class="sc">$</span>meanE <span class="ot">=</span> <span class="fu">rowMeans</span>(all.sub<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># remove non-expressed genes</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>stats.df <span class="ot">=</span> stats.df[stats.df<span class="sc">$</span>nC<span class="sc">&gt;</span><span class="dv">0</span> <span class="sc">&amp;</span> stats.df<span class="sc">$</span>meanE <span class="sc">&gt;</span> <span class="dv">0</span>,]</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>top200 <span class="ot">=</span> tmp</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(tmp)){</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  stats.df[[n]] <span class="ot">=</span> <span class="fu">rownames</span>(stats.df) <span class="sc">%in%</span> tmp[[n]]</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Number of cells detected.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>all <span class="ot">=</span> <span class="fu">ggplot</span>(stats.df, <span class="fu">aes</span>(<span class="at">x=</span>nC)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>, <span class="at">binwidth =</span> <span class="fl">0.05</span>)  <span class="sc">+</span> <span class="fu">scale_x_log10</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">0.9</span>,<span class="dv">2000</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"All genes"</span>) </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(tmp)){</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  plots[[n]] <span class="ot">=</span> <span class="fu">ggplot</span>(stats.df[stats.df[,n],], <span class="fu">aes</span>(<span class="at">x=</span>nC)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>, <span class="at">binwidth=</span><span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">0.9</span>,<span class="dv">2000</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(n) </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-29-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="comparison_dge_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Mean expression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>all <span class="ot">=</span> <span class="fu">ggplot</span>(stats.df, <span class="fu">aes</span>(<span class="at">x=</span>meanE)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>, <span class="at">binwidth =</span> <span class="fl">0.05</span>)  <span class="sc">+</span> <span class="fu">scale_x_log10</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">0.0001</span>,<span class="dv">5</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"All genes"</span>) <span class="sc">+</span><span class="fu">RotatedAxis</span>()</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(tmp)){</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  plots[[n]] <span class="ot">=</span> <span class="fu">ggplot</span>(stats.df[stats.df[,n],], <span class="fu">aes</span>(<span class="at">x=</span>meanE)) <span class="sc">+</span> <span class="fu">geom_histogram</span>( <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">position=</span><span class="st">"identity"</span>, <span class="at">binwidth=</span><span class="fl">0.05</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">0.0001</span>,<span class="dv">5</span>)) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(n) <span class="sc">+</span> <span class="fu">RotatedAxis</span>()</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-30-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="comparison_dge_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="separation-of-clusters" class="level3" data-number="6.4">
<h3 data-number="6.4" class="anchored" data-anchor-id="separation-of-clusters"><span class="header-section-number">6.4</span> Separation of clusters</h3>
<p>For each geneset with top 200 genes per cluster, test running dim reduction and clustering and see how well it overlaps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>all.sub <span class="ot">=</span> <span class="fu">FindVariableFeatures</span>(all.sub, <span class="at">verbose =</span> F)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>all.sub <span class="ot">=</span> all.sub <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">RunUMAP</span>(<span class="at">verbose =</span> F, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) <span class="sc">%&gt;%</span> <span class="fu">FindNeighbors</span>(<span class="at">verbose =</span> F, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) <span class="sc">%&gt;%</span> <span class="fu">FindClusters</span>(<span class="at">verbose =</span> F)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>all.sub<span class="sc">@</span>reductions<span class="sc">$</span>umap_hgv <span class="ot">=</span> all.sub<span class="sc">@</span>reductions<span class="sc">$</span>umap</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(top200)){</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">VariableFeatures</span>(all.sub) <span class="ot">=</span> top200[[n]]</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  all.sub <span class="ot">=</span> all.sub <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span> <span class="fu">RunUMAP</span>(<span class="at">verbose =</span> F, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) <span class="sc">%&gt;%</span> <span class="fu">FindNeighbors</span>(<span class="at">verbose =</span> F, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) <span class="sc">%&gt;%</span> <span class="fu">FindClusters</span>(<span class="at">verbose =</span> F)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  all.sub[[<span class="fu">paste0</span>(<span class="st">"clusters_"</span>,n)]] <span class="ot">=</span> all.sub<span class="sc">@</span>active.ident</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  all.sub<span class="sc">@</span>reductions[[<span class="fu">paste0</span>(<span class="st">"umap_"</span>,n)]] <span class="ot">=</span> all.sub<span class="sc">@</span>reductions<span class="sc">$</span>umap</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>hgv <span class="ot">=</span> <span class="fu">DimPlot</span>(all.sub, <span class="at">group.by =</span> <span class="st">"cluster"</span>, <span class="at">reduction =</span> <span class="st">"umap_hgv"</span>, <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"HGV"</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(top200)){</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  plots[[n]] <span class="ot">=</span> <span class="fu">DimPlot</span>(all.sub, <span class="at">group.by =</span> <span class="st">"cluster"</span>, <span class="at">reduction =</span> <span class="fu">paste0</span>(<span class="st">"umap_"</span>,n), <span class="at">label =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(n)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-32-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="comparison_dge_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="for-one-cluster" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="for-one-cluster"><span class="header-section-number">7</span> For one cluster</h2>
<p>Select one cluster and compare the DGEs for that cluster across all methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>clust <span class="ot">=</span> <span class="st">"CD4-7"</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">=</span> <span class="fu">which</span>(annot.tests<span class="sc">$</span>Cluster <span class="sc">==</span> clust)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>ranked.sel <span class="ot">=</span> all.rank[sel]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Top 50 genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(ranked.sel, <span class="cf">function</span>(x) x[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>])</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(topG,topG, <span class="at">nsize =</span> <span class="dv">5</span>, <span class="at">bg =</span> <span class="fu">nrow</span>(all), <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-34-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="comparison_dge_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(<span class="fu">table</span>(<span class="fu">unlist</span>(topG)),<span class="at">decreasing =</span> T),<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   CCR7    LEF1 PIK3IP1    TCF7 TRABD2A  GIMAP7    IL7R   RPLP2    RPS6   SATB1 
     16      16      14      14      14      13      13      12      12      12 </code></pre>
</div>
</div>
<p>Only 2 genes selected in all 16 tests.</p>
<p>Top 10 genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(ranked.sel, <span class="cf">function</span>(x) x[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">overlap_phyper2</span>(topG,topG, <span class="at">nsize =</span> <span class="dv">5</span>, <span class="at">bg =</span> <span class="fu">nrow</span>(all), <span class="at">remove.diag =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-35-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="comparison_dge_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(<span class="fu">table</span>(<span class="fu">unlist</span>(topG)),<span class="at">decreasing =</span> T),<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   CCR7 PIK3IP1    IL7R    TCF7    LEF1    RPS6   TSHZ2     LTB TRABD2A   ADTRP 
     15      12      11      11      10       7       7       6       6       5 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">unique</span>(<span class="fu">unlist</span>(topG)))</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">#D = all.sub@assays$RNA@layers$data[sort(unique(unlist(topG))),]</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> all.sub[genes,]</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> <span class="fu">as.matrix</span>(tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> genes </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(all.sub)</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="co"># reorder D</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> D[,<span class="fu">order</span>(all.sub<span class="sc">$</span>cluster)]</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"red"</span>,<span class="st">"blue"</span>)</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(coldef) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"FALSE"</span>,<span class="st">"TRUE"</span>)</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>annotG <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> genes)</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>annot.col <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(topG)){</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>  annotG[[n]] <span class="ot">=</span> <span class="fu">as.character</span>(<span class="fu">rownames</span>(annotG) <span class="sc">%in%</span> topG[[n]])</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>  annot.col[[n]] <span class="ot">=</span> coldef</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>}  </span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>annotC <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">cluster=</span>all.sub<span class="sc">$</span>cluster)</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annotC) <span class="ot">=</span> <span class="fu">colnames</span>(all.sub)</span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"blue"</span>,<span class="st">"red"</span>)</span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(coldef) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span>,<span class="st">"FALSE"</span>)</span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>cols2 <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">"Set3"</span>)</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cols2) <span class="ot">=</span> clusters</span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(coldef, cols2)</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(D, <span class="at">annotation_row =</span> annotG, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> T, <span class="at">labels_col =</span> F, <span class="at">annotation_colors =</span> annot.col, <span class="at">annotation_col =</span> annotC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-36-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="comparison_dge_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>The ribosomal proteins are high in all B/T cells.</p>
<p>Proportion ribosomal among the top200 across all clusters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>nR <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(rbind,<span class="fu">lapply</span>(top200, <span class="cf">function</span>(x) <span class="fu">c</span>(<span class="fu">sum</span>(<span class="fu">grepl</span>(<span class="st">"^RP[SL]"</span>,x)), <span class="fu">length</span>(x)))))</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(nR) <span class="ot">=</span> <span class="fu">names</span>(top200)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(nR) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"nR"</span>,<span class="st">"nTot"</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>nR<span class="sc">$</span>perc <span class="ot">=</span> nR<span class="sc">$</span>nR <span class="sc">/</span> nR<span class="sc">$</span>nTot <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>nR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["nR"],"name":[1],"type":["int"],"align":["right"]},{"label":["nTot"],"name":[2],"type":["int"],"align":["right"]},{"label":["perc"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"76","2":"1239","3":"6.13397902","_rn_":"bioc_AUC"},{"1":"1","2":"2000","3":"0.05000000","_rn_":"bioc_binom"},{"1":"76","2":"1227","3":"6.19396903","_rn_":"bioc_Cohen"},{"1":"1","2":"1376","3":"0.07267442","_rn_":"bioc_det"},{"1":"1","2":"1106","3":"0.09041591","_rn_":"bioc_SOME-binom"},{"1":"72","2":"1026","3":"7.01754386","_rn_":"bioc_SOME-Ttest"},{"1":"75","2":"1023","3":"7.33137830","_rn_":"bioc_SOME-wilc"},{"1":"26","2":"2000","3":"1.30000000","_rn_":"bioc_Ttest"},{"1":"46","2":"2000","3":"2.30000000","_rn_":"bioc_wilc"},{"1":"10","2":"1886","3":"0.53022269","_rn_":"scpy_logreg"},{"1":"77","2":"1221","3":"6.30630631","_rn_":"scpy_Ttest"},{"1":"76","2":"1341","3":"5.66741238","_rn_":"scpy_TtestO"},{"1":"76","2":"1240","3":"6.12903226","_rn_":"scpy_wilc"},{"1":"75","2":"1407","3":"5.33049041","_rn_":"seu_mast"},{"1":"77","2":"1223","3":"6.29599346","_rn_":"seu_Ttest"},{"1":"75","2":"1493","3":"5.02344273","_rn_":"seu_wilc"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(all.sub, <span class="st">"percent_ribo"</span>, <span class="at">group.by =</span> <span class="st">"cluster"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-37-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="comparison_dge_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Around 6% of the genes for Ttest, wilcox and mast. Except for bioc wilcox. Also high for AUC/Cohen.</p>
<p>Same but for a B-cell cluster, take “B-3”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>clust <span class="ot">=</span> <span class="st">"B-3"</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">=</span> <span class="fu">which</span>(annot.tests<span class="sc">$</span>Cluster <span class="sc">==</span> clust)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>topG <span class="ot">=</span> <span class="fu">lapply</span>(all.rank[sel], <span class="cf">function</span>(x) x[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">unique</span>(<span class="fu">unlist</span>(topG)))</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">#D = all.sub@assays$RNA@layers$data[sort(unique(unlist(topG))),]</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> all.sub[genes,]</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> <span class="fu">as.matrix</span>(tmp<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(D) <span class="ot">=</span> genes </span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(D) <span class="ot">=</span> <span class="fu">colnames</span>(all.sub)</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co"># reorder D</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>D <span class="ot">=</span> D[,<span class="fu">order</span>(all.sub<span class="sc">$</span>cluster)]</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"red"</span>,<span class="st">"blue"</span>)</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(coldef) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"FALSE"</span>,<span class="st">"TRUE"</span>)</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>annotG <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> genes)</span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>annot.col <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="fu">names</span>(topG)){</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>  annotG[[n]] <span class="ot">=</span> <span class="fu">as.character</span>(<span class="fu">rownames</span>(annotG) <span class="sc">%in%</span> topG[[n]])</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>  annot.col[[n]] <span class="ot">=</span> coldef</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>}  </span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>annotC <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">cluster=</span>all.sub<span class="sc">$</span>cluster)</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annotC) <span class="ot">=</span> <span class="fu">colnames</span>(all.sub)</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"blue"</span>,<span class="st">"red"</span>)</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(coldef) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span>,<span class="st">"FALSE"</span>)</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>cols2 <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">"Set3"</span>)</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cols2) <span class="ot">=</span> clusters</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span> <span class="fu">c</span>(coldef, cols2)</span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(D, <span class="at">annotation_row =</span> annotG, <span class="at">cluster_cols =</span> F, <span class="at">cluster_rows =</span> T, <span class="at">labels_col =</span> F, <span class="at">annotation_colors =</span> annot.col, <span class="at">annotation_col =</span> annotC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_dge_files/figure-html/unnamed-chunk-39-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="comparison_dge_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-session" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">8</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/asabjor/miniconda3/envs/seurat5/lib/libopenblasp-r0.3.27.dylib;  LAPACK version 3.12.0

locale:
[1] sv_SE.UTF-8/sv_SE.UTF-8/sv_SE.UTF-8/C/sv_SE.UTF-8/sv_SE.UTF-8

time zone: Europe/Stockholm
tzcode source: system (macOS)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] pheatmap_1.0.12             ComplexHeatmap_2.18.0      
 [3] zellkonverter_1.12.1        basilisk.utils_1.14.1      
 [5] basilisk_1.14.1             scran_1.30.0               
 [7] scuttle_1.12.0              SingleCellExperiment_1.24.0
 [9] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[11] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
[13] IRanges_2.36.0              S4Vectors_0.40.2           
[15] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[17] matrixStats_1.4.1           dplyr_1.1.4                
[19] patchwork_1.2.0             ggplot2_3.5.1              
[21] Matrix_1.6-5                Seurat_5.1.0               
[23] SeuratObject_5.0.2          sp_2.1-4                   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.3.3            
  [3] later_1.3.2               bitops_1.0-8             
  [5] filelock_1.0.3            tibble_3.2.1             
  [7] polyclip_1.10-7           fastDummies_1.7.4        
  [9] lifecycle_1.0.4           doParallel_1.0.17        
 [11] edgeR_4.0.16              globals_0.16.3           
 [13] lattice_0.22-6            MASS_7.3-60.0.1          
 [15] magrittr_2.0.3            limma_3.58.1             
 [17] plotly_4.10.4             rmarkdown_2.28           
 [19] remotes_2.5.0             yaml_2.3.10              
 [21] metapod_1.10.0            httpuv_1.6.15            
 [23] sctransform_0.4.1         spam_2.10-0              
 [25] sessioninfo_1.2.2         pkgbuild_1.4.4           
 [27] spatstat.sparse_3.1-0     reticulate_1.39.0        
 [29] cowplot_1.1.3             pbapply_1.7-2            
 [31] RColorBrewer_1.1-3        pkgload_1.4.0            
 [33] abind_1.4-5               zlibbioc_1.48.0          
 [35] Rtsne_0.17                purrr_1.0.2              
 [37] RCurl_1.98-1.16           circlize_0.4.16          
 [39] GenomeInfoDbData_1.2.11   ggrepel_0.9.6            
 [41] irlba_2.3.5.1             listenv_0.9.1            
 [43] spatstat.utils_3.1-0      goftest_1.2-3            
 [45] RSpectra_0.16-2           spatstat.random_3.2-3    
 [47] dqrng_0.3.2               fitdistrplus_1.2-1       
 [49] parallelly_1.38.0         DelayedMatrixStats_1.24.0
 [51] leiden_0.4.3.1            codetools_0.2-20         
 [53] DelayedArray_0.28.0       shape_1.4.6.1            
 [55] tidyselect_1.2.1          farver_2.1.2             
 [57] ScaledMatrix_1.10.0       spatstat.explore_3.2-6   
 [59] jsonlite_1.8.8            GetoptLong_1.0.5         
 [61] BiocNeighbors_1.20.0      ellipsis_0.3.2           
 [63] progressr_0.14.0          iterators_1.0.14         
 [65] ggridges_0.5.6            survival_3.7-0           
 [67] foreach_1.5.2             tools_4.3.3              
 [69] ica_1.0-3                 Rcpp_1.0.13              
 [71] glue_1.7.0                gridExtra_2.3            
 [73] SparseArray_1.2.2         xfun_0.47                
 [75] usethis_3.0.0             withr_3.0.1              
 [77] fastmap_1.2.0             bluster_1.12.0           
 [79] fansi_1.0.6               digest_0.6.37            
 [81] rsvd_1.0.5                R6_2.5.1                 
 [83] mime_0.12                 colorspace_2.1-1         
 [85] scattermore_1.2           tensor_1.5               
 [87] spatstat.data_3.1-2       utf8_1.2.4               
 [89] tidyr_1.3.1               generics_0.1.3           
 [91] data.table_1.15.4         httr_1.4.7               
 [93] htmlwidgets_1.6.4         S4Arrays_1.2.0           
 [95] uwot_0.1.16               pkgconfig_2.0.3          
 [97] gtable_0.3.5              lmtest_0.9-40            
 [99] XVector_0.42.0            htmltools_0.5.8.1        
[101] profvis_0.4.0             dotCall64_1.1-1          
[103] clue_0.3-65               scales_1.3.0             
[105] png_0.1-8                 knitr_1.48               
[107] rjson_0.2.21              reshape2_1.4.4           
[109] curl_5.2.1                nlme_3.1-165             
[111] cachem_1.1.0              GlobalOptions_0.1.2      
[113] zoo_1.8-12                stringr_1.5.1            
[115] KernSmooth_2.23-24        vipor_0.4.7              
[117] parallel_4.3.3            miniUI_0.1.1.1           
[119] ggrastr_1.0.2             pillar_1.9.0             
[121] vctrs_0.6.5               RANN_2.6.2               
[123] urlchecker_1.0.1          promises_1.3.0           
[125] BiocSingular_1.18.0       beachmat_2.18.0          
[127] xtable_1.8-4              cluster_2.1.6            
[129] beeswarm_0.4.0            evaluate_0.24.0          
[131] cli_3.6.3                 locfit_1.5-9.9           
[133] compiler_4.3.3            rlang_1.1.4              
[135] crayon_1.5.3              future.apply_1.11.2      
[137] labeling_0.4.3            ggbeeswarm_0.7.2         
[139] fs_1.6.4                  plyr_1.8.9               
[141] stringi_1.8.4             viridisLite_0.4.2        
[143] deldir_2.0-4              BiocParallel_1.36.0      
[145] munsell_0.5.1             lazyeval_0.2.2           
[147] devtools_2.4.5            spatstat.geom_3.2-9      
[149] dir.expiry_1.10.0         RcppHNSW_0.6.0           
[151] sparseMatrixStats_1.14.0  future_1.34.0            
[153] statmod_1.5.0             shiny_1.9.1              
[155] ROCR_1.0-11               memoise_2.0.1            
[157] igraph_2.0.3             </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"selector":".lightbox","closeEffect":"zoom","loop":true,"openEffect":"zoom","descPosition":"bottom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>