<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Overview of all three pipeline results.">

<title>Comparison of all pipelines</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Comparison of all pipelines</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat Toolkit <i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit <i class="fa-brands fa-python" aria-label="python"></i> Scanpy Toolkit</p>
                  <div>
        <div class="description">
          Overview of all three pipeline results.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">23-Oct-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-data" id="toc-load-data" class="nav-link active" data-scroll-target="#load-data"><span class="header-section-number">1</span> Load data</a></li>
  <li><a href="#umaps" id="toc-umaps" class="nav-link" data-scroll-target="#umaps"><span class="header-section-number">2</span> Umaps</a></li>
  <li><a href="#doublet-predictions" id="toc-doublet-predictions" class="nav-link" data-scroll-target="#doublet-predictions"><span class="header-section-number">3</span> Doublet predictions</a>
  <ul>
  <li><a href="#doublet-scores" id="toc-doublet-scores" class="nav-link" data-scroll-target="#doublet-scores"><span class="header-section-number">3.1</span> Doublet scores</a></li>
  </ul></li>
  <li><a href="#cell-cycle" id="toc-cell-cycle" class="nav-link" data-scroll-target="#cell-cycle"><span class="header-section-number">4</span> Cell cycle</a></li>
  <li><a href="#variable-features" id="toc-variable-features" class="nav-link" data-scroll-target="#variable-features"><span class="header-section-number">5</span> Variable features</a></li>
  <li><a href="#integrations" id="toc-integrations" class="nav-link" data-scroll-target="#integrations"><span class="header-section-number">6</span> Integrations</a>
  <ul>
  <li><a href="#clustering-on-integrations" id="toc-clustering-on-integrations" class="nav-link" data-scroll-target="#clustering-on-integrations"><span class="header-section-number">6.1</span> Clustering on integrations</a></li>
  </ul></li>
  <li><a href="#clustering-consistency" id="toc-clustering-consistency" class="nav-link" data-scroll-target="#clustering-consistency"><span class="header-section-number">7</span> Clustering consistency</a></li>
  <li><a href="#clustering-in-the-pipelines" id="toc-clustering-in-the-pipelines" class="nav-link" data-scroll-target="#clustering-in-the-pipelines"><span class="header-section-number">8</span> Clustering in the pipelines</a></li>
  <li><a href="#celltype-prediction." id="toc-celltype-prediction." class="nav-link" data-scroll-target="#celltype-prediction."><span class="header-section-number">9</span> Celltype prediction.</a></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">10</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>With data in place, now we can start loading libraries we will use in this tutorial.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(zellkonverter)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Matrix)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-data" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="load-data"><span class="header-section-number">1</span> Load data</h2>
<p>OBS! Zellkonverter installs conda env with basilisk! Takes a while to run first time!!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>path_results <span class="ot">&lt;-</span> <span class="st">"data/covid/results"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_results)) <span class="fu">dir.create</span>(path_results, <span class="at">recursive =</span> T)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>path_seurat <span class="ot">=</span> <span class="st">"../seurat/data/covid/results/"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>path_bioc <span class="ot">=</span> <span class="st">"../bioc/data/covid/results/"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>path_scanpy <span class="ot">=</span> <span class="st">"../scanpy/data/covid/results/"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># fetch the files with qc and dimred for each </span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_seurat,<span class="st">"seurat_covid_qc_dr_int_cl.rds"</span>))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co"># bioc</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_bioc,<span class="st">"bioc_covid_qc_dr_int_cl.rds"</span>))</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>bioc <span class="ot">=</span> <span class="fu">as.Seurat</span>(sce)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># scanpy</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>scanpy.sce <span class="ot">=</span> <span class="fu">readH5AD</span>(<span class="fu">file.path</span>(path_scanpy, <span class="st">"scanpy_covid_qc_dr_scanorama_cl.h5ad"</span>))</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>scanpy <span class="ot">=</span> <span class="fu">as.Seurat</span>(scanpy.sce, <span class="at">counts =</span> <span class="cn">NULL</span>, <span class="at">data =</span> <span class="st">"X"</span>) <span class="co"># only have the var.genes data that is scaled.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="umaps" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="umaps"><span class="header-section-number">2</span> Umaps</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(sobj, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Seurat"</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(bioc, <span class="at">group.by =</span> <span class="st">"sample"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bioc"</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(scanpy, <span class="at">group.by =</span> <span class="st">"sample"</span>, <span class="at">reduction =</span> <span class="st">"X_umap_uncorr"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Scanpy"</span>),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-3-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>Settings for umap differs:</p>
<ul>
<li><p>Seurat</p>
<ul>
<li>dims = 1:30,</li>
<li>n.neighbors = 30,</li>
<li>n.epochs = 200,</li>
<li>min.dist = 0.3,</li>
<li>learning.rate = 1,</li>
<li>spread = 1</li>
</ul></li>
<li><p>Bioc</p>
<ul>
<li>n_dimred = 30,</li>
<li>n_neighbors = 15,<br>
</li>
<li>spread = 1,</li>
<li>min_dist = 0.01,</li>
<li>n_epchs = 200</li>
</ul></li>
<li><p>Scanpy</p>
<ul>
<li>does not use the umap algorithm for neigbors, feeds the KNN to the umap function.</li>
<li>n_pcs = 30,</li>
<li>n_neighbors = 20,</li>
<li>min_dist=0.5,</li>
<li>spread=1.0,</li>
</ul></li>
</ul>
</section>
<section id="doublet-predictions" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="doublet-predictions"><span class="header-section-number">3</span> Doublet predictions</h2>
<p>In all 3 pipelines we run the filtering of cells in the exact same way, but doublet detection is different.</p>
<ul>
<li>In Seurat - <code>DoubletFinder</code> with predefined 4% doublets.</li>
<li>In Scanpy - <code>Scrublet</code>, done per batch. No predefined cutoff.</li>
<li>In Bioc - <code>scDblFinder</code> - default cutoffs, no batch separation.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Seurat: "</span>, <span class="fu">dim</span>(sobj),<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Seurat:  18854 7134 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Bioc: "</span>, <span class="fu">dim</span>(bioc),<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Bioc:  18854 6741 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Scanpy: "</span>, <span class="fu">dim</span>(scanpy),<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Scanpy:  19468 7222 </code></pre>
</div>
</div>
<p>Highest number of cells filtered with Bioc. Lowest with scanpy.</p>
<p>Cell names are different in all 3, have to make a conversion to match them.</p>
<ul>
<li>Seurat has the sample name merged to cell name.</li>
<li>Scanpy has sample number after cell name.</li>
<li>Bioc has just cell name.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>meta.seurat <span class="ot">=</span> sobj<span class="sc">@</span>meta.data</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>meta.scanpy <span class="ot">=</span> scanpy<span class="sc">@</span>meta.data</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>meta.bioc <span class="ot">=</span> bioc<span class="sc">@</span>meta.data</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>meta.bioc<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">rownames</span>(meta.bioc)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>meta.scanpy<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">rownames</span>(meta.scanpy), <span class="cf">function</span>(x) <span class="fu">substr</span>(x,<span class="dv">1</span>,<span class="fu">nchar</span>(x)<span class="sc">-</span><span class="dv">2</span>))</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>meta.seurat<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">rownames</span>(meta.seurat),<span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">3</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Visualize overlap of cells using an UpSet plot in the <code>ComplexHeatmap</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>l <span class="ot">=</span> <span class="fu">list</span>(<span class="at">seurat =</span> meta.seurat<span class="sc">$</span>cell, <span class="at">bioc =</span> meta.bioc<span class="sc">$</span>cell, <span class="at">scanpy =</span> meta.scanpy<span class="sc">$</span>cell)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>cmat <span class="ot">=</span> <span class="fu">make_comb_mat</span>(l)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A combination matrix with 3 sets and 7 combinations.
  ranges of combination set size: c(10, 6540).
  mode for the combination size: distinct.
  sets are on rows.

Combination sets are:
  seurat bioc scanpy code size
       x    x      x  111 6540
       x    x         110   36
       x           x  101  484
            x      x  011  155
       x              100   74
            x         010   10
                   x  001   43

Sets are:
     set size
  seurat 7134
    bioc 6741
  scanpy 7222</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UpSet</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-6-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<section id="doublet-scores" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="doublet-scores"><span class="header-section-number">3.1</span> Doublet scores</h3>
<p>Create one dataset with the cells that are present in all samples. Also add in umap from all 3 pipelines.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>in.all <span class="ot">=</span> <span class="fu">intersect</span>(<span class="fu">intersect</span>(meta.scanpy<span class="sc">$</span>cell, meta.seurat<span class="sc">$</span>cell), meta.bioc<span class="sc">$</span>cell)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>tmp1 <span class="ot">=</span> meta.bioc[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp1) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp1),<span class="st">"_bioc"</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">=</span> meta.scanpy[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp2) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp2),<span class="st">"_scpy"</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> sobj[,<span class="fu">match</span>(in.all, meta.seurat<span class="sc">$</span>cell)]</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>meta.all <span class="ot">=</span> <span class="fu">cbind</span>(all<span class="sc">@</span>meta.data, tmp1,tmp2)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>all<span class="sc">@</span>meta.data <span class="ot">=</span> meta.all</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "pca"               "umap"              "tsne"             
 [4] "UMAP10_on_PCA"     "UMAP_on_ScaleData" "integrated_cca"   
 [7] "umap_cca"          "tsne_cca"          "harmony"          
[10] "umap_harmony"      "scanorama"         "scanoramaC"       
[13] "umap_scanorama"    "umap_scanoramaC"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>reductions<span class="sc">$</span>UMAP_on_PCA<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_bioc"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapbioc_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>reductions<span class="sc">$</span>X_umap_uncorr<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_scpy"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapscpy_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "pca"               "umap"              "tsne"             
 [4] "UMAP10_on_PCA"     "UMAP_on_ScaleData" "integrated_cca"   
 [7] "umap_cca"          "tsne_cca"          "harmony"          
[10] "umap_harmony"      "scanorama"         "scanoramaC"       
[13] "umap_scanorama"    "umap_scanoramaC"   "umap_bioc"        
[16] "umap_scpy"        </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeatureScatter</span>(all, <span class="st">"pANN_0.25_0.09_297"</span>,<span class="st">"scDblFinder.score_bioc"</span>),</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeatureScatter</span>(all, <span class="st">"pANN_0.25_0.09_297"</span>,<span class="st">"doublet_scores_scpy"</span>),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a> <span class="fu">FeatureScatter</span>(all, <span class="st">"scDblFinder.score_bioc"</span>,<span class="st">"doublet_scores_scpy"</span>),</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">3</span> </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-8-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>Highest correlation for the doublet scores in Scrublet and scDblFinder, but still only 0.27 in correlation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">features =</span> <span class="st">"pANN_0.25_0.09_297"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">features =</span> <span class="st">"scDblFinder.score_bioc"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">features =</span> <span class="st">"doublet_scores_scpy"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_bioc"</span>, <span class="at">features =</span> <span class="st">"pANN_0.25_0.09_297"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_bioc"</span>, <span class="at">features =</span> <span class="st">"scDblFinder.score_bioc"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_bioc"</span>, <span class="at">features =</span> <span class="st">"doublet_scores_scpy"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),  </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_scpy"</span>, <span class="at">features =</span> <span class="st">"pANN_0.25_0.09_297"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_scpy"</span>, <span class="at">features =</span> <span class="st">"scDblFinder.score_bioc"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_scpy"</span>, <span class="at">features =</span> <span class="st">"doublet_scores_scpy"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),  </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-9-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>It seems like the DoubletFinder scores are mainly high in the monocyte populations, while it is more mixed in the other 2 methods.</p>
</section>
</section>
<section id="cell-cycle" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="cell-cycle"><span class="header-section-number">4</span> Cell cycle</h2>
<p>Cell cycle predictions are done in a similar way for Seurat and Scanpy using module scores, while in Bioc we are using Cyclone.</p>
<p>Visualize the phase predictions onto the Seurat umap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(all<span class="sc">$</span>Phase)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  G1  G2M    S 
2878 1514 2148 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(all<span class="sc">$</span>phase_bioc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  G1  G2M    S 
3771  734 1551 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(all<span class="sc">$</span>phase_scpy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  G1  G2M    S 
5010   76 1454 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"Phase"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"phase_bioc"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(all, <span class="at">group.by =</span> <span class="st">"phase_scpy"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">3</span> </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<p>In seurat, most T/B-cells look like cycling. With bioc, more mixed. With scanpy also more in B/T-cells, but much more G1 prediction.</p>
<p>Plot the scores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>cc.scores <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>,<span class="st">"G1.score_bioc"</span>, <span class="st">"G2M.score_bioc"</span>, <span class="st">"S.score_bioc"</span>,<span class="st">"S_score_scpy"</span>, <span class="st">"G2M_score_scpy"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># copied from pairs help section.</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>panel.cor <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">prefix =</span> <span class="st">""</span>, cex.cor, ...) {</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    usr <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="st">"usr"</span>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">on.exit</span>(<span class="fu">par</span>(usr))</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">usr =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    nas <span class="ot">=</span> <span class="fu">is.na</span>(x) <span class="sc">|</span> <span class="fu">is.na</span>(y)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    Cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(x[<span class="sc">!</span>nas], y[<span class="sc">!</span>nas], <span class="at">method =</span> <span class="st">"spearman"</span>) <span class="co"># Remove abs function if desired</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">paste0</span>(prefix, <span class="fu">format</span>(<span class="fu">c</span>(Cor, <span class="fl">0.123456789</span>), <span class="at">digits =</span> digits)[<span class="dv">1</span>])</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">missing</span>(cex.cor)) {</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>        cex.cor <span class="ot">&lt;-</span> <span class="fl">0.4</span> <span class="sc">/</span> <span class="fu">strwidth</span>(txt)</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt,</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">cex =</span> <span class="dv">1</span> <span class="sc">+</span> cex.cor) </span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>} </span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(all<span class="sc">@</span>meta.data[,cc.scores], <span class="at">a.action =</span> na.omit,</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">upper.panel =</span> panel.cor,    <span class="co"># Correlation panel</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">lower.panel =</span> panel.smooth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-11-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Use spearman cor as the distributions are very different</p>
<ul>
<li>S and G2M scores in Scanpy and Seurat are correlated to eachother</li>
<li>S vs G2M are also correlated.</li>
<li>In BioC all scores are anticorrelated.</li>
</ul>
</section>
<section id="variable-features" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="variable-features"><span class="header-section-number">5</span> Variable features</h2>
<p>In Seurat, <code>FindVariableFeatures</code> is not batch aware unless the data is split into layers by samples, here we have the variable genes created with layers. In Bioc <code>modelGeneVar</code> we used sample as a blocking parameter, e.g calculates the variable genes per sample and combines the variances. Similar in scanpy we used the samples as <code>batch_key</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>seurat <span class="ot">=</span> <span class="fu">VariableFeatures</span>(sobj)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>bioc <span class="ot">=</span> sce<span class="sc">@</span>metadata<span class="sc">$</span>hvgs</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># scanpy has no strict number on selection, instead uses dispersion cutoff. So select instead top 2000 dispersion genes.</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">rowData</span>(scanpy.sce)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>hvgs_scanpy <span class="ot">=</span> <span class="fu">rownames</span>(tmp)[tmp<span class="sc">$</span>highly_variable]</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>hvgs<span class="sc">$</span>scanpy <span class="ot">=</span> <span class="fu">rownames</span>(tmp)[<span class="fu">order</span>(tmp<span class="sc">$</span>dispersions_norm, <span class="at">decreasing =</span> T)][<span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>]</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>cmat <span class="ot">=</span> <span class="fu">make_comb_mat</span>(hvgs)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A combination matrix with 3 sets and 7 combinations.
  ranges of combination set size: c(211, 861).
  mode for the combination size: distinct.
  sets are on rows.

Combination sets are:
  seurat bioc scanpy code size
       x    x      x  111  641
       x    x         110  287
       x           x  101  211
            x      x  011  364
       x              100  861
            x         010  708
                   x  001  784

Sets are:
     set size
  seurat 2000
    bioc 2000
  scanpy 2000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UpSet</span>(cmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-12-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Surprisingly low overlap between the methods and many genes that are unique to one pipeline. With discrepancies in the doublet filtering the cells used differ to some extent, but otherwise the variation should be similar. Even if it is estimated in different ways.</p>
<p>Is the differences more due to the combination of ranks/dispersions or also found within a single dataset?</p>
<p>Only Seurat have the dispersions for each individual dataset stored in the object.</p>
<p>Explore more in the <code>comparison_hvg.qmd</code> script.</p>
</section>
<section id="integrations" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="integrations"><span class="header-section-number">6</span> Integrations</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>reductions <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"umap"</span>,<span class="st">"umap_cca"</span>, <span class="st">"umap_harmony"</span>,  <span class="st">"umap_scanorama"</span>,    <span class="st">"umap_scanoramaC"</span>, <span class="st">"umap_bioc"</span>,  <span class="st">"umap_bioc_mnn"</span>, <span class="st">"umap_bioc_harmony"</span>, <span class="st">"umap_bioc_scanorama"</span>, <span class="st">"umap_scpy"</span>,<span class="st">"umap_scpy_bbknn"</span>, <span class="st">"umap_scpy_scanorama"</span>, <span class="st">"umap_scpy_harmony"</span>   )</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>reductions<span class="sc">$</span>UMAP_on_MNN<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_bioc_mnn"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapbiocmnn_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>reductions<span class="sc">$</span>UMAP_on_Harmony<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_bioc_harmony"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapbiocharmony_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>reductions<span class="sc">$</span>UMAP_on_Scanorama<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_bioc_scanorama"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapbioscanorama_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>reductions<span class="sc">$</span>X_umap_bbknn<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_scpy_bbknn"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapscpybbknn_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>reductions<span class="sc">$</span>X_umap_scanorama<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_scpy_scanorama"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapscpyscanorama_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>reductions<span class="sc">$</span>X_umap_harmony<span class="sc">@</span>cell.embeddings[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tmp) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>all[[<span class="st">"umap_scpy_harmony"</span>]] <span class="ot">=</span> <span class="fu">CreateDimReducObject</span>(tmp, <span class="at">key =</span> <span class="st">"umapscpyharmony_"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>plotlist <span class="ot">=</span> <span class="fu">lapply</span>(reductions, <span class="cf">function</span>(x) <span class="fu">DimPlot</span>(all, <span class="at">reduction =</span> x, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(x))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  plotlist,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">4</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-14-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("carmonalab/scIntegrationMetrics", dependencies = FALSE), fails with compiler errors on mac. </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="clustering-on-integrations" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="clustering-on-integrations"><span class="header-section-number">6.1</span> Clustering on integrations</h3>
<p>Run clustering on each of the integrations but with same method. Use Seurat, with 30 first components, k=30, and louvain with a few different resolutions.</p>
<p>For BBKNN the full integration matrix is not in the reductions, skip for now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all seurat integrations.</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>integrations <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">seu_cca =</span> <span class="st">"integrated_cca"</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">seu_harm =</span> <span class="st">"harmony"</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">seu_scan =</span> <span class="st">"scanorama"</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">seu_scanC =</span> <span class="st">"scanoramaC"</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">0.4</span>,<span class="fl">0.6</span>,<span class="fl">0.8</span>,<span class="fl">1.0</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(integrations)){</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  sname <span class="ot">=</span> <span class="fu">names</span>(integrations)[i]</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  sobj <span class="ot">=</span> <span class="fu">FindNeighbors</span>(sobj, <span class="at">reduction =</span> integrations[[sname]], <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> F)</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (r <span class="cf">in</span> res){</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    sobj <span class="ot">=</span> <span class="fu">FindClusters</span>(sobj, <span class="at">resolution =</span> r, <span class="at">cluster.name =</span> <span class="fu">paste</span>(sname,r,<span class="at">sep=</span><span class="st">"_"</span>), <span class="at">verbose =</span> F)</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>meta.clust <span class="ot">=</span> sobj<span class="sc">@</span>meta.data[,<span class="fu">grepl</span>(<span class="st">"^seu_"</span>, <span class="fu">colnames</span>(sobj<span class="sc">@</span>meta.data))]</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>meta.clust <span class="ot">=</span> meta.clust[<span class="fu">match</span>(in.all, meta.seurat<span class="sc">$</span>cell),]</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a><span class="co"># all bioc</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>integrations <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">bio_mnn =</span> <span class="st">"MNN"</span>,</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">bio_harm =</span> <span class="st">"harmony"</span>,</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">bio_scan =</span> <span class="st">"Scanorama"</span></span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(integrations)){</span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>  sname <span class="ot">=</span> <span class="fu">names</span>(integrations)[i]</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>  bioc <span class="ot">=</span> <span class="fu">FindNeighbors</span>(bioc, <span class="at">reduction =</span> integrations[[sname]], <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> F)</span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (r <span class="cf">in</span> res){</span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>    bioc <span class="ot">=</span> <span class="fu">FindClusters</span>(bioc, <span class="at">resolution =</span> r, <span class="at">cluster.name =</span> <span class="fu">paste</span>(sname,r,<span class="at">sep=</span><span class="st">"_"</span>), <span class="at">verbose =</span> F)</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> bioc<span class="sc">@</span>meta.data[,<span class="fu">grepl</span>(<span class="st">"^bio_"</span>, <span class="fu">colnames</span>(bioc<span class="sc">@</span>meta.data))]</span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> tmp[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>meta.clust <span class="ot">=</span> <span class="fu">cbind</span>(tmp, meta.clust)</span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a><span class="co"># all scanpy</span></span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a>integrations <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>  <span class="at">scpy_harm =</span> <span class="st">"X_pca_harmony"</span>,</span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>  <span class="at">scpy_scan =</span> <span class="st">"Scanorama"</span></span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(integrations)){</span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a>  sname <span class="ot">=</span> <span class="fu">names</span>(integrations)[i]</span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a>  scanpy <span class="ot">=</span> <span class="fu">FindNeighbors</span>(scanpy, <span class="at">reduction =</span> integrations[[sname]], <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> F)</span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (r <span class="cf">in</span> res){</span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a>    scanpy <span class="ot">=</span> <span class="fu">FindClusters</span>(scanpy, <span class="at">resolution =</span> r, <span class="at">cluster.name =</span> <span class="fu">paste</span>(sname,r,<span class="at">sep=</span><span class="st">"_"</span>), <span class="at">verbose =</span> F)</span>
<span id="cb34-52"><a href="#cb34-52" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb34-53"><a href="#cb34-53" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-54"><a href="#cb34-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-55"><a href="#cb34-55" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> scanpy<span class="sc">@</span>meta.data[,<span class="fu">grepl</span>(<span class="st">"^scpy_"</span>, <span class="fu">colnames</span>(scanpy<span class="sc">@</span>meta.data))]</span>
<span id="cb34-56"><a href="#cb34-56" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> tmp[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb34-57"><a href="#cb34-57" aria-hidden="true" tabindex="-1"></a>meta.clust <span class="ot">=</span> <span class="fu">cbind</span>(tmp, meta.clust)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Calculate adjusted Rand index, with mclust package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ari <span class="ot">=</span> <span class="fu">mat.or.vec</span>(<span class="fu">ncol</span>(meta.clust), <span class="fu">ncol</span>(meta.clust))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(meta.clust)){</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(meta.clust)){</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  ari[i,j] <span class="ot">=</span> mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(meta.clust[,i], meta.clust[,j])</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ari) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ari) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>annot <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">Reduce</span>(rbind,<span class="fu">strsplit</span>(<span class="fu">colnames</span>(meta.clust), <span class="st">"_"</span>)))</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annot) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Pipe"</span>, <span class="st">"Integration"</span>, <span class="st">"Resolution"</span>)</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annot) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>nclust <span class="ot">=</span> <span class="fu">apply</span>(meta.clust, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>annot<span class="sc">$</span>nClust <span class="ot">=</span> nclust</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(ari, <span class="at">annotation_row =</span> annot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-17-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>All the scanorama ones are similar, except for the one with counts. Harmony on the bioc object also stands out as quite different.</p>
<p>Number of clusters per method:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>nclust <span class="ot">=</span> <span class="fu">apply</span>(meta.clust, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">10</span>,<span class="dv">3</span>,<span class="dv">3</span>))</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(nclust, <span class="at">horiz=</span>T, <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.names=</span><span class="fl">0.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-18-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="clustering-consistency" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="clustering-consistency"><span class="header-section-number">7</span> Clustering consistency</h2>
<p>Compare all these different clusterings by calculating how often each cell pair is in the same cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>calculateCLProb <span class="ot">=</span> <span class="cf">function</span>(nC,nS){</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    P <span class="ot">=</span> (nS<span class="sc">/</span>nC<span class="dv">-1</span>) <span class="sc">/</span> (nS<span class="dv">-1</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(P)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>clusterOverlapN <span class="ot">&lt;-</span> <span class="cf">function</span>(meta.df){</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># meta.df is a data frame with clustering results in each column </span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  dmats <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  probs <span class="ot">=</span> <span class="fu">mat.or.vec</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(meta.df))</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(meta.df)){</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>    tmp <span class="ot">=</span> <span class="fu">as</span>(<span class="fu">outer</span>(meta.df[,i],meta.df[,i], <span class="at">FUN =</span> <span class="st">"=="</span>)<span class="sc">+</span><span class="dv">0</span>, <span class="st">"dgCMatrix"</span>)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diag</span>(tmp) <span class="ot">=</span> <span class="dv">0</span> <span class="co">#score to self to zero.</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># normalize the counts by probability given number of clusters and cells.</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>    prob <span class="ot">=</span> <span class="fu">calculateCLProb</span>(<span class="fu">length</span>(<span class="fu">unique</span>(meta.df[,i])),<span class="fu">length</span>(meta.df[,i]))</span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>    probs[i] <span class="ot">=</span> prob</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>    tmp <span class="ot">=</span> tmp <span class="sc">/</span>prob</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>    dmats[[i]] <span class="ot">=</span> tmp</span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(dmats) <span class="ot">=</span> <span class="fu">colnames</span>(meta.df)</span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add all matrices together, divide by number of matrices.</span></span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a>  sum.dist <span class="ot">=</span> <span class="fu">Reduce</span>(<span class="st">"+"</span>,dmats)<span class="sc">/</span><span class="fu">ncol</span>(meta.df)</span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># scores each clustering by adding together the individual consensus scores for the cell pairs that we have. </span></span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Divide by number of cells. </span></span>
<span id="cb37-26"><a href="#cb37-26" aria-hidden="true" tabindex="-1"></a>  scores <span class="ot">=</span> <span class="fu">lapply</span>(dmats, <span class="cf">function</span>(x) <span class="fu">sum</span>(x<span class="sc">*</span>sum.dist)<span class="sc">/</span><span class="fu">nrow</span>(meta.df)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb37-27"><a href="#cb37-27" aria-hidden="true" tabindex="-1"></a>  scores <span class="ot">=</span> <span class="fu">unlist</span>(scores)</span>
<span id="cb37-28"><a href="#cb37-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(scores) <span class="ot">=</span> <span class="fu">colnames</span>(meta.df)</span>
<span id="cb37-29"><a href="#cb37-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">overlapMatrices =</span> dmats, <span class="at">sumMatrix =</span> sum.dist, <span class="at">scores =</span> scores, <span class="at">probabilities =</span> probs))</span>
<span id="cb37-30"><a href="#cb37-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">=</span> <span class="fu">clusterOverlapN</span>(meta.clust)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>cs <span class="ot">=</span> <span class="fu">colSums</span>(stats<span class="sc">$</span>sumMatrix)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cs) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">AddMetaData</span>(all, cs, <span class="at">col.name =</span> <span class="st">"overlapN"</span>)</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">AddMetaData</span>(all, <span class="fu">log10</span>(cs), <span class="at">col.name =</span> <span class="st">"overlapN.log"</span>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"overlapN"</span>, <span class="st">"overlapN.log"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-20-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>For each cell, we have stats on how often it is in the same cluster as another cell. Ranging from 0 to nclust.</p>
<p>Now we have a score for each pair of cells of how frequently they cluster together. So we can just multiply the matrices to see if high scoring cells are grouped in that clustering. But we need to adjust for cluster sizes, larger clusters have bigger chance of having a high score. So the scores are normalized by the number of clusters with <code>calculateCLProb</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>annot<span class="sc">$</span>Score <span class="ot">=</span> stats<span class="sc">$</span>scores</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>annot<span class="sc">$</span>Name <span class="ot">=</span> <span class="fu">rownames</span>(annot)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(annot, <span class="fu">aes</span>(<span class="at">x=</span>Name, <span class="at">y=</span>Score, <span class="at">color=</span>Pipe, <span class="at">fill=</span>Integration)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">RotatedAxis</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-21-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Generally high scores for harmony with Scanpy or Seurat, CCA clearly differs. But the scores are biased and only show what is closest to consensus.</p>
</section>
<section id="clustering-in-the-pipelines" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="clustering-in-the-pipelines"><span class="header-section-number">8</span> Clustering in the pipelines</h2>
<p>Above, we did the clustering with the same method on the different integrated spaces. Now we will instead explore the integrated clustering from the different pipelines. We still only use the cells that are in common from all 3 pipelines.</p>
<p>The graph clustering is implemented quite differently in the different pipelines.</p>
<ul>
<li>Seurat - runs detection on SNN, resolution implemented</li>
<li>Scanpy - runs detection on KNN, resolution implemented</li>
<li>Bioc - runs detection on SNN, but different scoring to Seurat, resolution parameter behaves strange. Instead use <code>k</code> to tweak cluster resolution.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>seu.columns <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"RNA_snn_res"</span>,<span class="st">"kmeans_"</span>,<span class="st">"hc_"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>bioc.columns <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"louvain_"</span>,<span class="st">"leiden_"</span>,<span class="st">"hc_"</span>,<span class="st">"kmeans_"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>scpy.columns <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"leiden_"</span>,<span class="st">"hclust_"</span>,<span class="st">"kmeans_"</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">rowSums</span>(<span class="fu">sapply</span>(seu.columns, grepl, <span class="fu">colnames</span>(sobj<span class="sc">@</span>meta.data)))<span class="sc">&gt;</span><span class="dv">0</span>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>tmp1 <span class="ot">=</span> sobj<span class="sc">@</span>meta.data[<span class="fu">match</span>(in.all, meta.seurat<span class="sc">$</span>cell), idx]</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp1) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"seur_"</span>, <span class="fu">colnames</span>(tmp1))</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">rowSums</span>(<span class="fu">sapply</span>(bioc.columns, grepl, <span class="fu">colnames</span>(bioc<span class="sc">@</span>meta.data)))<span class="sc">&gt;</span><span class="dv">0</span>)</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">=</span> bioc<span class="sc">@</span>meta.data[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell), idx]</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp2) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"bioc_"</span>, <span class="fu">colnames</span>(tmp2))</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>meta.clust2 <span class="ot">=</span> <span class="fu">cbind</span>(tmp1, tmp2)</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">rowSums</span>(<span class="fu">sapply</span>(scpy.columns, grepl, <span class="fu">colnames</span>(scanpy<span class="sc">@</span>meta.data)))<span class="sc">&gt;</span><span class="dv">0</span>)</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">=</span> scanpy<span class="sc">@</span>meta.data[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell), idx]</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp2) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"scpy_"</span>, <span class="fu">colnames</span>(tmp2))</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>meta.clust2 <span class="ot">=</span> <span class="fu">cbind</span>(meta.clust2, tmp2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Group based on adjusted Rand index.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>ari <span class="ot">=</span> <span class="fu">mat.or.vec</span>(<span class="fu">ncol</span>(meta.clust2), <span class="fu">ncol</span>(meta.clust2))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(meta.clust2)){</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(meta.clust2)){</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  ari[i,j] <span class="ot">=</span> mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(meta.clust2[,i], meta.clust2[,j])</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ari) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust2)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ari) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust2)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>m <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">colnames</span>(meta.clust2), <span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">2</span>]))</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">colnames</span>(meta.clust2), <span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>]))</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>nc <span class="ot">=</span> <span class="fu">apply</span>(meta.clust2, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>       </span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>annot <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Pipe=</span>p, <span class="at">Method=</span>m, <span class="at">nC=</span>nc)</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(annot) <span class="ot">=</span> <span class="fu">colnames</span>(meta.clust2)</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(ari, <span class="at">annotation_row =</span> annot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-23-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Clustering consistency per cell</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>stats2 <span class="ot">=</span> <span class="fu">clusterOverlapN</span>(meta.clust2)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>cs <span class="ot">=</span> <span class="fu">colSums</span>(stats2<span class="sc">$</span>sumMatrix)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cs) <span class="ot">=</span> <span class="fu">colnames</span>(all)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">AddMetaData</span>(all, cs, <span class="at">col.name =</span> <span class="st">"overlapN.all"</span>)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> <span class="fu">AddMetaData</span>(all, <span class="fu">log10</span>(cs), <span class="at">col.name =</span> <span class="st">"overlapN.all.log"</span>)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"overlapN.all"</span>, <span class="st">"overlapN.all.log"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-24-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Score each method based on how often they place high scoring cells in different clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>nclust2 <span class="ot">=</span> <span class="fu">apply</span>(meta.clust2,<span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nclust2, stats2<span class="sc">$</span>scores)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nclust2, stats2<span class="sc">$</span>scores<span class="sc">/</span>nclust2)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nclust2, stats2<span class="sc">$</span>probabilities)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>annot<span class="sc">$</span>score <span class="ot">=</span> stats2<span class="sc">$</span>scores</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>annot<span class="sc">$</span>Name <span class="ot">=</span> <span class="fu">rownames</span>(annot)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(annot, <span class="fu">aes</span>(<span class="at">x=</span>Name, <span class="at">y =</span> score, <span class="at">fill=</span>Pipe)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">RotatedAxis</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-26-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(annot, <span class="fu">aes</span>(<span class="at">x=</span>nC, <span class="at">y =</span> score, <span class="at">color=</span>Pipe)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-26-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(all, <span class="st">"data/covid/results/merged_all.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="celltype-prediction." class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="celltype-prediction."><span class="header-section-number">9</span> Celltype prediction.</h2>
<p>Done only for sample <code>ctrl13</code>, load all the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_seurat, <span class="st">"seurat_covid_qc_dr_int_cl_ct-ctrl13.rds"</span>))</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path_bioc, <span class="st">"bioc_covid_qc_dr_int_cl_ct-ctrl13.rds"</span>))</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>bioc <span class="ot">=</span> <span class="fu">as.Seurat</span>(sce)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>bioc<span class="sc">$</span>scmap_cluster <span class="ot">=</span> sce<span class="sc">$</span>scmap_cluster <span class="co">#for some reason that column dissappears when running as.Seurat.</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>scanpy.sce <span class="ot">=</span> <span class="fu">readH5AD</span>(<span class="fu">file.path</span>(path_scanpy, <span class="st">"scanpy_covid_qc_dr_int_cl_ct-ctrl13.h5ad"</span>))</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>scanpy <span class="ot">=</span> <span class="fu">as.Seurat</span>(scanpy.sce, <span class="at">counts =</span> <span class="cn">NULL</span>, <span class="at">data =</span> <span class="st">"X"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Merge into one object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>meta.seurat <span class="ot">=</span> sobj<span class="sc">@</span>meta.data</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>meta.scanpy <span class="ot">=</span> scanpy<span class="sc">@</span>meta.data</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>meta.bioc <span class="ot">=</span> bioc<span class="sc">@</span>meta.data</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>meta.bioc<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">rownames</span>(meta.bioc)</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>meta.scanpy<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">rownames</span>(meta.scanpy), <span class="cf">function</span>(x) <span class="fu">substr</span>(x,<span class="dv">1</span>,<span class="fu">nchar</span>(x)<span class="sc">-</span><span class="dv">2</span>))</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>meta.seurat<span class="sc">$</span>cell <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">rownames</span>(meta.seurat),<span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">3</span>]))</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>in.all <span class="ot">=</span> <span class="fu">intersect</span>(<span class="fu">intersect</span>(meta.scanpy<span class="sc">$</span>cell, meta.seurat<span class="sc">$</span>cell), meta.bioc<span class="sc">$</span>cell)</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>tmp1 <span class="ot">=</span> meta.bioc[<span class="fu">match</span>(in.all, meta.bioc<span class="sc">$</span>cell),]</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp1) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp1),<span class="st">"_bioc"</span>)</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">=</span> meta.scanpy[<span class="fu">match</span>(in.all, meta.scanpy<span class="sc">$</span>cell),]</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tmp2) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(tmp2),<span class="st">"_scpy"</span>)</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>all <span class="ot">=</span> sobj[,<span class="fu">match</span>(in.all, meta.seurat<span class="sc">$</span>cell)]</span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a>meta.all <span class="ot">=</span> <span class="fu">cbind</span>(all<span class="sc">@</span>meta.data, tmp1,tmp2)</span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a>all<span class="sc">@</span>meta.data <span class="ot">=</span> meta.all</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Celltype predictions:</p>
<ul>
<li><p>Seurat</p>
<ul>
<li><code>predicted.id</code> - <code>TransferData</code></li>
<li><code>singler.immune</code> - singleR with <code>DatabaseImmuneCellExpressionData</code></li>
<li><code>singler.hpca</code>- singleR with <code>HumanPrimaryCellAtlasData</code></li>
<li><code>singler.ref</code>- singleR with own ref.</li>
<li><code>predicted.celltype.l1</code>, <code>predicted.celltype.l2</code>, <code>predicted.celltype.l3</code> - Azimuth predictions at different levels.</li>
</ul></li>
<li><p>Bioc</p>
<ul>
<li><code>scmap_cluster</code> - scMap with clusters in ref data</li>
<li><code>scmap_cell</code> - scMap with all cells in ref data</li>
<li><code>singler.immune</code> - singleR with <code>DatabaseImmuneCellExpressionData</code></li>
<li><code>singler.hpca</code>- singleR with <code>HumanPrimaryCellAtlasData</code></li>
<li><code>singler.ref</code>- singleR with own ref.</li>
</ul></li>
<li><p>Scanpy</p>
<ul>
<li><code>label_trans</code> - label transfer with scanorama</li>
<li><code>louvain</code> - bbknn</li>
<li><code>predicted_labels</code>, <code>majority_voting</code> - celltypist “Immune_All_High” reference</li>
<li><code>predicted_labels_ref</code>, <code>majority_voting_ref</code> - celltypist with own reference</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>all.pred <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"label_trans_scpy"</span>,   <span class="st">"louvain_scpy"</span>,  <span class="st">"predicted_labels_scpy"</span>, <span class="st">"majority_voting_scpy"</span>, <span class="st">"predicted_labels_ref_scpy"</span>,  <span class="st">"majority_voting_ref_scpy"</span>,      <span class="st">"scmap_cluster_bioc"</span>, <span class="st">"scmap_cell_bioc"</span>,  <span class="st">"singler.immune_bioc"</span>, <span class="st">"singler.hpca_bioc"</span>,  <span class="st">"singler.ref_bioc"</span>, <span class="st">"predicted.id"</span>, <span class="st">"singler.immune"</span>, <span class="st">"singler.hpca"</span>,  <span class="st">"singler.ref"</span>, <span class="st">"predicted.celltype.l1"</span>, <span class="st">"predicted.celltype.l2"</span>, <span class="st">"predicted.celltype.l3"</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>ct <span class="ot">=</span> all<span class="sc">@</span>meta.data[,all.pred]</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">=</span> <span class="fu">sapply</span>(all.pred, <span class="cf">function</span>(x) <span class="fu">DimPlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">group.by =</span> x) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(x))</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots[<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>], <span class="at">ncol =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-30-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots[<span class="dv">10</span><span class="sc">:</span><span class="dv">18</span>], <span class="at">ncol =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-30-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-30-2.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<p>Subset for ones that are using same ref:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>ref.pred <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"label_trans_scpy"</span>,   <span class="st">"louvain_scpy"</span>, <span class="st">"predicted_labels_ref_scpy"</span>,  <span class="st">"majority_voting_ref_scpy"</span>,      <span class="st">"scmap_cluster_bioc"</span>, <span class="st">"scmap_cell_bioc"</span>,   <span class="st">"singler.ref_bioc"</span>, <span class="st">"predicted.id"</span>,  <span class="st">"singler.ref"</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>lev <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">unique</span>(<span class="fu">unlist</span>(<span class="fu">apply</span>(all<span class="sc">@</span>meta.data[,ref.pred],<span class="dv">2</span>,unique))))</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>lev <span class="ot">=</span> <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"cells"</span>,<span class="st">"cell"</span>,lev))</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co">#"FCGR3A+ Monocytes" -&gt; "ncMono"</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co">#"CD14+ Monocytes" -&gt; "cMono"</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co">#"Dendritic cell" -&gt; "cDC"</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>lev <span class="ot">=</span> lev[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">+"</span>,lev)]</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>lev <span class="ot">=</span> lev[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"Dendr"</span>,lev)]</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (rp <span class="cf">in</span> ref.pred){</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">as.character</span>(all<span class="sc">@</span>meta.data[,rp])</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"cells"</span>,<span class="st">"cell"</span>,tmp)</span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"FCGR3A</span><span class="sc">\\</span><span class="st">+ Monocytes"</span>,<span class="st">"ncMono"</span>,tmp)</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"CD14</span><span class="sc">\\</span><span class="st">+ Monocytes"</span>,<span class="st">"cMono"</span>,tmp)</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">=</span> <span class="fu">sub</span>(<span class="st">"Dendritic cell"</span>,<span class="st">"cDC"</span>,tmp)</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>  tmp[<span class="fu">is.na</span>(tmp)] <span class="ot">=</span> <span class="st">"unassigned"</span></span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>  all<span class="sc">@</span>meta.data[,<span class="fu">paste0</span>(rp,<span class="st">"2"</span>)] <span class="ot">=</span> <span class="fu">factor</span>(tmp,<span class="at">levels =</span> lev)</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>coldef <span class="ot">=</span>  RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">"Paired"</span>)</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(coldef) <span class="ot">=</span> lev</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>plots2 <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">paste0</span>(ref.pred,<span class="st">"2"</span>), <span class="cf">function</span>(x) <span class="fu">DimPlot</span>(all, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">group.by =</span> x) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(x) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> coldef))</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plots2, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-31-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="1152"></a></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>ref.pred2 <span class="ot">=</span> <span class="fu">paste0</span>(ref.pred,<span class="st">"2"</span>)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">=</span> all<span class="sc">@</span>meta.data[,ref.pred2]</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">=</span> <span class="fu">sapply</span>(ref.pred2,<span class="cf">function</span>(x) <span class="fu">table</span>(tmp[,x]))</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(counts)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(counts, <span class="fu">aes</span>(<span class="at">x=</span>Var2, <span class="at">y=</span>value, <span class="at">fill=</span>Var1)) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span> <span class="fu">RotatedAxis</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="comparison_pipelines_files/figure-html/unnamed-chunk-32-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="comparison_pipelines_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-session" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">10</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/asabjor/miniconda3/envs/seurat5/lib/libopenblasp-r0.3.27.dylib;  LAPACK version 3.12.0

locale:
[1] sv_SE.UTF-8/sv_SE.UTF-8/sv_SE.UTF-8/C/sv_SE.UTF-8/sv_SE.UTF-8

time zone: Europe/Stockholm
tzcode source: system (macOS)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ComplexHeatmap_2.18.0       scran_1.30.0               
 [3] scuttle_1.12.0              SingleCellExperiment_1.24.0
 [5] SummarizedExperiment_1.32.0 Biobase_2.62.0             
 [7] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
 [9] IRanges_2.36.0              S4Vectors_0.40.2           
[11] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[13] matrixStats_1.4.1           patchwork_1.2.0            
[15] ggplot2_3.5.1               Matrix_1.6-5               
[17] zellkonverter_1.12.1        Seurat_5.1.0               
[19] SeuratObject_5.0.2          sp_2.1-4                   

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.3.3            
  [3] later_1.3.2               bitops_1.0-8             
  [5] filelock_1.0.3            tibble_3.2.1             
  [7] polyclip_1.10-7           basilisk.utils_1.14.1    
  [9] fastDummies_1.7.4         lifecycle_1.0.4          
 [11] doParallel_1.0.17         edgeR_4.0.16             
 [13] globals_0.16.3            lattice_0.22-6           
 [15] MASS_7.3-60.0.1           magrittr_2.0.3           
 [17] limma_3.58.1              plotly_4.10.4            
 [19] rmarkdown_2.28            yaml_2.3.10              
 [21] metapod_1.10.0            httpuv_1.6.15            
 [23] sctransform_0.4.1         spam_2.10-0              
 [25] spatstat.sparse_3.1-0     reticulate_1.39.0        
 [27] cowplot_1.1.3             pbapply_1.7-2            
 [29] RColorBrewer_1.1-3        abind_1.4-5              
 [31] zlibbioc_1.48.0           Rtsne_0.17               
 [33] purrr_1.0.2               RCurl_1.98-1.16          
 [35] circlize_0.4.16           GenomeInfoDbData_1.2.11  
 [37] ggrepel_0.9.6             irlba_2.3.5.1            
 [39] listenv_0.9.1             spatstat.utils_3.1-0     
 [41] pheatmap_1.0.12           goftest_1.2-3            
 [43] RSpectra_0.16-2           spatstat.random_3.2-3    
 [45] dqrng_0.3.2               fitdistrplus_1.2-1       
 [47] parallelly_1.38.0         DelayedMatrixStats_1.24.0
 [49] leiden_0.4.3.1            codetools_0.2-20         
 [51] DelayedArray_0.28.0       shape_1.4.6.1            
 [53] tidyselect_1.2.1          farver_2.1.2             
 [55] ScaledMatrix_1.10.0       spatstat.explore_3.2-6   
 [57] jsonlite_1.8.8            GetoptLong_1.0.5         
 [59] BiocNeighbors_1.20.0      progressr_0.14.0         
 [61] iterators_1.0.14          ggridges_0.5.6           
 [63] survival_3.7-0            foreach_1.5.2            
 [65] tools_4.3.3               ica_1.0-3                
 [67] Rcpp_1.0.13               glue_1.7.0               
 [69] gridExtra_2.3             SparseArray_1.2.2        
 [71] xfun_0.47                 dplyr_1.1.4              
 [73] withr_3.0.1               fastmap_1.2.0            
 [75] basilisk_1.14.1           bluster_1.12.0           
 [77] fansi_1.0.6               digest_0.6.37            
 [79] rsvd_1.0.5                R6_2.5.1                 
 [81] mime_0.12                 colorspace_2.1-1         
 [83] Cairo_1.6-2               scattermore_1.2          
 [85] tensor_1.5                spatstat.data_3.1-2      
 [87] utf8_1.2.4                tidyr_1.3.1              
 [89] generics_0.1.3            data.table_1.15.4        
 [91] httr_1.4.7                htmlwidgets_1.6.4        
 [93] S4Arrays_1.2.0            uwot_0.1.16              
 [95] pkgconfig_2.0.3           gtable_0.3.5             
 [97] lmtest_0.9-40             XVector_0.42.0           
 [99] htmltools_0.5.8.1         dotCall64_1.1-1          
[101] clue_0.3-65               scales_1.3.0             
[103] png_0.1-8                 knitr_1.48               
[105] rjson_0.2.21              reshape2_1.4.4           
[107] nlme_3.1-165              GlobalOptions_0.1.2      
[109] zoo_1.8-12                stringr_1.5.1            
[111] KernSmooth_2.23-24        parallel_4.3.3           
[113] miniUI_0.1.1.1            pillar_1.9.0             
[115] vctrs_0.6.5               RANN_2.6.2               
[117] promises_1.3.0            BiocSingular_1.18.0      
[119] beachmat_2.18.0           xtable_1.8-4             
[121] cluster_2.1.6             evaluate_0.24.0          
[123] cli_3.6.3                 locfit_1.5-9.9           
[125] compiler_4.3.3            rlang_1.1.4              
[127] crayon_1.5.3              future.apply_1.11.2      
[129] labeling_0.4.3            mclust_6.1.1             
[131] plyr_1.8.9                stringi_1.8.4            
[133] viridisLite_0.4.2         deldir_2.0-4             
[135] BiocParallel_1.36.0       munsell_0.5.1            
[137] lazyeval_0.2.2            spatstat.geom_3.2-9      
[139] dir.expiry_1.10.0         RcppHNSW_0.6.0           
[141] sparseMatrixStats_1.14.0  future_1.34.0            
[143] statmod_1.5.0             shiny_1.9.1              
[145] ROCR_1.0-11               igraph_2.0.3             </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"selector":".lightbox","closeEffect":"zoom","descPosition":"bottom","loop":true,"openEffect":"zoom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>