---
title: "Scater/Scran: Quality control"
#CSS_ALL:
editor_options:
  chunk_output_type: console
---

#CHUNK_OPT:

***
#DATA_TITLE:

#DATA_ALL1:

```{r,message='hide',warning='hide',results='hold'}
webpath <- "https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/new_dataset/labs/data/covid_data_GSE149689/sub/"
dir.create("./data/raw",recursive = T)
file_list <- c("Normal_PBMC_13.h5","Normal_PBMC_14.h5","Normal_PBMC_5.h5",
               "nCoV_PBMC_15.h5","nCoV_PBMC_17.h5","nCoV_PBMC_1.h5")
for(i in file_list){
  download.file( url = paste0(webpath,i) ,
                 destfile = paste0("./data/raw/",i))
}
```

#DATA_ALL2:

```{r, message='hide',warning='hide',results='hold'}
suppressMessages(require(scater))
suppressMessages(require(scran))
suppressMessages(require(cowplot))
suppressMessages(require(org.Hs.eg.db))

if(!require(DoubletFinder)){
   remotes::install_github('chris-mcginnis-ucsf/DoubletFinder',upgrade = F, dependencies=F)
}
suppressMessages(require(DoubletFinder))
```

#DATA_ALL3:

```{r,message='hide',warning='hide',results='hold'}
cov.15 <- Seurat::Read10X_h5(
  filename = "data/raw/nCoV_PBMC_15.h5",
  use.names = T)
cov.1 <- Seurat::Read10X_h5(
  filename = "data/raw/nCoV_PBMC_1.h5",
  use.names = T)
cov.17 <- Seurat::Read10X_h5(
  filename = "data/raw/nCoV_PBMC_17.h5",
  use.names = T)

ctrl.5 <- Seurat::Read10X_h5(
  filename = "data/raw/Normal_PBMC_5.h5",
  use.names = T)
ctrl.13 <- Seurat::Read10X_h5(
  filename = "data/raw/Normal_PBMC_13.h5",
  use.names = T)
ctrl.14 <- Seurat::Read10X_h5(
  filename = "data/raw/Normal_PBMC_14.h5",
  use.names = T)

```

***
#OBJ_TITLE:

#OBJ_ALL1:

```{r}
sce <- SingleCellExperiment( assays = list(counts = cbind(cov.1,cov.15, cov.17, ctrl.5, ctrl.13, ctrl.14)) )
dim(sce)



#Adding metadata
sce@colData$sample <- unlist(sapply(c("cov.1","cov.15","cov.17", "ctrl.5","ctrl.13","ctrl.14"),function(x) rep(x,ncol(get(x)))))
sce@colData$type <- ifelse(grepl("cov",sce@colData$sample), "Covid","Control")

```

#OBJ_ALL1.1:

```{r}
# remove all objects that will not be used.
rm(cov.15, cov.1, cov.17, ctrl.5, ctrl.13, ctrl.14)

# run garbage collect to free up memory
gc()
```


#OBJ_ALL2:

```{r ,  results='hold'}
head(counts(sce)[,1:10])

head(sce@colData,10)
```


***
#QC_TITLE:

#QC_ALL1:

#QC_ALL1.1:

#QC_1_SCATER:

```{r}
# Mitochondrial genes
mito_genes <- rownames(sce)[grep("^MT-",rownames(sce))]

# Ribosomal genes
ribo_genes <- rownames(sce)[grep("^RP[SL]",rownames(sce))]

# Hemoglobin genes - includes all genes starting with HB except HBP.
hb_genes <- rownames(sce)[grep("^HB[^(P)]",rownames(sce))]

```

```{r, results='hold'}
sce <- addPerCellQC(sce, flatten = T, subsets = list(mt=mito_genes, hb=hb_genes, ribo=ribo_genes))

head(colData(sce))

```

#QC_1.1_SCATER:

```{r}
# Way2: Doing it manually
sce@colData$percent_mito <- Matrix::colSums(counts(sce)[mito_genes, ]) / sce@colData$total
```


***
#QC_TITLE2:

#QC_ALL3:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=6,fig.width=10}
# total is total UMIs per cell
# detected is number of detected genes.
# the different gene subset percentages are listed as subsets_mt_percent etc.

plot_grid(plotColData(sce,y = "detected",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "total",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_mt_percent",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_ribo_percent",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_hb_percent",x = "sample",colour_by = "sample"),ncol = 3)
```

#QC_ALL4:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=4,fig.width=16}
plotColData(sce,x = "total", y = "detected",colour_by = "sample")
```

#QC_ALL4.4:

***
#FILTERING_TITLE:

#FILTERING_TITLE1:

#FILTERING_SCATER0:

#FILTERING_ALL0:


```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=4}
dim(sce)

selected_c <-  colnames(sce)[sce$detected > 200]
selected_f <- rownames(sce)[ Matrix::rowSums(counts(sce)) > 3]

sce.filt <- sce[selected_f , selected_c]
dim(sce.filt)
```

#FILTERING_ALL3:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
# skip for now and run doublet detection instead...

#high.det.v3 <- sce.filt$nFeatures > 4100
#high.det.v2 <- (sce.filt$nFeatures > 2000) & (sce.filt$sample_id == "v2.1k")

# remove these cells
#sce.filt <- sce.filt[ , (!high.det.v3) & (!high.det.v2)]

# check number of cells
#ncol(sce.filt)
```

#FILTERING_ALL01:

#FILTERING_SCATER01:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=7,fig.width=12}
#Compute the relative expression of each gene per cell

#Use sparse matrix operations, if your dataset is large, doing matrix devisions the regular way will take a very long time.
C = counts(sce)
C@x = C@x / rep.int(colSums(C), diff(C@p))
most_expressed <- order(Matrix::rowSums( C ),decreasing = T)[20:1]
boxplot( as.matrix(t(C[most_expressed,])),cex=.1, las=1, xlab="% total count per cell",col=scales::hue_pal()(20)[20:1],horizontal=TRUE)

rm(C)

# also, there is the option of running the function "plotHighestExprs" in the scater package, however, this function takes very long to execute.
```

#FILTERING_ALL02:

#FILTERING_TITLE2:

#FILTERING_ALL1:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
selected_mito <- sce.filt$subsets_mt_percent < 30
selected_ribo <- sce.filt$subsets_ribo_percent > 5

# and subset the object to only keep those cells
sce.filt <- sce.filt[, selected_mito & selected_ribo ]
dim(sce.filt)
```

#FILTERING_ALL2:

#FILTERING_TITLE4:

#FILTERING_ALL5:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=6,fig.width=10}
plot_grid(plotColData(sce,y = "detected",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "total",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_mt_percent",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_ribo_percent",x = "sample",colour_by = "sample"),
          plotColData(sce,y = "subsets_hb_percent",x = "sample",colour_by = "sample"),ncol = 3)

```

#FILTERING_TITLE5:

#FILTERING_ALL6:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=4,fig.width=4}
dim(sce.filt)

# Filter MALAT1
sce.filt <- sce.filt[ ! grepl("MALAT1", rownames(sce.filt)), ]

# Filter Mitocondrial
sce.filt <- sce.filt[ ! grepl("^MT-", rownames(sce.filt)), ]

# Filter Ribossomal gene (optional if that is a problem on your data)
# sce.filt <- sce.filt[ ! grepl("^RP[SL]", rownames(sce.filt)), ]

# Filter Hemoglobin gene
sce.filt <- sce.filt[ ! grepl("^HB[^(P)]", rownames(sce.filt)), ]

dim(sce.filt)
```


#SEX_TITLE:

#SEX_ALL1:

#SEX_ALL2:

```{r}
genes.file = "data/results/genes.table.csv"

if (!file.exists(genes.file)){
  suppressMessages(require(biomaRt))

  # initialize connection to mart, may take some time if the sites are unresponsive.
  mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl")

  # fetch chromosome info plus some other annotations
  genes.table <- try(biomaRt::getBM(attributes = c("ensembl_gene_id",
              "external_gene_name", "description","gene_biotype", "chromosome_name","start_position"),
              mart = mart, useCache = F))

  if(!dir.exists("data/results")){dir.create("data/results")}
  if(is.data.frame(genes.table)){write.csv(genes.table, file = genes.file)}

  if (!file.exists(genes.file)){
  download.file("https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/misc/genes.table.csv",destfile = "data/results/genes.table.csv")
    genes.table = read.csv(genes.file)
    }
}else{
  genes.table = read.csv(genes.file)
}

genes.table <- genes.table[genes.table$external_gene_name %in% rownames(sce.filt),]
```

#SEX_ALL3:

```{r}
chrY.gene = genes.table$external_gene_name[genes.table$chromosome_name == "Y"]

sce.filt@colData$pct_chrY  = Matrix::colSums(counts(sce.filt)[chrY.gene,]) / colSums(counts(sce.filt))
```

#SEX_ALL4:

```{r}
# as plotColData cannot take an expression vs metadata, we need to add in XIST expression to colData
sce.filt@colData$XIST = counts(sce.filt)["XIST",] /  colSums(counts(sce.filt)) * 10000

plotColData(sce.filt,  "XIST",  "pct_chrY")
```

#SEX_ALL5:

```{r}
plot_grid(plotColData(sce.filt, y = "XIST", x = "sample", colour_by = "sample"),
          plotColData(sce.filt, y = "pct_chrY", x = "sample", colour_by = "sample"),  ncol = 2)

```

#SEX_ALL6:

#CELLCYCLE_TITLE:

#CELLCYCLE_ALL1:

```{r cc,message='hide',warning='hide', results='hold',results='hold',fig.height=4,fig.width=4}
hs.pairs <- readRDS(system.file("exdata", "human_cycle_markers.rds", package="scran"))
anno <- select(org.Hs.eg.db, keys=rownames(sce.filt), keytype="SYMBOL", column="ENSEMBL")
ensembl <- anno$ENSEMBL[match(rownames(sce.filt), anno$SYMBOL)]

#Use only genes related to biological process cell cycle to speed up
#https://www.ebi.ac.uk/QuickGO/term/GO:0007049 = cell cycle (BP,Biological Process)
GOs <- na.omit(select(org.Hs.eg.db, keys=na.omit(ensembl), keytype="ENSEMBL", column="GO"))
GOs <- GOs[GOs$GO == "GO:0007049","ENSEMBL"]
hs.pairs <- lapply(hs.pairs,function(x){ x[rowSums( apply(x, 2, function(i) i %in% GOs)) >= 1,]})
str(hs.pairs)
cc.ensembl <- ensembl[ensembl %in% GOs] #This is the fastest (less genes), but less accurate too
#cc.ensembl <- ensembl[ ensembl %in% unique(unlist(hs.pairs))]


assignments <- cyclone(sce.filt[ensembl %in% cc.ensembl,], hs.pairs, gene.names= ensembl[ ensembl %in% cc.ensembl])
sce.filt$G1.score <- assignments$scores$G1
sce.filt$G2M.score <- assignments$scores$G2M
sce.filt$S.score <- assignments$scores$S

```

#CELLCYCLE_ALL2:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
plot_grid(plotColData(sce.filt,y = "G2M.score",x = "G1.score",colour_by = "sample"),
          plotColData(sce.filt,y = "G2M.score",x = "sample",colour_by = "sample"),
          plotColData(sce.filt,y = "G1.score",x = "sample",colour_by = "sample"),
          plotColData(sce.filt,y = "S.score",x = "sample",colour_by = "sample"),ncol = 4)
```

#CELLCYCLE_SCATER3:

#DOUBLET_TITLE:

#DOUBLET_ALL1:

#DOUBLET_SCATER1:
#DOUBLET_SCATER1:

```{r}
sce.filt <- logNormCounts(sce.filt)
dec <- modelGeneVar(sce.filt, block = sce.filt$sample)
hvgs = getTopHVGs(dec, n=2000)

sce.filt <- runPCA(sce.filt, subset_row=hvgs)

sce.filt <- runUMAP(sce.filt, pca = 10)
```

```{r}
suppressPackageStartupMessages(require(scDblFinder))

# run computeDoubletDensity with 10 principal components.
sce.filt <- scDblFinder(sce.filt, dims = 10)
```

```{r}
plot_grid(plotUMAP(sce.filt, colour_by="scDblFinder.score"),
          plotUMAP(sce.filt, colour_by="scDblFinder.class"),
          plotUMAP(sce.filt, colour_by="sample"), ncol = 3)
```

#DOUBLET_ALL2:

```{r}
sce.filt = sce.filt[,sce.filt$scDblFinder.score < 2]

dim(sce.filt)
```

#DOUBLET_ALL3:

```{r,message='hide',warning='hide', results='hold',results='hold',fig.height=5,fig.width=16}
dir.create('data/results', showWarnings = F)

saveRDS(sce.filt,"data/results/covid_qc.rds")
```

#DOUBLET_ALL1.1:

#SESSION_INFO:

```{r}
sessionInfo()
```
