<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Jennifer Fransson">
  <title>Differential Gene Expression</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
  <link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">
</head>
<body class="quarto-light">

  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/images/home.jpg" data-background-opacity="1" data-background-size="cover" class="center">
  <div class="title-container">

        <div class="title-parent title-two">
      <div class="title-child-text">
        <h1 class="title">Differential Gene Expression</h1>
                <p class="subtitle">Single Cell RNA-Seq Analysis</p>
                        <p class="author">Jennifer Fransson</p>
                                <p class="date">27-Mar-2025</p>
              </div>
      <div class="title-child-hero">
                <div class="title-hero">
          <img class="hero" src="../../assets/images/slide-hero.png">
        </div>
              </div>
    </div>

    
  </div>
</section>

<section id="workflow" class="slide level2">
<h2>Workflow</h2>
<ul>
<li>Quality control</li>
<li>Cell cycle phase classification</li>
<li>Normalization</li>
<li>Select highly variable genes</li>
<li>Data integration</li>
<li>Clustering</li>
<li><strong>Differential gene expression</strong></li>
<li>GSA/GSEA</li>
<li>Cell typing</li>
</ul>
</section>
<section id="what-is-differential-gene-expression" class="slide level2">
<h2>What is differential gene expression?</h2>

<img data-src="index_files/figure-revealjs/diffexp_1-1.png" width="768" class="r-stretch"><p>Count data -&gt; statistical analysis -&gt; Are differences significant (greater than expected randomly)</p>
<aside class="notes">
<p>Sometimes, it’s easy to visually identify if the expression of a gene is significantly different between groups and in which direction. If the samples overlap or the spread is too high, it might be harder.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="what-is-differential-gene-expression-1" class="slide level2">
<h2>What is differential gene expression?</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/zeroexample-1.png" width="480"></p>
</div>
</div>
</div><div class="column">
<div class="fragment">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/zeroexample_vln-1.png" width="576"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="what-is-differential-gene-expression-2" class="slide level2">
<h2>What is differential gene expression?</h2>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">avg_log2FC</th>
<th style="text-align: right;">p_val_adj</th>
<th style="text-align: right;">pct.1</th>
<th style="text-align: right;">pct.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CD7</td>
<td style="text-align: right;">5.535220</td>
<td style="text-align: right;">0.0000001</td>
<td style="text-align: right;">0.714</td>
<td style="text-align: right;">0.058</td>
</tr>
<tr class="even">
<td style="text-align: left;">LCK</td>
<td style="text-align: right;">3.605886</td>
<td style="text-align: right;">0.0000046</td>
<td style="text-align: right;">0.679</td>
<td style="text-align: right;">0.077</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HLA-DPB1</td>
<td style="text-align: right;">-5.291575</td>
<td style="text-align: right;">0.0000051</td>
<td style="text-align: right;">0.107</td>
<td style="text-align: right;">0.769</td>
</tr>
<tr class="even">
<td style="text-align: left;">HLA-DRA</td>
<td style="text-align: right;">-4.128576</td>
<td style="text-align: right;">0.0000126</td>
<td style="text-align: right;">0.357</td>
<td style="text-align: right;">0.865</td>
</tr>
<tr class="odd">
<td style="text-align: left;">HLA-DRB1</td>
<td style="text-align: right;">-5.027130</td>
<td style="text-align: right;">0.0000172</td>
<td style="text-align: right;">0.107</td>
<td style="text-align: right;">0.731</td>
</tr>
<tr class="even">
<td style="text-align: left;">GNLY</td>
<td style="text-align: right;">8.198735</td>
<td style="text-align: right;">0.0000191</td>
<td style="text-align: right;">0.571</td>
<td style="text-align: right;">0.058</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GZMM</td>
<td style="text-align: right;">3.120563</td>
<td style="text-align: right;">0.0000767</td>
<td style="text-align: right;">0.571</td>
<td style="text-align: right;">0.058</td>
</tr>
<tr class="even">
<td style="text-align: left;">CD3D</td>
<td style="text-align: right;">2.255304</td>
<td style="text-align: right;">0.0000805</td>
<td style="text-align: right;">0.643</td>
<td style="text-align: right;">0.096</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GZMA</td>
<td style="text-align: right;">3.078594</td>
<td style="text-align: right;">0.0001174</td>
<td style="text-align: right;">0.571</td>
<td style="text-align: right;">0.058</td>
</tr>
<tr class="even">
<td style="text-align: left;">HLA-DPA1</td>
<td style="text-align: right;">-3.661491</td>
<td style="text-align: right;">0.0002595</td>
<td style="text-align: right;">0.179</td>
<td style="text-align: right;">0.712</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<aside class="notes">
<p>You can still miss bimodal distributions if the “negative” part is &gt; 0</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="what-is-differential-gene-expression-3" class="slide level2">
<h2>What is differential gene expression?</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p><img data-src="assets/distributions_Tiberi_2023.png"></p>
<div class="small">
<p><span class="citation" data-cites="tiberi2023distinct">(Modified from <a href="#/references" role="doc-biblioref" onclick="">Tiberi et al., 2023</a>)</span></p>
</div>
</div><div class="column" style="width:40%;">
<ul>
<li>Most methods focus on fold-change</li>
<li>Fold-change can reflect many differences</li>
<li>Not all differences are evident in fold-change or proportion of cells expressing gene</li>
</ul>
</div>
</div>
</section>
<section id="defining-groups-of-interest" class="slide level2">
<h2>Defining groups of interest</h2>
<ul>
<li><em>A priori</em> defined groups: Compare cells from different samples, e.g.:
<ul>
<li>Experimental groups (treatment, time points, clinical information etc)</li>
<li>Sorted cells</li>
</ul></li>
<li>Data-driven definition: Compare cells depending on analysis output, e.g.:
<ul>
<li>RNA-based clustering/identity</li>
<li>Identity based on other data from multi-omics</li>
</ul></li>
</ul>
<p><em>Warning: Performing DE on clusters defined with the same data (“double-dipping”) will inflate DE analysis. Be mindful of this when you interpret the results.</em></p>
</section>
<section id="functions" class="slide level2">
<h2>Functions</h2>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">Toolkit</th>
<th style="text-align: left;">Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat</td>
<td style="text-align: left;"><code>FindMarkers()</code>, <code>FindAllMarkers()</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Scran</td>
<td style="text-align: left;"><code>findMarkers()</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy</td>
<td style="text-align: left;"><code>scanpy.tl.rank_genes_groups()</code></td>
</tr>
</tbody>
</table>
</section>
<section id="findallmarkers" class="slide level2">
<h2>FindAllMarkers</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2" data-code-line-numbers="6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">FindAllMarkers</span>(</span>
<span id="cb2-2"><a href="#cb2-2"></a>  object,</span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="at">assay =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="at">features =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="at">logfc.threshold =</span> <span class="fl">0.1</span>,</span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="at">test.use =</span> <span class="st">"wilcox"</span>,</span>
<span id="cb2-7"><a href="#cb2-7"></a>  <span class="at">slot =</span> <span class="st">"data"</span>,</span>
<span id="cb2-8"><a href="#cb2-8"></a>  <span class="at">min.pct =</span> <span class="fl">0.01</span>,</span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="at">min.diff.pct =</span> <span class="sc">-</span><span class="cn">Inf</span>,</span>
<span id="cb2-10"><a href="#cb2-10"></a>  <span class="at">node =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-11"><a href="#cb2-11"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-12"><a href="#cb2-12"></a>  <span class="at">only.pos =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="at">max.cells.per.ident =</span> <span class="cn">Inf</span>,</span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="at">random.seed =</span> <span class="dv">1</span>,</span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="at">latent.vars =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="at">min.cells.feature =</span> <span class="dv">3</span>,</span>
<span id="cb2-17"><a href="#cb2-17"></a>  <span class="at">min.cells.group =</span> <span class="dv">3</span>,</span>
<span id="cb2-18"><a href="#cb2-18"></a>  <span class="at">mean.fxn =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-19"><a href="#cb2-19"></a>  <span class="at">fc.name =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-20"><a href="#cb2-20"></a>  <span class="at">base =</span> <span class="dv">2</span>,</span>
<span id="cb2-21"><a href="#cb2-21"></a>  <span class="at">return.thresh =</span> <span class="fl">0.01</span>,</span>
<span id="cb2-22"><a href="#cb2-22"></a>  <span class="at">densify =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-23"><a href="#cb2-23"></a>  ...</span>
<span id="cb2-24"><a href="#cb2-24"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span class="small">Seurat 5.1.0</span></p>
<aside class="notes">
<p>FindAllMarkers() does a 1 vs rest comparison for all Identities. <code>assay</code> must be RNA. <code>slot</code> is usually data.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="tests" class="slide level2">
<h2>Tests</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3" data-code-line-numbers="1,7"><pre class="sourceCode numberSource yaml number-lines code-with-copy"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">"wilcox" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a Wilcoxon Rank Sum test (default)</span></span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="fu">"bimod" </span><span class="kw">:</span><span class="at"> Likelihood-ratio test for single cell gene expression, (McDavid et al., Bioinformatics, 2013)</span></span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="fu">"roc" </span><span class="kw">:</span><span class="at"> Identifies 'markers' of gene expression using ROC analysis. For each gene, evaluates (using AUC) a classifier built on that gene alone, to classify between two groups of cells. An AUC value of 1 means that expression values for this gene alone can perfectly classify the two groupings (i.e. Each of the cells in cells.1 exhibit a higher level than each of the cells in cells.2). An AUC value of 0 also means there is perfect classification, but in the other direction. A value of 0.5 implies that the gene has no predictive power to classify the two groups. Returns a 'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially expressed genes.</span></span>
<span id="cb3-6"><a href="#cb3-6"></a></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="fu">"t" </span><span class="kw">:</span><span class="at"> Identify differentially expressed genes between two groups of cells using the Student's t-test.</span></span>
<span id="cb3-8"><a href="#cb3-8"></a></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="fu">"negbinom" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a negative binomial generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb3-10"><a href="#cb3-10"></a></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="fu">"poisson" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a poisson generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb3-12"><a href="#cb3-12"></a></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="fu">"LR" </span><span class="kw">:</span><span class="at"> Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</span></span>
<span id="cb3-14"><a href="#cb3-14"></a></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="fu">"MAST" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST package to run the DE testing.</span></span>
<span id="cb3-16"><a href="#cb3-16"></a></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="fu">"DESeq2" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (Love et al, Genome Biology, 2014).This test does not support pre-filtering of genes based on average difference (or percent detection rate) between cell groups. However, genes may be pre-filtered based on their minimum detection rate (min.pct) across both cell groups. To use this method, please install DESeq2, using the instructions at https://bioconductor.org/packages/release/bioc/html/DESeq2.html</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tests-1" class="slide level2">
<h2>Tests</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4" data-code-line-numbers="9,11,17"><pre class="sourceCode numberSource yaml number-lines code-with-copy"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">"wilcox" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a Wilcoxon Rank Sum test (default)</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="fu">"bimod" </span><span class="kw">:</span><span class="at"> Likelihood-ratio test for single cell gene expression, (McDavid et al., Bioinformatics, 2013)</span></span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="fu">"roc" </span><span class="kw">:</span><span class="at"> Identifies 'markers' of gene expression using ROC analysis. For each gene, evaluates (using AUC) a classifier built on that gene alone, to classify between two groups of cells. An AUC value of 1 means that expression values for this gene alone can perfectly classify the two groupings (i.e. Each of the cells in cells.1 exhibit a higher level than each of the cells in cells.2). An AUC value of 0 also means there is perfect classification, but in the other direction. A value of 0.5 implies that the gene has no predictive power to classify the two groups. Returns a 'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially expressed genes.</span></span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="fu">"t" </span><span class="kw">:</span><span class="at"> Identify differentially expressed genes between two groups of cells using the Student's t-test.</span></span>
<span id="cb4-8"><a href="#cb4-8"></a></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="fu">"negbinom" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a negative binomial generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb4-10"><a href="#cb4-10"></a></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="fu">"poisson" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a poisson generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="fu">"LR" </span><span class="kw">:</span><span class="at"> Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</span></span>
<span id="cb4-14"><a href="#cb4-14"></a></span>
<span id="cb4-15"><a href="#cb4-15"></a><span class="fu">"MAST" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST package to run the DE testing.</span></span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="fu">"DESeq2" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (Love et al, Genome Biology, 2014).This test does not support pre-filtering of genes based on average difference (or percent detection rate) between cell groups. However, genes may be pre-filtered based on their minimum detection rate (min.pct) across both cell groups. To use this method, please install DESeq2, using the instructions at https://bioconductor.org/packages/release/bioc/html/DESeq2.html</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<aside class="notes">
<p>wilcox: non-parametric, fast, no covariates t-test: normal distribution, fast, no covariates mast: parametric, slow, complex modelling negbinom: neg bionomial distribution, only umi data poisson: poisson distribution, only umi data</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="distributions" class="slide level2">
<h2>Distributions</h2>
<div class="columns">
<div class="column">
<p><img data-src="assets/distributions.png"></p>
</div><div class="column">
<ul>
<li>High noise (technical + biology)</li>
<li>Low library sizes</li>
<li>Low mRNA quantity</li>
<li>Amplification bias, drop-outs</li>
<li>3’ bias, partial coverage</li>
<li>Bursting</li>
<li>Mixed cell types</li>
</ul>
</div>
</div>
<aside class="notes">
<p>What distribution is single-cell data? Distributions differ gene to gene.</p>
<p>Data is zero inflated. Some distributions capture this closely: ZINB, Poisson-Beta</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="tests-2" class="slide level2">
<h2>Tests</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb5" data-code-line-numbers="15"><pre class="sourceCode numberSource yaml number-lines code-with-copy"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">"wilcox" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a Wilcoxon Rank Sum test (default)</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="fu">"bimod" </span><span class="kw">:</span><span class="at"> Likelihood-ratio test for single cell gene expression, (McDavid et al., Bioinformatics, 2013)</span></span>
<span id="cb5-4"><a href="#cb5-4"></a></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="fu">"roc" </span><span class="kw">:</span><span class="at"> Identifies 'markers' of gene expression using ROC analysis. For each gene, evaluates (using AUC) a classifier built on that gene alone, to classify between two groups of cells. An AUC value of 1 means that expression values for this gene alone can perfectly classify the two groupings (i.e. Each of the cells in cells.1 exhibit a higher level than each of the cells in cells.2). An AUC value of 0 also means there is perfect classification, but in the other direction. A value of 0.5 implies that the gene has no predictive power to classify the two groups. Returns a 'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially expressed genes.</span></span>
<span id="cb5-6"><a href="#cb5-6"></a></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="fu">"t" </span><span class="kw">:</span><span class="at"> Identify differentially expressed genes between two groups of cells using the Student's t-test.</span></span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="fu">"negbinom" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a negative binomial generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb5-10"><a href="#cb5-10"></a></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="fu">"poisson" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a poisson generalized linear model. Use only for UMI-based datasets</span></span>
<span id="cb5-12"><a href="#cb5-12"></a></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="fu">"LR" </span><span class="kw">:</span><span class="at"> Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</span></span>
<span id="cb5-14"><a href="#cb5-14"></a></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="fu">"MAST" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST package to run the DE testing.</span></span>
<span id="cb5-16"><a href="#cb5-16"></a></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="fu">"DESeq2" </span><span class="kw">:</span><span class="at"> Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (Love et al, Genome Biology, 2014).This test does not support pre-filtering of genes based on average difference (or percent detection rate) between cell groups. However, genes may be pre-filtered based on their minimum detection rate (min.pct) across both cell groups. To use this method, please install DESeq2, using the instructions at https://bioconductor.org/packages/release/bioc/html/DESeq2.html</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="hurdle-models" class="slide level2">
<h2>Hurdle models</h2>
<blockquote>
<p>…most computational methods still stick with the old mentality of viewing differential expression as a simple ‘up or down’ phenomenon. We advocate that we should fully embrace the features of single cell data, which allows us to observe binary (from Off to On) as well as continuous (the amount of expression) regulations. <span class="citation" data-cites="wu2018two">Wu et al. (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span></p>
</blockquote>
<aside class="notes">
<p>But single cell expression looks more like this. With lots of zeros. Now, it’s harder to say if the difference is significant. Apart from the level of expression, the number of samples (cells) that express the gene also becomes relevant. If we only look at expression level, they might be similar, but if we look at number of cells, one might be higher than the other.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="fragment">
<p>MAST</p>
<ul>
<li>Two part GLM (Hurdle model)</li>
<li>Models the continuous nature of gene expression and the discrete binary nature of gene detection</li>
<li>Detection hurdle
<ul>
<li>Expression detected or not?</li>
<li>Logistic regression</li>
<li>If gene is not detected, stop, else move to next hurdle</li>
</ul></li>
<li>Expression hurdle
<ul>
<li>Genes with positive expression levels modelled using GLM</li>
</ul></li>
<li>Hurdle model is able to handle drop-outs</li>
<li>Support complex modelling</li>
</ul>
<p><span class="small"><span class="citation" data-cites="finak2015mast">Finak et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span></span></p>
</div>
</section>
<section id="complex-designs" class="slide level2">
<h2>Complex designs</h2>
<ul>
<li>Comparing groups of samples (e.g.&nbsp;patients vs controls)</li>
<li>Including batch effects</li>
<li>Correcting for covariates (e.g.&nbsp;age)</li>
</ul>
</section>
<section id="complex-designs-groups-of-samples" class="slide level2">
<h2>Complex designs: Groups of samples</h2>
<div class="columns">
<p>Example: 3 patients vs 3 controls</p><p><span class="math inline">\(n\)</span>: Number of cells (1000s) or number of individuals (3)?</p>
</div>
<div class="columns">
<div class="column" style="width:35%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simple_complex-1.png" width="288"></p>
</div>
</div>
</div><div class="column" style="width:65%;">
<div class="fragment">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/patients-1.png" width="480"></p>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<p>Many tests assume independence!</p>
</div>
</section>
<section id="complex-designs-approaches" class="slide level2">
<h2>Complex designs: Approaches</h2>

<img data-src="index_files/figure-revealjs/complexapproaches-1.png" width="960" class="r-stretch"></section>
<section id="complex-designs-approaches-1" class="slide level2">
<h2>Complex designs: Approaches</h2>
<table>
<colgroup>
<col style="width: 20%">
<col style="width: 40%">
<col style="width: 40%">
</colgroup>
<thead>
<tr class="header">
<th>Approach</th>
<th>Advantages</th>
<th>Disadvantages</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Naive model</td>
<td>- Fast</td>
<td>- Sensitive to false positives<br>- Assumptions are often violated</td>
</tr>
<tr class="even">
<td>GLM</td>
<td>- Can account for covariates</td>
<td>- Slower<br>- Not recommended to account for separate individuals</td>
</tr>
<tr class="odd">
<td>Mixed models</td>
<td>- Can account for separate individuals<br>- Lower risk for false positives</td>
<td>- Slow</td>
</tr>
<tr class="even">
<td>Pseudobulk</td>
<td>- Accounts for separate individuals/samples<br>- Low risk for false positives<br>- Fast</td>
<td>- Often has lower power<br>- Cannot include information about intrasample distribution</td>
</tr>
</tbody>
</table>
</section>
<section id="complex-designs-approaches-2" class="slide level2">
<h2>Complex designs: Approaches</h2>
<table>
<colgroup>
<col style="width: 20%">
<col style="width: 13%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 13%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th>Approach</th>
<th>Speed</th>
<th>Can include covariates</th>
<th>Can account for multilevel design</th>
<th>Sensitivity</th>
<th>Specificity</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Naive model</td>
<td>Fast</td>
<td>-</td>
<td>-</td>
<td>High</td>
<td>Low</td>
</tr>
<tr class="even">
<td>GLM</td>
<td>Slow</td>
<td>✓</td>
<td>Not recommended</td>
<td>High</td>
<td>Low</td>
</tr>
<tr class="odd">
<td>Mixed models</td>
<td>Slow</td>
<td>✓</td>
<td>✓</td>
<td>Medium</td>
<td>Medium</td>
</tr>
<tr class="even">
<td>Pseudobulk</td>
<td>Fast</td>
<td>✓</td>
<td>✓</td>
<td>Low</td>
<td>High</td>
</tr>
</tbody>
</table>
<p><br> <br> <em>NB: This table broadly summarizes each approach, but each approach includes many methods with their own advantages and disadvantages.</em></p>
</section>
<section id="vs-1-and-1-vs-all" class="slide level2">
<h2>1-vs-1 and 1-vs-all</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simpledesign_datadriven-1.png" width="576"></p>
</div>
</div>
</div><div class="column" style="width:60%;">
<ul>
<li>1-vs-1: C1 vs C2</li>
<li>1-vs-all: C1 vs C0 + C2 + C3</li>
</ul>
</div>
</div>
</section>
<section id="vs-all-analysis" class="slide level2">
<h2>1-vs-all analysis</h2>
<div class="columns">
<div class="column">
<p><img data-src="assets/unbalanced-clusters-2.png"></p>
</div><div class="column">
<ul>
<li>Larger clusters will be over-represented unless subsampled</li>
<li>Highly similar clusters
<ul>
<li>Will have most of their DEGs overlapping</li>
<li>Pairwise comparisons might help rather than 1 vs rest</li>
</ul></li>
</ul>
</div>
</div>
<aside class="notes">
<p>When doing a C4 vs all comparison, very few DEGs are found because Clusters 1 and 4 might share a lot of similar genes. Cluster C3 vs all will be mostly C3 vs C1 because C1 dominates the second group. Subsample to bring all clusters to equal number of cells.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="considerations---what-does-the-p-value-mean" class="slide level2">
<h2>Considerations - what does the p-value mean?</h2>
<ul>
<li>Null hypothesis: Mean/median/distribution is equal between group A and group B</li>
<li><span class="math inline">\(p\)</span> &lt; 0.05 : if null hypothesis is true, we can expect the measured result in &lt; 5% of cases where group A and group B have been sampled at <span class="math inline">\(n\)</span></li>
<li><span class="math inline">\(p\)</span> depends on <span class="math inline">\(n\)</span>, variance and intergroup difference
<ul>
<li>As <span class="math inline">\(n\)</span> increases, variance can increase and difference can decrease without losing power</li>
</ul></li>
</ul>

<img data-src="index_files/figure-revealjs/whatissignificant-1.png" width="768" class="r-stretch"><ul>
<li>Likelihood of differences reflecting a true difference in distribution vs random chance</li>
</ul>
</section>
<section id="statistics---what-does-the-p-value-mean" class="slide level2">
<h2>Statistics - what does the p-value mean?</h2>

<img data-src="index_files/figure-revealjs/pvalueexample-1.png" width="576" class="r-stretch"><p>Are all statistically significant differences of interest?</p>
<aside class="notes">
<p>Not as simple as setting a threshold for difference - variance also matters</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="considerations-composition-vs-expression" class="slide level2">
<h2>Considerations: Composition vs expression</h2>

<img data-src="index_files/figure-revealjs/composition-1.png" width="576" class="r-stretch"></section>
<section id="comparing-methods" class="slide level2">
<h2>Comparing Methods</h2>
<div class="columns">
<div class="column" style="width:35%;">
<p><img data-src="assets/soneson-2018-deg-methods.jpg"></p>
</div><div class="column" style="width:65%;">
<ul>
<li>T-test and Wilcoxon work well enough given sufficient number of samples</li>
<li>Bulk methods are not worse than single-cell specific methods</li>
<li>Pre-filtering lowly expressed genes is important for bulk methods</li>
</ul>
<p><span class="small"><span class="citation" data-cites="soneson2018bias">Soneson &amp; Robinson (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span></span></p>
</div>
</div>
<aside class="notes">
<ul>
<li>36 datasets (31 full-length, 5 UMI based)</li>
<li>edgeRQLFDetRate: EdgeR when normalized for detection rate</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="comparing-methods-1" class="slide level2">
<h2>Comparing Methods</h2>

<img data-src="assets/soneson-2018-gene-ranking.jpg" class="r-stretch"><p>Similarity between methods based on genes identified</p>
<p><span class="small"><span class="citation" data-cites="soneson2018bias">Soneson &amp; Robinson (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span></span></p>
<aside class="notes">

<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="assessing-results" class="slide level2">
<h2>Assessing results</h2>
<ul>
<li>We don’t know the ground truth</li>
<li>Using known data (positive controls)</li>
<li>Simulated data by modelling</li>
<li>Intersect of multiple methods</li>
<li>Visual inspection</li>
</ul>
</section>
<section id="assessing-results-1" class="slide level2">
<h2>Assessing results</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="assets/violin.png"></p>
<p>Violin plots are good to visualize distribution</p>
<aside class="notes">
<ul>
<li>Violin plots are great to visually confirm DGE results</li>
<li>Gm15645: Significant DEG</li>
<li>Lgals2, Pla2g4a: Not convincing</li>
<li>Sp110: Low fold change?</li>
<li>Ly6g6e: Multiple distributions?</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/unnamed-chunk-5-1.png" width="480"></p>
</div>
</div>
<p>Dot plots give a quick overview of both expression and % of cells expressing a gene</p>
</div>
</div>
</section>
<section id="things-to-think-about" class="slide level2">
<h2>Things to think about</h2>
<ul>
<li>Which data should I use? Raw? Normalized? Log Normalized?
<ul>
<li>Depends on test/method</li>
</ul></li>
<li>Check that DEGs are not just a result of some batch effect</li>
<li>Batch effects can be corrected using covariates</li>
<li>Consider independence between cells
<ul>
<li>Use a method that accounts for non-independence or proceed with caution</li>
</ul></li>
<li>How many cells/samples do I need for reliable DGE?
<ul>
<li>Also depends on quality of library prep</li>
</ul></li>
<li>Distinctness of cell types</li>
<li>Differentiate between noise and signal
<ul>
<li>Any comparison will produce some DEGs</li>
</ul></li>
<li>Keep in mind what you are comparing</li>
<li>DE results are always relative to other cells</li>
</ul>
</section>
<section id="conclusion" class="slide level2">
<h2>Conclusion</h2>
<ul>
<li>Important to understand what you are trying to compare: mean expression vs probability of being expressed</li>
<li>Important to understand the data</li>
<li>Take into account single-cell specific nuances</li>
<li>Models can get complicated</li>
<li>Important to assess and validate the results</li>
<li>Most tests give similar results for top genes but p values might differ.</li>
<li>Fold changes can’t be interpreted the same way as bulk rnaseq</li>
<li>Too many cells can give extremely low p values</li>
</ul>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-finak2015mast" class="csl-entry" role="listitem">
Finak, G., McDavid, A., Yajima, M., Deng, J., Gersuk, V., Shalek, A. K., Slichter, C. K., Miller, H. W., McElrath, M. J., Prlic, M., et al. (2015). MAST: A flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell RNA sequencing data. <em>Genome Biology</em>, <em>16</em>(1), 1–13. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5</a>
</div>
<div id="ref-soneson2018bias" class="csl-entry" role="listitem">
Soneson, C., &amp; Robinson, M. D. (2018). Bias, robustness and scalability in single-cell differential expression analysis. <em>Nature Methods</em>, <em>15</em>(4), 255–261. <a href="https://www.nature.com/articles/nmeth.4612">https://www.nature.com/articles/nmeth.4612</a>
</div>
<div id="ref-tiberi2023distinct" class="csl-entry" role="listitem">
Tiberi, S., Crowell, H. L., Samartsidis, P., Weber, L. M., &amp; Robinson, M. (2023). Distinct: A novel approach to differential distribution analyses. <em>The Annals of Applied Statistics</em>, <em>17</em>(2), 1681–1700. <a href="https://doi.org/10.1214/22-AOAS1689">https://doi.org/10.1214/22-AOAS1689</a>
</div>
<div id="ref-wu2018two" class="csl-entry" role="listitem">
Wu, Z., Zhang, Y., Stitzel, M. L., &amp; Wu, H. (2018). Two-phase differential expression analysis for single cell RNA-seq. <em>Bioinformatics</em>, <em>34</em>(19), 3340–3348. <a href="https://academic.oup.com/bioinformatics/article/34/19/3340/4984507">https://academic.oup.com/bioinformatics/article/34/19/3340/4984507</a>
</div>
</div>
</section>
<section id="acknowledgements" class="slide level2">
<h2>Acknowledgements</h2>
<p>Slides adapted from previous presentations by Olga Dethlefson, Åsa Björklund, Vincent van Hoef and Roy Francis.</p>

<div class="footer footer-default">

</div>
<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-scrnaseq/home_schedule.html"><img data-src="../../assets/logos/nbis.png" style="height:25px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<a href="https://nbisweden.github.io/workshop-scrnaseq/home_contents.html"><img data-src="../../assets/logos/scilifelab.png" style="height:25px;max-width:none;"></a>
</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'fast',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>