---
title: "Seurat analysis"
output:
  md_document:
    variant: markdown_github
---

# Seurat analysis
Author: Åsa Björklund


Analysis of data using Seurat package, following tutorial at:
http://satijalab.org/seurat/pbmc3k_tutorial.html

Seurat has specific functions for loading and working with drop-seq data. To follow that tutorial, please use the provided dataset for PBMCs that comes with the tutorial. 

It has been downloaded in the course uppmax folder with subfolder:

`scrnaseq_course/data/PBMC_10x/pbmc3k_filtered_gene_bc_matrices.tar.gz`


But you can also run with your own data or with our example data following the example below. But keep in mind that some functions assume that the count data is UMIs, but we run it with RPKMs. This is an example with human innate lympoid cells (ILCs) from Bjorklund et al. 2016.

All data you need is available in the course uppmax folder with subfolder:

`scrnaseq_course/data/ILC/`


### Load packages
```{r load, warning=FALSE}
suppressMessages(require(Seurat))
suppressMessages(require(gridExtra))

```

### Load expression values and metadata

Also fetch ensembl annotations with gene symbols (created in the [biomart lab](biomart)) and include those in the gene names. 

```{r}
R <- read.table("data/ILC/ensembl_rpkmvalues_ILC.csv",sep=",",header=T,row.names=1)
M <- read.table("data/ILC/Metadata_ILC.csv",sep=",",header=T,row.names=1)

# in this case it may be wise to translate ensembl IDs to gene names 
# to make plots with genes more understandable
# the file was created in the biomart exercise.
TR <- read.table("data/ILC/gene_name_translation_biotype.tab",sep="\t")

# find the correct entries in TR and merge ensembl name and gene id.
m <- match(rownames(R),TR$ensembl_gene_id)
newnames <- apply(cbind(as.vector(TR$external_gene_name)[m],rownames(R)),1,paste,collapse=":")
rownames(R)<-newnames

```

### Create seurat object

Will automatically filter out genes/cells that do not meet the criteria specified to save space.

```{r create, warning=FALSE}
# in seurat we will not make use of the spike-ins, so remove them from the expression matrix before creating the Seurat object. 
ercc <- grep("ERCC_",rownames(R))

# when you create a Seurat object, you can specify filtering of 
# genes and cells with the min.cells and min.genes parameters
data <- CreateSeuratObject(raw.data = R[-ercc,], 
                           min.cells = 3, min.genes = 200, 
                           project = "ILC", is.expr=1, meta.data=M)

# plot number of genes and nUMI (rpkms in this case) for each Donor
VlnPlot(object = data, features.plot = c("nGene", "nUMI"), nCol = 2)
# same for celltype
VlnPlot(object = data, features.plot = c("nGene", "nUMI"), nCol = 2, group.by="Celltype")

# scatterplot with detected genes vs RPKMs
GenePlot(object = data, gene1 = "nUMI", gene2 = "nGene")

# the slot data@ident defines the classes of cells, which is automatically set to plate id, To instead plot by celltype, data@ident needs to be changed.
data <- SetAllIdent(object = data, id = "Celltype")
GenePlot(object = data, gene1 = "nUMI", gene2 = "nGene")

# change ident back to Donor
data <- SetAllIdent(object = data, id = "Donor")
```

OBS! Each time you want to change colors in a gene plot, you need to change the identity class value in the seurat object in the slot data@ident. Perhaps there is a better way, but I did not find a solution.

In many of the other Seurat plotting functions like TNSEPlot and PCAPlot you can use “group.by” to define which meta data variable the cells should be coloured by.


### Mitochondrial content and rRNAs

Other common QC-measures are proportion mitochondrial and rRNA mapping.  First calculate them and add to the metadata. 

```{r}
ensnames <- unlist(lapply(strsplit(rownames(data@data),":"), function(x) x[2]))

# calculate mito proportion
mito.genes <- which(ensnames %in% TR$ensembl_gene_id[TR$chromosome_name == "MT"])
p.mito <- colSums(data@raw.data[mito.genes, ])/colSums(data@raw.data)
data <- AddMetaData(object = data, metadata = p.mito, col.name = "p.mito")



# and rRNA proportion
rRNA.genes <- which(ensnames %in% TR$ensembl_gene_id[TR$gene_biotype %in% c("rRNA","rRNA_pseudogene")])
p.rRNA <- colSums(data@raw.data[rRNA.genes, ])/colSums(data@raw.data)
data <- AddMetaData(object = data, metadata = p.rRNA, col.name = "p.rRNA")

# plot as violins
VlnPlot(object = data, features.plot = c("nGene", "nUMI","p.mito","p.rRNA"), nCol = 2)

# or as scatter plot
par(mfrow = c(1, 2))
GenePlot(data, gene1 = "nGene", gene2 = "p.mito")
GenePlot(data, gene1 = "nGene", gene2 = "p.rRNA")

```

In this case the data has already been filtered for low quality cells, but you may want to remove cells that have high proportion of mitochondrial or rRNA reads from your analysis using the `FilterCells` function. 


## Data normalization and scaling

Next step is to normalize the data, detect variable genes and to scale it. 

The scaling step is done to remove unwanted confounders. It is quite common to regress out the number of detected genes (`nGene`), that quite often will drive the variation in your data due to library quality. 
We will also run one version of scaling where we include the `Donor` batch information and compare. 

```{r}
# set scale factor according to mean library size
# Seurat multiplies the  by a scale factor (10,000 by default)
scale.factor <- mean(colSums(R))
data <- NormalizeData(object = data, normalization.method = "LogNormalize", 
    scale.factor = scale.factor)

# look at the plot for suitable cutoffs for your dataset rerun
# you can define the lower/upper bound of mean expression with x.low.cutoff/x.high.cutoff
# and the limit of dispersion with y.cutoff.
data <- FindVariableGenes(object = data, mean.function = ExpMean, 
                          dispersion.function = LogVMR, 
                          x.low.cutoff = 0.5, x.high.cutoff = 10, y.cutoff = 0.5)
length(x = data@var.genes)


# regress out number of detected genes.
data <- ScaleData(object = data, vars.to.regress = c("nGene"), display.progress=F)

# also with batch info + detected genes.
dataB <- ScaleData(object = data, vars.to.regress = c("nGene","Donor"), display.progress=F)

```


## PCA

Run PCA based on the variable genes. The pca is stored in the slot `dr`

```{r warning=F}

data <- RunPCA(object = data, pc.genes = data@var.genes, do.print = TRUE, pcs.print = 1:5,  genes.print = 5)
dataB <- RunPCA(object = dataB, pc.genes = data@var.genes, do.print = TRUE, pcs.print = 1:5,  genes.print = 5)

# plot gene loadings
VizPCA(object = data, pcs.use = 1:4)

# Plot pca for both normalizatinos, 
# since we want to compare them it is a good idea to put multiple 
# plots in the same window, do.return=T was used to return the ggplot2 object
# that is then plotted in one window with grid.arrange
p1 <- PCAPlot(object = data, dim.1 = 1, dim.2 = 2, do.return=T)
p2 <- PCAPlot(object = dataB, dim.1 = 1, dim.2 = 2, do.return=T)

# and with both color by Celltype, here you can use group.by 
p3 <- PCAPlot(object = data, dim.1 = 1, dim.2 = 2, do.return=T,group.by="Celltype")
p4 <- PCAPlot(object = dataB, dim.1 = 1, dim.2 = 2, do.return=T,group.by="Celltype")

# plot together
grid.arrange(p1,p2,p3,p4,ncol=2)

# heatmap with top loading genes
# OBS! margins too large to display well in R-studio, save to pdf instead.
pdf("data/ILC/seurat_pc_loadings_heatmaps.pdf")
PCHeatmap(object = data, pc.use = 1, do.balanced = TRUE, label.columns = FALSE)
PCHeatmap(object = data, pc.use = 1:5, do.balanced = TRUE, label.columns = FALSE)
dev.off()
```

As you can see, the batch effect is not as strong in this PCA as it was in PCAs that we did in other labs, so the PCA plot with batch correction does look quite similar. 

This is mainly due to the fact that we are only using top variable genes, and it seems that the batch effect is mainly seen among genes that are not highly variable. 

Still, if you know you have a clear batch effect, it may be a good idea to remove it with regression. So from now on, we will continue with the `dataB` object. 

## Determine statistically significant principal components

Now we use the `JackStraw` function to check which of the principal components that are significant. If dataset is large, you can instead use `PCElbowPlot()`. 

As a default, JackStraw is only run on the first 20 PCs, if you want to include more PCs in your tSNE and clustering, run JackStraw with `num.pc=30` or similar. 

```{r}
dataB <- JackStraw(object = dataB, num.replicate = 100, display.progress = FALSE)
JackStrawPlot(object = dataB, PCs = 1:12)
```

In this case, only PCs 1,2,3,4,6,7 & 12 are significant, so we will only use those in subsequent steps. 

## Find clusters

In this case, we use the PCs as suggested by the JackStrawPlot. FindClusters constructs a KNN-graph based on distances in PCA space using the defined principal components. This graph is split into clusters using modularity optimization techniques. 

You can tweak the clustring with the `resolution` parameter to get more/less clusters and also with parameters `k` and `k.scale` for the construction of the graph. 

*OBS!* Any function that depends on random start positions, like the KNN graph and tSNE will not give identical results each time you run it. So it is adviced to set the random seed with `set.seed` function before running the function. 

```{r}
use.pcs <- c(1,2,3,4,6,7,12)
set.seed(1)
dataB <- FindClusters(object = dataB, reduction.type = "pca", dims.use = use.pcs, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)

PrintFindClustersParams(object = dataB)
```

## tSNE

For visualization, we use a tSNE with the same PCs as in the clustering. 

```{r}
set.seed(1)
dataB <- RunTSNE(object = dataB, dims.use = use.pcs, do.fast = TRUE)

# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = dataB,do.label = T)

# compare to celltype identities, colour instead by Celltype with group.by
TSNEPlot(object = dataB, group.by = "Celltype")

# colour instead by Donor 
TSNEPlot(object = dataB, group.by = "Donor")

```

Color now automatically changes to the cluster identities, since the slot `ident` in the seurat object is automatically set to the cluster ids after clusering. 

### Cluster markers

Now we can find and plot some of the cluster markers to check if our clustering makes sense. The default method in Seurat is a Wilcoxon rank sum test. All Differential expression tests that are implemented in Seurat are:

* "wilcox" : Wilcoxon rank sum test (default)
* "bimod" : Likelihood-ratio test for single cell gene expression, (McDavid et al., Bioinformatics, 2013)
* "roc" : Standard AUC classifier
* "t" : Student's t-test
* "tobit" : Tobit-test for differential gene expression (Trapnell et al., Nature Biotech, 2014)
* "poisson" : Likelihood ratio test assuming an underlying poisson distribution. Use only for UMI-based datasets
* "negbinom" : Likelihood ratio test assuming an underlying  negative binomial distribution. Use only for UMI-based  datasets
* "MAST" : GLM-framework that treates cellular detection rate as a covariate (Finak et al, Genome Biology, 2015)
* "DESeq2" : DE based on a model using the negative binomial distribution (Love et al, Genome Biology, 2014)



```{r}
# find all genes that defines cluster1, we comare cluster1 to all the other clusters. 
cluster1.markers <- FindMarkers(object = dataB, ident.1 = 1, min.pct = 0.25)
print(head(cluster1.markers),5)
```

As you can see, some of the genes are higher in cluster1, and have a positive fold change (avg_logFC) while others are lower in cluster1 such as CD2. 

```{r}
# plot top cluster1 markers as violins
VlnPlot(object = dataB, features.plot = rownames(cluster1.markers)[1:6],nCol=3,size.title.use=10)

# or plot them onto tSNE
FeaturePlot(object = dataB, features.plot = rownames(cluster1.markers)[1:6], cols.use = c("yellow", "red","black"), reduction.use = "tsne")

# or as a ridge plot
RidgePlot(object = dataB, features.plot = rownames(cluster1.markers)[1:6])
```


You can also specify specific cells, or clusters that you want to compare:

```{r}
# find genes that separates cluster 0 & cluster 3 by specifying both ident.1 and ident.2 
cluster03.markers <- FindMarkers(object = dataB, ident.1 = 0, ident.2 = 3, min.pct = 0.25)
print(x = head(x = cluster03.markers, n = 5))
```

You can also run function `FindAllMarkers` wich will run DE detection for all the clusters vs the rest (or any other classification you may have in the `ident` slot).

# Compare clustering settings

Using the clustree package (https://cran.r-project.org/web/packages/clustree/index.html) you can visualize the clustering from Seurat with different resolutions that can help you understand how the clustering changes whith different settings. 

```{r clustree, warning=FALSE}
suppressMessages(require(clustree))

# run clustering with different resolutions, test values from 0.1-1, we used resolution 0.6 before
res <- c(0.1,0.2,0.3,0.4,0.5,0.7,0.8,0.9,1)

dataB <- FindClusters(dataB, resolution = res, print.output = 0)

```

In this case it seems that the clustering is quite consistent regardless of the resolution settings. Only when going from resolution 0.4 to 0.5 there is a split of cluster 0. 

##### Session info

```{r}
sessionInfo()
```

