<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Assignment of cell identities based on gene expression patterns using reference data.">

<title> Celltype prediction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-bacterium" aria-label="bacterium"></i> Celltype prediction</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat Toolkit</p>
                  <div>
        <div class="description">
          Assignment of cell identities based on gene expression patterns using reference data.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">06-Feb-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-ct_read" id="toc-meta-ct_read" class="nav-link active" data-scroll-target="#meta-ct_read"><span class="header-section-number">1</span> Read data</a></li>
  <li><a href="#meta-ct_ref" id="toc-meta-ct_ref" class="nav-link" data-scroll-target="#meta-ct_ref"><span class="header-section-number">2</span> Reference data</a></li>
  <li><a href="#label-transfer" id="toc-label-transfer" class="nav-link" data-scroll-target="#label-transfer"><span class="header-section-number">3</span> Label transfer</a></li>
  <li><a href="#meta-ct_scpred" id="toc-meta-ct_scpred" class="nav-link" data-scroll-target="#meta-ct_scpred"><span class="header-section-number">4</span> scPred</a></li>
  <li><a href="#meta-ct_azimuth" id="toc-meta-ct_azimuth" class="nav-link" data-scroll-target="#meta-ct_azimuth"><span class="header-section-number">5</span> Azimuth</a></li>
  <li><a href="#meta-ct_compare" id="toc-meta-ct_compare" class="nav-link" data-scroll-target="#meta-ct_compare"><span class="header-section-number">6</span> Compare results</a></li>
  <li><a href="#meta-ct_gsea" id="toc-meta-ct_gsea" class="nav-link" data-scroll-target="#meta-ct_gsea"><span class="header-section-number">7</span> GSEA with celltype markers</a>
  <ul>
  <li><a href="#meta-ct_gsea_deg" id="toc-meta-ct_gsea_deg" class="nav-link" data-scroll-target="#meta-ct_gsea_deg"><span class="header-section-number">7.1</span> DEG overlap</a></li>
  <li><a href="#meta-ct_gsea_annot" id="toc-meta-ct_gsea_annot" class="nav-link" data-scroll-target="#meta-ct_gsea_annot"><span class="header-section-number">7.2</span> With annotated gene sets</a></li>
  </ul></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">8</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<p>Celltype prediction can either be performed on indiviudal cells where each cell gets a predicted celltype label, or on the level of clusters. All methods are based on similarity to other datasets, single cell or sorted bulk RNAseq, or uses known marker genes for each cell type.<br>
We will select one sample from the Covid data, <code>ctrl_13</code> and predict celltype by cell on that sample.<br>
Some methods will predict a celltype to each cell based on what it is most similar to, even if that celltype is not included in the reference. Other methods include an uncertainty so that cells with low similarity scores will be unclassified.<br>
There are multiple different methods to predict celltypes, here we will just cover a few of those.</p>
<p>We will use a reference PBMC dataset from the <code>scPred</code> package which is provided as a Seurat object with counts. And we will test classification based on the <code>scPred</code> and <code>scMap</code> methods. Finally we will use gene set enrichment predict celltype based on the DEGs of each cluster.</p>
<section id="meta-ct_read" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-ct_read"><span class="header-section-number">1</span> Read data</h2>
<p>First, lets load required libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pheatmap)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># remotes::install_github("powellgenomicslab/scPred")</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scPred)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s read in the saved Covid-19 data object from the clustering step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download pre-computed data if missing or long compute</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>fetch_data <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># url for source and intermediate data</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>path_data <span class="ot">&lt;-</span> <span class="st">"https://export.uppmax.uu.se/naiss2023-23-3/workshops/workshop-scrnaseq"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>path_file <span class="ot">&lt;-</span> <span class="st">"data/covid/results/seurat_covid_qc_dr_int_cl.rds"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">dirname</span>(path_file))) <span class="fu">dir.create</span>(<span class="fu">dirname</span>(path_file), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (fetch_data <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">file.exists</span>(path_file)) <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">file.path</span>(path_data, <span class="st">"covid/results/seurat_covid_qc_dr_int_cl.rds"</span>), <span class="at">destfile =</span> path_file)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>alldata <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(path_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subset one patient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> alldata[, alldata<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"ctrl_13"</span>]</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set active assay to RNA and remove the CCA assay</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">@</span>active.assay <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>ctrl[[<span class="st">"CCA"</span>]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>ctrl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
18863 features across 1125 samples within 1 assay 
Active assay: RNA (18863 features, 2000 variable features)
 8 dimensional reductions calculated: umap, tsne, umap_raw, pca_harmony, harmony, umap_harmony, scanorama, umap_scanorama</code></pre>
</div>
</div>
</section>
<section id="meta-ct_ref" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-ct_ref"><span class="header-section-number">2</span> Reference data</h2>
<p>Load the reference dataset with annotated labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> scPred<span class="sc">::</span>pbmc_1</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>reference</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
32838 features across 3500 samples within 1 assay 
Active assay: RNA (32838 features, 0 variable features)</code></pre>
</div>
</div>
<p>Rerun analysis pipeline. Run normalization, feature selection and dimensionality reduction</p>
<p>Here, we will run all the steps that we did in previous labs in one go using the <code>magittr</code> package with the pipe-operator <code>%&gt;%</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> reference <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunUMAP</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(reference, <span class="at">group.by =</span> <span class="st">"cell_type"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-6-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Run all steps of the analysis for the <strong>ctrl</strong> sample as well. Use the clustering from the integration lab with resolution 0.5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the identity as louvain with resolution 0.3</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(ctrl, <span class="at">value =</span> <span class="st">"CCA_snn_res.0.5"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> ctrl <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunPCA</span>(<span class="at">verbose =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunUMAP</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-8-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="label-transfer" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="label-transfer"><span class="header-section-number">3</span> Label transfer</h2>
<p>First we will run label transfer using a similar method as in the integration exercise. But, instead of CCA, which is the default for the <code>FindTransferAnchors()</code> function, we will use <code>pcaproject</code>, ie; the query dataset is projected onto the PCA of the reference dataset. Then, the labels of the reference data are predicted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>transfer.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">reference =</span> reference, <span class="at">query =</span> ctrl,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">anchorset =</span> transfer.anchors, <span class="at">refdata =</span> reference<span class="sc">$</span>cell_type,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(<span class="at">object =</span> ctrl, <span class="at">metadata =</span> predictions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-10-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Now plot how many cells of each celltypes can be found in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x =</span> CCA_snn_res.<span class="fl">0.5</span>, <span class="at">fill =</span> predicted.id)) <span class="sc">+</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-11-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_scpred" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-ct_scpred"><span class="header-section-number">4</span> scPred</h2>
<p>scPred will train a classifier based on all principal components. First, <code>getFeatureSpace()</code> will create a scPred object stored in the <code>@misc</code> slot where it extracts the PCs that best separates the different celltypes. Then <code>trainModel()</code> will do the actual training for each celltype.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">getFeatureSpace</span>(reference, <span class="st">"cell_type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Extracting feature space for each cell type...
DONE!</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">trainModel</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Training models for each cell type...
maximum number of iterations reached 0.0001152056 -0.0001143117DONE!</code></pre>
</div>
</div>
<p>We can then print how well the training worked for the different celltypes by printing the number of PCs used, the ROC value and Sensitivity/Specificity. Which celltypes do you think are harder to classify based on this dataset?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_scpred</span>(reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'scPred' object
✔  Prediction variable = cell_type 
✔  Discriminant features per cell type
✔  Training model(s)
Summary

|Cell type   |    n| Features|Method    |   ROC|  Sens|  Spec|
|:-----------|----:|--------:|:---------|-----:|-----:|-----:|
|B cell      |  280|       50|svmRadial | 1.000| 0.964| 1.000|
|CD4 T cell  | 1620|       50|svmRadial | 0.997| 0.972| 0.975|
|CD8 T cell  |  945|       50|svmRadial | 0.985| 0.899| 0.978|
|cDC         |   26|       50|svmRadial | 0.995| 0.547| 1.000|
|cMono       |  212|       50|svmRadial | 0.994| 0.958| 0.970|
|ncMono      |   79|       50|svmRadial | 0.998| 0.570| 1.000|
|NK cell     |  312|       50|svmRadial | 0.999| 0.933| 0.996|
|pDC         |   20|       50|svmRadial | 1.000| 0.700| 1.000|
|Plasma cell |    6|       50|svmRadial | 1.000| 0.800| 1.000|</code></pre>
</div>
</div>
<p>You can optimize parameters for each dataset by chaining parameters and testing different types of models, see more at: <a href="https://powellgenomicslab.github.io/scPred/articles/introduction.html" class="uri">https://powellgenomicslab.github.io/scPred/articles/introduction.html</a>. But for now, we will continue with this model. Now, let’s predict celltypes on our data, where scPred will align the two datasets with Harmony and then perform classification.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">scPredict</span>(ctrl, reference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>●  Matching reference with new dataset...
     ─ 2000 features present in reference loadings
     ─ 1786 features shared between reference and new dataset
     ─ 89.3% of features in the reference are present in new dataset
●  Aligning new data to reference...
●  Classifying cells...
DONE!</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-15-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<p>Now plot how many cells of each celltypes can be found in each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ctrl<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x =</span> CCA_snn_res.<span class="fl">0.5</span>, <span class="at">fill =</span> scpred_prediction)) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-16-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_azimuth" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-ct_azimuth"><span class="header-section-number">5</span> Azimuth</h2>
<p>There are multiple online resources with large curated datasets with methods to integrate and do label transfer. One such resource is <a href="https://azimuth.hubmapconsortium.org/">Azimuth</a> another one is <a href="https://www.immunesinglecell.org/">Disco</a>.</p>
<p>Azimuth is also possible to install and run locally, but in this case we have used the online app and ran the predictions for you. So you just have to download the prediction file.</p>
<p>This is how to save the count matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># No need to run now.</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>C <span class="ot">=</span> ctrl<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(C, <span class="at">file =</span> <span class="st">"data/covid/results/ctrl13_count_matrix.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead load the results and visualize:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>path_data <span class="ot">&lt;-</span> <span class="st">"https://export.uppmax.uu.se/naiss2023-23-3/workshops/workshop-scrnaseq"</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>path_file <span class="ot">&lt;-</span> <span class="st">"data/covid/results/azimuth_pred.tsv"</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">dirname</span>(path_file))) <span class="fu">dir.create</span>(<span class="fu">dirname</span>(path_file), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (fetch_data <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">file.exists</span>(path_file)) <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">file.path</span>(path_data, <span class="st">"covid/results/azimuth_pred.tsv"</span>), <span class="at">destfile =</span> path_file)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>azimuth_pred <span class="ot">&lt;-</span> <span class="fu">read.table</span>(path_file, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> T)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co"># add predictions to the seurat object</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">$</span>azimuth <span class="ot">=</span> azimuth_pred<span class="sc">$</span>predicted.celltype.l2</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ctrl, <span class="at">group.by =</span> <span class="st">"azimuth"</span>, <span class="at">label =</span> T, <span class="at">repel =</span> T) <span class="sc">+</span> <span class="fu">NoAxes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-18-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_compare" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-ct_compare"><span class="header-section-number">6</span> Compare results</h2>
<p>Now we will compare the output of the two methods using the convenient function in scPred <code>crossTab()</code> that prints the overlap between two metadata slots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">crossTab</span>(ctrl, <span class="st">"predicted.id"</span>, <span class="st">"scpred_prediction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["B cell"],"name":[1],"type":["int"],"align":["right"]},{"label":["CD4 T cell"],"name":[2],"type":["int"],"align":["right"]},{"label":["CD8 T cell"],"name":[3],"type":["int"],"align":["right"]},{"label":["cDC"],"name":[4],"type":["int"],"align":["right"]},{"label":["cMono"],"name":[5],"type":["int"],"align":["right"]},{"label":["ncMono"],"name":[6],"type":["int"],"align":["right"]},{"label":["NK cell"],"name":[7],"type":["int"],"align":["right"]}],"data":[{"1":"101","2":"0","3":"0","4":"0","5":"1","6":"0","7":"0","_rn_":"B cell"},{"1":"0","2":"185","3":"0","4":"0","5":"0","6":"0","7":"0","_rn_":"CD4 T cell"},{"1":"0","2":"10","3":"239","4":"0","5":"0","6":"0","7":"29","_rn_":"CD8 T cell"},{"1":"0","2":"0","3":"0","4":"1","5":"12","6":"0","7":"0","_rn_":"cDC"},{"1":"0","2":"3","3":"4","4":"0","5":"210","6":"6","7":"0","_rn_":"cMono"},{"1":"0","2":"0","3":"0","4":"0","5":"5","6":"117","7":"0","_rn_":"ncMono"},{"1":"0","2":"0","3":"1","4":"0","5":"0","6":"0","7":"156","_rn_":"NK cell"},{"1":"0","2":"5","3":"5","4":"1","5":"29","6":"1","7":"4","_rn_":"unassigned"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>We can also plot all the different predictions side by side</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"azimuth"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-20-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="meta-ct_gsea"><span class="header-section-number">7</span> GSEA with celltype markers</h2>
<p>Another option, where celltype can be classified on cluster level is to use gene set enrichment among the DEGs with known markers for different celltypes. Similar to how we did functional enrichment for the DEGs in the differential expression exercise. There are some resources for celltype gene sets that can be used. Such as <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/">CellMarker</a>, <a href="https://panglaodb.se/">PanglaoDB</a> or celltype gene sets at <a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>. We can also look at overlap between DEGs in a reference dataset and the dataset you are analyzing.</p>
<section id="meta-ct_gsea_deg" class="level3" data-number="7.1">
<h3 data-number="7.1" class="anchored" data-anchor-id="meta-ct_gsea_deg"><span class="header-section-number">7.1</span> DEG overlap</h3>
<p>First, lets extract top DEGs for our Covid-19 dataset and the reference dataset. When we run differential expression for our dataset, we want to report as many genes as possible, hence we set the cutoffs quite lenient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run differential expression in our dataset, using clustering at resolution 0.5</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>alldata <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(alldata, <span class="at">value =</span> <span class="st">"CCA_snn_res.0.5"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>DGE_table <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    alldata,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">logfc.threshold =</span> <span class="dv">0</span>,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">test.use =</span> <span class="st">"wilcox"</span>,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.pct =</span> <span class="fl">0.1</span>,</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.diff.pct =</span> <span class="dv">0</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.cells.per.ident =</span> <span class="dv">20</span>,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">return.thresh =</span> <span class="dv">1</span>,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"RNA"</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="co"># split into a list</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>DGE_list <span class="ot">&lt;-</span> <span class="fu">split</span>(DGE_table, DGE_table<span class="sc">$</span>cluster)</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(<span class="fu">lapply</span>(DGE_list, nrow))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   0    1    2    3    4    5    6    7    8    9 
3355 4127 3221 2497 2082 2065 2427 2328 3459 3613 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute differential gene expression in reference dataset (that has cell annotation)</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(reference, <span class="at">value =</span> <span class="st">"cell_type"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    reference,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.pct =</span> .<span class="dv">1</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.diff.pct =</span> .<span class="dv">2</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">only.pos =</span> T,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">max.cells.per.ident =</span> <span class="dv">20</span>,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">return.thresh =</span> <span class="dv">1</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the top cell marker genes in reference dataset</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="co"># select top 50 with hihgest foldchange among top 100 signifcant genes.</span></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="ot">&lt;-</span> reference_markers[<span class="fu">order</span>(reference_markers<span class="sc">$</span>avg_log2FC, <span class="at">decreasing =</span> T), ]</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>reference_markers <span class="sc">%&gt;%</span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="sc">-</span><span class="dv">100</span>, p_val) <span class="sc">%&gt;%</span></span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="dv">50</span>, avg_log2FC) <span class="ot">-&gt;</span> top50_cell_selection</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform the markers into a list</span></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>ref_list <span class="ot">&lt;-</span> <span class="fu">split</span>(top50_cell_selection<span class="sc">$</span>gene, top50_cell_selection<span class="sc">$</span>cluster)</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(<span class="fu">lapply</span>(ref_list, length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CD8 T cell  CD4 T cell       cMono      B cell     NK cell         pDC 
         30          15          50          50          50          50 
     ncMono         cDC Plasma cell 
         50          50          50 </code></pre>
</div>
</div>
<p>Now we can run GSEA for the DEGs from our dataset and check for enrichment of top DEGs in the reference dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(fgsea))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x) {</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>avg_log2FC, x<span class="sc">$</span>gene)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(<span class="at">pathways =</span> ref_list, <span class="at">stats =</span> gene_rank, <span class="at">nperm =</span> <span class="dv">10000</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(fgseaRes)</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>    x[x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.1</span>, ]</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>    x[x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">2</span>, ]</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>    x[<span class="fu">order</span>(x<span class="sc">$</span>NES, <span class="at">decreasing =</span> T), ]</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
   pathway        pval        padj        ES      NES nMoreExtreme size
1:   cMono 0.000099990 0.000299970 0.9598121 2.069210            0   48
2:     cDC 0.000099990 0.000299970 0.8424705 1.803285            0   41
3:  ncMono 0.000099990 0.000299970 0.8382748 1.798336            0   43
4:     pDC 0.002205735 0.004962904 0.7461808 1.533175           21   21
5: NK cell 0.028176336 0.050717405 0.7528355 1.433674          270   10
6:  B cell 0.064405063 0.096607595 0.6648106 1.324808          635   15
                                    leadingEdge
1:      S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
2:            LYZ,GRN,TYMP,CST3,AIF1,LGALS2,...
3:     CTSS,TYMP,CST3,S100A11,AIF1,SERPINA1,...
4:         GRN,MS4A6A,CST3,MPEG1,CTSB,TGFBI,...
5:      TYROBP,FCER1G,SRGN,CCL3,MYO1F,ITGB2,...
6: NCF1,LY86,MARCH1,POU2F2,HLA-DMB,HLA-DRB5,...

$`1`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:     NK cell 0.0001000100 0.0004012439 0.9454377 2.374190            0   48
2:  CD8 T cell 0.0001003110 0.0004012439 0.9277198 2.216579            0   25
3:      ncMono 0.0005728033 0.0015274755 0.9085109 1.756729            4    6
4:         pDC 0.0054134381 0.0086615009 0.7811335 1.662083           50   10
5: Plasma cell 0.0050035025 0.0086615009 0.6337936 1.537820           49   30
                                     leadingEdge
1:          GNLY,GZMB,FGFBP2,PRF1,NKG7,SPON2,...
2:           GNLY,GZMB,FGFBP2,PRF1,NKG7,CST7,...
3:                            FCGR3A,IFITM2,RHOC
4: GZMB,C12orf75,HSP90B1,ALOX5AP,PLAC8,RRBP1,...
5:      FKBP11,PRDM1,SDF2L1,CD38,SLAMF7,PPIB,...

$`2`
      pathway         pval         padj        ES      NES nMoreExtreme size
1: CD8 T cell 0.0001001201 0.0003504205 0.9272147 2.125533            0   29
2:    NK cell 0.0001000700 0.0003504205 0.8194827 1.886789            0   31
3: CD4 T cell 0.0021071310 0.0049166389 0.8715169 1.686304           18    7
                         leadingEdge
1: CD3D,CD8A,CD3G,CD8B,CCL5,GZMH,...
2: CCL5,GZMA,CCL4,GZMM,NKG7,CST7,...
3:       CD3D,CD3G,CD3E,IL7R,PIK3IP1

$`3`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:      B cell 0.0000999900 0.0004062976 0.9063218 1.996177            0   46
2:         cDC 0.0001015744 0.0004062976 0.8903699 1.798339            0   14
3:         pDC 0.0004028197 0.0010741860 0.8030843 1.662144            3   18
4: Plasma cell 0.0759263226 0.1214821161 0.7286200 1.370511          708    8
                                            leadingEdge
1:      CD79A,LINC00926,TCL1A,MS4A1,CD79B,TNFRSF13C,...
2: CD74,HLA-DQB1,HLA-DRA,HLA-DPB1,HLA-DRB1,HLA-DQA1,...
3:            CD74,BCL11A,TCF4,IRF8,HERPUD1,TSPAN13,...
4:                 PLPP5,ISG20,HERPUD1,MZB1,ITM2C,DERL3

$`4`
      pathway         pval         padj        ES      NES nMoreExtreme size
1: CD4 T cell 0.0001015847 0.0006095083 0.9094929 1.769999            0   14
2: CD8 T cell 0.0004244032 0.0012732095 0.8911313 1.624905            3    8
                        leadingEdge
1: IL7R,LTB,LDHB,MAL,RCAN3,CD3D,...
2:     CD3D,IL32,CD3G,CD2,CD3E,CD8B

$`5`
Empty data.table (0 rows and 8 cols): pathway,pval,padj,ES,NES,nMoreExtreme...

$`6`
       pathway         pval         padj        ES      NES nMoreExtreme size
1:      B cell 0.0000999900 0.0004061738 0.8925237 1.835880            0   45
2:         cDC 0.0001015435 0.0004061738 0.8892494 1.696261            0   14
3:         pDC 0.0005049485 0.0013465293 0.8286587 1.600294            4   16
4: Plasma cell 0.0468718364 0.0535678130 0.7059931 1.355775          462   15
                                            leadingEdge
1:        CD79A,MS4A1,BANK1,HLA-DQA1,CD74,TNFRSF13C,...
2: HLA-DQA1,CD74,HLA-DRA,HLA-DPB1,HLA-DQB1,HLA-DPA1,...
3:             CD74,JCHAIN,SPIB,TCF4,HERPUD1,CCDC50,...
4:            JCHAIN,HERPUD1,ISG20,PEBP1,MZB1,DERL3,...

$`7`
      pathway         pval         padj        ES      NES nMoreExtreme size
1: CD4 T cell 0.0001014816 0.0006088898 0.9375405 1.993694            0   14
2: CD8 T cell 0.0213283876 0.0639851629 0.8316899 1.484307          183    5
                            leadingEdge
1: IL7R,TCF7,PIK3IP1,LTB,TSHZ2,LEF1,...
2:                   CD3G,CD3E,CD3D,CD2

$`8`
   pathway        pval         padj        ES      NES nMoreExtreme size
1:  ncMono 0.000099990 0.0002666933 0.9648025 2.033501            0   49
2:   cMono 0.000100000 0.0002666933 0.8832827 1.832286            0   36
3:     cDC 0.000100010 0.0002666933 0.8272452 1.718416            0   37
4: NK cell 0.007492912 0.0149858242 0.7651983 1.485395           73   14
5:     pDC 0.021491272 0.0343860357 0.7210796 1.408835          212   15
6:  B cell 0.073965986 0.0986213143 0.6644604 1.305989          734   16
                                              leadingEdge
1:                CDKN1C,LST1,FCGR3A,MS4A7,AIF1,COTL1,...
2:               LST1,AIF1,COTL1,SERPINA1,FCER1G,CST3,...
3:                   LST1,AIF1,COTL1,FCER1G,CST3,SPI1,...
4:              FCGR3A,FCER1G,RHOC,TYROBP,IFITM2,CCL3,...
5:                    CST3,NPC2,CTSB,PLD4,MPEG1,TGFBI,...
6: HLA-DPA1,POU2F2,HLA-DRB5,HLA-DRB1,HLA-DRA,HLA-DPB1,...

$`9`
       pathway        pval         padj        ES      NES nMoreExtreme size
1:         pDC 0.000101102 0.0004549591 0.8242356 1.802655            0   16
2: Plasma cell 0.000099990 0.0004549591 0.7455072 1.775404            0   44
3:  CD8 T cell 0.023773547 0.0713206407 0.6449076 1.427016          235   18
4:      B cell 0.057355090 0.1290489528 0.8116906 1.399605          470    4
                                  leadingEdge
1: JCHAIN,MZB1,HSP90B1,TXN,ITM2C,C12orf75,...
2:   JCHAIN,MZB1,HSP90B1,SUB1,MANF,SEC11C,...
3:      GZMA,C12orf75,MT2A,GZMB,IL32,GNLY,...
4:                              JCHAIN,PDLIM1</code></pre>
</div>
</div>
<p>Selecting top significant overlap per cluster, we can now rename the clusters according to the predicted labels. OBS! Be aware that if you have some clusters that have non-significant p-values for all the gene sets, the cluster label will not be very reliable. Also, the gene sets you are using may not cover all the celltypes you have in your dataset and hence predictions may just be the most similar celltype. Also, some of the clusters have very similar p-values to multiple celltypes, for instance the ncMono and cMono celltypes are equally good for some clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>, <span class="dv">1</span>]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">$</span>ref_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>active.ident)]</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(alldata, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"CCA_snn_res.0.5"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(alldata, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-24-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Compare the results with the other celltype prediction methods in the <strong>ctrl_13</strong> sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ctrl<span class="sc">$</span>ref_gsea <span class="ot">&lt;-</span> alldata<span class="sc">$</span>ref_gsea[alldata<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"ctrl_13"</span>]</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"GSEA"</span>),</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"LabelTransfer"</span>),</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(ctrl, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"scpred_prediction"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scPred"</span>),</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="1536"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-ct_gsea_annot" class="level3" data-number="7.2">
<h3 data-number="7.2" class="anchored" data-anchor-id="meta-ct_gsea_annot"><span class="header-section-number">7.2</span> With annotated gene sets</h3>
<p>We have downloaded the celltype gene lists from http://bio-bigdata.hrbmu.edu.cn/CellMarker/CellMarker_download.html and converted the excel file to a csv for you. Read in the gene lists and do some filtering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>path_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"data/cell_marker_human.csv"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(path_file)) <span class="fu">download.file</span>(<span class="fu">file.path</span>(path_data, <span class="st">"cell_marker_human.csv"</span>), <span class="at">destfile =</span> path_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the human marker table</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"data/cell_marker_human.csv"</span>, <span class="at">sep =</span> <span class="st">";"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers[markers<span class="sc">$</span>species <span class="sc">==</span> <span class="st">"Human"</span>, ]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers[markers<span class="sc">$</span>cancer_type <span class="sc">==</span> <span class="st">"Normal"</span>, ]</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter by tissue (to reduce computational time and have tissue-specific classification)</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">unique</span>(markers<span class="sc">$</span>tissue_type))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "Abdomen"                        "Abdominal adipose tissue"      
  [3] "Abdominal fat pad"              "Acinus"                        
  [5] "Adipose tissue"                 "Adrenal gland"                 
  [7] "Adventitia"                     "Airway"                        
  [9] "Airway epithelium"              "Allocortex"                    
 [11] "Alveolus"                       "Amniotic fluid"                
 [13] "Amniotic membrane"              "Ampullary"                     
 [15] "Anogenital tract"               "Antecubital vein"              
 [17] "Anterior cruciate ligament"     "Anterior presomitic mesoderm"  
 [19] "Aorta"                          "Aortic valve"                  
 [21] "Artery"                         "Arthrosis"                     
 [23] "Articular&nbsp;Cartilage"            "Ascites"                       
 [25] "Atrium"                         "Auditory cortex"               
 [27] "Basilar membrane"               "Beige Fat"                     
 [29] "Bile duct"                      "Biliary tract"                 
 [31] "Bladder"                        "Blood"                         
 [33] "Blood vessel"                   "Bone"                          
 [35] "Bone marrow"                    "Brain"                         
 [37] "Breast"                         "Bronchial vessel"              
 [39] "Bronchiole"                     "Bronchoalveolar lavage"        
 [41] "Bronchoalveolar system"         "Bronchus"                      
 [43] "Brown adipose tissue"           "Calvaria"                      
 [45] "Capillary"                      "Cardiac atrium"                
 [47] "Cardiovascular system"          "Carotid artery"                
 [49] "Carotid plaque"                 "Cartilage"                     
 [51] "Caudal cortex"                  "Caudal forebrain"              
 [53] "Caudal ganglionic eminence"     "Cavernosum"                    
 [55] "Central amygdala"               "Central nervous system"        
 [57] "Central Nervous System"         "Cerebellum"                    
 [59] "Cerebral organoid"              "Cerebrospinal fluid"           
 [61] "Choriocapillaris"               "Chorionic villi"               
 [63] "Chorionic villus"               "Choroid"                       
 [65] "Choroid plexus"                 "Colon"                         
 [67] "Colon epithelium"               "Colorectum"                    
 [69] "Cornea"                         "Corneal endothelium"           
 [71] "Corneal epithelium"             "Coronary artery"               
 [73] "Corpus callosum"                "Corpus luteum"                 
 [75] "Cortex"                         "Cortical layer"                
 [77] "Cortical thymus"                "Decidua"                       
 [79] "Deciduous tooth"                "Dental pulp"                   
 [81] "Dermis"                         "Diencephalon"                  
 [83] "Distal airway"                  "Dorsal forebrain"              
 [85] "Dorsal root ganglion"           "Dorsolateral prefrontal cortex"
 [87] "Ductal tissue"                  "Duodenum"                      
 [89] "Ectocervix"                     "Ectoderm"                      
 [91] "Embryo"                         "Embryoid body"                 
 [93] "Embryonic brain"                "Embryonic heart"               
 [95] "Embryonic Kidney"               "Embryonic prefrontal cortex"   
 [97] "Embryonic stem cell"            "Endocardium"                   
 [99] "Endocrine"                      "Endoderm"                      
[101] "Endometrium"                    "Endometrium stroma"            
[103] "Entorhinal cortex"              "Epidermis"                     
[105] "Epithelium"                     "Esophagus"                     
[107] "Eye"                            "Fat pad"                       
[109] "Fetal brain"                    "Fetal gonad"                   
[111] "Fetal heart"                    "Fetal ileums"                  
[113] "Fetal kidney"                   "Fetal Leydig"                  
[115] "Fetal liver"                    "Fetal lung"                    
[117] "Fetal pancreas"                 "Fetal thymus"                  
[119] "Fetal umbilical cord"           "Fetus"                         
[121] "Foreskin"                       "Frontal cortex"                
[123] "Fundic gland"                   "Gall bladder"                  
[125] "Gastric corpus"                 "Gastric epithelium"            
[127] "Gastric gland"                  "Gastrointestinal tract"        
[129] "Germ"                           "Gingiva"                       
[131] "Gonad"                          "Gut"                           
[133] "Hair follicle"                  "Heart"                         
[135] "Heart muscle"                   "Hippocampus"                   
[137] "Ileum"                          "Inferior colliculus"           
[139] "Interfollicular epidermis"      "Intervertebral disc"           
[141] "Intestinal crypt"               "Intestine"                     
[143] "Intrahepatic cholangio"         "Jejunum"                       
[145] "Kidney"                         "Lacrimal gland"                
[147] "Large intestine"                "Laryngeal squamous epithelium" 
[149] "Lateral ganglionic eminence"    "Ligament"                      
[151] "Limb bud"                       "Limbal epithelium"             
[153] "Liver"                          "Lumbar vertebra"               
[155] "Lung"                           "Lymph"                         
[157] "Lymph node"                     "Lymphatic vessel"              
[159] "Lymphoid tissue"                "Malignant pleural effusion"    
[161] "Mammary epithelium"             "Mammary gland"                 
[163] "Medial ganglionic eminence"     "Medullary thymus"              
[165] "Meniscus"                       "Mesoblast"                     
[167] "Mesoderm"                       "Microvascular endothelium"     
[169] "Microvessel"                    "Midbrain"                      
[171] "Middle temporal gyrus"          "Milk"                          
[173] "Molar"                          "Muscle"                        
[175] "Myenteric plexus"               "Myocardium"                    
[177] "Myometrium"                     "Nasal concha"                  
[179] "Nasal epithelium"               "Nasal mucosa"                  
[181] "Nasal polyp"                    "Neocortex"                     
[183] "Nerve"                          "Nose"                          
[185] "Nucleus pulposus"               "Olfactory neuroepithelium"     
[187] "Optic nerve"                    "Oral cavity"                   
[189] "Oral mucosa"                    "Osteoarthritic cartilage"      
[191] "Ovarian cortex"                 "Ovarian follicle"              
[193] "Ovary"                          "Oviduct"                       
[195] "Pancreas"                       "Pancreatic acinar tissue"      
[197] "Pancreatic duct"                "Pancreatic islet"              
[199] "Periodontal ligament"           "Periodontium"                  
[201] "Periosteum"                     "Peripheral blood"              
[203] "Peritoneal fluid"               "Peritoneum"                    
[205] "Pituitary"                      "Placenta"                      
[207] "Plasma"                         "Pluripotent stem cell"         
[209] "Polyp"                          "Posterior presomitic mesoderm" 
[211] "Prefrontal cortex"              "Premolar"                      
[213] "Presomitic mesoderm"            "Primitive streak"              
[215] "Prostate"                       "Pulmonary arteriy"             
[217] "Pyloric gland"                  "Rectum"                        
[219] "Renal glomerulus"               "Respiratory tract"             
[221] "Retina"                         "Retinal organoid"              
[223] "Retinal pigment epithelium"     "Right ventricle"               
[225] "Saliva"                         "Salivary gland"                
[227] "Scalp"                          "Sclerocorneal tissue"          
[229] "Seminal plasma"                 "Septum transversum"            
[231] "Serum"                          "Sinonasal mucosa"              
[233] "Sinus tissue"                   "Skeletal muscle"               
[235] "Skin"                           "Small intestinal crypt"        
[237] "Small intestine"                "Soft tissue"                   
[239] "Sperm"                          "Spinal cord"                   
[241] "Spleen"                         "Splenic red pulp"              
[243] "Sputum"                         "Stomach"                       
[245] "Subcutaneous adipose tissue"    "Submandibular gland"           
[247] "Subpallium"                     "Subplate"                      
[249] "Subventricular zone"            "Superior frontal gyrus"        
[251] "Sympathetic ganglion"           "Synovial fluid"                
[253] "Synovium"                       "Taste bud"                     
[255] "Tendon"                         "Testis"                        
[257] "Thalamus"                       "Thymus"                        
[259] "Thyroid"                        "Tonsil"                        
[261] "Tooth"                          "Trachea"                       
[263] "Tracheal airway epithelium"     "Transformed artery"            
[265] "Trophoblast"                    "Umbilical cord"                
[267] "Umbilical cord blood"           "Umbilical vein"                
[269] "Undefined"                      "Urine"                         
[271] "Urothelium"                     "Uterine cervix"                
[273] "Uterus"                         "Vagina"                        
[275] "Vein"                           "Venous blood"                  
[277] "Ventral thalamus"               "Ventricle"                     
[279] "Ventricular and atrial"         "Ventricular zone"              
[281] "Visceral adipose tissue"        "Vocal fold"                    
[283] "Whartons jelly"                 "White adipose tissue"          
[285] "White matter"                   "Yolk sac"                      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">"blood"</span>, <span class="fu">unique</span>(markers<span class="sc">$</span>tissue_type), <span class="at">value =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Peripheral blood"     "Umbilical cord blood" "Venous blood"        </code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> markers[markers<span class="sc">$</span>tissue_type <span class="sc">%in%</span> <span class="fu">c</span>(</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Blood"</span>, <span class="st">"Venous blood"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Serum"</span>, <span class="st">"Plasma"</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Spleen"</span>, <span class="st">"Bone marrow"</span>, <span class="st">"Lymph node"</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>), ]</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove strange characters etc.</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(markers<span class="sc">$</span>cell_name), <span class="cf">function</span>(x) {</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">paste</span>(markers<span class="sc">$</span>Symbol[markers<span class="sc">$</span>cell_name <span class="sc">==</span> x], <span class="at">sep =</span> <span class="st">","</span>)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"[[]|[]]| |-"</span>, <span class="st">","</span>, x)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(x, <span class="at">split =</span> <span class="st">","</span>))</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">unique</span>(x[<span class="sc">!</span>x <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">"NA"</span>, <span class="st">"family"</span>)])</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">casefold</span>(x, <span class="at">upper =</span> T)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(celltype_list) <span class="ot">&lt;-</span> <span class="fu">unique</span>(markers<span class="sc">$</span>cell_name)</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[<span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list, length)) <span class="sc">&lt;</span> <span class="dv">100</span>]</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>celltype_list <span class="ot">&lt;-</span> celltype_list[<span class="fu">unlist</span>(<span class="fu">lapply</span>(celltype_list, length)) <span class="sc">&gt;</span> <span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run fgsea for each of the clusters in the list</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(DGE_list, <span class="cf">function</span>(x) {</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    gene_rank <span class="ot">&lt;-</span> <span class="fu">setNames</span>(x<span class="sc">$</span>avg_log2FC, x<span class="sc">$</span>gene)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(<span class="at">pathways =</span> celltype_list, <span class="at">stats =</span> gene_rank, <span class="at">nperm =</span> <span class="dv">10000</span>, <span class="at">scoreType =</span> <span class="st">"pos"</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(fgseaRes)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res) <span class="ot">&lt;-</span> <span class="fu">names</span>(DGE_list)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="co"># You can filter and resort the table based on ES, NES or pvalue</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    x[x<span class="sc">$</span>pval <span class="sc">&lt;</span> <span class="fl">0.01</span>, ]</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>    x[x<span class="sc">$</span>size <span class="sc">&gt;</span> <span class="dv">5</span>, ]</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    x[<span class="fu">order</span>(x<span class="sc">$</span>NES, <span class="at">decreasing =</span> T), ]</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a><span class="co"># show top 3 for each cluster.</span></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(res, head, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`0`
      pathway       pval        padj        ES      NES nMoreExtreme size
1: Neutrophil 0.00009999 0.002274773 0.8599522 1.772801            0   22
2:   Monocyte 0.00009999 0.002274773 0.8174774 1.746745            0   40
3: Eosinophil 0.00019998 0.003639636 0.8662784 1.720519            1   13
                                 leadingEdge
1: S100A8,S100A9,CD14,CSF3R,S100A6,PLAUR,...
2:   S100A8,S100A9,LYZ,S100A12,VCAN,FCN1,...
3:    S100A8,S100A9,LYZ,RETN,CSF3R,ICAM1,...

$`1`
                                                            pathway      pval
1:                                              Natural killer cell 9.999e-05
2: CD4+ recently activated effector memory or effector T cell (CTL) 9.999e-05
3:                    Finally highly effector (TEMRA) memory T cell 9.999e-05
         padj        ES      NES nMoreExtreme size
1: 0.00076356 0.8992559 2.222949            0   38
2: 0.00076356 0.9582220 2.110878            0   10
3: 0.00076356 0.9917476 2.105673            0    7
                            leadingEdge
1: GNLY,GZMB,FGFBP2,PRF1,NKG7,SPON2,...
2:   GNLY,PRF1,NKG7,CTSW,GZMH,S1PR5,...
3:    GNLY,NKG7,CST7,GZMA,GZMH,CCL5,...

$`2`
       pathway      pval        padj        ES      NES nMoreExtreme size
1: CD8+ T cell 9.999e-05 0.000579942 0.9460816 2.021043            0   12
2: CD4+ T cell 9.999e-05 0.000579942 0.9469272 2.007422            0   11
3:      T cell 9.999e-05 0.000579942 0.8607694 1.997743            0   34
                          leadingEdge
1: CD3D,CD8A,CD8B,TRGC2,TRAC,CD3E,...
2:  CD3D,CD8A,TRAC,CD3E,IL32,IL7R,...
3:  GZMK,CD3D,CD8A,CD3G,CD8B,GZMH,...

$`3`
             pathway      pval       padj        ES      NES nMoreExtreme size
1:      Naive B cell 9.999e-05 0.00253308 0.9492283 1.916533            0   13
2:            B cell 9.999e-05 0.00253308 0.8919383 1.911438            0   30
3: Follicular B cell 9.999e-05 0.00253308 0.9545637 1.889150            0   10
                             leadingEdge
1:  IGHM,IGHD,TCL1A,MS4A1,FCER2,YBX3,...
2:  IGHM,IGHD,CD79A,IGKC,TCL1A,MS4A1,...
3: IGHD,CD79A,TCL1A,MS4A1,CD79B,CD74,...

$`4`
             pathway      pval        padj        ES      NES nMoreExtreme size
1: Naive T(Th0) cell 9.999e-05 0.001639836 0.9082195 1.746627            0   10
2:       CD8+ T cell 9.999e-05 0.001639836 0.8801025 1.703183            0   11
3:            B cell 9.999e-05 0.001639836 0.8680466 1.685321            0   12
                         leadingEdge
1: IL7R,CD3D,IL32,TCF7,CD3E,NPM1,...
2: IL7R,TRAC,CD3D,IL32,CD28,CD3E,...
3:  IL7R,CD5,CD28,JUNB,CD27,BCL2,...

$`5`
                        pathway       pval       padj        ES      NES
1:                Megakaryocyte 0.00009999 0.00679932 0.9785849 1.826302
2: Red blood cell (erythrocyte) 0.00259974 0.08839116 0.9250879 1.682181
3:                     Platelet 0.00539946 0.09179082 0.7964424 1.515682
   nMoreExtreme size                         leadingEdge
1:            0   12    PPBP,PF4,NRGN,MYL9,GNG11,GP9,...
2:           25    6             HBB,HBA2,HBA1,SPI1,CD36
3:           53   18 PPBP,PF4,GP9,ITGA2B,CD151,ICAM2,...

$`6`
       pathway      pval       padj        ES      NES nMoreExtreme size
1:      B cell 9.999e-05 0.00249975 0.8860207 1.798112            0   35
2: Plasma cell 9.999e-05 0.00249975 0.9281515 1.771493            0   12
3: Plasmablast 9.999e-05 0.00249975 0.9189098 1.722143            0    9
                             leadingEdge
1:  IGKC,CD79A,MS4A1,BANK1,IGHM,CD74,...
2: IGKC,CD79A,IGLC2,IGLC3,IGHM,IGHG1,...
3: IGKC,IGLC2,IGLC3,IGHM,IGHG1,IGHA1,...

$`7`
                      pathway      pval       padj        ES      NES
1: Central memory CD8+ T cell 9.999e-05 0.00085547 0.9134742 1.935042
2:          Naive CD8+ T cell 9.999e-05 0.00085547 0.8709884 1.895299
3:           Naive CD8 T cell 9.999e-05 0.00085547 0.9021189 1.875505
   nMoreExtreme size                         leadingEdge
1:            0   12 CCR7,IL7R,TCF7,TSHZ2,LEF1,RCAN3,...
2:            0   16  CCR7,TCF7,TSHZ2,LEF1,RCAN3,MAL,...
3:            0   10  CCR7,TCF7,TSHZ2,LEF1,RCAN3,MAL,...

$`8`
          pathway       pval       padj        ES      NES nMoreExtreme size
1: CD16+ monocyte 0.00029997 0.00899910 0.9431821 1.746453            2    6
2:       Monocyte 0.00009999 0.00449955 0.8137464 1.668767            0   27
3:     Macrophage 0.00009999 0.00449955 0.8068860 1.644543            0   24
                              leadingEdge
1:           FCGR3A,TCF7L2,HES4,LYN,MTSS1
2: LST1,FCGR3A,MS4A7,CST3,PECAM1,CD68,...
3:  FCGR3A,MS4A7,FCER1G,CD68,FTL,C1QA,...

$`9`
       pathway      pval       padj        ES      NES nMoreExtreme size
1: Plasmablast 9.999e-05 0.00273306 0.9256975 2.022903            0   15
2: Plasma cell 9.999e-05 0.00273306 0.9219160 1.988355            0   13
3:      B cell 9.999e-05 0.00273306 0.8738002 1.938684            0   18
                               leadingEdge
1:  IGHA1,IGLC2,IGKC,IGHG1,IGLC3,IGHA2,...
2: IGHA1,IGLC2,IGKC,JCHAIN,IGHG1,IGLC3,...
3: IGHA1,IGKC,JCHAIN,IGHG1,IGHA2,IGHG2,...</code></pre>
</div>
</div>
<p>Let’s plot the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(res, <span class="cf">function</span>(x) {</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(x)[<span class="dv">1</span>, <span class="dv">1</span>]</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>alldata<span class="sc">$</span>cellmarker_gsea <span class="ot">&lt;-</span> new.cluster.ids[<span class="fu">as.character</span>(alldata<span class="sc">@</span>active.ident)]</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(alldata, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"ref_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DimPlot</span>(alldata, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"cellmarker_gsea"</span>) <span class="sc">+</span> <span class="fu">NoAxes</span>(),</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="seurat_06_celltyping_files/figure-html/unnamed-chunk-29-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="seurat_06_celltyping_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="1056"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you think that the methods overlap well? Where do you see the most inconsistencies?</p>
</div>
</div>
<p>In this case we do not have any ground truth, and we cannot say which method performs best. You should keep in mind, that any celltype classification method is just a prediction, and you still need to use your common sense and knowledge of the biological system to judge if the results make sense.</p>
<p>Finally, lets save the data with predictions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ctrl, <span class="st">"data/covid/results/seurat_covid_qc_dr_int_cl_ct-ctrl13.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="meta-session" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">8</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] fgsea_1.28.0       caret_6.0-94       lattice_0.21-8     scPred_1.9.2      
 [5] pheatmap_1.0.12    ggplot2_3.4.2      patchwork_1.1.2    dplyr_1.1.2       
 [9] SeuratObject_4.1.3 Seurat_4.3.0      

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.14        jsonlite_1.8.5        
  [4] magrittr_2.0.3         ggbeeswarm_0.7.2       spatstat.utils_3.0-3  
  [7] farver_2.1.1           rmarkdown_2.22         vctrs_0.6.2           
 [10] ROCR_1.0-11            spatstat.explore_3.2-1 htmltools_0.5.5       
 [13] pROC_1.18.2            sctransform_0.3.5      parallelly_1.36.0     
 [16] KernSmooth_2.23-20     htmlwidgets_1.6.2      ica_1.0-3             
 [19] plyr_1.8.8             lubridate_1.9.2        plotly_4.10.2         
 [22] zoo_1.8-12             igraph_1.4.3           mime_0.12             
 [25] lifecycle_1.0.3        iterators_1.0.14       pkgconfig_2.0.3       
 [28] Matrix_1.5-4           R6_2.5.1               fastmap_1.1.1         
 [31] fitdistrplus_1.1-11    future_1.32.0          shiny_1.7.4           
 [34] digest_0.6.31          colorspace_2.1-0       tensor_1.5            
 [37] irlba_2.3.5.1          labeling_0.4.2         progressr_0.13.0      
 [40] timechange_0.2.0       fansi_1.0.4            spatstat.sparse_3.0-1 
 [43] httr_1.4.6             polyclip_1.10-4        abind_1.4-5           
 [46] compiler_4.3.0         withr_2.5.0            BiocParallel_1.36.0   
 [49] lava_1.7.2.1           MASS_7.3-58.4          ModelMetrics_1.2.2.2  
 [52] tools_4.3.0            vipor_0.4.5            lmtest_0.9-40         
 [55] beeswarm_0.4.0         httpuv_1.6.11          future.apply_1.11.0   
 [58] nnet_7.3-18            goftest_1.2-3          glue_1.6.2            
 [61] nlme_3.1-162           promises_1.2.0.1       grid_4.3.0            
 [64] Rtsne_0.16             cluster_2.1.4          reshape2_1.4.4        
 [67] generics_0.1.3         recipes_1.0.6          gtable_0.3.3          
 [70] spatstat.data_3.0-1    class_7.3-21           tidyr_1.3.0           
 [73] data.table_1.14.8      sp_1.6-1               utf8_1.2.3            
 [76] spatstat.geom_3.2-1    RcppAnnoy_0.0.20       ggrepel_0.9.3         
 [79] RANN_2.6.1             foreach_1.5.2          pillar_1.9.0          
 [82] stringr_1.5.0          limma_3.58.1           later_1.3.1           
 [85] splines_4.3.0          survival_3.5-5         deldir_1.0-9          
 [88] tidyselect_1.2.0       miniUI_0.1.1.1         pbapply_1.7-0         
 [91] knitr_1.43             gridExtra_2.3          scattermore_1.2       
 [94] RhpcBLASctl_0.23-42    stats4_4.3.0           xfun_0.39             
 [97] statmod_1.5.0          hardhat_1.3.0          timeDate_4022.108     
[100] matrixStats_1.0.0      stringi_1.7.12         lazyeval_0.2.2        
[103] yaml_2.3.7             evaluate_0.21          codetools_0.2-19      
[106] kernlab_0.9-32         tibble_3.2.1           cli_3.6.1             
[109] uwot_0.1.14            rpart_4.1.19           xtable_1.8-4          
[112] reticulate_1.30        munsell_0.5.0          harmony_1.2.0         
[115] Rcpp_1.0.10            globals_0.16.2         spatstat.random_3.1-5 
[118] png_0.1-8              parallel_4.3.0         ellipsis_0.3.2        
[121] gower_1.0.1            listenv_0.9.0          viridisLite_0.4.2     
[124] ipred_0.9-14           prodlim_2023.03.31     scales_1.2.1          
[127] ggridges_0.5.4         crayon_1.5.2           leiden_0.4.3          
[130] purrr_1.0.1            rlang_1.1.1            fastmatch_1.1-3       
[133] cowplot_1.1.1         </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2024 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"descPosition":"bottom","openEffect":"zoom","loop":true,"closeEffect":"zoom","selector":".lightbox"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>