---
title: Gene set analysis
author: Jennifer Fransson
format:
  revealjs:
    df-print: kable
execute:
  echo: false
bibliography: references.bib
date: 2025-04-02
---




## What is gene set analysis?

![](assets/functional.png)

. . .

Gene-level data -> Gene set data

. . .

We focus on transcriptomics and DGE, but in principle applies to any genome-wide data

:::{.notes}
Many different kinds of genome wide analyses such differential gene expression, differential binding (chip-seq), differential methylation etc. could all end up here at functional analysis.
:::

## Why gene set analysis?
<br>
<br>
<br>

### Predict the functional changes of cells based on differential gene expression analysis

- Make sense of a long list of DEGs
  - What is the function of those genes?
  - What is the biological consequence of over/under expression of genes?
- Connect your DEGs and thereby your experiment to pathway activity
- Small differences in many genes may have a bigger impact than large differences in one genes
- Less sensitive to false positive DEGs


## Requirements
<br>
<br>
<br>

::: {.center}

### DE results 
 
### +

### Gene set(s) (list(s) of genes) 

### + 

### Statistical test

:::

## Where to get gene sets?


- Databases
  - Gene Ontology
  - KEGG
  - Reactome
  - MSigDB
  - ...
- Previous studies
  - Markers of a population of interest
  - Targets of a transcription factor of interest
  - DE genes from another analysis



## Gene ontology 

::: {.columns}
::: {.column}

![](assets/hexose-biosynthetic-process.png){width=85%}

:::
::: {.column}

![](assets/db-go.png)

- Network graph, loosely hierarchical
- Three ontologies
  - Biological process (e.g. Neutrophil Chemotaxis, Cell proliferation)
  - Molecular Function (e.g. Histone acetylation, Phosphorylation)
  - Cellular compartment (e.g. Nucleus, Cytoplasm, Plasma membrane)
- Genes can belong to multiple terms

:::
:::

## Kegg 

![](assets/kegg.png)

. . .

::: {}

- Fewer and smaller ontology
- Highly curated

:::

[[Pathview](https://www.bioconductor.org/packages/release/bioc/html/pathview.html)]{.small .right}



## Overrepresentation analysis (ORA)

![](assets/ora.png)

- Hypergeometric test (Fisher's exact test)

## Overrepresentation analysis (ORA)

- Background can be all genes or all genes expressed in your cell population
- Requires arbitrary cut-off
- Omits actual gene-level statistics
- Computationally fast
- Generally works for few genes with strong effects

## Gene set enrichment analysis (GSEA)

![](assets/subramanium-2005-gsa.webp)

[@subramanian2005gene]{.small}

## Gene set enrichment analysis (GSEA)

::: {.columns}
::: {.column}

![](assets/gsea_image009.jpg)

:::
::: {.column}

- Enrichment score (ES)
- Normalized enrichment score (NES)
- No need for cut-offs
- Takes gene-level stats into account
- More sensitive to subtle changes

[[GSEA User Guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html)]{.small}

:::
:::


## Tools

**Online**  

- [Enrichr](https://maayanlab.cloud/Enrichr/)
- [Gorilla](http://cbl-gorilla.cs.technion.ac.il/)
- [Webgestalt](https://www.webgestalt.org/)

**Code**  

- [fgsea (R)](https://bioconductor.org/packages/release/bioc/html/fgsea.html)
- [clusterProfiler (R)](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
- [msigdbr (R)](https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html)
- [gseapy (Python)](https://gseapy.readthedocs.io/en/latest/introduction.html)


## Considerations

- Bias in curation - highly researched topics will be over-represented
- Gene set names can be misleading
- Specific vs general gene sets
- Multifunctional genes
- Translation of gene ids
- Databases change
- Curation is organism-specific
- Critical evaluation is required

## References

::: {#refs}
:::

## Acknowledgements

Adapted from previous presentations by Leif Wigge, Paulo Czarnewski and Roy Francis.