---
title: Functional Analysis
author: Roy Francis, Paulo Czarnewski
format:
  revealjs:
    df-print: kable
execute:
  echo: false
bibliography: references.bib
---

## Workflow

- Quality control
- Cell cycle phase classification
- Normalization
- Select highly variable genes
- Data integration
- Clustering
- Cell typing
- Differential gene expression
- **Functional analysis (GSA/GSEA)**

## Many names for functional analysis

- Functional analysis
- Pathway analysis
- Gene set analyses (GSA)
- Gene set enrichment analysis (GSEA)
- Gene list enrichment analysis
- Go term analysis
- Over representation analysis
- Hypergeometric test

...

## What is functional analysis?

![](assets/functional.png)

. . .

Gene-level data -> Gene set data

. . .

We focus on transcriptiomics and DGE, but in principle applies to any genome-wide data

:::{.notes}
Many different kinds of genome wide analyses such differential gene expression, differential binding (chip-seq), differential methylation etc. could all end up here at functional analysis.
:::

## Why functional analysis?

- Make sense of long list of DEGs
- What is the function of those genes?
- What is the biological consequence of over/under expression of genes?
- Connect your DEGs and thereby your experiment to pathway activity
- Top genes in itself might be be the most interesting inference
- Less prone to false positive DEGs

## Gene sets

::: {.columns}
::: {.column}

![](assets/db-go.png)  
Gene Ontology ([GO](https://geneontology.org/))

![](assets/db-kegg.png)  
Kyoto encyclopedia of genes and genomes ([KEGG](https://www.genome.jp/kegg/))

![](assets/db-reactome.png)  
[Reactome](https://reactome.org/)

:::
::: {.column}

![](assets/db-wikipathways.png)  
[Wikipathways](https://www.wikipathways.org/)

![](assets/db-msigdb.png)  
Molecular signatures database ([MSigDB](https://www.gsea-msigdb.org/gsea/msigdb))

:::
:::

## Gene ontology 

::: {.columns}
::: {.column}

![](assets/hexose-biosynthetic-process.png){width=85%}

:::
::: {.column}

![](assets/db-go.png)

- Network graph, loosely hierarchical
- Disjoint graphs
  - Biological process (Neutrophil Chemotaxis, Cell proliferation)
  - Molecular Function (Histone acetylation, Phosphorylation)
  - Cellular compartment (Nucleus, Cytoplasm, Plasma membrane)
- Genes can belong to multiple terms

:::
:::

## Kegg 

![](assets/kegg.png)

. . .

::: {}

- Fewer and smaller ontology
- Better curated
- Metabolic pathways

:::

[[Pathview](https://www.bioconductor.org/packages/release/bioc/html/pathview.html)]{.small .right}

## Tools

**Online**  

- [Enrichr](https://maayanlab.cloud/Enrichr/)
- [Gorilla](http://cbl-gorilla.cs.technion.ac.il/)
- [Webgestalt](https://www.webgestalt.org/)

**Code**  

- [clusterProfiler (R)](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
- [fgsea (R)](https://bioconductor.org/packages/release/bioc/html/fgsea.html)
- [msigdbr (R)](https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html)
- [gseapy (Python)](https://gseapy.readthedocs.io/en/latest/introduction.html)

## Overrepresentation analysis (ORA/GSA)

![](assets/ora.png)

- Hypergeometric test (Fisher's exact test)

## Overrepresentation analysis (ORA/GSA)

- Background can be all genes or all genes expressed in your cell population
- Requires arbitrary cut-off
- Omits actual gene-level statistics
- Computationally fast
- Generally works for few genes with strong effects

## Gene set enrichment analysis (GSEA)

![](assets/subramanium-2005-gsa.webp)

[@subramanian2005gene]{.small}

## Gene set enrichment analysis (GSEA)

::: {.columns}
::: {.column}

![](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideTEXT.fld/image009.jpg)

:::
::: {.column}

- Enrichment score (ES)
- Normalized enrichment score (NES)
- No need for cut-offs
- Takes gene-level stats into account
- More sensitive to subtle changes

[[GSEA User Guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html)]{.small}

:::
:::

## Considerations

- Bias in gene sets
- Multifunctional genes, set size
- Well studied topics/diseases will be overrepresented (eg. covid)
- Translation of gene ids
- Changes in database
- Incomplete information for your organism
- Critical evaluation is required

## References

::: {#refs}
:::

## Acknowledgements

Adapted from previous presentations by Leif Wigge & Paulo Czarnewski.