<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Åsa Björklund">
<meta name="author" content="Paulo Czarnewski">
<meta name="author" content="Susanne Reinsbach">
<meta name="author" content="Roy Francis">
<meta name="description" content="Quality control of single cell RNA-Seq data. Inspection of QC metrics including number of UMIs, number of genes expressed, mitochondrial and ribosomal expression, sex and cell cycle state.">

<title> Quality Control</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">

<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-archive" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Archive</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-archive">    
        <li>
    <a class="dropdown-item" href="../../archive/2025/index.html" rel="" target="">
 <span class="dropdown-text">2025</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../archive/2024/index.html" rel="" target="">
 <span class="dropdown-text">2024</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../archive/2023/index.html" rel="" target="">
 <span class="dropdown-text">2023</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/susrei/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title"><i class="fa-solid fa-clipboard-list" aria-label="clipboard-list"></i> Quality Control</h1>
            <p class="subtitle lead"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor Toolkit</p>
                  <div>
        <div class="description">
          Quality control of single cell RNA-Seq data. Inspection of QC metrics including number of UMIs, number of genes expressed, mitochondrial and ribosomal expression, sex and cell cycle state.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Authors</div>
      <div class="quarto-title-meta-contents">
               <p>Åsa Björklund </p>
               <p>Paulo Czarnewski </p>
               <p>Susanne Reinsbach </p>
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">19-Oct-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#meta-qc_data" id="toc-meta-qc_data" class="nav-link active" data-scroll-target="#meta-qc_data"><span class="header-section-number">1</span> Get data</a></li>
  <li><a href="#meta-qc_collate" id="toc-meta-qc_collate" class="nav-link" data-scroll-target="#meta-qc_collate"><span class="header-section-number">2</span> Collate</a></li>
  <li><a href="#meta-qc_calqc" id="toc-meta-qc_calqc" class="nav-link" data-scroll-target="#meta-qc_calqc"><span class="header-section-number">3</span> Calculate QC</a></li>
  <li><a href="#meta-qc_plotqc" id="toc-meta-qc_plotqc" class="nav-link" data-scroll-target="#meta-qc_plotqc"><span class="header-section-number">4</span> Plot QC</a></li>
  <li><a href="#meta-qc_filter" id="toc-meta-qc_filter" class="nav-link" data-scroll-target="#meta-qc_filter"><span class="header-section-number">5</span> Filtering</a>
  <ul>
  <li><a href="#meta-qc_filter_detect" id="toc-meta-qc_filter_detect" class="nav-link" data-scroll-target="#meta-qc_filter_detect"><span class="header-section-number">5.1</span> Detection-based filtering</a></li>
  <li><a href="#meta-qc_filter_mr" id="toc-meta-qc_filter_mr" class="nav-link" data-scroll-target="#meta-qc_filter_mr"><span class="header-section-number">5.2</span> Mito/Ribo filtering</a></li>
  <li><a href="#meta-qc_filter_plot" id="toc-meta-qc_filter_plot" class="nav-link" data-scroll-target="#meta-qc_filter_plot"><span class="header-section-number">5.3</span> Plot filtered QC</a></li>
  <li><a href="#meta-qc_filter_genes" id="toc-meta-qc_filter_genes" class="nav-link" data-scroll-target="#meta-qc_filter_genes"><span class="header-section-number">5.4</span> Filter genes</a></li>
  </ul></li>
  <li><a href="#meta-qc_sex" id="toc-meta-qc_sex" class="nav-link" data-scroll-target="#meta-qc_sex"><span class="header-section-number">6</span> Sample sex</a></li>
  <li><a href="#meta-qc_cellcycle" id="toc-meta-qc_cellcycle" class="nav-link" data-scroll-target="#meta-qc_cellcycle"><span class="header-section-number">7</span> Cell cycle state</a></li>
  <li><a href="#meta-qc_doublet" id="toc-meta-qc_doublet" class="nav-link" data-scroll-target="#meta-qc_doublet"><span class="header-section-number">8</span> Predict doublets</a></li>
  <li><a href="#meta-qc_save" id="toc-meta-qc_save" class="nav-link" data-scroll-target="#meta-qc_save"><span class="header-section-number">9</span> Save data</a></li>
  <li><a href="#meta-session" id="toc-meta-session" class="nav-link" data-scroll-target="#meta-session"><span class="header-section-number">10</span> Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks run R commands unless otherwise specified.</p>
</div>
</div>
<section id="meta-qc_data" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="meta-qc_data"><span class="header-section-number">1</span> Get data</h2>
<p>In this tutorial, we will run all tutorials with a set of 8 PBMC 10x datasets from 4 covid-19 patients and 4 healthy controls, the samples have been subsampled to 1500 cells per sample. We can start by defining our paths.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download pre-computed annotation</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>fetch_annotation <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># url for source and intermediate data</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>path_data <span class="ot">&lt;-</span> <span class="st">"https://nextcloud.dc.scilifelab.se/public.php/webdav"</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>curl_upass <span class="ot">&lt;-</span> <span class="st">"-u zbC5fr2LbEZ9rSE:scRNAseq2025"</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>path_covid <span class="ot">&lt;-</span> <span class="st">"./data/covid/raw"</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_covid)) <span class="fu">dir.create</span>(path_covid, <span class="at">recursive =</span> T)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>path_results <span class="ot">&lt;-</span> <span class="st">"./data/covid/results"</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_results)) <span class="fu">dir.create</span>(path_results, <span class="at">recursive =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"normal_pbmc_13.h5"</span>, <span class="st">"normal_pbmc_14.h5"</span>, <span class="st">"normal_pbmc_19.h5"</span>, <span class="st">"normal_pbmc_5.h5"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ncov_pbmc_15.h5"</span>, <span class="st">"ncov_pbmc_16.h5"</span>, <span class="st">"ncov_pbmc_17.h5"</span>, <span class="st">"ncov_pbmc_1.h5"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> file_list) {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    path_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_covid, i)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(path_file)) {</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">download.file</span>(<span class="at">url =</span> <span class="fu">file.path</span>(<span class="fu">file.path</span>(path_data, <span class="st">"covid/raw"</span>), i), </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">destfile =</span> path_file, <span class="at">method =</span> <span class="st">"curl"</span>, <span class="at">extra =</span> curl_upass)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With data in place, now we can start loading libraries we will use in this tutorial.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scater)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork) <span class="co"># combining figures</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scDblFinder)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can first load the data individually by reading directly from HDF5 file format (.h5).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>cov<span class="fl">.15</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"ncov_pbmc_15.h5"</span>),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>cov<span class="fl">.1</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"ncov_pbmc_1.h5"</span>),</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>cov<span class="fl">.16</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"ncov_pbmc_16.h5"</span>),</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>cov<span class="fl">.17</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"ncov_pbmc_17.h5"</span>),</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>ctrl<span class="fl">.5</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"normal_pbmc_5.h5"</span>),</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>ctrl<span class="fl">.13</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"normal_pbmc_13.h5"</span>),</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>ctrl<span class="fl">.14</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"normal_pbmc_14.h5"</span>),</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>ctrl<span class="fl">.19</span> <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">Read10X_h5</span>(</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">filename =</span> <span class="fu">file.path</span>(path_covid, <span class="st">"normal_pbmc_19.h5"</span>),</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.names =</span> T</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-qc_collate" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="meta-qc_collate"><span class="header-section-number">2</span> Collate</h2>
<p>We can now merge them objects into a single object. Each analysis workflow (Seurat, Scater, Scanpy, etc) has its own way of storing data. We will add dataset labels as <strong>cell.ids</strong> just in case you have overlapping barcodes between the datasets. After that we add a column <strong>type</strong> in the metadata to define covid and ctrl samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> <span class="fu">cbind</span>(cov<span class="fl">.1</span>, cov<span class="fl">.15</span>, cov<span class="fl">.16</span>, cov<span class="fl">.17</span>, ctrl<span class="fl">.5</span>, ctrl<span class="fl">.13</span>, ctrl<span class="fl">.14</span>,ctrl<span class="fl">.19</span>)))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33538 12000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding metadata</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>colData<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">sapply</span>(<span class="fu">c</span>(<span class="st">"cov.1"</span>, <span class="st">"cov.15"</span>, <span class="st">"cov.16"</span>, <span class="st">"cov.17"</span>, <span class="st">"ctrl.5"</span>, <span class="st">"ctrl.13"</span>, <span class="st">"ctrl.14"</span>,<span class="st">"ctrl.19"</span>), <span class="cf">function</span>(x) <span class="fu">rep</span>(x, <span class="fu">ncol</span>(<span class="fu">get</span>(x)))))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>colData<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"cov"</span>, sce<span class="sc">@</span>colData<span class="sc">$</span>sample), <span class="st">"Covid"</span>, <span class="st">"Control"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once you have created the merged Seurat object, the count matrices and individual count matrices and objects are not needed anymore. It is a good idea to remove them and run garbage collect to free up some memory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove all objects that will not be used.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(cov<span class="fl">.15</span>, cov<span class="fl">.1</span>, cov<span class="fl">.17</span>, cov<span class="fl">.16</span>, ctrl<span class="fl">.5</span>, ctrl<span class="fl">.13</span>, ctrl<span class="fl">.14</span>, ctrl<span class="fl">.19</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># run garbage collect to free up memory</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           used  (Mb) gc trigger  (Mb)  max used  (Mb)
Ncells 11938606 637.6   17315390 924.8  16951466 905.4
Vcells 47505463 362.5  119270751 910.0 119024704 908.1</code></pre>
</div>
</div>
<p>Here is how the count matrix and the metadata look like for every cell.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">counts</span>(sce)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6 x 10 sparse Matrix of class "dgCMatrix"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                               
MIR1302-2HG . . . . . . . . . .
FAM138A     . . . . . . . . . .
OR4F5       . . . . . . . . . .
AL627309.1  . . . . . . . . . .
AL627309.3  . . . . . . . . . .
AL627309.2  . . . . . . . . . .</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sce<span class="sc">@</span>colData, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 10 rows and 2 columns
                        sample        type
                   &lt;character&gt; &lt;character&gt;
AGGTAGGTCGTTGTTT-1       cov.1       Covid
TAGAGTCGTCCTCCAT-1       cov.1       Covid
CCCTGATAGCGAACTG-1       cov.1       Covid
TCATCATTCCACGTAA-1       cov.1       Covid
ATTTACCCAAGCCTGC-1       cov.1       Covid
GTTGTCCTCTAGAACC-1       cov.1       Covid
CCTCCAACAAGAGATT-1       cov.1       Covid
AATAGAGGTGTGAGCA-1       cov.1       Covid
GGTGGCTAGCGAATGC-1       cov.1       Covid
TCGGGCACAGTGTGGA-1       cov.1       Covid</code></pre>
</div>
</div>
</section>
<section id="meta-qc_calqc" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="meta-qc_calqc"><span class="header-section-number">3</span> Calculate QC</h2>
<p>Having the data in a suitable format, we can start calculating some quality metrics. We can for example calculate the percentage of mitochondrial and ribosomal genes per cell and add to the metadata. The proportion of hemoglobin genes can give an indication of red blood cell contamination, but in some tissues it can also be the case that some celltypes have higher content of hemoglobin. This will be helpful to visualize them across different metadata parameters (i.e.&nbsp;datasetID and chemistry version). There are several ways of doing this. The QC metrics are finally added to the metadata table.</p>
<p>Citing from Simple Single Cell workflows (Lun, McCarthy &amp; Marioni, 2017): High proportions are indicative of poor-quality cells (Islam et al.&nbsp;2014; Ilicic et al.&nbsp;2016), possibly because of loss of cytoplasmic RNA from perforated cells. The reasoning is that mitochondria are larger than individual transcript molecules and less likely to escape through tears in the cell membrane.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mitochondrial genes</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^MT-"</span>, <span class="fu">rownames</span>(sce))]</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Ribosomal genes</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>ribo_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^RP[SL]"</span>, <span class="fu">rownames</span>(sce))]</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Hemoglobin genes - includes all genes starting with HB except HBP.</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>hb_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[<span class="fu">grep</span>(<span class="st">"^HB[^(P|E|S)]"</span>, <span class="fu">rownames</span>(sce))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First, let Scran calculate some general qc-stats for genes and cells with the function <code>perCellQCMetrics</code>. It can also calculate proportion of counts for specific gene subsets, so first we need to define which genes are mitochondrial, ribosomal and hemoglobin.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(sce, <span class="at">flatten =</span> T, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mt =</span> mito_genes, <span class="at">hb =</span> hb_genes, <span class="at">ribo =</span> ribo_genes))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Way2: Doing it manually</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>colData<span class="sc">$</span>percent_mito <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(<span class="fu">counts</span>(sce)[mito_genes, ]) <span class="sc">/</span> sce<span class="sc">@</span>colData<span class="sc">$</span>total <span class="sc">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now you can see that we have additional data in the metadata slot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(sce))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 15 columns
                        sample        type       sum  detected subsets_mt_sum
                   &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;      &lt;numeric&gt;
AGGTAGGTCGTTGTTT-1       cov.1       Covid      1396       656             26
TAGAGTCGTCCTCCAT-1       cov.1       Covid      1613       779            186
CCCTGATAGCGAACTG-1       cov.1       Covid      9482      2036            761
TCATCATTCCACGTAA-1       cov.1       Covid      4357       875           2960
ATTTACCCAAGCCTGC-1       cov.1       Covid     12466      3290            686
GTTGTCCTCTAGAACC-1       cov.1       Covid      5541      1606            707
                   subsets_mt_detected subsets_mt_percent subsets_hb_sum
                             &lt;integer&gt;          &lt;numeric&gt;      &lt;numeric&gt;
AGGTAGGTCGTTGTTT-1                   8            1.86246              1
TAGAGTCGTCCTCCAT-1                  10           11.53131              0
CCCTGATAGCGAACTG-1                  11            8.02573              3
TCATCATTCCACGTAA-1                  13           67.93665              2
ATTTACCCAAGCCTGC-1                  12            5.50297              1
GTTGTCCTCTAGAACC-1                  12           12.75943              3
                   subsets_hb_detected subsets_hb_percent subsets_ribo_sum
                             &lt;integer&gt;          &lt;numeric&gt;        &lt;numeric&gt;
AGGTAGGTCGTTGTTT-1                   1         0.07163324               11
TAGAGTCGTCCTCCAT-1                   0         0.00000000               96
CCCTGATAGCGAACTG-1                   1         0.03163889             4157
TCATCATTCCACGTAA-1                   2         0.04590314               99
ATTTACCCAAGCCTGC-1                   1         0.00802182             2281
GTTGTCCTCTAGAACC-1                   2         0.05414185             1664
                   subsets_ribo_detected subsets_ribo_percent     total
                               &lt;integer&gt;            &lt;numeric&gt; &lt;numeric&gt;
AGGTAGGTCGTTGTTT-1                     9             0.787966      1396
TAGAGTCGTCCTCCAT-1                    45             5.951643      1613
CCCTGATAGCGAACTG-1                    85            43.840962      9482
TCATCATTCCACGTAA-1                    52             2.272206      4357
ATTTACCCAAGCCTGC-1                    82            18.297770     12466
GTTGTCCTCTAGAACC-1                    80            30.030680      5541
                   percent_mito
                      &lt;numeric&gt;
AGGTAGGTCGTTGTTT-1      1.86246
TAGAGTCGTCCTCCAT-1     11.53131
CCCTGATAGCGAACTG-1      8.02573
TCATCATTCCACGTAA-1     67.93665
ATTTACCCAAGCCTGC-1      5.50297
GTTGTCCTCTAGAACC-1     12.75943</code></pre>
</div>
</div>
</section>
<section id="meta-qc_plotqc" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="meta-qc_plotqc"><span class="header-section-number">4</span> Plot QC</h2>
<p>Now we can plot some of the QC variables as violin plots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># total is total UMIs per cell</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># detected is number of detected genes.</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># the different gene subset percentages are listed as subsets_mt_percent etc.</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"detected"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"total"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_mt_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_ribo_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_hb_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/qc-vln-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="bioc_01_qc_files/figure-html/qc-vln-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Looking at the violin plots, what do you think are appropriate cutoffs for filtering these samples</p>
</div>
</div>
<p>As you can see, there is quite some difference in quality for these samples, with for instance the covid_15 and covid_16 samples having cells with fewer detected genes and more mitochondrial content. As the ribosomal proteins are highly expressed they will make up a larger proportion of the transcriptional landscape when fewer of the lowly expressed genes are detected. We can also plot the different QC-measures as scatter plots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(sce, <span class="at">x =</span> <span class="st">"total"</span>, <span class="at">y =</span> <span class="st">"detected"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/qc-scatter-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="bioc_01_qc_files/figure-html/qc-scatter-1.png" class="img-fluid figure-img" width="576"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Plot additional QC stats that we have calculated as scatter plots. How are the different measures correlated? Can you explain why?</p>
</div>
</div>
</section>
<section id="meta-qc_filter" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="meta-qc_filter"><span class="header-section-number">5</span> Filtering</h2>
<section id="meta-qc_filter_detect" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="meta-qc_filter_detect"><span class="header-section-number">5.1</span> Detection-based filtering</h3>
<p>A standard approach is to filter cells with low number of reads as well as genes that are present in at least a given number of cells. Here we will only consider cells with at least 200 detected genes and genes need to be expressed in at least 3 cells. Please note that those values are highly dependent on the library preparation method used.</p>
<p>In Scran, we can use the function <code>quickPerCellQC</code> to filter out outliers from distributions of qc stats, such as detected genes, gene subsets etc. But in this case, we will take one setting at a time and run through the steps of filtering cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33538 12000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>selected_c <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sce)[sce<span class="sc">$</span>detected <span class="sc">&gt;</span> <span class="dv">200</span>]</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>selected_f <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)[Matrix<span class="sc">::</span><span class="fu">rowSums</span>(<span class="fu">counts</span>(sce)) <span class="sc">&gt;</span> <span class="dv">3</span>]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce[selected_f, selected_c]</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce.filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18877 10656</code></pre>
</div>
</div>
<p>Extremely high number of detected genes could indicate doublets. However, depending on the cell type composition in your sample, you may have cells with higher number of genes (and also higher counts) from one cell type. In this case, we will run doublet prediction further down, so we will skip this step now, but the code below is an example of how it can be run:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># skip for now and run doublet detection instead...</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># high.det.v3 &lt;- sce.filt$nFeatures &gt; 4100</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># high.det.v2 &lt;- (sce.filt$nFeatures &gt; 2000) &amp; (sce.filt$sample_id == "v2.1k")</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># remove these cells</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># sce.filt &lt;- sce.filt[ , (!high.det.v3) &amp; (!high.det.v2)]</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co"># check number of cells</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ncol(sce.filt)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Additionally, we can also see which genes contribute the most to such reads. We can for instance plot the percentage of counts per gene.</p>
<p>In Scater, you can also use the function <code>plotHighestExprs()</code> to plot the gene contribution, but the function is quite slow, so we will do it on our own instead..</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the relative expression of each gene per cell</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Use sparse matrix operations, if your dataset is large, doing matrix devisions the regular way will take a very long time.</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>C <span class="ot">&lt;-</span> <span class="fu">counts</span>(sce.filt)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>C<span class="sc">@</span>x <span class="ot">&lt;-</span> C<span class="sc">@</span>x <span class="sc">/</span> <span class="fu">rep.int</span>(<span class="fu">colSums</span>(C), <span class="fu">diff</span>(C<span class="sc">@</span>p)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>most_expressed <span class="ot">&lt;-</span> <span class="fu">order</span>(Matrix<span class="sc">::</span><span class="fu">rowSums</span>(C), <span class="at">decreasing =</span> T)[<span class="dv">20</span><span class="sc">:</span><span class="dv">1</span>]</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">as.matrix</span>(<span class="fu">t</span>(C[most_expressed, ])), <span class="at">cex =</span> .<span class="dv">1</span>, <span class="at">las =</span> <span class="dv">1</span>, <span class="at">xlab =</span> <span class="st">"% total count per cell"</span>, <span class="at">col =</span> scales<span class="sc">::</span><span class="fu">hue_pal</span>()(<span class="dv">20</span>)[<span class="dv">20</span><span class="sc">:</span><span class="dv">1</span>], <span class="at">horizontal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/top-genes-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="bioc_01_qc_files/figure-html/top-genes-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(C)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># also, there is the option of running the function "plotHighestExprs" in the scater package, however, this function takes very long to execute.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you can see, MALAT1 constitutes up to 30% of the UMIs from a single cell and the other top genes are mitochondrial and ribosomal genes. It is quite common that nuclear lincRNAs have correlation with quality and mitochondrial reads, so high detection of MALAT1 may be a technical issue. Let us assemble some information about such genes, which are important for quality control and downstream filtering.</p>
</section>
<section id="meta-qc_filter_mr" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="meta-qc_filter_mr"><span class="header-section-number">5.2</span> Mito/Ribo filtering</h3>
<p>We also have quite a lot of cells with high proportion of mitochondrial and low proportion of ribosomal reads. It would be wise to remove those cells, if we have enough cells left after filtering. Another option would be to either remove all mitochondrial reads from the dataset and hope that the remaining genes still have enough biological signal. A third option would be to just regress out the <code>percent_mito</code> variable during scaling. In this case we had as much as 99.7% mitochondrial reads in some of the cells, so it is quite unlikely that there is much cell type signature left in those. Looking at the plots, make reasonable decisions on where to draw the cutoff. In this case, the bulk of the cells are below 20% mitochondrial reads and that will be used as a cutoff. We will also remove cells with less than 5% ribosomal reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>selected_mito <span class="ot">&lt;-</span> sce.filt<span class="sc">$</span>subsets_mt_percent <span class="sc">&lt;</span> <span class="dv">20</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>selected_ribo <span class="ot">&lt;-</span> sce.filt<span class="sc">$</span>subsets_ribo_percent <span class="sc">&gt;</span> <span class="dv">5</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># and subset the object to only keep those cells</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce.filt[, selected_mito <span class="sc">&amp;</span> selected_ribo]</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce.filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18877  7431</code></pre>
</div>
</div>
<p>As you can see, a large proportion of sample covid_15 is filtered out. Also, there is still quite a lot of variation in <code>percent_mito</code>, so it will have to be dealt with in the data analysis step. We can also notice that the <code>percent_ribo</code> are also highly variable, but that is expected since different cell types have different proportions of ribosomal content, according to their function.</p>
</section>
<section id="meta-qc_filter_plot" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="meta-qc_filter_plot"><span class="header-section-number">5.3</span> Plot filtered QC</h3>
<p>Lets plot the same QC-stats once more.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"detected"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"total"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_mt_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_ribo_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce, <span class="at">y =</span> <span class="st">"subsets_hb_percent"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/qc-vln-filt-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="bioc_01_qc_files/figure-html/qc-vln-filt-1.png" class="img-fluid figure-img" width="960"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="meta-qc_filter_genes" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="meta-qc_filter_genes"><span class="header-section-number">5.4</span> Filter genes</h3>
<p>As the level of expression of mitochondrial and MALAT1 genes are judged as mainly technical, it can be wise to remove them from the dataset before any further analysis. In this case we will also remove the HB genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce.filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18877  7431</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter MALAT1</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce.filt[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"MALAT1"</span>, <span class="fu">rownames</span>(sce.filt)), ]</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Mitocondrial</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce.filt[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^MT-"</span>, <span class="fu">rownames</span>(sce.filt)), ]</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Ribossomal gene (optional if that is a problem on your data)</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="co"># sce.filt &lt;- sce.filt[ ! grepl("^RP[SL]", rownames(sce.filt)), ]</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter Hemoglobin gene  (optional if that is a problem on your data)</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce.filt[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^HB[^(P|E|S)]"</span>, <span class="fu">rownames</span>(sce.filt)), ]</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce.filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18854  7431</code></pre>
</div>
</div>
</section>
</section>
<section id="meta-qc_sex" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="meta-qc_sex"><span class="header-section-number">6</span> Sample sex</h2>
<p>When working with human or animal samples, you should ideally constrain your experiments to a single sex to avoid including sex bias in the conclusions. However this may not always be possible. By looking at reads from chromosomeY (males) and XIST (X-inactive specific transcript) expression (mainly female) it is quite easy to determine per sample which sex it is. It can also be a good way to detect if there has been any mislabelling in which case, the sample metadata sex does not agree with the computational predictions.</p>
<p>To get chromosome information for all genes, you should ideally parse the information from the gtf file that you used in the mapping pipeline as it has the exact same annotation version/gene naming. However, it may not always be available, as in this case where we have downloaded public data. R package biomaRt can be used to fetch annotation information. The code to run biomaRt is provided. As the biomart instances are quite often unresponsive, we will download and use a file that was created in advance.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Here is the code to download annotation data from Ensembl using biomaRt. We will not run this now and instead use a pre-computed file in the step below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fetch_annotation is defined at the top of this document</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>fetch_annotation) {</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">suppressMessages</span>(<span class="fu">library</span>(biomaRt))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># initialize connection to mart, may take some time if the sites are unresponsive.</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  mart <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>, <span class="at">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># fetch chromosome info plus some other annotations</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  genes_table <span class="ot">&lt;-</span> <span class="fu">try</span>(biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"description"</span>, <span class="st">"gene_biotype"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"start_position"</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  ), <span class="at">mart =</span> mart, <span class="at">useCache =</span> F))</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(genes_table, <span class="at">file =</span> <span class="st">"data/covid/results/genes_table.csv"</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>Download precomputed data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fetch_annotation is defined at the top of this document</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (fetch_annotation) {</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  genes_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_results, <span class="st">"genes_table.csv"</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(genes_file)) <span class="fu">download.file</span>(<span class="fu">file.path</span>(path_data, <span class="st">"covid/results_bioc/genes_table.csv"</span>), <span class="at">destfile =</span> genes_file,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">method =</span> <span class="st">"curl"</span>, <span class="at">extra =</span> curl_upass)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>genes.table <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(genes_file)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>genes.table <span class="ot">&lt;-</span> genes.table[genes.table<span class="sc">$</span>external_gene_name <span class="sc">%in%</span> <span class="fu">rownames</span>(sce.filt), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have the chromosome information, we can calculate the proportion of reads that comes from chromosome Y per cell.But first we have to remove all genes in the pseudoautosmal regions of chrY that are: * chromosome:GRCh38:Y:10001 - 2781479 is shared with X: 10001 - 2781479 (PAR1) * chromosome:GRCh38:Y:56887903 - 57217415 is shared with X: 155701383 - 156030895 (PAR2)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>par1 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">10001</span>, <span class="dv">2781479</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>par2 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">56887903</span>, <span class="dv">57217415</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>p1.gene <span class="ot">=</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>start_position <span class="sc">&gt;</span> par1[<span class="dv">1</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>start_position <span class="sc">&lt;</span> par1[<span class="dv">2</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>p2.gene <span class="ot">=</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>start_position <span class="sc">&gt;</span> par2[<span class="dv">1</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>start_position <span class="sc">&lt;</span> par2[<span class="dv">2</span>] <span class="sc">&amp;</span> genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>chrY.gene <span class="ot">&lt;-</span> genes.table<span class="sc">$</span>external_gene_name[genes.table<span class="sc">$</span>chromosome_name <span class="sc">==</span> <span class="st">"Y"</span>]</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>chrY.gene <span class="ot">=</span> <span class="fu">setdiff</span>(chrY.gene, <span class="fu">c</span>(p1.gene, p2.gene))</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">@</span>colData<span class="sc">$</span>pct_chrY <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(<span class="fu">counts</span>(sce.filt)[chrY.gene, ]) <span class="sc">/</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(sce.filt))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then plot XIST expression vs chrY proportion. As you can see, the samples are clearly on either side, even if some cells do not have detection of either.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># as plotColData cannot take an expression vs metadata, we need to add in XIST expression to colData</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">@</span>colData<span class="sc">$</span>XIST <span class="ot">&lt;-</span> <span class="fu">counts</span>(sce.filt)[<span class="st">"XIST"</span>, ] <span class="sc">/</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(sce.filt)) <span class="sc">*</span> <span class="dv">10000</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(sce.filt, <span class="st">"XIST"</span>, <span class="st">"pct_chrY"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/sex-scatter-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="bioc_01_qc_files/figure-html/sex-scatter-1.png" class="img-fluid figure-img" width="480"></a></p>
</figure>
</div>
</div>
</div>
<p>Plot as violins.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"XIST"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"pct_chrY"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/sex-vln-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="bioc_01_qc_files/figure-html/sex-vln-1.png" class="img-fluid figure-img" width="768"></a></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here, we can see clearly that we have three males and five females, can you see which samples they are? Do you think this will cause any problems for downstream analysis? Discuss with your group: what would be the best way to deal with this type of sex bias?</p>
</div>
</div>
</section>
<section id="meta-qc_cellcycle" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="meta-qc_cellcycle"><span class="header-section-number">7</span> Cell cycle state</h2>
<p>We here perform cell cycle scoring. To score a gene list, the algorithm calculates the difference of mean expression of the given list and the mean expression of reference genes. To build the reference, the function randomly chooses a bunch of genes matching the distribution of the expression of the given list. Cell cycle scoring adds three slots in the metadata, a score for S phase, a score for G2M phase and the predicted cell cycle phase.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>hs.pairs <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">"exdata"</span>, <span class="st">"human_cycle_markers.rds"</span>, <span class="at">package =</span> <span class="st">"scran"</span>))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">&lt;-</span> <span class="fu">select</span>(org.Hs.eg.db, <span class="at">keys =</span> <span class="fu">rownames</span>(sce.filt), <span class="at">keytype =</span> <span class="st">"SYMBOL"</span>, <span class="at">column =</span> <span class="st">"ENSEMBL"</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> anno<span class="sc">$</span>ENSEMBL[<span class="fu">match</span>(<span class="fu">rownames</span>(sce.filt), anno<span class="sc">$</span>SYMBOL)]</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Use only genes related to biological process cell cycle to speed up</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># https://www.ebi.ac.uk/QuickGO/term/GO:0007049 = cell cycle (BP,Biological Process)</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>GOs <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(<span class="fu">select</span>(org.Hs.eg.db, <span class="at">keys =</span> <span class="fu">na.omit</span>(ensembl), <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>, <span class="at">column =</span> <span class="st">"GO"</span>))</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>GOs <span class="ot">&lt;-</span> GOs[GOs<span class="sc">$</span>GO <span class="sc">==</span> <span class="st">"GO:0007049"</span>, <span class="st">"ENSEMBL"</span>]</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>hs.pairs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(hs.pairs, <span class="cf">function</span>(x) {</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>    x[<span class="fu">rowSums</span>(<span class="fu">apply</span>(x, <span class="dv">2</span>, <span class="cf">function</span>(i) i <span class="sc">%in%</span> GOs)) <span class="sc">&gt;=</span> <span class="dv">1</span>, ]</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(hs.pairs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ G1 :'data.frame':    6633 obs. of  2 variables:
  ..$ first : chr [1:6633] "ENSG00000100519" "ENSG00000100519" "ENSG00000100519" "ENSG00000100519" ...
  ..$ second: chr [1:6633] "ENSG00000065135" "ENSG00000080345" "ENSG00000101266" "ENSG00000135679" ...
 $ S  :'data.frame':    8308 obs. of  2 variables:
  ..$ first : chr [1:8308] "ENSG00000255302" "ENSG00000119969" "ENSG00000179051" "ENSG00000127586" ...
  ..$ second: chr [1:8308] "ENSG00000100519" "ENSG00000100519" "ENSG00000100519" "ENSG00000136856" ...
 $ G2M:'data.frame':    6235 obs. of  2 variables:
  ..$ first : chr [1:6235] "ENSG00000100519" "ENSG00000136856" "ENSG00000136856" "ENSG00000136856" ...
  ..$ second: chr [1:6235] "ENSG00000146457" "ENSG00000138028" "ENSG00000227268" "ENSG00000101265" ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>cc.ensembl <span class="ot">&lt;-</span> ensembl[ensembl <span class="sc">%in%</span> GOs] <span class="co"># This is the fastest (less genes), but less accurate too</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cc.ensembl &lt;- ensembl[ ensembl %in% unique(unlist(hs.pairs))]</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>assignments <span class="ot">&lt;-</span> <span class="fu">cyclone</span>(sce.filt[ensembl <span class="sc">%in%</span> cc.ensembl, ], hs.pairs, <span class="at">gene.names =</span> ensembl[ensembl <span class="sc">%in%</span> cc.ensembl])</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">$</span>G1.score <span class="ot">&lt;-</span> assignments<span class="sc">$</span>scores<span class="sc">$</span>G1</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">$</span>G2M.score <span class="ot">&lt;-</span> assignments<span class="sc">$</span>scores<span class="sc">$</span>G2M</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">$</span>S.score <span class="ot">&lt;-</span> assignments<span class="sc">$</span>scores<span class="sc">$</span>S</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>sce.filt<span class="sc">$</span>phase <span class="ot">&lt;-</span> assignments<span class="sc">$</span>phases</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce.filt<span class="sc">$</span>phase)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  G1  G2M    S 
4337  828 1746 </code></pre>
</div>
</div>
<p>We can now create a violin plot for the cell cycle scores as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"G2M.score"</span>, <span class="at">x =</span> <span class="st">"G1.score"</span>, <span class="at">colour_by =</span> <span class="st">"phase"</span>),</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"G2M.score"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"G1.score"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"S.score"</span>, <span class="at">x =</span> <span class="st">"sample"</span>, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">4</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/cc-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="bioc_01_qc_files/figure-html/cc-plot-1.png" class="img-fluid figure-img" width="1344"></a></p>
</figure>
</div>
</div>
</div>
<p>Cyclone predicts most cells as G1, but also quite a lot of cells with high S-Phase scores. Compare to results with Seurat and Scanpy and see how different predictors will give clearly different results.</p>
<p>Cyclone does an automatic prediction of cell cycle phase with a default cutoff of the scores at 0.5 As you can see this does not fit this data very well, so be cautious with using these predictions. Instead we suggest that you look at the scores.</p>
</section>
<section id="meta-qc_doublet" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="meta-qc_doublet"><span class="header-section-number">8</span> Predict doublets</h2>
<p>Doublets/Multiples of cells in the same well/droplet is a common issue in scRNAseq protocols. Especially in droplet-based methods with overloading of cells. In a typical 10x experiment the proportion of doublets is linearly dependent on the amount of loaded cells. As indicated from the Chromium user guide, doublet rates are about as follows:<br>
<a href="../figs/10x_doublet_rate.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="../figs/10x_doublet_rate.png" class="img-fluid"></a><br>
Most doublet detectors simulates doublets by merging cell counts and predicts doublets as cells that have similar embeddings as the simulated doublets. Most such packages need an assumption about the number/proportion of expected doublets in the dataset. The data you are using is subsampled, but the original datasets contained about 5 000 cells per sample, hence we can assume that they loaded about 9 000 cells and should have a doublet rate at about 4%.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Ideally doublet prediction should be run on each sample separately, especially if your samples have different proportions of cell types. In this case, the data is subsampled so we have very few cells per sample and all samples are sorted PBMCs, so it is okay to run them together.</p>
</div>
</div>
<p>There is a method to predict if a cluster consists of mainly doublets <code>findDoubletClusters()</code>, but we can also predict individual cells based on simulations using the function <code>computeDoubletDensity()</code> which we will do here. Doublet detection will be performed using PCA, so we need to first normalize the data and run variable gene detection, as well as UMAP for visualization. These steps will be explored in more detail in coming exercises.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce.filt)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce.filt, <span class="at">block =</span> sce.filt<span class="sc">$</span>sample)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">n =</span> <span class="dv">2000</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce.filt, <span class="at">subset_row =</span> hvgs)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(sce.filt, <span class="at">pca =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run computeDoubletDensity with 10 principal components.</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> <span class="fu">scDblFinder</span>(sce.filt, <span class="at">dims =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotUMAP</span>(sce.filt, <span class="at">colour_by =</span> <span class="st">"scDblFinder.score"</span>),</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotUMAP</span>(sce.filt, <span class="at">colour_by =</span> <span class="st">"scDblFinder.class"</span>),</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plotUMAP</span>(sce.filt, <span class="at">colour_by =</span> <span class="st">"sample"</span>),</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/doublet-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="bioc_01_qc_files/figure-html/doublet-plot-1.png" class="img-fluid figure-img" width="1344"></a></p>
</figure>
</div>
</div>
</div>
<p>We should expect that two cells have more detected genes than a single cell, lets check if our predicted doublets also have more detected genes in general.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(sce.filt, <span class="at">y =</span> <span class="st">"detected"</span>, <span class="at">x =</span> <span class="st">"scDblFinder.class"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><a href="bioc_01_qc_files/figure-html/doublet-vln-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="bioc_01_qc_files/figure-html/doublet-vln-1.png" class="img-fluid figure-img" width="672"></a></p>
</figure>
</div>
</div>
</div>
<p>Now, lets remove all predicted doublets from our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>sce.filt <span class="ot">&lt;-</span> sce.filt[, sce.filt<span class="sc">$</span>scDblFinder.class <span class="sc">==</span> <span class="st">"singlet"</span>]</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sce.filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18854  6924</code></pre>
</div>
</div>
<p>To summarize, lets check how many cells we have removed per sample, we started with 1500 cells per sample. Looking back at the intitial QC plots does it make sense that some samples have much fewer cells now?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  cov.1  cov.15  cov.16  cov.17 ctrl.13 ctrl.14 ctrl.19  ctrl.5 
   1500    1500    1500    1500    1500    1500    1500    1500 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce.filt<span class="sc">$</span>sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  cov.1  cov.15  cov.16  cov.17 ctrl.13 ctrl.14 ctrl.19  ctrl.5 
    836     538     355    1062    1111     952    1086     984 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Discuss">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discuss
</div>
</div>
<div class="callout-body-container callout-body">
<p>“In this case we ran doublet detection with all samples together since we have very small subsampled datasets. But in a real scenario it should be run one sample at a time. Why is this important do you think?”</p>
</div>
</div>
</section>
<section id="meta-qc_save" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="meta-qc_save"><span class="header-section-number">9</span> Save data</h2>
<p>Finally, lets save the QC-filtered data for further analysis. Create output directory <code>data/covid/results</code> and save data to that folder. This will be used in downstream labs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sce.filt, <span class="fu">file.path</span>(path_results, <span class="st">"bioc_covid_qc.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="meta-session" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="meta-session"><span class="header-section-number">10</span> Session info</h2>
<details>
<summary>
Click here
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS/LAPACK: /usr/local/conda/envs/seurat/lib/libopenblasp-r0.3.28.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scDblFinder_1.16.0          org.Hs.eg.db_3.18.0        
 [3] AnnotationDbi_1.64.1        patchwork_1.3.0            
 [5] scran_1.30.0                scater_1.30.1              
 [7] ggplot2_3.5.1               scuttle_1.12.0             
 [9] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0
[11] Biobase_2.62.0              GenomicRanges_1.54.1       
[13] GenomeInfoDb_1.38.1         IRanges_2.36.0             
[15] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[17] MatrixGenerics_1.14.0       matrixStats_1.5.0          

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.3.3            
  [3] later_1.4.1               BiocIO_1.12.0            
  [5] bitops_1.0-9              tibble_3.2.1             
  [7] polyclip_1.10-7           XML_3.99-0.17            
  [9] fastDummies_1.7.5         lifecycle_1.0.4          
 [11] edgeR_4.0.16              hdf5r_1.3.11             
 [13] globals_0.16.3            lattice_0.22-6           
 [15] MASS_7.3-60.0.1           magrittr_2.0.3           
 [17] limma_3.58.1              plotly_4.10.4            
 [19] rmarkdown_2.29            yaml_2.3.10              
 [21] metapod_1.10.0            httpuv_1.6.15            
 [23] Seurat_5.1.0              sctransform_0.4.1        
 [25] spam_2.11-1               spatstat.sparse_3.1-0    
 [27] sp_2.2-0                  reticulate_1.40.0        
 [29] cowplot_1.1.3             pbapply_1.7-2            
 [31] DBI_1.2.3                 RColorBrewer_1.1-3       
 [33] abind_1.4-5               zlibbioc_1.48.0          
 [35] Rtsne_0.17                purrr_1.0.2              
 [37] RCurl_1.98-1.16           GenomeInfoDbData_1.2.11  
 [39] ggrepel_0.9.6             irlba_2.3.5.1            
 [41] spatstat.utils_3.1-2      listenv_0.9.1            
 [43] goftest_1.2-3             RSpectra_0.16-2          
 [45] spatstat.random_3.3-2     dqrng_0.3.2              
 [47] fitdistrplus_1.2-2        parallelly_1.42.0        
 [49] DelayedMatrixStats_1.24.0 leiden_0.4.3.1           
 [51] codetools_0.2-20          DelayedArray_0.28.0      
 [53] tidyselect_1.2.1          farver_2.1.2             
 [55] ScaledMatrix_1.10.0       viridis_0.6.5            
 [57] spatstat.explore_3.3-4    GenomicAlignments_1.38.0 
 [59] jsonlite_1.8.9            BiocNeighbors_1.20.0     
 [61] progressr_0.15.1          ggridges_0.5.6           
 [63] survival_3.8-3            tools_4.3.3              
 [65] ica_1.0-3                 Rcpp_1.0.14              
 [67] glue_1.8.0                gridExtra_2.3            
 [69] SparseArray_1.2.2         xfun_0.50                
 [71] dplyr_1.1.4               withr_3.0.2              
 [73] fastmap_1.2.0             bluster_1.12.0           
 [75] digest_0.6.37             rsvd_1.0.5               
 [77] R6_2.6.1                  mime_0.12                
 [79] colorspace_2.1-1          scattermore_1.2          
 [81] tensor_1.5                spatstat.data_3.1-4      
 [83] RSQLite_2.3.9             tidyr_1.3.1              
 [85] generics_0.1.3            data.table_1.16.4        
 [87] rtracklayer_1.62.0        httr_1.4.7               
 [89] htmlwidgets_1.6.4         S4Arrays_1.2.0           
 [91] uwot_0.2.2                pkgconfig_2.0.3          
 [93] gtable_0.3.6              blob_1.2.4               
 [95] lmtest_0.9-40             XVector_0.42.0           
 [97] htmltools_0.5.8.1         dotCall64_1.2            
 [99] SeuratObject_5.0.2        scales_1.3.0             
[101] png_0.1-8                 spatstat.univar_3.1-1    
[103] knitr_1.49                reshape2_1.4.4           
[105] rjson_0.2.23              nlme_3.1-167             
[107] cachem_1.1.0              zoo_1.8-12               
[109] stringr_1.5.1             KernSmooth_2.23-26       
[111] parallel_4.3.3            miniUI_0.1.1.1           
[113] vipor_0.4.7               restfulr_0.0.15          
[115] pillar_1.10.1             grid_4.3.3               
[117] vctrs_0.6.5               RANN_2.6.2               
[119] promises_1.3.2            BiocSingular_1.18.0      
[121] beachmat_2.18.0           xtable_1.8-4             
[123] cluster_2.1.8             beeswarm_0.4.0           
[125] evaluate_1.0.3            cli_3.6.4                
[127] locfit_1.5-9.11           compiler_4.3.3           
[129] Rsamtools_2.18.0          rlang_1.1.5              
[131] crayon_1.5.3              future.apply_1.11.3      
[133] labeling_0.4.3            plyr_1.8.9               
[135] ggbeeswarm_0.7.2          stringi_1.8.4            
[137] deldir_2.0-4              viridisLite_0.4.2        
[139] BiocParallel_1.36.0       munsell_0.5.1            
[141] Biostrings_2.70.1         lazyeval_0.2.2           
[143] spatstat.geom_3.3-5       Matrix_1.6-5             
[145] RcppHNSW_0.6.0            sparseMatrixStats_1.14.0 
[147] bit64_4.5.2               future_1.34.0            
[149] KEGGREST_1.42.0           statmod_1.5.0            
[151] shiny_1.10.0              ROCR_1.0-11              
[153] igraph_2.0.3              memoise_2.0.1            
[155] bit_4.5.0.1               xgboost_2.1.4.1          </code></pre>
</div>
</div>
</details>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"openEffect":"zoom","descPosition":"bottom","selector":".lightbox","loop":true,"closeEffect":"zoom"});</script>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>