<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="">
<meta name="description" content="Tutorials on some of the common steps during scRNAseq data analysis using three popular toolkits:  Seurat,  Bioconductor and  Scanpy.">

<title>Contents</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./assets/favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(assets/images/banner.jpg);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">


</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./assets/logos/nbis-scilifelab.png" alt="nbis-scilifelab-logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./home_schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./home_contents.html" rel="" target="" aria-current="page">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./home_precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./home_syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./home_info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-scRNAseq/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Contents</h1>
                  <div>
        <div class="description">
          Tutorials on some of the common steps during scRNAseq data analysis using three popular toolkits: <a href="https://satijalab.org/seurat/"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor</a> and <a href="https://scanpy.readthedocs.io/en/stable/"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy</a>.
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>You can run the labs either using SciLifeLab Serve applications (recommended during course) or using Docker locally on your system. Instructions on running the labs are provided <a href="./other/containers.html"><span class="badge badge-primary">here</span></a></p>
<p>A short description of the data used in the tutorials is provided <a href="./other/data.html">here</a>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>We perform the same steps with all three toolkits, but there are some small differences as all methods are not implemented everywhere. It’s up to you which toolkit you want to work with. To download file, <em>Right click &gt; Save Link As…</em>.</p>
</div>
</div>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Topic</th>
<th style="text-align: center;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat</th>
<th style="text-align: center;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor</th>
<th style="text-align: center;"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1 <i class="fa-solid fa-clipboard-list" aria-label="clipboard-list"></i> Quality Control</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_01_qc.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_01_qc.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_01_qc.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_01_qc.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_01_qc.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_01_qc.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="even">
<td style="text-align: left;">2 <i class="fa-brands fa-hubspot" aria-label="hubspot"></i> Dimensionality reduction</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_02_dimred.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_02_dimred.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_02_dimred.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_02_dimred.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_02_dimred.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_02_dimred.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="odd">
<td style="text-align: left;">3 <i class="fa-solid fa-diagram-project" aria-label="diagram-project"></i> Data integration</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_03_integration.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_03_integration.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_03_integration.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_03_integration.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_03_integration.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_03_integration.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="even">
<td style="text-align: left;">4 <i class="fa-solid fa-circle-nodes" aria-label="circle-nodes"></i> Clustering</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_04_clustering.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_04_clustering.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_04_clustering.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_04_clustering.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_04_clustering.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_04_clustering.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="odd">
<td style="text-align: left;">5 <i class="fa-solid fa-chart-simple" aria-label="chart-simple"></i> Differential expression</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_05_dge.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_05_dge.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_05_dge.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_05_dge.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_05_dge.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_05_dge.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="even">
<td style="text-align: left;">6 <i class="fa-solid fa-bacterium" aria-label="bacterium"></i> Celltype prediction</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_06_celltyping.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_06_celltyping.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_06_celltyping.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_06_celltyping.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_06_celltyping.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_06_celltyping.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
<tr class="odd">
<td style="text-align: left;">7 <i class="fa-solid fa-route" aria-label="route"></i> Trajectory inference</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_07_trajectory.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_07_trajectory.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_07_trajectory.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_07_trajectory.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
</tbody>
</table>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <em>Spatial Transcriptomics</em> labs are not covered in the workshop and are therefore optional. See <a href="./other/containers-spatial.html">instructions</a> for running these labs.</p>
</div>
</div>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Optional Topic</th>
<th style="text-align: center;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Seurat</th>
<th style="text-align: center;"><i class="fa-brands fa-r-project" aria-label="r-project"></i> Bioconductor</th>
<th style="text-align: center;"><i class="fa-brands fa-python" aria-label="python"></i> Scanpy</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">8 <i class="fa-solid fa-microscope" aria-label="microscope"></i> Spatial transcriptomics</td>
<td style="text-align: center;"><a href="labs/seurat/seurat_08_spatial.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/seurat/seurat_08_spatial.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/bioc/bioc_08_spatial.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/bioc/bioc_08_spatial.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
<td style="text-align: center;"><a href="labs/scanpy/scanpy_08_spatial.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/compiled/labs/scanpy/scanpy_08_spatial.ipynb"><i class="fa-solid fa-download" aria-label="download"></i></a></td>
</tr>
</tbody>
</table>
<section id="lecture-recordings" class="level2">
<h2 class="anchored" data-anchor-id="lecture-recordings">Lecture recordings</h2>
<p>The lectures from the 2025 course can be found on the <a href="NBIS YouTube channel.">https://www.youtube.com/playlist?list=PLBsJUKzoJTHTqY4FOO8R_8qcZb1hgB2QX</a></p>
</section>
<section id="toolkit-comparisons" class="level2">
<h2 class="anchored" data-anchor-id="toolkit-comparisons">Toolkit comparisons</h2>
<p>Below are a few reports where the results from the 3 toolkits and multiple methods in the toolkits are analysed. <em>OBS!</em> Several of these scripts requires additional packages that are not available in the Docker containers used for the exercises.</p>
<ul>
<li>General comparison of the main steps of the pipelines: <a href="labs/comparison/comparison_pipelines.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_pipelines.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></li>
<li>Comparison of Variable gene selection methods: <a href="labs/comparison/comparison_hvg.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercont%20ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_hvg.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></li>
<li>Differential gene expression detection: <a href="labs/comparison/comparison_dge.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercont%20ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_dge.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></li>
<li>Normalization methods: <a href="labs/comparison/comparison_normalization.html"><i class="fa-solid fa-file-lines" aria-label="file-lines"></i></a> <a href="https://raw.githubusercont%20ent.com/NBISweden/workshop-scRNAseq/master/labs/comparison/comparison_normalization.qmd"><i class="fa-solid fa-download" aria-label="download"></i></a></li>
</ul>
</section>
<section id="previous-course-iterations" class="level2">
<h2 class="anchored" data-anchor-id="previous-course-iterations">Previous course iterations</h2>
<ul>
<li>scRNAseq 2024 <a href="https://nbisweden.github.io/workshop-scrnaseq-2024"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>scRNAseq 2023 <a href="https://nbisweden.github.io/workshop-scrnaseq-2023"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Recorded videos of lectures (from 2022) are available on Youtube <a href="https://youtube.com/playlist?list=PLBsJUKzoJTHQA4Qg1yc1RRY2Km4t4vEeN&amp;si=p37W8NwQREqf617q"><i class="fa-brands fa-youtube" aria-label="youtube"></i></a></li>
</ul>
<p><em>NB: Seurat exercises in previous courses used Seurat v4</em></p>
</section>
<section id="useful-resources" class="level2">
<h2 class="anchored" data-anchor-id="useful-resources">Useful resources</h2>
<ul>
<li>The github repository for this course <a href="https://github.com/NBISweden/workshop-scRNAseq"><i class="fa-brands fa-github" aria-label="github"></i></a></li>
<li>Single Cell Glossary <a href="https://nbisweden.github.io/single-cell-pbl/glossary_of_terms_single_cell.html"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Single cell RNA-seq course at from Hemberg lab <a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Single cell RNA-seq course in Python <a href="https://chanzuckerberg.github.io/scRNA-python-workshop/intro/about"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Single cell RNA-seq course at Broad <a href="https://broadinstitute.github.io/2019_scWorkshop/"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Repository listing many scRNA-seq tools <a href="https://github.com/seandavi/awesome-single-cell"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>SingleCellExperiment objects for many datasets <a href="https://hemberg-lab.github.io/scRNA.seq.datasets/"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>Conquer datasets - many different datasets based on a salmon pipeline <a href="http://imlspenticton.uzh.ch:3838/conquer/"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
<li>The Human Cell Atlas project <a href="https://www.humancellatlas.org/"><i class="fa-solid fa-square-arrow-up-right" aria-label="square-arrow-up-right"></i></a></li>
</ul>
<hr>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2025 <a href="https://nbis.se">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v1.3.450
</div>
  </div>
</footer>



<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>